{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4a51acd3-1d9b-4920-8d5f-d0536f63c1e1",
   "metadata": {},
   "source": [
    "## Useful libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "id": "9b382b8b-c7f2-4ab5-a486-ad3949c22dec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import torch_geometric\n",
    "import sklearn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "id": "cb66ecc2-e662-4244-b1c7-f401a6729f8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## TODO: https://medium.com/math-simplified/checking-for-linear-dependence-in-sympy-9776b66dbe75"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "815e7d14-8b0d-4986-8fe3-0fe2ab1ecc51",
   "metadata": {},
   "source": [
    "## Utility functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "id": "211a7c02-ef90-4fdb-b990-a489fb91cab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Idea: Adding categroized features might a problem -> So lets completely remove them for now since they dont have an inherent order\n",
    "def categorize_columns(df, columns):\n",
    "    for column in columns:\n",
    "        #df[column] = df[column].astype(\"category\")\n",
    "        df.pop(column)\n",
    "    # cat_columns = df.select_dtypes(['category']).columns\n",
    "    # df[cat_columns] = df[cat_columns].apply(lambda x: x.cat.codes)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "id": "ceac2ce7-18d8-4af9-bc95-b912bf979089",
   "metadata": {},
   "outputs": [],
   "source": [
    "def z_scale(df_z_scaled, columns): ## is a bit better for neural networks\n",
    "    # for column in columns:\n",
    "    #     df_z_scaled[column] = (df_z_scaled[column] - df_z_scaled[column].mean()) / df_z_scaled[column].std()   \n",
    "    return df_z_scaled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "id": "dbe70eb0-f7b2-4176-8396-c0927e128da1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandas.api.types import is_numeric_dtype\n",
    "def remove_uninformative_columns(df, columns):\n",
    "    ## removes categorcial columns where we have no duplicates (i.e., columns with all different categorical values like ids)\n",
    "    filtered_columns = []\n",
    "    for column in columns:\n",
    "        all_values_in_column_identical = df[column].unique().shape[0] < 2\n",
    "        if all_values_in_column_identical:\n",
    "            continue\n",
    "        category_has_duplicates = df.duplicated(subset=column).sum()> 0 ##is informative\n",
    "        if category_has_duplicates or is_numeric_dtype(df[column]) :\n",
    "            filtered_columns.append(column)\n",
    "    return filtered_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "id": "3151e4a1-9cbc-472b-9570-95c1875dea43",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "def reverse_edge_index(torch_tensor):\n",
    "    reversed_torch_tensor = torch.zeros_like(torch_tensor)\n",
    "    reversed_torch_tensor[:, 0] = torch_tensor[:, 1]\n",
    "    reversed_torch_tensor[:, 1] = torch_tensor[:, 0]\n",
    "    return reversed_torch_tensor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "843dee8e-c226-44e0-a6a9-23bb4374c923",
   "metadata": {},
   "source": [
    "## What we need\n",
    "- Gene node feature #genes x #features\n",
    "- Model node feature #models x features (for now includes drug inofrmation and the drug IC50 target for simplicity) (more complicated than I expected)\n",
    "  -> later edge attribute prediction?\n",
    "- Gene_model_edge_index_cn #edges x 2\n",
    "- Gene_model_edge_index_mutations #edges x 2\n",
    "- Gene_model_edge_index_geneexpressions #edges x 2\n",
    "- Gene_model_edge_attributes_cn #edges x #attributes\n",
    "- Gene_model_edge_attributes_mutations #edges x #attributes\n",
    "- Gene_model_edge_attributes_geneexpressions #edges x #attributes\n",
    "\n",
    "  Het-GAT for regression task with MSE as loss"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d8fff4c-8aa3-44f3-a6fb-e131cde8971e",
   "metadata": {},
   "source": [
    "## Read human genome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03df2aa8-d19b-4f82-bfac-59bb75be0f4e",
   "metadata": {},
   "source": [
    "## Get gene symbols to human genome index dataframe (is in nested list semicolon, comma separated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "id": "29e589c8-0962-49fa-82d6-ed31273f874b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_idx_to_gene_symbol(human_genome):\n",
    "    sem_split = human_genome[3].str.split(\"; \")\n",
    "    comma_split = sem_split.map(lambda n: n[0]).str.split(\",\")\n",
    "    exploded_gene_symbols = comma_split.explode().str.strip().to_frame()\n",
    "    exploded_gene_symbols = exploded_gene_symbols.rename(columns={3:\"symbol\"})\n",
    "    exploded_gene_symbols = exploded_gene_symbols.drop_duplicates(subset = \"symbol\")\n",
    "    exploded_gene_symbols[\"human_genome_index\"] = exploded_gene_symbols.index\n",
    "    exploded_gene_symbols = exploded_gene_symbols.reset_index(drop=True)\n",
    "    return exploded_gene_symbols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "id": "152e224a-646f-434f-afdb-761c23f3ac92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>human_genome_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-mir-6859-1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIRN1302-2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-mir-1302-2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78333</th>\n",
       "      <td>G antigen 4</td>\n",
       "      <td>24683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78334</th>\n",
       "      <td>putative TAF11-like protein ENSP05220103537</td>\n",
       "      <td>24685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78335</th>\n",
       "      <td>Uncharacterized LOC132932482</td>\n",
       "      <td>24686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78336</th>\n",
       "      <td>POC1B-DUSP6</td>\n",
       "      <td>24687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78337</th>\n",
       "      <td>LYPLA1-TCEA1</td>\n",
       "      <td>24688</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78338 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            symbol  human_genome_index\n",
       "0                                        MIR6859-1                   0\n",
       "1                                   hsa-mir-6859-1                   0\n",
       "2                                        MIR1302-2                   1\n",
       "3                                       MIRN1302-2                   1\n",
       "4                                   hsa-mir-1302-2                   1\n",
       "...                                            ...                 ...\n",
       "78333                                  G antigen 4               24683\n",
       "78334  putative TAF11-like protein ENSP05220103537               24685\n",
       "78335                 Uncharacterized LOC132932482               24686\n",
       "78336                                  POC1B-DUSP6               24687\n",
       "78337                                 LYPLA1-TCEA1               24688\n",
       "\n",
       "[78338 rows x 2 columns]"
      ]
     },
     "execution_count": 359,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_idx_to_gene_symbol_df = get_gene_idx_to_gene_symbol(human_genome)\n",
    "gene_idx_to_gene_symbol_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "id": "f42afdff-0849-43cd-be7d-9070603e6245",
   "metadata": {},
   "outputs": [],
   "source": [
    "perfect_gene_features = ['FIGN', 'FAR2P2', 'UBAC2-AS1', 'PKM', 'LINC01752', 'FOXCUT',\n",
    "       'COL9A3', 'SPP1', 'CEP170P1', 'NF2', 'INPP1', 'ARL9', 'BTBD19',\n",
    "       'LYPD1', 'MYH14', 'ELOVL7', 'ROMO1', 'HSPE1P21', 'RPS26P21',\n",
    "       'FUT8', 'TMEM127', 'POU5F1P5', 'MANCR', 'GLO1', 'VCP', 'LPGAT1',\n",
    "       'C4BPB', 'ADGRF4', 'CCDC57', 'KLK8', 'LINC02551', 'NGEF',\n",
    "       'ANKDD1B', 'SNHG30', 'LINC01133', 'FAM189A2', 'HMMR', 'KRTAP3-1',\n",
    "       'TIAM1', 'SHKBP1', 'FASTKD3', 'CARD14', 'SPATS2L', 'TMED7-TICAM2',\n",
    "       'PABPC1P3', 'MTND6P4', 'EPS8L1', 'MIR99AHG', 'PCSK6-AS1', 'IRF6',\n",
    "       'SNORD124', 'A2M-AS1', 'SMOC1', 'SLC22A3', 'IGFBP6', 'ANXA2R',\n",
    "       'PPFIA3', 'CD9', 'GAS2L1', 'HEIH', 'GGACT', 'ANKRA2', 'NUP210',\n",
    "       'FAM43B', 'MIR574', 'KRT23', 'CTU1', 'PELO', 'ERICH5', 'LRRC2',\n",
    "       'ZNF17', 'LPP-AS2', 'HLA-F-AS1', 'RNU6-37P', 'MTMR11', 'TAGLN',\n",
    "       'AKAP12', 'PROM2', 'LYPLAL1-DT', 'PCDHB17P', 'KCNH3', 'CCDC61',\n",
    "       'RPS2P55', 'SCARNA8', 'LINC01694', 'SNHG25', 'INHA', 'CHMP6',\n",
    "       'SLC44A4', 'PIGC', 'ERCC6', 'MAGEH1', 'OR7E91P', 'RAPGEF4',\n",
    "       'RPS20', 'NECTIN3-AS1', 'LINC00622', 'RCN1P2', 'IPO4', 'VSIG10L',\n",
    "       'IMPA1P1', 'KLK7', 'IGFBP5', 'GJC2', 'NFRKB', 'RPL35P6', 'ADAM33',\n",
    "       'GDI2', 'BEND7', 'USPL1', 'FLRT3', 'RBM23', 'KLK14', 'AGAP14P',\n",
    "       'LINC02331', 'M1AP', 'UCN2', 'OSBPL9', 'SEMA4F', 'ERMP1', 'OPLAH',\n",
    "       'TMEM87A', 'ZNF674-AS1', 'INSIG2', 'MIR3192', 'NELFA', 'RPSAP58',\n",
    "       'WBP4', 'TCIRG1', 'AMZ1', 'FGFBP1', 'CDC27P1', 'COLGALT2', 'TGM2',\n",
    "       'PEDS1', 'MMP10', 'RPL7AP11', 'MFSD14A', 'TMEM253', 'DNAH5',\n",
    "       'TMLHE', 'HERC5', 'CNPY2', 'SNAPC2', 'GJB7', 'OCRL', 'MGLL',\n",
    "       'ANO1', 'CNOT3', 'SLC5A6', 'NAT1', 'ANKRD40', 'PLEKHG4B', 'FLG',\n",
    "       'LINC02340', 'ABHD11', 'LINC00659', 'RN7SL552P', 'IHO1', 'AP1M2',\n",
    "       'CCNB1IP1', 'AK4', 'SULT1C4', 'IGFL2-AS1', 'SVIL-AS1', 'PTPA',\n",
    "       'PLXDC1', 'SLC30A3', 'SH3BP2', 'KCNK2', 'MTCO3P12', 'MIR6835',\n",
    "       'MRPL37', 'MAGEF1', 'PRR34-AS1', 'HDLBP', 'SDR42E1', 'CASP7',\n",
    "       'NKILA', 'AFF1', 'NUPR1', 'CUBN', 'TGFBR3', 'PPP1R10', 'MIR181B2',\n",
    "       'LIF', 'IFRD1', 'TLL2', 'ERICH6-AS1', 'RNU6-589P', 'A1BG-AS1',\n",
    "       'MSX2', 'MFSD3', 'LINC02102', 'LDHAL6FP', 'PTGR1', 'CDYL2', 'PEX3',\n",
    "       'EEF1A1P11', 'LSMEM2', 'LINC01040', 'ZNF776', 'SLC44A2', 'CCDC9',\n",
    "       'EXOSC2', 'ATG101', 'MYD88', 'SNORA79B', 'SLC3A2', 'CERNA1',\n",
    "       'LYPD3', 'USP2', 'RPS20P2', 'VNN1', 'KREMEN2', 'HTR1D', 'TBC1D3C',\n",
    "       'RPS6KA2', 'CHKB-CPT1B', 'RNU6-469P', 'CXADR', 'BAZ1A',\n",
    "       'LINC00960', 'FTH1P23', 'RNASEK-C17orf49', 'VILL', 'BNIP3P5',\n",
    "       'PARP3', 'COBL', 'TGFB3', 'CDO1', 'CT45A11P', 'LASP1', 'HYI-AS1',\n",
    "       'NRBP2', 'CYB561A3', 'C1QTNF12', 'IQUB', 'MIR6499', 'TAFA3',\n",
    "       'TRPV6', 'CCL20', 'SLC16A3', 'RASA1', 'POMT1', 'RBM17', 'CYP4X1',\n",
    "       'SPDEF', 'VEGFC', 'APOBEC3A', 'MIR4754', 'METTL15P1', 'PNKP',\n",
    "       'UBE2L3', 'ARSK', 'BACH1-AS1', 'TRPM4', 'KCTD14', 'NNT-AS1',\n",
    "       'B3GALNT1', 'PRSS35', 'UGT1A3', 'MIR6826', 'PRSS41', 'MT-CO3',\n",
    "       'SIRT2', 'AGAP12P', 'SNX18P9', 'YBX1P1', 'ADRA1B', 'ACTL11P',\n",
    "       'MINDY3', 'CEP250', 'STAG3L5P', 'TGFA-IT1', 'IER5', 'UNC5CL',\n",
    "       'HNRNPA3P12', 'FGB', 'SOD3', 'OBSL1', 'STX10', 'PDRG1', 'SPATA17',\n",
    "       'PCED1B', 'EFNB2', 'CHCHD2P2', 'GYG1P1', 'CD109', 'EEF2K',\n",
    "       'ADAMTSL4', 'PERM1', 'CDC20B', 'RGS5', 'WNT2', 'PCBP2', 'HOMEZ',\n",
    "       'RPS12P27', 'ZNF615', 'BMS1P11', 'ZNF577', 'CFAP298', 'NRSN2',\n",
    "       'TMEM139', 'SEPTIN7P3', 'RNU11-2P', 'RPL31P59', 'RASSF3', 'EFHD1',\n",
    "       'ORAI3', 'RNU6-100P', 'ZNF662', 'VAMP8', 'HEPHL1', 'FSTL5',\n",
    "       'FAM3D', 'HS3ST1', 'NUMB', 'MYLK4', 'CLDN16', 'LINC00460', 'PROM1',\n",
    "       'PRRG2', 'CCDC6', 'GSC', 'CCDC157', 'GABARAPL1', 'RN7SKP160',\n",
    "       'PGLYRP1', 'RPS5', 'BCAN', 'FAM107A', 'ADAMTS13', 'OR6A2',\n",
    "       'ARPC1B', 'D2HGDH', 'JAZF1', 'KRTAP2-1', 'ARHGAP23', 'DEFB1',\n",
    "       'ZNF584', 'LRP1B', 'POTEI', 'NLRP1', 'RNY1', 'C6orf15', 'B4GAT1',\n",
    "       'SCARF1', 'LINC00643', 'CSGALNACT1', 'MIR1262', 'RNU6-377P',\n",
    "       'RTN4RL1', 'JOSD2', 'ATXN7L3B', 'KCP', 'MAP2K1', 'C2orf88',\n",
    "       'LAPTM4A', 'SSTR5-AS1', 'DAPK3', 'ZNF75D', 'PIGV', 'SAV1', 'PCGF2',\n",
    "       'TOB2P1', 'DNAJB12', 'AK2P2', 'YWHABP2', 'CPT1B', 'PLK3', 'PODXL',\n",
    "       'RPSAP18', 'PRPF6', 'VPS26BP1', 'CD47', 'PRR36', 'HERC2P3',\n",
    "       'TMEM134', 'ATP8B1', 'WNT4', 'ERV3-1', 'PSMD10P2', 'SCRN3',\n",
    "       'CDK10', 'CTSV', 'RNA5SP440', 'EIF1AD']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "id": "30cf5de9-20d0-429f-9d59-52015ad8a743",
   "metadata": {},
   "outputs": [],
   "source": [
    "# perfect_gene_features = features\n",
    "perfect_genes = gene_idx_to_gene_symbol_df[gene_idx_to_gene_symbol_df[\"symbol\"].isin(perfect_gene_features)].reset_index(drop=True).groupby(\"human_genome_index\")[\"symbol\"].apply(list).reset_index(drop= True).explode()\n",
    "perfect_gene_df = pd.DataFrame(columns = [\"human_genome_index\", \"symbol\"])\n",
    "perfect_gene_df[\"symbol\"] = perfect_genes\n",
    "perfect_gene_df[\"human_genome_index\"] = perfect_genes.index\n",
    "gene_idx_to_gene_symbol_df = perfect_gene_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daade5af-e566-4297-ab2d-e5dc06e78ee9",
   "metadata": {},
   "source": [
    "## Generate nodes features for gene nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "id": "6c9e6013-ece8-42de-9ecb-e57984394c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_features(human_genome):\n",
    "    categorize_columns(human_genome, human_genome.columns[:3])\n",
    "    z_scale(human_genome, human_genome.columns[:3])\n",
    "    print(human_genome)\n",
    "    gene_features = human_genome[\"human_genome_index\"].values #human_genome.loc[:, :2].values\n",
    "    return gene_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "id": "6702079a-f995-4020-bdd6-71a0eb17ce10",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "gene_features = np.eye(gene_idx_to_gene_symbol_df.shape[0]) # get_gene_features(human_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "id": "ab8f3b36-3a1a-49e8-8429-0afc0632f9bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(295, 295)"
      ]
     },
     "execution_count": 364,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_features.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1e43de4-16ad-4df2-8ffe-9876bff1e0e4",
   "metadata": {},
   "source": [
    "## Generate node features for models (with target) and target labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "id": "76e4ef3e-fad7-4389-b2d9-049d1c025b4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>model_name</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>tissue</th>\n",
       "      <th>cancer_type</th>\n",
       "      <th>cancer_type_ncit_id</th>\n",
       "      <th>tissue_status</th>\n",
       "      <th>...</th>\n",
       "      <th>msh6_expression_by_ihc</th>\n",
       "      <th>braf_mutation_identified</th>\n",
       "      <th>braf_expression_by_ihc</th>\n",
       "      <th>pik3ca_mutation_identified</th>\n",
       "      <th>pten_expression_by_ihc</th>\n",
       "      <th>pten_mutation_identified</th>\n",
       "      <th>kras_mutation_identified</th>\n",
       "      <th>mismatch_repair_status</th>\n",
       "      <th>preoperative_ce_alevel</th>\n",
       "      <th>crispr_ko_data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDM01774</td>\n",
       "      <td>SIDS01659</td>\n",
       "      <td>SIDP01578</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PK-59</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Pancreatic Carcinoma</td>\n",
       "      <td>C3850</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDM00192</td>\n",
       "      <td>SIDS00612</td>\n",
       "      <td>SIDP00541</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNU-1033</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Large Intestine</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>C9383</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDM01447</td>\n",
       "      <td>SIDS01466</td>\n",
       "      <td>SIDP01347</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNU-466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Central Nervous System</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>C3058</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDM01554</td>\n",
       "      <td>SIDS01363</td>\n",
       "      <td>SIDP01247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IST-MES-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Mesothelioma</td>\n",
       "      <td>C45662</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDM01689</td>\n",
       "      <td>SIDS01631</td>\n",
       "      <td>SIDP01557</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MUTZ-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haematopoietic and Lymphoid</td>\n",
       "      <td>B-Lymphoblastic Leukemia</td>\n",
       "      <td>C8644</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 97 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    model_id  sample_id patient_id parent_id model_name synonyms  \\\n",
       "0  SIDM01774  SIDS01659  SIDP01578       NaN      PK-59      NaN   \n",
       "1  SIDM00192  SIDS00612  SIDP00541       NaN   SNU-1033      NaN   \n",
       "2  SIDM01447  SIDS01466  SIDP01347       NaN    SNU-466      NaN   \n",
       "3  SIDM01554  SIDS01363  SIDP01247       NaN  IST-MES-2      NaN   \n",
       "4  SIDM01689  SIDS01631  SIDP01557       NaN     MUTZ-5      NaN   \n",
       "\n",
       "                        tissue               cancer_type cancer_type_ncit_id  \\\n",
       "0                     Pancreas      Pancreatic Carcinoma               C3850   \n",
       "1              Large Intestine      Colorectal Carcinoma               C9383   \n",
       "2       Central Nervous System              Glioblastoma               C3058   \n",
       "3                         Lung              Mesothelioma              C45662   \n",
       "4  Haematopoietic and Lymphoid  B-Lymphoblastic Leukemia               C8644   \n",
       "\n",
       "  tissue_status  ... msh6_expression_by_ihc braf_mutation_identified  \\\n",
       "0    Metastasis  ...                    NaN                      NaN   \n",
       "1        Tumour  ...                    NaN                      NaN   \n",
       "2        Tumour  ...                    NaN                      NaN   \n",
       "3       Unknown  ...                    NaN                      NaN   \n",
       "4        Tumour  ...                    NaN                      NaN   \n",
       "\n",
       "  braf_expression_by_ihc pik3ca_mutation_identified pten_expression_by_ihc  \\\n",
       "0                    NaN                        NaN                    NaN   \n",
       "1                    NaN                        NaN                    NaN   \n",
       "2                    NaN                        NaN                    NaN   \n",
       "3                    NaN                        NaN                    NaN   \n",
       "4                    NaN                        NaN                    NaN   \n",
       "\n",
       "   pten_mutation_identified  kras_mutation_identified  mismatch_repair_status  \\\n",
       "0                       NaN                       NaN                     NaN   \n",
       "1                       NaN                       NaN                     NaN   \n",
       "2                       NaN                       NaN                     NaN   \n",
       "3                       NaN                       NaN                     NaN   \n",
       "4                       NaN                       NaN                     NaN   \n",
       "\n",
       "  preoperative_ce_alevel  crispr_ko_data  \n",
       "0                    NaN           False  \n",
       "1                    NaN            True  \n",
       "2                    NaN           False  \n",
       "3                    NaN            True  \n",
       "4                    NaN           False  \n",
       "\n",
       "[5 rows x 97 columns]"
      ]
     },
     "execution_count": 365,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets = pd.read_csv(\"target.csv\",sep=\";\")\n",
    "models = pd.read_csv(\"model_list_20240110.csv\")\n",
    "models.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 422,
   "id": "cfed20b0-e8e0-41e7-830a-2acb82b4f584",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>model_name</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>tissue</th>\n",
       "      <th>cancer_type</th>\n",
       "      <th>cancer_type_ncit_id</th>\n",
       "      <th>tissue_status</th>\n",
       "      <th>...</th>\n",
       "      <th>msh6_expression_by_ihc</th>\n",
       "      <th>braf_mutation_identified</th>\n",
       "      <th>braf_expression_by_ihc</th>\n",
       "      <th>pik3ca_mutation_identified</th>\n",
       "      <th>pten_expression_by_ihc</th>\n",
       "      <th>pten_mutation_identified</th>\n",
       "      <th>kras_mutation_identified</th>\n",
       "      <th>mismatch_repair_status</th>\n",
       "      <th>preoperative_ce_alevel</th>\n",
       "      <th>crispr_ko_data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDM01774</td>\n",
       "      <td>SIDS01659</td>\n",
       "      <td>SIDP01578</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PK-59</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Pancreatic Carcinoma</td>\n",
       "      <td>C3850</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDM00192</td>\n",
       "      <td>SIDS00612</td>\n",
       "      <td>SIDP00541</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNU-1033</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Large Intestine</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>C9383</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDM01447</td>\n",
       "      <td>SIDS01466</td>\n",
       "      <td>SIDP01347</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNU-466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Central Nervous System</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>C3058</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDM01554</td>\n",
       "      <td>SIDS01363</td>\n",
       "      <td>SIDP01247</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IST-MES-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Mesothelioma</td>\n",
       "      <td>C45662</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDM01689</td>\n",
       "      <td>SIDS01631</td>\n",
       "      <td>SIDP01557</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MUTZ-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haematopoietic and Lymphoid</td>\n",
       "      <td>B-Lymphoblastic Leukemia</td>\n",
       "      <td>C8644</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2150</th>\n",
       "      <td>SIDM02155</td>\n",
       "      <td>SIDS02010</td>\n",
       "      <td>SIDP01906</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NZM3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>C3224</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2151</th>\n",
       "      <td>SIDM02157</td>\n",
       "      <td>SIDS02040</td>\n",
       "      <td>SIDP01935</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WM3772F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eye</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>C7712</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2152</th>\n",
       "      <td>SIDM02159</td>\n",
       "      <td>SIDS02033</td>\n",
       "      <td>SIDP01928</td>\n",
       "      <td>SIDM02121</td>\n",
       "      <td>HAP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Haematopoietic and Lymphoid</td>\n",
       "      <td>Chronic Myelogenous Leukemia</td>\n",
       "      <td>C3174</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2153</th>\n",
       "      <td>SIDM01339</td>\n",
       "      <td>SIDS01788</td>\n",
       "      <td>SIDP01690</td>\n",
       "      <td>SIDM01851</td>\n",
       "      <td>SLR24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>Kidney Carcinoma</td>\n",
       "      <td>C4033</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154</th>\n",
       "      <td>SIDM02009</td>\n",
       "      <td>SIDS01892</td>\n",
       "      <td>SIDP01788</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP550hTERT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Head and Neck</td>\n",
       "      <td>Non-Cancerous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Normal</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2155 rows × 97 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       model_id  sample_id patient_id  parent_id  model_name synonyms  \\\n",
       "0     SIDM01774  SIDS01659  SIDP01578        NaN       PK-59      NaN   \n",
       "1     SIDM00192  SIDS00612  SIDP00541        NaN    SNU-1033      NaN   \n",
       "2     SIDM01447  SIDS01466  SIDP01347        NaN     SNU-466      NaN   \n",
       "3     SIDM01554  SIDS01363  SIDP01247        NaN   IST-MES-2      NaN   \n",
       "4     SIDM01689  SIDS01631  SIDP01557        NaN      MUTZ-5      NaN   \n",
       "...         ...        ...        ...        ...         ...      ...   \n",
       "2150  SIDM02155  SIDS02010  SIDP01906        NaN        NZM3      NaN   \n",
       "2151  SIDM02157  SIDS02040  SIDP01935        NaN     WM3772F      NaN   \n",
       "2152  SIDM02159  SIDS02033  SIDP01928  SIDM02121        HAP1      NaN   \n",
       "2153  SIDM01339  SIDS01788  SIDP01690  SIDM01851       SLR24      NaN   \n",
       "2154  SIDM02009  SIDS01892  SIDP01788        NaN  NP550hTERT      NaN   \n",
       "\n",
       "                           tissue                   cancer_type  \\\n",
       "0                        Pancreas          Pancreatic Carcinoma   \n",
       "1                 Large Intestine          Colorectal Carcinoma   \n",
       "2          Central Nervous System                  Glioblastoma   \n",
       "3                            Lung                  Mesothelioma   \n",
       "4     Haematopoietic and Lymphoid      B-Lymphoblastic Leukemia   \n",
       "...                           ...                           ...   \n",
       "2150                         Skin                      Melanoma   \n",
       "2151                          Eye                      Melanoma   \n",
       "2152  Haematopoietic and Lymphoid  Chronic Myelogenous Leukemia   \n",
       "2153                       Kidney              Kidney Carcinoma   \n",
       "2154                Head and Neck                 Non-Cancerous   \n",
       "\n",
       "     cancer_type_ncit_id tissue_status  ... msh6_expression_by_ihc  \\\n",
       "0                  C3850    Metastasis  ...                    NaN   \n",
       "1                  C9383        Tumour  ...                    NaN   \n",
       "2                  C3058        Tumour  ...                    NaN   \n",
       "3                 C45662       Unknown  ...                    NaN   \n",
       "4                  C8644        Tumour  ...                    NaN   \n",
       "...                  ...           ...  ...                    ...   \n",
       "2150               C3224    Metastasis  ...                    NaN   \n",
       "2151               C7712    Metastasis  ...                    NaN   \n",
       "2152               C3174        Tumour  ...                    NaN   \n",
       "2153               C4033        Tumour  ...                    NaN   \n",
       "2154                 NaN        Normal  ...                    NaN   \n",
       "\n",
       "     braf_mutation_identified braf_expression_by_ihc  \\\n",
       "0                         NaN                    NaN   \n",
       "1                         NaN                    NaN   \n",
       "2                         NaN                    NaN   \n",
       "3                         NaN                    NaN   \n",
       "4                         NaN                    NaN   \n",
       "...                       ...                    ...   \n",
       "2150                      NaN                    NaN   \n",
       "2151                      NaN                    NaN   \n",
       "2152                      NaN                    NaN   \n",
       "2153                      NaN                    NaN   \n",
       "2154                      NaN                    NaN   \n",
       "\n",
       "     pik3ca_mutation_identified pten_expression_by_ihc  \\\n",
       "0                           NaN                    NaN   \n",
       "1                           NaN                    NaN   \n",
       "2                           NaN                    NaN   \n",
       "3                           NaN                    NaN   \n",
       "4                           NaN                    NaN   \n",
       "...                         ...                    ...   \n",
       "2150                        NaN                    NaN   \n",
       "2151                        NaN                    NaN   \n",
       "2152                        NaN                    NaN   \n",
       "2153                        NaN                    NaN   \n",
       "2154                        NaN                    NaN   \n",
       "\n",
       "      pten_mutation_identified  kras_mutation_identified  \\\n",
       "0                          NaN                       NaN   \n",
       "1                          NaN                       NaN   \n",
       "2                          NaN                       NaN   \n",
       "3                          NaN                       NaN   \n",
       "4                          NaN                       NaN   \n",
       "...                        ...                       ...   \n",
       "2150                       NaN                       NaN   \n",
       "2151                       NaN                       NaN   \n",
       "2152                       NaN                       NaN   \n",
       "2153                       NaN                       NaN   \n",
       "2154                       NaN                       NaN   \n",
       "\n",
       "      mismatch_repair_status preoperative_ce_alevel  crispr_ko_data  \n",
       "0                        NaN                    NaN           False  \n",
       "1                        NaN                    NaN            True  \n",
       "2                        NaN                    NaN           False  \n",
       "3                        NaN                    NaN            True  \n",
       "4                        NaN                    NaN           False  \n",
       "...                      ...                    ...             ...  \n",
       "2150                     NaN                    NaN           False  \n",
       "2151                     NaN                    NaN           False  \n",
       "2152                     NaN                    NaN           False  \n",
       "2153                     NaN                    NaN            True  \n",
       "2154                     NaN                    NaN           False  \n",
       "\n",
       "[2155 rows x 97 columns]"
      ]
     },
     "execution_count": 422,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 366,
   "id": "2589f1bf-13bd-4379-8b6e-90674d079ad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transformed_merged_drug_model_df(models, targets):\n",
    "    columns_with_most_val_filled = models.columns[models.isnull().sum(axis=0)/ models.shape[0] <= 0.1]\n",
    "    filtered_columns = remove_uninformative_columns(models, columns_with_most_val_filled) ##all categorical\n",
    "    filtered_columns.append(\"model_id\") ## required for later join\n",
    "    print(filtered_columns)\n",
    "    models = models[filtered_columns].fillna(\"Unkown\")\n",
    "    columns_besides_id = list(filter(lambda x: x != \"model_id\", filtered_columns))\n",
    "    models = categorize_columns(models, columns_besides_id)\n",
    "    z_scale(models, columns_besides_id)\n",
    "    models = models.rename(columns={\"model_id\":\"SANGER_MODEL_ID\"}) ##rename for join\n",
    "    limited_drug_target = targets[targets[\"DRUG_ID\"] == 1803] ##filter specific drug #1862\n",
    "    limited_drug_target_filtered = limited_drug_target.loc[:, [\"CELL_LINE_NAME\", \"SANGER_MODEL_ID\", \"MIN_CONC\", \"LN_IC50\"]]\n",
    "    limited_drug_target_filtered = categorize_columns(limited_drug_target_filtered, [\"CELL_LINE_NAME\"])\n",
    "    z_scale(limited_drug_target_filtered, list(filter(lambda x: x not in [\"LN_IC50\", \"SANGER_MODEL_ID\"], limited_drug_target_filtered.columns)))\n",
    "    model_drug_information = pd.merge(limited_drug_target_filtered, models, how=\"inner\", on=\"SANGER_MODEL_ID\") ## merge drug and model information\n",
    "\n",
    "    model_drug_information[\"model_idx\"] = model_drug_information.index ##store index for later joins to retrieve edge index\n",
    "    return model_drug_information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "id": "aefb43ef-a45a-4f20-9bd2-9b430e22e0f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['sample_id', 'patient_id', 'model_name', 'tissue', 'cancer_type', 'cancer_type_ncit_id', 'tissue_status', 'sample_site', 'cancer_type_detail', 'model_type', 'growth_properties', 'gender', 'ethnicity', 'smoking_status', 'sample_treatment', 'crispr_ko_data', 'model_id']\n"
     ]
    }
   ],
   "source": [
    "merged_drug_model_df = get_transformed_merged_drug_model_df(models,  targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 368,
   "id": "4ca00747-6fe0-448f-ad39-c4b02cc69b41",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_target_label_and_model_with_target_feature(model_drug_information):\n",
    "    target_labels = model_drug_information.loc[:, \"LN_IC50\"].values # model_drug_information.pop(\"LN_IC50\").values\n",
    "    model_feature_columns = list(filter(lambda x: x not in [\"model_idx\", \"SANGER_MODEL_ID\"], model_drug_information.columns))\n",
    "    print(model_feature_columns)\n",
    "    model_and_target_features = model_drug_information[model_feature_columns].values\n",
    "    return (target_labels, model_and_target_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "id": "3e27ec9f-5956-4778-9a10-e8e2d6799898",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['MIN_CONC', 'LN_IC50']\n"
     ]
    }
   ],
   "source": [
    "target_labels, model_and_target_features = get_target_label_and_model_with_target_feature(merged_drug_model_df) ## If using this pop IC50 otherwise label leakage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "id": "ac9787ae-7d44-4f54-a468-2d7847282c4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MIN_CONC</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>model_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>7.170000e+02</td>\n",
       "      <td>717.000000</td>\n",
       "      <td>717.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>3.001600e-02</td>\n",
       "      <td>4.960259</td>\n",
       "      <td>358.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>1.597060e-16</td>\n",
       "      <td>1.856540</td>\n",
       "      <td>207.124359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>3.001600e-02</td>\n",
       "      <td>-0.737397</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.001600e-02</td>\n",
       "      <td>3.993791</td>\n",
       "      <td>179.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>3.001600e-02</td>\n",
       "      <td>5.316799</td>\n",
       "      <td>358.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>3.001600e-02</td>\n",
       "      <td>6.294390</td>\n",
       "      <td>537.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>3.001600e-02</td>\n",
       "      <td>9.829512</td>\n",
       "      <td>716.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           MIN_CONC     LN_IC50   model_idx\n",
       "count  7.170000e+02  717.000000  717.000000\n",
       "mean   3.001600e-02    4.960259  358.000000\n",
       "std    1.597060e-16    1.856540  207.124359\n",
       "min    3.001600e-02   -0.737397    0.000000\n",
       "25%    3.001600e-02    3.993791  179.000000\n",
       "50%    3.001600e-02    5.316799  358.000000\n",
       "75%    3.001600e-02    6.294390  537.000000\n",
       "max    3.001600e-02    9.829512  716.000000"
      ]
     },
     "execution_count": 370,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_drug_model_df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "id": "61469b8d-39b4-4b17-b6c5-75a7bbbf6cda",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_and_target_features = np.eye(model_and_target_features.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 372,
   "id": "347c749c-587b-47a7-ba1f-d7c6242116ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((717,), (717, 717))"
      ]
     },
     "execution_count": 372,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_labels.shape, model_and_target_features.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdfdd5f8-1289-422e-b4f6-b43ba89a7eb9",
   "metadata": {},
   "source": [
    "## Generate edge indices and edge attributes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36927cda-6a32-4ceb-9f08-53e69f5f4b9e",
   "metadata": {},
   "source": [
    "### Read copy numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "id": "ac410ccb-6350-46c5-b5fb-2604c5b1e928",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model_name</th>\n",
       "      <th>model_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>chr_start</th>\n",
       "      <th>chr_end</th>\n",
       "      <th>total_copy_number</th>\n",
       "      <th>minor_copy_number</th>\n",
       "      <th>loh</th>\n",
       "      <th>...</th>\n",
       "      <th>seg_mean</th>\n",
       "      <th>gene_mean</th>\n",
       "      <th>num_targets</th>\n",
       "      <th>focal</th>\n",
       "      <th>breakpoints</th>\n",
       "      <th>num_snps</th>\n",
       "      <th>gatk_mean_log2_copy_ratio</th>\n",
       "      <th>comment</th>\n",
       "      <th>source</th>\n",
       "      <th>data_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MEC-1</td>\n",
       "      <td>SIDM00001</td>\n",
       "      <td>ABCB1</td>\n",
       "      <td>SIDG00064</td>\n",
       "      <td>chr7</td>\n",
       "      <td>87503966</td>\n",
       "      <td>87600461</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.599167</td>\n",
       "      <td>0.414125</td>\n",
       "      <td>28.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1124.0</td>\n",
       "      <td>0.574127</td>\n",
       "      <td>POOR GOF (70.4%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MEC-1</td>\n",
       "      <td>SIDM00001</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>SIDG00150</td>\n",
       "      <td>chr9</td>\n",
       "      <td>130714099</td>\n",
       "      <td>130885948</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.599167</td>\n",
       "      <td>0.472693</td>\n",
       "      <td>12.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>977.0</td>\n",
       "      <td>0.585569</td>\n",
       "      <td>POOR GOF (70.4%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MEC-1</td>\n",
       "      <td>SIDM00001</td>\n",
       "      <td>BRD3</td>\n",
       "      <td>SIDG02504</td>\n",
       "      <td>chr9</td>\n",
       "      <td>134033313</td>\n",
       "      <td>134053689</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.599167</td>\n",
       "      <td>0.705704</td>\n",
       "      <td>11.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>977.0</td>\n",
       "      <td>0.585569</td>\n",
       "      <td>POOR GOF (70.4%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MEC-1</td>\n",
       "      <td>SIDM00001</td>\n",
       "      <td>CARD11</td>\n",
       "      <td>SIDG03455</td>\n",
       "      <td>chr7</td>\n",
       "      <td>2906361</td>\n",
       "      <td>2958783</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.599167</td>\n",
       "      <td>0.696245</td>\n",
       "      <td>23.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1124.0</td>\n",
       "      <td>0.574127</td>\n",
       "      <td>POOR GOF (70.4%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MEC-1</td>\n",
       "      <td>SIDM00001</td>\n",
       "      <td>CDK6</td>\n",
       "      <td>SIDG04156</td>\n",
       "      <td>chr7</td>\n",
       "      <td>92614869</td>\n",
       "      <td>92833599</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.599167</td>\n",
       "      <td>0.750720</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1124.0</td>\n",
       "      <td>0.574127</td>\n",
       "      <td>POOR GOF (70.4%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  model_name   model_id  symbol    gene_id chr_name  chr_start    chr_end  \\\n",
       "0      MEC-1  SIDM00001   ABCB1  SIDG00064     chr7   87503966   87600461   \n",
       "1      MEC-1  SIDM00001    ABL1  SIDG00150     chr9  130714099  130885948   \n",
       "2      MEC-1  SIDM00001    BRD3  SIDG02504     chr9  134033313  134053689   \n",
       "3      MEC-1  SIDM00001  CARD11  SIDG03455     chr7    2906361    2958783   \n",
       "4      MEC-1  SIDM00001    CDK6  SIDG04156     chr7   92614869   92833599   \n",
       "\n",
       "   total_copy_number  minor_copy_number    loh  ...  seg_mean  gene_mean  \\\n",
       "0                3.0                1.0  False  ...  0.599167   0.414125   \n",
       "1                3.0                1.0  False  ...  0.599167   0.472693   \n",
       "2                3.0                1.0  False  ...  0.599167   0.705704   \n",
       "3                3.0                1.0  False  ...  0.599167   0.696245   \n",
       "4                3.0                1.0  False  ...  0.599167   0.750720   \n",
       "\n",
       "  num_targets  focal  breakpoints  num_snps gatk_mean_log2_copy_ratio  \\\n",
       "0        28.0  False          0.0    1124.0                  0.574127   \n",
       "1        12.0  False          0.0     977.0                  0.585569   \n",
       "2        11.0  False          0.0     977.0                  0.585569   \n",
       "3        23.0  False          0.0    1124.0                  0.574127   \n",
       "4         7.0  False          0.0    1124.0                  0.574127   \n",
       "\n",
       "            comment  source  data_type  \n",
       "0  POOR GOF (70.4%)  Sanger        WES  \n",
       "1  POOR GOF (70.4%)  Sanger        WES  \n",
       "2  POOR GOF (70.4%)  Sanger        WES  \n",
       "3  POOR GOF (70.4%)  Sanger        WES  \n",
       "4  POOR GOF (70.4%)  Sanger        WES  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 373,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_numbers = pd.read_csv(\"WES_pureCN_CNV_genes_20220623.csv\")\n",
    "copy_numbers.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ee4345d-a3f9-472d-a466-15142d6afe83",
   "metadata": {},
   "source": [
    "### Get edge index (model-[CopyNumbers]->gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "id": "d4a88811-c2ee-45d3-a73a-c8e766cef098",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_and_model_merged_df(gene_idx_to_gene_symbol_df, merged_drug_model_df, df):\n",
    "    human_genome_index_merged_df = pd.merge(gene_idx_to_gene_symbol_df, df, on=\"symbol\", how = \"inner\")\n",
    "    human_genome_index_merged_df = human_genome_index_merged_df.rename(columns={\"model_id\": \"SANGER_MODEL_ID\"}) ##rename for join\n",
    "    model_gene_merged_df = pd.merge(human_genome_index_merged_df, merged_drug_model_df, on=\"SANGER_MODEL_ID\", how=\"inner\")\n",
    "    return model_gene_merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "id": "c62a20f9-3395-4d81-b891-0ffa5bc3ccb7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9372, 2)"
      ]
     },
     "execution_count": 375,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_gene_merged_copy_number_df = get_gene_and_model_merged_df(gene_idx_to_gene_symbol_df, merged_drug_model_df, copy_numbers)\n",
    "model_to_gene_copy_number_edge_index = model_gene_merged_copy_number_df.loc[:, [\"model_idx\", \"human_genome_index\"]].values.astype(np.int64)\n",
    "model_to_gene_copy_number_edge_index.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "id": "cc78836f-01bc-45a7-8665-bec52239b444",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model_name</th>\n",
       "      <th>model_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>chr_start</th>\n",
       "      <th>chr_end</th>\n",
       "      <th>total_copy_number</th>\n",
       "      <th>minor_copy_number</th>\n",
       "      <th>loh</th>\n",
       "      <th>...</th>\n",
       "      <th>seg_mean</th>\n",
       "      <th>gene_mean</th>\n",
       "      <th>num_targets</th>\n",
       "      <th>focal</th>\n",
       "      <th>breakpoints</th>\n",
       "      <th>num_snps</th>\n",
       "      <th>gatk_mean_log2_copy_ratio</th>\n",
       "      <th>comment</th>\n",
       "      <th>source</th>\n",
       "      <th>data_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>679</th>\n",
       "      <td>MEC-1</td>\n",
       "      <td>SIDM00001</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.018974</td>\n",
       "      <td>0.203659</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3166.0</td>\n",
       "      <td>0.001309</td>\n",
       "      <td>POOR GOF (70.4%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>848</th>\n",
       "      <td>MEC-1</td>\n",
       "      <td>SIDM00001</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.149810</td>\n",
       "      <td>0.001817</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>164.0</td>\n",
       "      <td>-0.012460</td>\n",
       "      <td>POOR GOF (57.6%);NOISY LOG-RATIO;NOISY SEGMENT...</td>\n",
       "      <td>Broad</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222</th>\n",
       "      <td>NBsusSR</td>\n",
       "      <td>SIDM00002</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.221313</td>\n",
       "      <td>-0.192888</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>758.0</td>\n",
       "      <td>-0.236619</td>\n",
       "      <td>POOR GOF (78.3%);POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2845</th>\n",
       "      <td>M14</td>\n",
       "      <td>SIDM00003</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.340054</td>\n",
       "      <td>0.354836</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2993.0</td>\n",
       "      <td>0.357526</td>\n",
       "      <td>POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3182</th>\n",
       "      <td>MCAS</td>\n",
       "      <td>SIDM00008</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.179029</td>\n",
       "      <td>0.223966</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>211.0</td>\n",
       "      <td>0.020543</td>\n",
       "      <td>POOR GOF (73.2%);NOISY LOG-RATIO;NOISY SEGMENT...</td>\n",
       "      <td>Broad</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4126</th>\n",
       "      <td>Malme-3M</td>\n",
       "      <td>SIDM00011</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.169298</td>\n",
       "      <td>-0.077307</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1427.0</td>\n",
       "      <td>-0.312813</td>\n",
       "      <td>POOR GOF (71.7%);NOISY LOG-RATIO;NOISY SEGMENT...</td>\n",
       "      <td>Broad</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4951</th>\n",
       "      <td>M120327_3</td>\n",
       "      <td>SIDM00013</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.020660</td>\n",
       "      <td>-0.109530</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>-0.122533</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5729</th>\n",
       "      <td>KYSE-30</td>\n",
       "      <td>SIDM00015</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.028441</td>\n",
       "      <td>-0.076286</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>318.0</td>\n",
       "      <td>-0.139138</td>\n",
       "      <td>POOR GOF (76.6%);NOISY LOG-RATIO;NOISY SEGMENT...</td>\n",
       "      <td>Broad</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6257</th>\n",
       "      <td>KP-N-S19s</td>\n",
       "      <td>SIDM00016</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.076784</td>\n",
       "      <td>-1.090163</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>390.0</td>\n",
       "      <td>-1.243077</td>\n",
       "      <td>POOR GOF (70.8%);NOISY LOG-RATIO;NOISY SEGMENT...</td>\n",
       "      <td>Broad</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7267</th>\n",
       "      <td>NCI-H2286</td>\n",
       "      <td>SIDM00017</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.280241</td>\n",
       "      <td>0.579134</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1073.0</td>\n",
       "      <td>0.117469</td>\n",
       "      <td>POOR GOF (61.6%);NOISY LOG-RATIO;NOISY SEGMENT...</td>\n",
       "      <td>Broad</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7921</th>\n",
       "      <td>K052</td>\n",
       "      <td>SIDM00018</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.506914</td>\n",
       "      <td>0.777213</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2756.0</td>\n",
       "      <td>0.553279</td>\n",
       "      <td>POOR GOF (77.1%);POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8843</th>\n",
       "      <td>TE-12</td>\n",
       "      <td>SIDM00023</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.054266</td>\n",
       "      <td>-0.138818</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>421.0</td>\n",
       "      <td>-0.017144</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8991</th>\n",
       "      <td>UE6E7T-3</td>\n",
       "      <td>SIDM00024</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.054295</td>\n",
       "      <td>-0.015314</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5688.0</td>\n",
       "      <td>0.057005</td>\n",
       "      <td>NON-ABERRANT;POOR GOF (77.5%);POTENTIAL SAMPLE...</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9969</th>\n",
       "      <td>TASK1</td>\n",
       "      <td>SIDM00026</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.158679</td>\n",
       "      <td>0.152795</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3099.0</td>\n",
       "      <td>0.109565</td>\n",
       "      <td>POOR GOF (79.5%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11162</th>\n",
       "      <td>SK-NEP-1</td>\n",
       "      <td>SIDM00027</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.239190</td>\n",
       "      <td>0.264059</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5401.0</td>\n",
       "      <td>0.203719</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11373</th>\n",
       "      <td>SCaBER</td>\n",
       "      <td>SIDM00032</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.029134</td>\n",
       "      <td>-0.047131</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3025.0</td>\n",
       "      <td>-0.000824</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12490</th>\n",
       "      <td>S-117</td>\n",
       "      <td>SIDM00035</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.349462</td>\n",
       "      <td>0.246671</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>433.0</td>\n",
       "      <td>0.290633</td>\n",
       "      <td>POOR GOF (74.9%);POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13232</th>\n",
       "      <td>IOSE-75-16SV40</td>\n",
       "      <td>SIDM00039</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.026243</td>\n",
       "      <td>0.055357</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3170.0</td>\n",
       "      <td>0.013484</td>\n",
       "      <td>NON-ABERRANT;POOR GOF (75.6%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13674</th>\n",
       "      <td>TMK-1</td>\n",
       "      <td>SIDM00040</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.111053</td>\n",
       "      <td>0.074953</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>549.0</td>\n",
       "      <td>0.058586</td>\n",
       "      <td>POOR GOF (78.7%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14632</th>\n",
       "      <td>STS-0421</td>\n",
       "      <td>SIDM00041</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.099238</td>\n",
       "      <td>0.137666</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2740.0</td>\n",
       "      <td>0.085457</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15431</th>\n",
       "      <td>PL4</td>\n",
       "      <td>SIDM00042</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.523902</td>\n",
       "      <td>-0.305600</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1628.0</td>\n",
       "      <td>-0.553531</td>\n",
       "      <td>RARE KARYOTYPE</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15855</th>\n",
       "      <td>PCI-4B</td>\n",
       "      <td>SIDM00043</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.181295</td>\n",
       "      <td>-0.346051</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2154.0</td>\n",
       "      <td>-0.210099</td>\n",
       "      <td>POOR GOF (72.9%);POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16588</th>\n",
       "      <td>PCI-30</td>\n",
       "      <td>SIDM00044</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.268235</td>\n",
       "      <td>0.207081</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2143.0</td>\n",
       "      <td>0.272169</td>\n",
       "      <td>POOR GOF (77.8%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17132</th>\n",
       "      <td>HSC-39</td>\n",
       "      <td>SIDM00045</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>SIDG38237</td>\n",
       "      <td>chr1</td>\n",
       "      <td>2556388</td>\n",
       "      <td>2563527</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006485</td>\n",
       "      <td>-0.044981</td>\n",
       "      <td>7.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5297.0</td>\n",
       "      <td>-0.028126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           model_name   model_id    symbol    gene_id chr_name  chr_start  \\\n",
       "679             MEC-1  SIDM00001  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "848             MEC-1  SIDM00001  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "2222          NBsusSR  SIDM00002  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "2845              M14  SIDM00003  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "3182             MCAS  SIDM00008  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "4126         Malme-3M  SIDM00011  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "4951        M120327_3  SIDM00013  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "5729          KYSE-30  SIDM00015  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "6257        KP-N-S19s  SIDM00016  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "7267        NCI-H2286  SIDM00017  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "7921             K052  SIDM00018  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "8843            TE-12  SIDM00023  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "8991         UE6E7T-3  SIDM00024  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "9969            TASK1  SIDM00026  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "11162        SK-NEP-1  SIDM00027  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "11373          SCaBER  SIDM00032  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "12490           S-117  SIDM00035  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "13232  IOSE-75-16SV40  SIDM00039  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "13674           TMK-1  SIDM00040  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "14632        STS-0421  SIDM00041  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "15431             PL4  SIDM00042  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "15855          PCI-4B  SIDM00043  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "16588          PCI-30  SIDM00044  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "17132          HSC-39  SIDM00045  TNFRSF14  SIDG38237     chr1    2556388   \n",
       "\n",
       "       chr_end  total_copy_number  minor_copy_number    loh  ...  seg_mean  \\\n",
       "679    2563527                2.0                1.0  False  ...  0.018974   \n",
       "848    2563527                2.0                1.0  False  ...  0.149810   \n",
       "2222   2563527                3.0                1.0  False  ... -0.221313   \n",
       "2845   2563527                4.0                2.0  False  ...  0.340054   \n",
       "3182   2563527                4.0                2.0  False  ...  0.179029   \n",
       "4126   2563527                3.0                1.0  False  ... -0.169298   \n",
       "4951   2563527                3.0                1.0  False  ... -0.020660   \n",
       "5729   2563527                3.0                0.0   True  ...  0.028441   \n",
       "6257   2563527                1.0                0.0   True  ... -1.076784   \n",
       "7267   2563527                2.0                1.0  False  ...  0.280241   \n",
       "7921   2563527                3.0                1.0  False  ...  0.506914   \n",
       "8843   2563527                3.0                1.0  False  ...  0.054266   \n",
       "8991   2563527                2.0                1.0  False  ...  0.054295   \n",
       "9969   2563527                3.0                1.0  False  ...  0.158679   \n",
       "11162  2563527                3.0                1.0  False  ...  0.239190   \n",
       "11373  2563527                3.0                1.0  False  ...  0.029134   \n",
       "12490  2563527                5.0                1.0  False  ...  0.349462   \n",
       "13232  2563527                2.0                1.0  False  ...  0.026243   \n",
       "13674  2563527                3.0                1.0  False  ...  0.111053   \n",
       "14632  2563527                4.0                2.0  False  ...  0.099238   \n",
       "15431  2563527                3.0                1.0  False  ... -0.523902   \n",
       "15855  2563527                3.0                1.0  False  ... -0.181295   \n",
       "16588  2563527                3.0                1.0  False  ...  0.268235   \n",
       "17132  2563527                3.0                1.0  False  ...  0.006485   \n",
       "\n",
       "       gene_mean num_targets  focal  breakpoints  num_snps  \\\n",
       "679     0.203659         7.0  False          0.0    3166.0   \n",
       "848     0.001817         7.0  False          0.0     164.0   \n",
       "2222   -0.192888         7.0  False          0.0     758.0   \n",
       "2845    0.354836         7.0  False          0.0    2993.0   \n",
       "3182    0.223966         7.0  False          0.0     211.0   \n",
       "4126   -0.077307         7.0  False          0.0    1427.0   \n",
       "4951   -0.109530         7.0  False          0.0     108.0   \n",
       "5729   -0.076286         7.0  False          0.0     318.0   \n",
       "6257   -1.090163         7.0  False          0.0     390.0   \n",
       "7267    0.579134         7.0  False          0.0    1073.0   \n",
       "7921    0.777213         7.0  False          0.0    2756.0   \n",
       "8843   -0.138818         7.0  False          0.0     421.0   \n",
       "8991   -0.015314         7.0  False          0.0    5688.0   \n",
       "9969    0.152795         7.0  False          0.0    3099.0   \n",
       "11162   0.264059         7.0  False          0.0    5401.0   \n",
       "11373  -0.047131         7.0  False          0.0    3025.0   \n",
       "12490   0.246671         7.0  False          0.0     433.0   \n",
       "13232   0.055357         7.0  False          0.0    3170.0   \n",
       "13674   0.074953         7.0  False          0.0     549.0   \n",
       "14632   0.137666         7.0  False          0.0    2740.0   \n",
       "15431  -0.305600         7.0  False          0.0    1628.0   \n",
       "15855  -0.346051         7.0  False          0.0    2154.0   \n",
       "16588   0.207081         7.0  False          0.0    2143.0   \n",
       "17132  -0.044981         7.0  False          0.0    5297.0   \n",
       "\n",
       "      gatk_mean_log2_copy_ratio  \\\n",
       "679                    0.001309   \n",
       "848                   -0.012460   \n",
       "2222                  -0.236619   \n",
       "2845                   0.357526   \n",
       "3182                   0.020543   \n",
       "4126                  -0.312813   \n",
       "4951                  -0.122533   \n",
       "5729                  -0.139138   \n",
       "6257                  -1.243077   \n",
       "7267                   0.117469   \n",
       "7921                   0.553279   \n",
       "8843                  -0.017144   \n",
       "8991                   0.057005   \n",
       "9969                   0.109565   \n",
       "11162                  0.203719   \n",
       "11373                 -0.000824   \n",
       "12490                  0.290633   \n",
       "13232                  0.013484   \n",
       "13674                  0.058586   \n",
       "14632                  0.085457   \n",
       "15431                 -0.553531   \n",
       "15855                 -0.210099   \n",
       "16588                  0.272169   \n",
       "17132                 -0.028126   \n",
       "\n",
       "                                                 comment  source  data_type  \n",
       "679                                     POOR GOF (70.4%)  Sanger        WES  \n",
       "848    POOR GOF (57.6%);NOISY LOG-RATIO;NOISY SEGMENT...   Broad        WES  \n",
       "2222     POOR GOF (78.3%);POTENTIAL SAMPLE CONTAMINATION  Sanger        WES  \n",
       "2845                      POTENTIAL SAMPLE CONTAMINATION  Sanger        WES  \n",
       "3182   POOR GOF (73.2%);NOISY LOG-RATIO;NOISY SEGMENT...   Broad        WES  \n",
       "4126   POOR GOF (71.7%);NOISY LOG-RATIO;NOISY SEGMENT...   Broad        WES  \n",
       "4951                                                 NaN  Sanger        WES  \n",
       "5729   POOR GOF (76.6%);NOISY LOG-RATIO;NOISY SEGMENT...   Broad        WES  \n",
       "6257   POOR GOF (70.8%);NOISY LOG-RATIO;NOISY SEGMENT...   Broad        WES  \n",
       "7267   POOR GOF (61.6%);NOISY LOG-RATIO;NOISY SEGMENT...   Broad        WES  \n",
       "7921     POOR GOF (77.1%);POTENTIAL SAMPLE CONTAMINATION  Sanger        WES  \n",
       "8843                                                 NaN  Sanger        WES  \n",
       "8991   NON-ABERRANT;POOR GOF (77.5%);POTENTIAL SAMPLE...  Sanger        WES  \n",
       "9969                                    POOR GOF (79.5%)  Sanger        WES  \n",
       "11162                                                NaN  Sanger        WES  \n",
       "11373                                                NaN  Sanger        WES  \n",
       "12490    POOR GOF (74.9%);POTENTIAL SAMPLE CONTAMINATION  Sanger        WES  \n",
       "13232                      NON-ABERRANT;POOR GOF (75.6%)  Sanger        WES  \n",
       "13674                                   POOR GOF (78.7%)  Sanger        WES  \n",
       "14632                                                NaN  Sanger        WES  \n",
       "15431                                     RARE KARYOTYPE  Sanger        WES  \n",
       "15855    POOR GOF (72.9%);POTENTIAL SAMPLE CONTAMINATION  Sanger        WES  \n",
       "16588                                   POOR GOF (77.8%)  Sanger        WES  \n",
       "17132                                                NaN  Sanger        WES  \n",
       "\n",
       "[24 rows x 23 columns]"
      ]
     },
     "execution_count": 376,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "copy_numbers[copy_numbers[\"symbol\"] == \"TNFRSF14\"].head(24)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "id": "a7a4e8d7-276d-4b7c-9bbf-557292096209",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_genome_index</th>\n",
       "      <th>symbol</th>\n",
       "      <th>model_name</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr_name</th>\n",
       "      <th>chr_start</th>\n",
       "      <th>chr_end</th>\n",
       "      <th>total_copy_number</th>\n",
       "      <th>minor_copy_number</th>\n",
       "      <th>...</th>\n",
       "      <th>focal</th>\n",
       "      <th>breakpoints</th>\n",
       "      <th>num_snps</th>\n",
       "      <th>gatk_mean_log2_copy_ratio</th>\n",
       "      <th>comment</th>\n",
       "      <th>source</th>\n",
       "      <th>data_type</th>\n",
       "      <th>MIN_CONC</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>model_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>M14</td>\n",
       "      <td>SIDM00003</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3702.0</td>\n",
       "      <td>-0.025029</td>\n",
       "      <td>POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>6.662953</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>TE-12</td>\n",
       "      <td>SIDM00023</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2535.0</td>\n",
       "      <td>0.058900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>3.234855</td>\n",
       "      <td>503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>PL4</td>\n",
       "      <td>SIDM00042</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1713.0</td>\n",
       "      <td>-0.072202</td>\n",
       "      <td>RARE KARYOTYPE</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>3.001589</td>\n",
       "      <td>643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>PCI-4B</td>\n",
       "      <td>SIDM00043</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>497.0</td>\n",
       "      <td>-0.284538</td>\n",
       "      <td>POOR GOF (72.9%);POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>4.778426</td>\n",
       "      <td>642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>H3255</td>\n",
       "      <td>SIDM00046</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3959.0</td>\n",
       "      <td>0.019040</td>\n",
       "      <td>POOR GOF (74.7%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>2.772832</td>\n",
       "      <td>595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>EMC-BAC-2</td>\n",
       "      <td>SIDM00047</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3896.0</td>\n",
       "      <td>0.005817</td>\n",
       "      <td>POOR GOF (79.2%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>5.359479</td>\n",
       "      <td>705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>DiFi</td>\n",
       "      <td>SIDM00049</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>517.0</td>\n",
       "      <td>-0.731585</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>1.573649</td>\n",
       "      <td>716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>K5</td>\n",
       "      <td>SIDM00056</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4305.0</td>\n",
       "      <td>-0.020164</td>\n",
       "      <td>POOR GOF (74.8%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>6.810090</td>\n",
       "      <td>486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>HLE</td>\n",
       "      <td>SIDM00064</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>970.0</td>\n",
       "      <td>-0.454736</td>\n",
       "      <td>POOR GOF (72.6%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>5.189920</td>\n",
       "      <td>241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>BICR10</td>\n",
       "      <td>SIDM00077</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1395.0</td>\n",
       "      <td>0.165381</td>\n",
       "      <td>POOR GOF (76.7%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>4.570718</td>\n",
       "      <td>598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>SK-OV-3</td>\n",
       "      <td>SIDM00079</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1252.0</td>\n",
       "      <td>0.014300</td>\n",
       "      <td>POOR GOF (73.5%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>7.840833</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>SK-MEL-5</td>\n",
       "      <td>SIDM00080</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>0.003297</td>\n",
       "      <td>RARE KARYOTYPE;POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>6.649351</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>SK-MEL-2</td>\n",
       "      <td>SIDM00082</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1541.0</td>\n",
       "      <td>-0.039908</td>\n",
       "      <td>POOR GOF (71.9%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>6.554291</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>SF539</td>\n",
       "      <td>SIDM00083</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>0.074819</td>\n",
       "      <td>POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>6.211021</td>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>SF268</td>\n",
       "      <td>SIDM00085</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1977.0</td>\n",
       "      <td>0.093663</td>\n",
       "      <td>POOR GOF (76.1%);POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>6.708500</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>RXF393</td>\n",
       "      <td>SIDM00086</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1873.0</td>\n",
       "      <td>0.429839</td>\n",
       "      <td>POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>4.232755</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>RPMI-8226</td>\n",
       "      <td>SIDM00087</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3156.0</td>\n",
       "      <td>-0.027947</td>\n",
       "      <td>POOR GOF (79.1%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>5.485755</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>PC-3</td>\n",
       "      <td>SIDM00088</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>0.006740</td>\n",
       "      <td>POOR GOF (78.8%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>4.061616</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>OVCAR-8</td>\n",
       "      <td>SIDM00090</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>501.0</td>\n",
       "      <td>0.849006</td>\n",
       "      <td>POOR GOF (78.3%);EXCESSIVE LOH</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>3.588933</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>OVCAR-5</td>\n",
       "      <td>SIDM00091</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3598.0</td>\n",
       "      <td>0.150751</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>8.155687</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>OVCAR-4</td>\n",
       "      <td>SIDM00092</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>647.0</td>\n",
       "      <td>0.231847</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>0.980648</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>SN12C</td>\n",
       "      <td>SIDM00094</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2508.0</td>\n",
       "      <td>-0.064742</td>\n",
       "      <td>POOR GOF (78%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>3.335386</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>SR</td>\n",
       "      <td>SIDM00096</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3934.0</td>\n",
       "      <td>-0.050211</td>\n",
       "      <td>POOR GOF (78.9%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>5.321334</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>T47D</td>\n",
       "      <td>SIDM00097</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>277.0</td>\n",
       "      <td>-0.268251</td>\n",
       "      <td>POOR GOF (75.9%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>3.701566</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>NCI-H2731</td>\n",
       "      <td>SIDM00098</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4952.0</td>\n",
       "      <td>0.073499</td>\n",
       "      <td>POOR GOF (77.7%);POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>6.269917</td>\n",
       "      <td>539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>NCI-H2722</td>\n",
       "      <td>SIDM00099</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2058.0</td>\n",
       "      <td>0.126428</td>\n",
       "      <td>POOR GOF (69.6%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>6.516027</td>\n",
       "      <td>609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>NCI-H2595</td>\n",
       "      <td>SIDM00100</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2099.0</td>\n",
       "      <td>-0.031144</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>6.050840</td>\n",
       "      <td>538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>NCI-H2591</td>\n",
       "      <td>SIDM00101</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4194.0</td>\n",
       "      <td>-0.006577</td>\n",
       "      <td>POOR GOF (76.1%);POTENTIAL SAMPLE CONTAMINATION</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>7.336872</td>\n",
       "      <td>537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>NCI-H2461</td>\n",
       "      <td>SIDM00102</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3617.0</td>\n",
       "      <td>0.028060</td>\n",
       "      <td>POOR GOF (75.9%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>7.046642</td>\n",
       "      <td>608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>39</td>\n",
       "      <td>TMEM127</td>\n",
       "      <td>NCI-H2373</td>\n",
       "      <td>SIDM00103</td>\n",
       "      <td>SIDG37976</td>\n",
       "      <td>chr2</td>\n",
       "      <td>96253561</td>\n",
       "      <td>96265619</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3739.0</td>\n",
       "      <td>-0.001297</td>\n",
       "      <td>POOR GOF (75.4%)</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>WES</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>5.906272</td>\n",
       "      <td>607</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    human_genome_index   symbol model_name SANGER_MODEL_ID    gene_id  \\\n",
       "0                   39  TMEM127        M14       SIDM00003  SIDG37976   \n",
       "1                   39  TMEM127      TE-12       SIDM00023  SIDG37976   \n",
       "2                   39  TMEM127        PL4       SIDM00042  SIDG37976   \n",
       "3                   39  TMEM127     PCI-4B       SIDM00043  SIDG37976   \n",
       "4                   39  TMEM127      H3255       SIDM00046  SIDG37976   \n",
       "5                   39  TMEM127  EMC-BAC-2       SIDM00047  SIDG37976   \n",
       "6                   39  TMEM127       DiFi       SIDM00049  SIDG37976   \n",
       "7                   39  TMEM127         K5       SIDM00056  SIDG37976   \n",
       "8                   39  TMEM127        HLE       SIDM00064  SIDG37976   \n",
       "9                   39  TMEM127     BICR10       SIDM00077  SIDG37976   \n",
       "10                  39  TMEM127    SK-OV-3       SIDM00079  SIDG37976   \n",
       "11                  39  TMEM127   SK-MEL-5       SIDM00080  SIDG37976   \n",
       "12                  39  TMEM127   SK-MEL-2       SIDM00082  SIDG37976   \n",
       "13                  39  TMEM127      SF539       SIDM00083  SIDG37976   \n",
       "14                  39  TMEM127      SF268       SIDM00085  SIDG37976   \n",
       "15                  39  TMEM127     RXF393       SIDM00086  SIDG37976   \n",
       "16                  39  TMEM127  RPMI-8226       SIDM00087  SIDG37976   \n",
       "17                  39  TMEM127       PC-3       SIDM00088  SIDG37976   \n",
       "18                  39  TMEM127    OVCAR-8       SIDM00090  SIDG37976   \n",
       "19                  39  TMEM127    OVCAR-5       SIDM00091  SIDG37976   \n",
       "20                  39  TMEM127    OVCAR-4       SIDM00092  SIDG37976   \n",
       "21                  39  TMEM127      SN12C       SIDM00094  SIDG37976   \n",
       "22                  39  TMEM127         SR       SIDM00096  SIDG37976   \n",
       "23                  39  TMEM127       T47D       SIDM00097  SIDG37976   \n",
       "24                  39  TMEM127  NCI-H2731       SIDM00098  SIDG37976   \n",
       "25                  39  TMEM127  NCI-H2722       SIDM00099  SIDG37976   \n",
       "26                  39  TMEM127  NCI-H2595       SIDM00100  SIDG37976   \n",
       "27                  39  TMEM127  NCI-H2591       SIDM00101  SIDG37976   \n",
       "28                  39  TMEM127  NCI-H2461       SIDM00102  SIDG37976   \n",
       "29                  39  TMEM127  NCI-H2373       SIDM00103  SIDG37976   \n",
       "\n",
       "   chr_name  chr_start   chr_end  total_copy_number  minor_copy_number  ...  \\\n",
       "0      chr2   96253561  96265619                3.0                1.0  ...   \n",
       "1      chr2   96253561  96265619                3.0                1.0  ...   \n",
       "2      chr2   96253561  96265619                5.0                2.0  ...   \n",
       "3      chr2   96253561  96265619                3.0                0.0  ...   \n",
       "4      chr2   96253561  96265619                4.0                2.0  ...   \n",
       "5      chr2   96253561  96265619                2.0                1.0  ...   \n",
       "6      chr2   96253561  96265619                2.0                0.0  ...   \n",
       "7      chr2   96253561  96265619                2.0                1.0  ...   \n",
       "8      chr2   96253561  96265619                2.0                1.0  ...   \n",
       "9      chr2   96253561  96265619                2.0                1.0  ...   \n",
       "10     chr2   96253561  96265619                2.0                1.0  ...   \n",
       "11     chr2   96253561  96265619                5.0                2.0  ...   \n",
       "12     chr2   96253561  96265619                2.0                1.0  ...   \n",
       "13     chr2   96253561  96265619                3.0                1.0  ...   \n",
       "14     chr2   96253561  96265619                3.0                1.0  ...   \n",
       "15     chr2   96253561  96265619                4.0                2.0  ...   \n",
       "16     chr2   96253561  96265619                3.0                1.0  ...   \n",
       "17     chr2   96253561  96265619                3.0                1.0  ...   \n",
       "18     chr2   96253561  96265619                5.0                1.0  ...   \n",
       "19     chr2   96253561  96265619                3.0                1.0  ...   \n",
       "20     chr2   96253561  96265619                3.0                1.0  ...   \n",
       "21     chr2   96253561  96265619                3.0                1.0  ...   \n",
       "22     chr2   96253561  96265619                2.0                1.0  ...   \n",
       "23     chr2   96253561  96265619                2.0                1.0  ...   \n",
       "24     chr2   96253561  96265619                3.0                1.0  ...   \n",
       "25     chr2   96253561  96265619                2.0                1.0  ...   \n",
       "26     chr2   96253561  96265619                3.0                1.0  ...   \n",
       "27     chr2   96253561  96265619                2.0                1.0  ...   \n",
       "28     chr2   96253561  96265619                3.0                1.0  ...   \n",
       "29     chr2   96253561  96265619                2.0                1.0  ...   \n",
       "\n",
       "    focal breakpoints  num_snps gatk_mean_log2_copy_ratio  \\\n",
       "0   False         0.0    3702.0                 -0.025029   \n",
       "1   False         0.0    2535.0                  0.058900   \n",
       "2   False         0.0    1713.0                 -0.072202   \n",
       "3   False         0.0     497.0                 -0.284538   \n",
       "4   False         0.0    3959.0                  0.019040   \n",
       "5   False         0.0    3896.0                  0.005817   \n",
       "6   False         0.0     517.0                 -0.731585   \n",
       "7   False         0.0    4305.0                 -0.020164   \n",
       "8   False         0.0     970.0                 -0.454736   \n",
       "9   False         0.0    1395.0                  0.165381   \n",
       "10  False         0.0    1252.0                  0.014300   \n",
       "11  False         0.0    2017.0                  0.003297   \n",
       "12  False         0.0    1541.0                 -0.039908   \n",
       "13  False         0.0     255.0                  0.074819   \n",
       "14  False         0.0    1977.0                  0.093663   \n",
       "15  False         0.0    1873.0                  0.429839   \n",
       "16  False         0.0    3156.0                 -0.027947   \n",
       "17  False         0.0    2018.0                  0.006740   \n",
       "18  False         0.0     501.0                  0.849006   \n",
       "19  False         0.0    3598.0                  0.150751   \n",
       "20  False         0.0     647.0                  0.231847   \n",
       "21  False         0.0    2508.0                 -0.064742   \n",
       "22  False         0.0    3934.0                 -0.050211   \n",
       "23  False         0.0     277.0                 -0.268251   \n",
       "24  False         0.0    4952.0                  0.073499   \n",
       "25  False         0.0    2058.0                  0.126428   \n",
       "26  False         0.0    2099.0                 -0.031144   \n",
       "27  False         0.0    4194.0                 -0.006577   \n",
       "28  False         0.0    3617.0                  0.028060   \n",
       "29  False         0.0    3739.0                 -0.001297   \n",
       "\n",
       "                                            comment  source  data_type  \\\n",
       "0                    POTENTIAL SAMPLE CONTAMINATION  Sanger        WES   \n",
       "1                                               NaN  Sanger        WES   \n",
       "2                                    RARE KARYOTYPE  Sanger        WES   \n",
       "3   POOR GOF (72.9%);POTENTIAL SAMPLE CONTAMINATION  Sanger        WES   \n",
       "4                                  POOR GOF (74.7%)  Sanger        WES   \n",
       "5                                  POOR GOF (79.2%)  Sanger        WES   \n",
       "6                                               NaN  Sanger        WES   \n",
       "7                                  POOR GOF (74.8%)  Sanger        WES   \n",
       "8                                  POOR GOF (72.6%)  Sanger        WES   \n",
       "9                                  POOR GOF (76.7%)  Sanger        WES   \n",
       "10                                 POOR GOF (73.5%)  Sanger        WES   \n",
       "11    RARE KARYOTYPE;POTENTIAL SAMPLE CONTAMINATION  Sanger        WES   \n",
       "12                                 POOR GOF (71.9%)  Sanger        WES   \n",
       "13                   POTENTIAL SAMPLE CONTAMINATION  Sanger        WES   \n",
       "14  POOR GOF (76.1%);POTENTIAL SAMPLE CONTAMINATION  Sanger        WES   \n",
       "15                   POTENTIAL SAMPLE CONTAMINATION  Sanger        WES   \n",
       "16                                 POOR GOF (79.1%)  Sanger        WES   \n",
       "17                                 POOR GOF (78.8%)  Sanger        WES   \n",
       "18                   POOR GOF (78.3%);EXCESSIVE LOH  Sanger        WES   \n",
       "19                                              NaN  Sanger        WES   \n",
       "20                                              NaN  Sanger        WES   \n",
       "21                                   POOR GOF (78%)  Sanger        WES   \n",
       "22                                 POOR GOF (78.9%)  Sanger        WES   \n",
       "23                                 POOR GOF (75.9%)  Sanger        WES   \n",
       "24  POOR GOF (77.7%);POTENTIAL SAMPLE CONTAMINATION  Sanger        WES   \n",
       "25                                 POOR GOF (69.6%)  Sanger        WES   \n",
       "26                                              NaN  Sanger        WES   \n",
       "27  POOR GOF (76.1%);POTENTIAL SAMPLE CONTAMINATION  Sanger        WES   \n",
       "28                                 POOR GOF (75.9%)  Sanger        WES   \n",
       "29                                 POOR GOF (75.4%)  Sanger        WES   \n",
       "\n",
       "    MIN_CONC   LN_IC50  model_idx  \n",
       "0   0.030016  6.662953        157  \n",
       "1   0.030016  3.234855        503  \n",
       "2   0.030016  3.001589        643  \n",
       "3   0.030016  4.778426        642  \n",
       "4   0.030016  2.772832        595  \n",
       "5   0.030016  5.359479        705  \n",
       "6   0.030016  1.573649        716  \n",
       "7   0.030016  6.810090        486  \n",
       "8   0.030016  5.189920        241  \n",
       "9   0.030016  4.570718        598  \n",
       "10  0.030016  7.840833        143  \n",
       "11  0.030016  6.649351        140  \n",
       "12  0.030016  6.554291        139  \n",
       "13  0.030016  6.211021        164  \n",
       "14  0.030016  6.708500        165  \n",
       "15  0.030016  4.232755        160  \n",
       "16  0.030016  5.485755        147  \n",
       "17  0.030016  4.061616        121  \n",
       "18  0.030016  3.588933        168  \n",
       "19  0.030016  8.155687        151  \n",
       "20  0.030016  0.980648        167  \n",
       "21  0.030016  3.335386        161  \n",
       "22  0.030016  5.321334        148  \n",
       "23  0.030016  3.701566        131  \n",
       "24  0.030016  6.269917        539  \n",
       "25  0.030016  6.516027        609  \n",
       "26  0.030016  6.050840        538  \n",
       "27  0.030016  7.336872        537  \n",
       "28  0.030016  7.046642        608  \n",
       "29  0.030016  5.906272        607  \n",
       "\n",
       "[30 rows x 27 columns]"
      ]
     },
     "execution_count": 377,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_gene_merged_copy_number_df.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 378,
   "id": "567cd356-bd52-49bd-849f-b81bdbc47b6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>MIN_CONC</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>model_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>SIDM00023</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>3.234855</td>\n",
       "      <td>503</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    SANGER_MODEL_ID  MIN_CONC   LN_IC50  model_idx\n",
       "503       SIDM00023  0.030016  3.234855        503"
      ]
     },
     "execution_count": 378,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_drug_model_df[merged_drug_model_df[\"model_idx\"] == 503]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 379,
   "id": "cfd10d8f-d652-4859-8db4-661a3bda70a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>MIN_CONC</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>model_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDM01132</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>5.078148</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDM00848</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>2.849687</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>5.814401</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDM00269</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>3.390164</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDM00203</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>6.030742</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>712</th>\n",
       "      <td>SIDM00214</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>4.973348</td>\n",
       "      <td>712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713</th>\n",
       "      <td>SIDM00194</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>3.812519</td>\n",
       "      <td>713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>714</th>\n",
       "      <td>SIDM00193</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>2.698528</td>\n",
       "      <td>714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>SIDM00498</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>7.079313</td>\n",
       "      <td>715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>SIDM00049</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>1.573649</td>\n",
       "      <td>716</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>717 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    SANGER_MODEL_ID  MIN_CONC   LN_IC50  model_idx\n",
       "0         SIDM01132  0.030016  5.078148          0\n",
       "1         SIDM00848  0.030016  2.849687          1\n",
       "2         SIDM00263  0.030016  5.814401          2\n",
       "3         SIDM00269  0.030016  3.390164          3\n",
       "4         SIDM00203  0.030016  6.030742          4\n",
       "..              ...       ...       ...        ...\n",
       "712       SIDM00214  0.030016  4.973348        712\n",
       "713       SIDM00194  0.030016  3.812519        713\n",
       "714       SIDM00193  0.030016  2.698528        714\n",
       "715       SIDM00498  0.030016  7.079313        715\n",
       "716       SIDM00049  0.030016  1.573649        716\n",
       "\n",
       "[717 rows x 4 columns]"
      ]
     },
     "execution_count": 379,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_drug_model_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0f5f95a-c45a-4350-8f3a-290a63ebd1b2",
   "metadata": {},
   "source": [
    "### Get edge attributes (model-[CopyNumbers]->gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "id": "5bb18868-44e0-4471-9cba-7700d8e7c982",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9372, 7)"
      ]
     },
     "execution_count": 380,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reduced_model_gene_attributes = model_gene_merged_copy_number_df.loc[:, ['total_copy_number',\n",
    "       'minor_copy_number', 'cn_category', 'seg_mean', 'gene_mean', 'num_targets', 'num_snps', 'gatk_mean_log2_copy_ratio']]\n",
    "reduced_model_gene_attributes = categorize_columns(reduced_model_gene_attributes, [\"cn_category\"])\n",
    "reduced_model_gene_attributes = reduced_model_gene_attributes.fillna(0)\n",
    "reduced_model_gene_attributes = z_scale(reduced_model_gene_attributes, reduced_model_gene_attributes.columns)\n",
    "copy_number_edge_attributes = reduced_model_gene_attributes.values\n",
    "copy_number_edge_attributes.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e38b19ee-40f1-4dc3-9e7d-d72f06e3850c",
   "metadata": {},
   "source": [
    "### Read mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "id": "24eb7bf2-0828-437d-8d47-7f0b8b99cebf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDG13960</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>SIDM02080</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>r.373a&gt;c</td>\n",
       "      <td>c.183A&gt;C</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>missense</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1308-C25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG38363</td>\n",
       "      <td>TP53</td>\n",
       "      <td>SIDM02080</td>\n",
       "      <td>p.E204*</td>\n",
       "      <td>r.800g&gt;u</td>\n",
       "      <td>c.610G&gt;T</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>nonsense</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1308-C25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG35070</td>\n",
       "      <td>SMAD4</td>\n",
       "      <td>SIDM02080</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.1494-2a&gt;g</td>\n",
       "      <td>c.956-2A&gt;G</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ess_splice</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1308-C25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDG25750</td>\n",
       "      <td>PREX2</td>\n",
       "      <td>SIDM02080</td>\n",
       "      <td>p.E553K</td>\n",
       "      <td>r.2006g&gt;a</td>\n",
       "      <td>c.1657G&gt;A</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>missense</td>\n",
       "      <td>0.2911</td>\n",
       "      <td>True</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1308-C25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDG01214</td>\n",
       "      <td>APC</td>\n",
       "      <td>SIDM02066</td>\n",
       "      <td>p.R790fs*8</td>\n",
       "      <td>r.2425_2426delAG</td>\n",
       "      <td>c.2366_2367delAG</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1313-C18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id symbol   model_id protein_mutation      rna_mutation  \\\n",
       "0  SIDG13960   KRAS  SIDM02080           p.Q61H          r.373a>c   \n",
       "1  SIDG38363   TP53  SIDM02080          p.E204*          r.800g>u   \n",
       "2  SIDG35070  SMAD4  SIDM02080              p.?       r.1494-2a>g   \n",
       "3  SIDG25750  PREX2  SIDM02080          p.E553K         r.2006g>a   \n",
       "4  SIDG01214    APC  SIDM02066       p.R790fs*8  r.2425_2426delAG   \n",
       "\n",
       "      cdna_mutation  cancer_driver  cancer_predisposition_variant      effect  \\\n",
       "0          c.183A>C           True                          False    missense   \n",
       "1          c.610G>T           True                          False    nonsense   \n",
       "2        c.956-2A>G           True                          False  ess_splice   \n",
       "3         c.1657G>A           True                          False    missense   \n",
       "4  c.2366_2367delAG           True                          False  frameshift   \n",
       "\n",
       "      vaf  coding  source         model_name  \n",
       "0  1.0000    True  Sanger  HCM-SANG-1308-C25  \n",
       "1  1.0000    True  Sanger  HCM-SANG-1308-C25  \n",
       "2  1.0000    True  Sanger  HCM-SANG-1308-C25  \n",
       "3  0.2911    True  Sanger  HCM-SANG-1308-C25  \n",
       "4  1.0000    True  Sanger  HCM-SANG-1313-C18  "
      ]
     },
     "execution_count": 381,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations = pd.read_csv(\"mutations_summary_20221018.csv\")\n",
    "mutations = mutations.rename(columns={\"gene_symbol\": \"symbol\"})\n",
    "mutations.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ebc93e5-1763-4867-bcdc-e8467e65ca1d",
   "metadata": {},
   "source": [
    "### Get edge index (model-[Mutations]->gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 382,
   "id": "8ceb435d-01a5-4c96-874d-b1f1bf7e5530",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(121, 2)"
      ]
     },
     "execution_count": 382,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_gene_merged_mutations_df = get_gene_and_model_merged_df(gene_idx_to_gene_symbol_df, merged_drug_model_df, mutations)\n",
    "model_to_gene_mutations_edge_index = model_gene_merged_mutations_df.loc[:, [\"model_idx\", \"human_genome_index\"]].values.astype(np.int64)\n",
    "model_to_gene_mutations_edge_index.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "id": "033e8217-1ba4-4e22-b3e8-686a5184b769",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein_mutation                 0\n",
       "rna_mutation                     0\n",
       "cdna_mutation                    0\n",
       "cancer_driver                    0\n",
       "cancer_predisposition_variant    0\n",
       "effect                           0\n",
       "vaf                              0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 383,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_gene_merged_mutations_df.loc[:, [\"protein_mutation\", \"rna_mutation\", \"cdna_mutation\", \"cancer_driver\", \"cancer_predisposition_variant\", \"effect\", \"vaf\"]].isnull().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11fb6dbe-2fd2-4915-8f94-8817c1744e70",
   "metadata": {},
   "source": [
    "### Get edge attributes (model-[Mutations]->gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "id": "b4327178-6920-41d4-81ea-576c4f9e41de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(121, 1)"
      ]
     },
     "execution_count": 384,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_edge_attributes = model_gene_merged_mutations_df.loc[:, [\"protein_mutation\", \"rna_mutation\", \"cdna_mutation\", \"cancer_driver\", \"cancer_predisposition_variant\", \"effect\", \"vaf\"]]\n",
    "mutations_edge_attributes = categorize_columns(mutations_edge_attributes,[\"protein_mutation\", \"rna_mutation\", \"cdna_mutation\", \"cancer_driver\", \"cancer_predisposition_variant\", \"effect\"])\n",
    "mutations_edge_attributes = z_scale(mutations_edge_attributes, mutations_edge_attributes.columns).values\n",
    "mutations_edge_attributes.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3550c6da-5eab-475e-80cf-46de873e31ff",
   "metadata": {},
   "source": [
    "### Read Gene expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "id": "f767e84a-e9e4-4232-851b-0471b192677c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>...</th>\n",
       "      <th>1423</th>\n",
       "      <th>1424</th>\n",
       "      <th>1425</th>\n",
       "      <th>1426</th>\n",
       "      <th>1427</th>\n",
       "      <th>1428</th>\n",
       "      <th>1429</th>\n",
       "      <th>1430</th>\n",
       "      <th>1431</th>\n",
       "      <th>1432</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>model_id</th>\n",
       "      <td>NaN</td>\n",
       "      <td>SIDM00001</td>\n",
       "      <td>SIDM00002</td>\n",
       "      <td>SIDM00003</td>\n",
       "      <td>SIDM00005</td>\n",
       "      <td>SIDM00006</td>\n",
       "      <td>SIDM00007</td>\n",
       "      <td>SIDM00008</td>\n",
       "      <td>SIDM00009</td>\n",
       "      <td>SIDM00011</td>\n",
       "      <td>...</td>\n",
       "      <td>SIDM02071</td>\n",
       "      <td>SIDM02072</td>\n",
       "      <td>SIDM02073</td>\n",
       "      <td>SIDM02074</td>\n",
       "      <td>SIDM02075</td>\n",
       "      <td>SIDM02076</td>\n",
       "      <td>SIDM02077</td>\n",
       "      <td>SIDM02078</td>\n",
       "      <td>SIDM02079</td>\n",
       "      <td>SIDM02080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model_name</th>\n",
       "      <td>NaN</td>\n",
       "      <td>MEC-1</td>\n",
       "      <td>NBsusSR</td>\n",
       "      <td>M14</td>\n",
       "      <td>MDA-MB-134-VI</td>\n",
       "      <td>MCC26</td>\n",
       "      <td>MCC13</td>\n",
       "      <td>MCAS</td>\n",
       "      <td>MC-1010</td>\n",
       "      <td>Malme-3M</td>\n",
       "      <td>...</td>\n",
       "      <td>HCM-SANG-1325-C15</td>\n",
       "      <td>HCM-SANG-1331-C18</td>\n",
       "      <td>HCM-SANG-1337-C18</td>\n",
       "      <td>HCM-SANG-1326-C15</td>\n",
       "      <td>HCM-SANG-1300-C18</td>\n",
       "      <td>HCM-SANG-1322-C15</td>\n",
       "      <td>HCM-SANG-1332-C18</td>\n",
       "      <td>HCM-SANG-1095-C25</td>\n",
       "      <td>HCM-SANG-1336-C15</td>\n",
       "      <td>HCM-SANG-1308-C25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dataset_name</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Sanger &amp; Broad Cell Lines RNASeq</td>\n",
       "      <td>Sanger &amp; Broad Cell Lines RNASeq</td>\n",
       "      <td>Sanger &amp; Broad Cell Lines RNASeq</td>\n",
       "      <td>Sanger &amp; Broad Cell Lines RNASeq</td>\n",
       "      <td>Sanger &amp; Broad Cell Lines RNASeq</td>\n",
       "      <td>Sanger &amp; Broad Cell Lines RNASeq</td>\n",
       "      <td>Sanger &amp; Broad Cell Lines RNASeq</td>\n",
       "      <td>Sanger &amp; Broad Cell Lines RNASeq</td>\n",
       "      <td>Sanger &amp; Broad Cell Lines RNASeq</td>\n",
       "      <td>...</td>\n",
       "      <td>Sanger Organoid RNASeq</td>\n",
       "      <td>Sanger Organoid RNASeq</td>\n",
       "      <td>Sanger Organoid RNASeq</td>\n",
       "      <td>Sanger Organoid RNASeq</td>\n",
       "      <td>Sanger Organoid RNASeq</td>\n",
       "      <td>Sanger Organoid RNASeq</td>\n",
       "      <td>Sanger Organoid RNASeq</td>\n",
       "      <td>Sanger Organoid RNASeq</td>\n",
       "      <td>Sanger Organoid RNASeq</td>\n",
       "      <td>Sanger Organoid RNASeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>data_source</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Broad</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Broad</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Broad</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Broad</td>\n",
       "      <td>...</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Sanger</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <td>symbol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1432 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                1                                 2     \\\n",
       "0                                                        \n",
       "model_id         NaN                         SIDM00001   \n",
       "model_name       NaN                             MEC-1   \n",
       "dataset_name     NaN  Sanger & Broad Cell Lines RNASeq   \n",
       "data_source      NaN                             Broad   \n",
       "gene_id       symbol                               NaN   \n",
       "\n",
       "                                          3     \\\n",
       "0                                                \n",
       "model_id                             SIDM00002   \n",
       "model_name                             NBsusSR   \n",
       "dataset_name  Sanger & Broad Cell Lines RNASeq   \n",
       "data_source                             Sanger   \n",
       "gene_id                                    NaN   \n",
       "\n",
       "                                          4     \\\n",
       "0                                                \n",
       "model_id                             SIDM00003   \n",
       "model_name                                 M14   \n",
       "dataset_name  Sanger & Broad Cell Lines RNASeq   \n",
       "data_source                             Sanger   \n",
       "gene_id                                    NaN   \n",
       "\n",
       "                                          5     \\\n",
       "0                                                \n",
       "model_id                             SIDM00005   \n",
       "model_name                       MDA-MB-134-VI   \n",
       "dataset_name  Sanger & Broad Cell Lines RNASeq   \n",
       "data_source                              Broad   \n",
       "gene_id                                    NaN   \n",
       "\n",
       "                                          6     \\\n",
       "0                                                \n",
       "model_id                             SIDM00006   \n",
       "model_name                               MCC26   \n",
       "dataset_name  Sanger & Broad Cell Lines RNASeq   \n",
       "data_source                             Sanger   \n",
       "gene_id                                    NaN   \n",
       "\n",
       "                                          7     \\\n",
       "0                                                \n",
       "model_id                             SIDM00007   \n",
       "model_name                               MCC13   \n",
       "dataset_name  Sanger & Broad Cell Lines RNASeq   \n",
       "data_source                             Sanger   \n",
       "gene_id                                    NaN   \n",
       "\n",
       "                                          8     \\\n",
       "0                                                \n",
       "model_id                             SIDM00008   \n",
       "model_name                                MCAS   \n",
       "dataset_name  Sanger & Broad Cell Lines RNASeq   \n",
       "data_source                              Broad   \n",
       "gene_id                                    NaN   \n",
       "\n",
       "                                          9     \\\n",
       "0                                                \n",
       "model_id                             SIDM00009   \n",
       "model_name                             MC-1010   \n",
       "dataset_name  Sanger & Broad Cell Lines RNASeq   \n",
       "data_source                             Sanger   \n",
       "gene_id                                    NaN   \n",
       "\n",
       "                                          10    ...                    1423  \\\n",
       "0                                               ...                           \n",
       "model_id                             SIDM00011  ...               SIDM02071   \n",
       "model_name                            Malme-3M  ...       HCM-SANG-1325-C15   \n",
       "dataset_name  Sanger & Broad Cell Lines RNASeq  ...  Sanger Organoid RNASeq   \n",
       "data_source                              Broad  ...                  Sanger   \n",
       "gene_id                                    NaN  ...                     NaN   \n",
       "\n",
       "                                1424                    1425  \\\n",
       "0                                                              \n",
       "model_id                   SIDM02072               SIDM02073   \n",
       "model_name         HCM-SANG-1331-C18       HCM-SANG-1337-C18   \n",
       "dataset_name  Sanger Organoid RNASeq  Sanger Organoid RNASeq   \n",
       "data_source                   Sanger                  Sanger   \n",
       "gene_id                          NaN                     NaN   \n",
       "\n",
       "                                1426                    1427  \\\n",
       "0                                                              \n",
       "model_id                   SIDM02074               SIDM02075   \n",
       "model_name         HCM-SANG-1326-C15       HCM-SANG-1300-C18   \n",
       "dataset_name  Sanger Organoid RNASeq  Sanger Organoid RNASeq   \n",
       "data_source                   Sanger                  Sanger   \n",
       "gene_id                          NaN                     NaN   \n",
       "\n",
       "                                1428                    1429  \\\n",
       "0                                                              \n",
       "model_id                   SIDM02076               SIDM02077   \n",
       "model_name         HCM-SANG-1322-C15       HCM-SANG-1332-C18   \n",
       "dataset_name  Sanger Organoid RNASeq  Sanger Organoid RNASeq   \n",
       "data_source                   Sanger                  Sanger   \n",
       "gene_id                          NaN                     NaN   \n",
       "\n",
       "                                1430                    1431  \\\n",
       "0                                                              \n",
       "model_id                   SIDM02078               SIDM02079   \n",
       "model_name         HCM-SANG-1095-C25       HCM-SANG-1336-C15   \n",
       "dataset_name  Sanger Organoid RNASeq  Sanger Organoid RNASeq   \n",
       "data_source                   Sanger                  Sanger   \n",
       "gene_id                          NaN                     NaN   \n",
       "\n",
       "                                1432  \n",
       "0                                     \n",
       "model_id                   SIDM02080  \n",
       "model_name         HCM-SANG-1308-C25  \n",
       "dataset_name  Sanger Organoid RNASeq  \n",
       "data_source                   Sanger  \n",
       "gene_id                          NaN  \n",
       "\n",
       "[5 rows x 1432 columns]"
      ]
     },
     "execution_count": 385,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq_df = pd.read_csv(\"rnaseq_tpm_20220624.csv\",index_col=0, header = None,low_memory=False)\n",
    "rna_seq_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 386,
   "id": "599bf06c-7e33-4c62-9256-4bfa362a5afe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transformed_df_in_model_gene_value_format(rna_seq_df):\n",
    "    rna_seq_df_transposed = rna_seq_df.transpose().iloc[0:, :]\n",
    "    new_header = rna_seq_df_transposed.iloc[0, :] \n",
    "    rna_seq_df_transposed = rna_seq_df_transposed.iloc[1:, :] \n",
    "    rna_seq_df_transposed.columns = new_header\n",
    "    rna_seq_df_transposed = rna_seq_df_transposed.iloc[:, [0, *list(range(5, new_header.shape[0]))]]\n",
    "    rna_seq_df_transposed = rna_seq_df_transposed.rename(columns={rna_seq_df_transposed.columns[0]: \"SANGER_MODEL_ID\"})\n",
    "    rna_seq_df_transposed_stacked = rna_seq_df_transposed.set_index('SANGER_MODEL_ID').stack().reset_index(name='ExpressionValue')\n",
    "    rna_seq_df_transposed_stacked = rna_seq_df_transposed_stacked.rename(columns={1:'symbol'})\n",
    "    return rna_seq_df_transposed_stacked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 387,
   "id": "10fa0820-d74b-4b14-aab2-ffc5ff45cdff",
   "metadata": {},
   "outputs": [],
   "source": [
    "transformed_seq_data = get_transformed_df_in_model_gene_value_format(rna_seq_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 388,
   "id": "d4ae6e0c-4914-46ac-8471-b2b109326484",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4b779c16685142aa914ebc84d0e27985",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/295 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ABHD11: -0.2603464562771736\n",
      "ADGRF4: -0.21832599431678892\n",
      "ANKRD40: 0.18066823048933575\n",
      "AP1M2: -0.3342653365759664\n",
      "BTBD19: 0.23667903938357338\n",
      "CARD14: -0.23217649204441665\n",
      "CCDC6: -0.19417654811259208\n",
      "CCDC9: -0.19246329996662295\n",
      "CD9: -0.259732418680272\n",
      "CEP250: 0.19533947085984696\n",
      "CNPY2: 0.19032832830059648\n",
      "CXADR: -0.18112641682070177\n",
      "CYP4X1: -0.19684435815501675\n",
      "EFHD1: -0.25972140315958625\n",
      "EPS8L1: -0.320805654862061\n",
      "ERMP1: -0.2673739424835304\n",
      "FGFBP1: -0.2065707598624642\n",
      "GAS2L1: -0.18878950859159796\n",
      "HOMEZ: -0.25126823492188394\n",
      "IFRD1: 0.21580474043149722\n",
      "JAZF1: 0.1982106191228681\n",
      "KLK8: -0.24739992578825482\n",
      "KREMEN2: -0.2356507342622787\n",
      "LYPD3: -0.31326846430309047\n",
      "MAP2K1: 0.1866270511929771\n",
      "MFSD3: -0.23222267670592697\n",
      "MIR6499: -0.2222250494434576\n",
      "MSX2: -0.23343550175413866\n",
      "MYH14: -0.35718004187357133\n",
      "OPLAH: -0.255798525508033\n",
      "PCBP2: 0.18281860322806642\n",
      "PELO: 0.21314562197564788\n",
      "PERM1: -0.1913978133108724\n",
      "PKM: 0.2350738006538517\n",
      "PPFIA3: -0.29509336755783133\n",
      "PROM2: -0.33581733645263817\n",
      "PRRG2: -0.36931885395982983\n",
      "RAPGEF4: 0.20215218952108185\n",
      "RTN4RL1: -0.32259758106885805\n",
      "SLC44A2: -0.28221835125277667\n",
      "SLC44A4: -0.1872192807194368\n",
      "SPDEF: -0.3214320343956796\n",
      "TGFB3: -0.19904367208122814\n",
      "TRPM4: -0.5016377783623077\n",
      "TRPV6: -0.24484813661938962\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from scipy import stats\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "def select_genes_by_corr(corr_threshold, sparsity_threshold = 100):\n",
    "    selected_features = []\n",
    "    train_indices, test_indices = train_test_split(torch.arange(target_labels.shape[0]), random_state = 42, test_size=0.25)\n",
    "    train_merged_drug_model_df = merged_drug_model_df.iloc[train_indices.numpy(), :]\n",
    "    train_model_gene_merged_df = get_gene_and_model_merged_df(gene_idx_to_gene_symbol_df, train_merged_drug_model_df, transformed_seq_data)\n",
    "\n",
    "    for id, group in tqdm(train_model_gene_merged_df.groupby(\"symbol\")):\n",
    "        if group.shape[0] < sparsity_threshold:\n",
    "            continue\n",
    "        res = stats.pearsonr(group[\"ExpressionValue\"].astype(np.float32), group[\"LN_IC50\"])\n",
    "\n",
    "        corr_factor = res.statistic\n",
    "        if np.isnan(res.statistic) or np.abs(res.statistic) < corr_threshold: continue\n",
    "        print(f\"{id}: \" + f\"{corr_factor}\")\n",
    "        selected_features.append(id)        \n",
    "    return selected_features\n",
    "    \n",
    "features = select_genes_by_corr(0.18, 500) #0.001, 500"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 389,
   "id": "54c4de95-b206-4d7e-9f96-32ba7ae42279",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 389,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def intersection(lst1, lst2):\n",
    "    lst3 = [value for value in lst1 if value in lst2]\n",
    "    return lst3\n",
    "    \n",
    "len(intersection(features, perfect_gene_features)) / min(len(features), len(perfect_gene_features))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 390,
   "id": "1c1c5c82-b20e-4991-a77d-4392413e4d31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "45"
      ]
     },
     "execution_count": 390,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(features)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f69bfb0e-ba6b-4960-acec-18a3686c49cf",
   "metadata": {},
   "source": [
    "### Get edge index (model-[Expression]->gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 391,
   "id": "8fe472ab-2599-4bfb-bec4-420ca6b12d3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(210925, 2)"
      ]
     },
     "execution_count": 391,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_gene_merged_exprssion_df = get_gene_and_model_merged_df(gene_idx_to_gene_symbol_df, merged_drug_model_df, transformed_seq_data)\n",
    "gene_expression_model_edge_index = model_gene_merged_exprssion_df.loc[:,[\"model_idx\", \"human_genome_index\"]].values.astype(np.int64)\n",
    "gene_expression_model_edge_index.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72d4715a-4074-4592-964b-586826ed06db",
   "metadata": {},
   "source": [
    "### Get edge attributes (model-[Expression]->gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 392,
   "id": "533b582f-9dab-482c-bd98-13636e064ec2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(210925, 1)"
      ]
     },
     "execution_count": 392,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import torch\n",
    "device = torch.device(\"cuda:1\")\n",
    "gene_expression_torch = torch.from_numpy(model_gene_merged_exprssion_df.loc[:, [\"ExpressionValue\"]].astype(float).values).to(device)\n",
    "gene_expression_model_edge_attributes_torch = (gene_expression_torch - gene_expression_torch.mean()) / gene_expression_torch.std()\n",
    "gene_expression_model_edge_attributes = gene_expression_model_edge_attributes_torch.cpu().numpy()\n",
    "gene_expression_model_edge_attributes.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d85d6b48-498b-46ee-a25d-b8c2fab1e35d",
   "metadata": {},
   "source": [
    "## Protein expression data\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 393,
   "id": "95caf931-63fa-4c84-97ff-a067fa50526d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>model_id</th>\n",
       "      <th>model_name</th>\n",
       "      <th>protein_intensity</th>\n",
       "      <th>zscore</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A075B6K4</td>\n",
       "      <td>SIDM00483</td>\n",
       "      <td>SK-GT-4</td>\n",
       "      <td>-0.577590</td>\n",
       "      <td>-1.284310</td>\n",
       "      <td>IGLV3-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A075B6K4</td>\n",
       "      <td>SIDM00689</td>\n",
       "      <td>JM1</td>\n",
       "      <td>3.408550</td>\n",
       "      <td>0.129406</td>\n",
       "      <td>IGLV3-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A075B6K4</td>\n",
       "      <td>SIDM01259</td>\n",
       "      <td>GR-ST</td>\n",
       "      <td>6.300060</td>\n",
       "      <td>1.154900</td>\n",
       "      <td>IGLV3-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A075B6N1</td>\n",
       "      <td>SIDM00846</td>\n",
       "      <td>HeLa</td>\n",
       "      <td>0.127721</td>\n",
       "      <td>-1.034210</td>\n",
       "      <td>TRBV19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A075B6N1</td>\n",
       "      <td>SIDM00958</td>\n",
       "      <td>CML-T1</td>\n",
       "      <td>4.283890</td>\n",
       "      <td>1.352450</td>\n",
       "      <td>TRBV19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4919677</th>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>SIDM01248</td>\n",
       "      <td>CGTH-W-1</td>\n",
       "      <td>3.812190</td>\n",
       "      <td>0.723363</td>\n",
       "      <td>SEC23IP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4919678</th>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>SIDM01251</td>\n",
       "      <td>H9</td>\n",
       "      <td>3.854180</td>\n",
       "      <td>0.787636</td>\n",
       "      <td>SEC23IP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4919679</th>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>SIDM01259</td>\n",
       "      <td>GR-ST</td>\n",
       "      <td>3.802910</td>\n",
       "      <td>0.709159</td>\n",
       "      <td>SEC23IP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4919680</th>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>SIDM01261</td>\n",
       "      <td>YMB-1-E</td>\n",
       "      <td>4.540440</td>\n",
       "      <td>1.838070</td>\n",
       "      <td>SEC23IP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4919681</th>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>SIDM01265</td>\n",
       "      <td>MM1S</td>\n",
       "      <td>3.873330</td>\n",
       "      <td>0.816948</td>\n",
       "      <td>SEC23IP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4919682 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         uniprot_id   model_id model_name  protein_intensity    zscore  \\\n",
       "0        A0A075B6K4  SIDM00483    SK-GT-4          -0.577590 -1.284310   \n",
       "1        A0A075B6K4  SIDM00689        JM1           3.408550  0.129406   \n",
       "2        A0A075B6K4  SIDM01259      GR-ST           6.300060  1.154900   \n",
       "3        A0A075B6N1  SIDM00846       HeLa           0.127721 -1.034210   \n",
       "4        A0A075B6N1  SIDM00958     CML-T1           4.283890  1.352450   \n",
       "...             ...        ...        ...                ...       ...   \n",
       "4919677      Q9Y6Y8  SIDM01248   CGTH-W-1           3.812190  0.723363   \n",
       "4919678      Q9Y6Y8  SIDM01251         H9           3.854180  0.787636   \n",
       "4919679      Q9Y6Y8  SIDM01259      GR-ST           3.802910  0.709159   \n",
       "4919680      Q9Y6Y8  SIDM01261    YMB-1-E           4.540440  1.838070   \n",
       "4919681      Q9Y6Y8  SIDM01265       MM1S           3.873330  0.816948   \n",
       "\n",
       "           symbol  \n",
       "0        IGLV3-10  \n",
       "1        IGLV3-10  \n",
       "2        IGLV3-10  \n",
       "3          TRBV19  \n",
       "4          TRBV19  \n",
       "...           ...  \n",
       "4919677   SEC23IP  \n",
       "4919678   SEC23IP  \n",
       "4919679   SEC23IP  \n",
       "4919680   SEC23IP  \n",
       "4919681   SEC23IP  \n",
       "\n",
       "[4919682 rows x 6 columns]"
      ]
     },
     "execution_count": 393,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteomic_to_model_df = pd.read_csv(\"proteomics_all_20221214.csv\")\n",
    "proteomic_to_model_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 394,
   "id": "bc55103d-5d6a-4b90-88a2-376688953e28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_idx</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A0A075B6K4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A0A075B6N1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>A0A075B7B8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>A0A075B7D8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>A0A087X0M5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8452</th>\n",
       "      <td>8452</td>\n",
       "      <td>Q9Y6X5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8453</th>\n",
       "      <td>8453</td>\n",
       "      <td>Q9Y6X8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8454</th>\n",
       "      <td>8454</td>\n",
       "      <td>Q9Y6X9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8455</th>\n",
       "      <td>8455</td>\n",
       "      <td>Q9Y6Y0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8456</th>\n",
       "      <td>8456</td>\n",
       "      <td>Q9Y6Y8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8457 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      protein_idx  uniprot_id\n",
       "0               0  A0A075B6K4\n",
       "1               1  A0A075B6N1\n",
       "2               2  A0A075B7B8\n",
       "3               3  A0A075B7D8\n",
       "4               4  A0A087X0M5\n",
       "...           ...         ...\n",
       "8452         8452      Q9Y6X5\n",
       "8453         8453      Q9Y6X8\n",
       "8454         8454      Q9Y6X9\n",
       "8455         8455      Q9Y6Y0\n",
       "8456         8456      Q9Y6Y8\n",
       "\n",
       "[8457 rows x 2 columns]"
      ]
     },
     "execution_count": 394,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_data = pd.DataFrame(columns=[\"protein_idx\", \"uniprot_id\"])\n",
    "protein_data[\"uniprot_id\"] = proteomic_to_model_df[\"uniprot_id\"].unique()\n",
    "protein_data[\"protein_idx\"] = np.arange(proteomic_to_model_df[\"uniprot_id\"].unique().shape[0])\n",
    "protein_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "id": "c0f46e8e-e690-4d23-bb3e-ff3434516612",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_proteomic_and_model_merged_df(protein_data, merged_drug_model_df, proteomic_to_model_df):\n",
    "    protein_edge_merged_df = pd.merge(protein_data, proteomic_to_model_df, on=\"uniprot_id\", how = \"inner\")\n",
    "    protein_edge_merged_df = protein_edge_merged_df.rename(columns={\"model_id\": \"SANGER_MODEL_ID\"}) ##rename for join\n",
    "    model_protein_merged_df = pd.merge(protein_edge_merged_df, merged_drug_model_df, on=\"SANGER_MODEL_ID\", how=\"inner\")\n",
    "    return model_protein_merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 396,
   "id": "50ad33f6-fd21-47f4-a66d-1b602f0a9e2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from scipy import stats\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "def select_by_corr(corr_threshold, sparsity_threshold = 100):\n",
    "    selected_features = []\n",
    "    train_indices, test_indices = train_test_split(torch.arange(target_labels.shape[0]), random_state = 42, test_size=0.25)\n",
    "    train_merged_drug_model_df = merged_drug_model_df.iloc[train_indices.numpy(), :]\n",
    "    train_model_protein_merged_df = get_proteomic_and_model_merged_df(protein_data, train_merged_drug_model_df, proteomic_to_model_df)\n",
    "\n",
    "    for id, group in tqdm(train_model_protein_merged_df.groupby(\"uniprot_id\")):\n",
    "        if group.shape[0] < sparsity_threshold:\n",
    "            continue\n",
    "        res = stats.pearsonr(group[\"protein_intensity\"], group[\"LN_IC50\"])\n",
    "        \n",
    "        if np.abs(res.statistic) < corr_threshold: continue\n",
    "        print(res.statistic)\n",
    "        selected_features.append(id)        \n",
    "    return selected_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "id": "0753e859-5cea-48ff-944f-c1931449b236",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ca8b9f4c7005421db1a273541c9680b0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/8450 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.1425689593086887\n",
      "-0.15264111177302064\n",
      "-0.17599966045011647\n",
      "0.14822871481724\n",
      "0.14209093172740245\n",
      "0.12653596905479905\n",
      "-0.16179686046070194\n",
      "0.20960816463081886\n",
      "0.1566801021093475\n",
      "-0.10320486408322492\n",
      "0.14459183460946248\n",
      "0.11791617285505497\n",
      "0.19767652439474293\n",
      "0.20328501650050643\n",
      "0.1041055743324344\n",
      "-0.15358420421869592\n",
      "0.10306418014030153\n",
      "0.1454237667339961\n",
      "0.18666177054812436\n",
      "-0.1606449956290917\n",
      "0.14365497290441814\n",
      "-0.1024706508172239\n",
      "0.157152174157493\n",
      "0.15102596162169923\n",
      "0.213731976373527\n",
      "-0.14155139506210337\n",
      "-0.12466506935804011\n",
      "0.15844336970658066\n",
      "0.13265349420826258\n",
      "-0.11683355895243504\n",
      "0.14649882915071336\n",
      "0.10901336111259853\n",
      "0.11583675328935013\n",
      "0.11781810858008603\n",
      "0.15973447061117757\n",
      "-0.1523446904139372\n",
      "0.18666539172424237\n",
      "0.12781651801796365\n",
      "0.1639131647308169\n",
      "-0.1978770609761334\n",
      "-0.1071164055737107\n",
      "0.1069696430960415\n",
      "0.13903122378912647\n",
      "-0.1330090120322887\n",
      "0.1559744114148788\n",
      "0.21618592578846318\n",
      "0.15538783192603367\n",
      "0.18365050090403612\n",
      "-0.11056015631230968\n",
      "-0.1376102299949433\n",
      "0.13859999359528305\n",
      "0.1656229118735303\n",
      "-0.13382638486262274\n",
      "0.1870779827671789\n",
      "0.1430424995205918\n",
      "-0.10157087027056493\n",
      "0.1178004068946029\n",
      "0.1249291538836457\n",
      "0.16559066418210247\n",
      "-0.17724203320620333\n",
      "0.14386695046976672\n",
      "0.19783524495176763\n",
      "0.1600713158779757\n",
      "0.14852377265917413\n",
      "-0.15137860086941915\n",
      "0.18045162624875544\n",
      "0.16556439734314934\n",
      "0.1312049855804672\n",
      "0.12792638521358993\n",
      "0.10515129323872668\n",
      "0.13579393622553965\n",
      "0.14397198727751923\n",
      "0.19419318614975425\n",
      "0.2277258414950262\n",
      "0.23868395245444973\n",
      "-0.14900555799822143\n",
      "0.11434502806319871\n",
      "-0.12759677383273155\n",
      "0.18476946683386494\n",
      "0.1509676437721003\n",
      "-0.1644920255485154\n",
      "-0.11239052550007632\n",
      "0.13026611627418574\n",
      "-0.10480530185023898\n",
      "0.12248663041769806\n",
      "0.19298015762737192\n",
      "0.16164262868877138\n",
      "-0.13881211463638812\n",
      "0.1294889533364562\n",
      "0.13370941007908163\n",
      "-0.18766736950574356\n",
      "0.17477605530291315\n",
      "0.1344882129930057\n",
      "-0.14405971771523948\n",
      "0.10895486596440036\n",
      "-0.1503122106347385\n",
      "0.10645385919820988\n",
      "0.12765069757294334\n",
      "0.12748520565568475\n",
      "-0.12663618424954123\n",
      "-0.12593914039228182\n",
      "-0.12692033807758202\n",
      "0.2194200812499114\n",
      "0.17744968992529755\n",
      "-0.1496619054345539\n",
      "-0.11506603655964653\n",
      "0.1442468600822876\n",
      "0.16187055098919417\n",
      "-0.18196311840459675\n",
      "0.1446213580841239\n",
      "0.11066510531120159\n",
      "0.14497829722283567\n",
      "0.18104352931426357\n",
      "-0.22276538963066833\n",
      "0.16098136999467477\n",
      "-0.17655999968331082\n",
      "-0.1272037385938904\n",
      "-0.1696275155027175\n",
      "-0.14646497280764806\n",
      "-0.10198778772570292\n",
      "-0.16975099968730284\n",
      "-0.14550216745175304\n",
      "0.2483755600045654\n",
      "0.14176873566585502\n",
      "0.20258773912764244\n",
      "-0.1080783601839201\n",
      "-0.12280546473811277\n",
      "0.1956856728768102\n",
      "-0.148287979690906\n",
      "0.2010672497675129\n",
      "0.11897126884014252\n",
      "-0.17701970211492524\n",
      "-0.1118715469074039\n",
      "0.19036045001782634\n",
      "-0.192304130070684\n",
      "-0.15792635459185816\n",
      "-0.1677172330261535\n",
      "0.13254888164019693\n",
      "-0.17292445731850214\n",
      "0.12601099395783386\n",
      "0.1327945245618277\n",
      "-0.11362993914875412\n",
      "0.1231423554150743\n",
      "0.20790977843040526\n",
      "0.12754805496642285\n",
      "0.18196523528348502\n",
      "-0.14257646258926693\n",
      "0.2357528410515462\n",
      "0.1841229600472745\n",
      "0.12940572029971964\n",
      "-0.17421898534448427\n",
      "-0.11938095141085342\n",
      "-0.19409045601670963\n",
      "0.12356026917586813\n",
      "-0.20870873102141094\n",
      "0.24067694599419256\n",
      "0.13273233243880478\n",
      "0.21779115747135852\n",
      "0.16113983340166987\n",
      "-0.21889262243071692\n",
      "0.1316040150509347\n",
      "0.17241372888695594\n",
      "-0.12862632740545954\n",
      "0.1445348537365224\n",
      "-0.12077359579792517\n",
      "0.12219873873609036\n",
      "0.15393210258080578\n",
      "-0.15146105813057337\n",
      "0.1905438892887707\n",
      "-0.10371697891109323\n",
      "0.24394407806581467\n",
      "0.19810440036204774\n",
      "-0.14435914708738407\n",
      "0.1447842429484998\n",
      "-0.17343333845838918\n",
      "0.21802033406059435\n",
      "-0.10808684004214479\n",
      "-0.17806429857336464\n",
      "0.13613346777174803\n",
      "0.15304890260213544\n",
      "0.14232521334340018\n",
      "0.10325763464100861\n",
      "0.1011374683403123\n",
      "0.18320719785830852\n",
      "0.11377653678768537\n",
      "-0.12906612161889067\n",
      "0.13408321645186427\n",
      "0.1423489170294335\n",
      "0.1626291342722025\n",
      "-0.14803916640604192\n",
      "0.13310259599403607\n",
      "0.1576688823658207\n",
      "-0.13291073300465361\n",
      "-0.10613247630326106\n",
      "-0.20861121105133018\n",
      "0.1260777037423087\n",
      "-0.10964515009356467\n",
      "0.11675437618504268\n",
      "-0.10712923369423308\n",
      "0.10381364072379477\n",
      "0.2929309263666905\n",
      "-0.11747867501252314\n",
      "0.21187260755818027\n",
      "-0.21830960018592468\n",
      "0.23034665407978278\n",
      "0.15428091201607347\n",
      "0.17808796965163973\n",
      "0.10841737731223935\n",
      "0.13973524414015623\n",
      "-0.22841795857702757\n",
      "0.15191326245332953\n",
      "-0.19036602977840864\n",
      "0.18025540526616055\n",
      "-0.1409970474927011\n",
      "0.23440020822264634\n",
      "0.12036206965484386\n",
      "0.1758555106720186\n",
      "0.1691083435002196\n",
      "-0.1883372748884129\n",
      "0.10937538048029923\n",
      "0.2283703311906761\n",
      "-0.19460706362666066\n",
      "-0.1524556711938191\n",
      "0.14570590337344674\n",
      "-0.10031291654009697\n",
      "0.17367813588849476\n",
      "0.10112611436529974\n",
      "0.13302083830816527\n",
      "-0.11232454567295093\n",
      "0.15912576210161702\n",
      "-0.11312561738937568\n",
      "-0.1385760473684256\n",
      "0.27765777153479015\n",
      "0.1496741004224382\n",
      "0.10679378264736454\n",
      "-0.14416382608625014\n",
      "-0.21595643839517345\n",
      "-0.14152838825736516\n",
      "0.13080194461464423\n",
      "0.10089921258184542\n",
      "0.17726769228806058\n",
      "-0.1620483843557667\n",
      "0.1743480362502041\n",
      "0.12875208732565094\n",
      "-0.16917427929037396\n",
      "-0.1759008515408161\n",
      "0.1427354917101963\n",
      "0.13874250097739255\n",
      "0.14859532804920436\n",
      "-0.1520060588486594\n",
      "0.16545600695200668\n",
      "0.16481988675636564\n",
      "-0.15124073718929534\n",
      "0.18751288948319567\n",
      "0.1567928096592334\n",
      "0.1287033800087121\n",
      "0.13547616476212884\n",
      "0.1025587055476778\n",
      "-0.10030417925649764\n",
      "0.18652020910104056\n",
      "0.17509711129766214\n",
      "0.15132304151456707\n",
      "0.1881705331698359\n",
      "0.12998592509521162\n",
      "0.10529094094656874\n",
      "0.16655267305773502\n",
      "0.10405510891919315\n",
      "0.12235478382402096\n",
      "0.14496046258673592\n",
      "-0.11045965528064075\n",
      "0.177375446006578\n",
      "0.139202405922004\n",
      "0.12770329546288617\n",
      "0.1703635815435896\n",
      "0.12113398912133506\n",
      "-0.10244887987293379\n",
      "0.14253710138011963\n",
      "0.14039520264688657\n",
      "0.20180553541113286\n",
      "0.11560121855476492\n",
      "0.10503419492106968\n",
      "-0.11630866700112327\n",
      "0.1090289319079118\n",
      "0.11698773322585322\n",
      "0.12788041404550352\n",
      "0.1419197182448222\n",
      "0.162786620733754\n",
      "-0.14355003967700322\n",
      "-0.1782634153251065\n",
      "0.13505253030760794\n",
      "0.14136323520907057\n",
      "-0.1903798030747523\n",
      "0.14404383235561774\n",
      "0.19872824337341904\n",
      "-0.1267067635043614\n",
      "0.15327060569568102\n",
      "-0.11831851876566535\n",
      "0.15386752508090878\n",
      "0.13079059746940133\n",
      "-0.19123522478704488\n",
      "0.1520973339612338\n",
      "0.1732766671368323\n",
      "0.1358115765554906\n",
      "0.12233874892415728\n",
      "0.1301567307886505\n",
      "0.10297692861640985\n",
      "0.1560709561821266\n",
      "0.13123201157134975\n",
      "0.1380348053569095\n",
      "0.14708794740703973\n",
      "0.13641378718802522\n",
      "0.13021966951775432\n",
      "0.1402000420088016\n",
      "0.12441170159538469\n",
      "0.11150685616581005\n",
      "0.14291749784676958\n",
      "0.19796695088994182\n",
      "0.18084732482349716\n",
      "0.14050257813463998\n",
      "-0.10131506722700691\n",
      "0.21759340701295252\n",
      "0.12314712074854067\n",
      "-0.13462000948809927\n",
      "0.13639030784829478\n",
      "0.15534735758492746\n",
      "-0.2054403815254372\n",
      "0.20573998601016136\n",
      "-0.10012479190966869\n",
      "0.10358976758531178\n",
      "0.11577616518888559\n",
      "-0.1322741682793978\n",
      "0.157317576470007\n",
      "0.24276548166667988\n",
      "0.14998277826590306\n",
      "0.1589898357771599\n",
      "-0.14947560170613364\n",
      "0.14244696030292336\n",
      "0.1002320555447144\n",
      "-0.1317646545112574\n",
      "-0.10295923936596042\n",
      "-0.16861384409549057\n",
      "0.1385584782164618\n",
      "-0.12871147634980615\n",
      "0.13403570020623212\n",
      "0.19810603855701242\n",
      "0.17371105187489894\n",
      "0.13519523246325374\n",
      "-0.13135415797784897\n",
      "0.11669642251326923\n",
      "0.1273998883652661\n",
      "0.11537223807090724\n",
      "0.1170587157946701\n",
      "0.10072774492437053\n",
      "-0.11527778973781756\n",
      "-0.10972951296795336\n",
      "0.16878085412042448\n",
      "0.13013913009444444\n",
      "-0.12624103247045318\n",
      "0.13863737382600416\n",
      "0.19665936410290097\n",
      "0.15853559998687672\n",
      "-0.25653862326867244\n",
      "0.15111537402011765\n",
      "0.15579425853572104\n",
      "-0.10596717608702479\n",
      "-0.1014015694338444\n",
      "0.1145972990314294\n",
      "0.13090179558022116\n",
      "0.18725929416464657\n",
      "0.1849380797028171\n",
      "0.14558377101659453\n",
      "-0.14417721894893126\n",
      "-0.17715135603084972\n",
      "-0.10082532922221217\n",
      "-0.12453082288443318\n",
      "0.23173301304052404\n",
      "-0.10543310566585205\n",
      "0.1367038031097055\n",
      "0.11784868657910087\n",
      "-0.12990539670671927\n",
      "-0.10828828414338472\n",
      "0.14699028198614278\n",
      "-0.13989717761209713\n",
      "0.12616572026837167\n",
      "0.14761041057957028\n",
      "0.16310022701219454\n",
      "0.13114357112307817\n",
      "0.10826828883821281\n",
      "-0.1194578116056136\n",
      "-0.14105762866896418\n",
      "0.10110285697460304\n",
      "-0.16647178960019895\n",
      "0.19400992552254837\n",
      "0.1056807185949163\n",
      "0.15352073235721028\n",
      "0.2559521060913605\n",
      "0.2560889881316188\n",
      "-0.10153969595696748\n",
      "-0.1403720113568013\n",
      "0.21138660528043154\n",
      "-0.12062771998168804\n",
      "0.11758034016898726\n",
      "0.1331478418637269\n",
      "0.12039864901188913\n",
      "0.15082505160846604\n",
      "0.10809221687098289\n",
      "0.11079376975524549\n",
      "-0.17597501533406468\n",
      "0.14472276765294018\n",
      "-0.23391420263167143\n",
      "0.12180808120890935\n",
      "0.11181097280630231\n",
      "-0.10553223380956024\n",
      "0.20198508987309569\n",
      "-0.12765871015222646\n",
      "0.21845776291763105\n",
      "0.16020054686152224\n",
      "0.1170020808928149\n",
      "-0.10842395646979294\n",
      "0.11307395944752371\n",
      "-0.13555752218199624\n",
      "-0.11389523779449325\n",
      "0.15964302745103148\n",
      "0.10768884949208797\n",
      "0.18230293512970422\n",
      "0.11404913294089189\n",
      "0.17985652207440717\n",
      "0.12080991661435198\n",
      "0.17584837859669974\n",
      "0.14310795008284244\n",
      "-0.12737422154915284\n",
      "0.16283617428686017\n",
      "0.163353165408855\n",
      "0.1652493659969007\n",
      "0.10932845618289849\n",
      "0.10828336138386171\n",
      "0.18842636160888132\n",
      "0.11026382305210121\n",
      "0.13864316355398912\n",
      "0.13007323420784603\n",
      "0.17315881311307113\n",
      "0.15187778029875174\n",
      "0.16970444051143038\n",
      "-0.11178283515272995\n",
      "0.13530977477381834\n",
      "-0.11046290522817916\n",
      "0.1513333799066801\n",
      "0.126198940906634\n",
      "0.19065872114932295\n",
      "-0.10138797593905534\n",
      "0.11848325990069822\n",
      "0.15305208992627442\n",
      "0.15090144664292687\n",
      "0.17388848815443306\n",
      "0.12580283569865117\n",
      "0.11272956536877479\n",
      "0.13775521066766783\n",
      "0.10583615628550971\n",
      "0.14205267216680142\n",
      "0.13306631277647787\n",
      "0.12710668549053186\n",
      "0.10582912993208192\n",
      "0.1419860096064031\n",
      "0.13189353040940616\n",
      "0.14168263161453218\n",
      "0.18094394219280857\n",
      "0.11708758759076883\n",
      "0.10954097471627995\n",
      "0.19712621414814321\n",
      "0.1297466843005734\n",
      "0.11450176641162506\n",
      "0.14472596803267204\n",
      "0.12674793112401836\n",
      "0.12148966787291182\n",
      "0.10100394256498042\n",
      "0.11520875621231332\n",
      "0.16001735205493897\n",
      "0.18120272281476255\n",
      "0.12338074625754226\n",
      "0.1997366952293994\n",
      "0.13419364404275108\n",
      "0.19323486611681912\n",
      "0.1271227298715914\n",
      "0.20804771963497795\n",
      "0.17891034559544727\n",
      "0.12771023561396486\n",
      "0.16386301018977706\n",
      "0.1575837713541832\n",
      "0.16904126643423228\n",
      "0.2296595571472594\n",
      "-0.1484059365375902\n",
      "-0.10098366221996295\n",
      "-0.14357719208584765\n",
      "-0.10960447843514781\n",
      "-0.13998554787827244\n",
      "-0.14223296948853886\n",
      "-0.10008320303596174\n",
      "-0.13149281941831642\n",
      "0.12040632003017263\n",
      "-0.12762272915223902\n",
      "0.10329253375164492\n",
      "0.10537505826646722\n",
      "0.1371372647330809\n",
      "0.11577604072653043\n",
      "0.11964397166485846\n",
      "-0.15618490219378559\n",
      "0.118417856514428\n",
      "0.18560464103273677\n",
      "-0.18534038539635758\n",
      "0.16925006438665385\n",
      "0.24017351689261318\n",
      "-0.11480266884939748\n",
      "-0.15710918953185282\n",
      "-0.1244335079455897\n",
      "-0.14238500845727414\n",
      "0.1762690440784431\n",
      "0.12288878754058385\n",
      "0.1353287318999184\n",
      "0.19194366523102122\n",
      "-0.16060749199124819\n",
      "0.13770565235498922\n",
      "-0.26161126667052054\n",
      "0.10859604869661568\n",
      "-0.1550319105779484\n",
      "-0.10163412778970989\n",
      "-0.14901674992648384\n",
      "-0.12035706868117681\n",
      "0.10062592805803826\n",
      "-0.14674980871052748\n",
      "-0.11753852023629599\n",
      "0.11901054262920809\n",
      "0.20901384294691566\n",
      "-0.10320013159677376\n",
      "0.12323888353893683\n",
      "0.11148798611860278\n",
      "-0.11076618163994015\n",
      "0.1678865629647456\n",
      "-0.1696494491150864\n",
      "0.15208458320271842\n",
      "0.18268247510810628\n",
      "0.17039148077302865\n",
      "0.2385629755303511\n",
      "0.13269768426723513\n",
      "-0.155620104851054\n",
      "0.12710389193005078\n",
      "0.15320712335333278\n",
      "-0.11106509976257647\n",
      "0.1461346540314881\n",
      "-0.1406957356780688\n",
      "0.10379747030056265\n",
      "0.19976032861725582\n",
      "0.16309219381637838\n",
      "0.18407362525465198\n",
      "0.12478381336373728\n",
      "-0.10410552287201637\n",
      "0.17542532159035384\n",
      "0.18351730361188665\n",
      "0.1583543343833986\n",
      "0.15771727917686662\n",
      "-0.2193227210685666\n",
      "-0.11675341646784382\n",
      "0.22552525862229159\n",
      "0.17812754153010943\n",
      "0.15505040001965734\n",
      "0.2011414776927479\n",
      "0.11032464069671438\n",
      "0.14100736573790945\n",
      "0.10334078339616519\n",
      "-0.17876339927189033\n",
      "0.1494455110016122\n",
      "0.23309897446854236\n",
      "0.26319708663622154\n",
      "-0.1338646723604144\n",
      "-0.12611939054939403\n",
      "-0.13458912208055057\n",
      "-0.12163136868255894\n",
      "0.15288021275872407\n",
      "0.21901492886429078\n",
      "0.1560812351584878\n",
      "0.1338219989328482\n",
      "-0.21598739418067314\n",
      "-0.11484992396823193\n",
      "0.18142603681245756\n",
      "-0.12010909088396109\n",
      "0.15662616086982456\n",
      "-0.10771641189980838\n",
      "0.19389171541487535\n",
      "0.17402599590206275\n",
      "0.15838762834478545\n",
      "-0.1758921645470936\n",
      "0.11124926444511157\n",
      "0.2024398732039833\n",
      "0.12033655348772299\n",
      "0.2178636145083483\n",
      "-0.1485382525828815\n",
      "-0.26882078006917864\n",
      "-0.19153262721915984\n",
      "0.1509609502433004\n",
      "0.23091027580061252\n",
      "0.18588051269827535\n",
      "0.13906983569975337\n",
      "0.2722231290901185\n",
      "0.11979230672765645\n",
      "0.14380137036401075\n",
      "0.13140909221103442\n",
      "-0.13659557674761666\n",
      "0.14446685510461696\n",
      "-0.2542733150236686\n",
      "-0.15428913106147574\n",
      "0.1363101506091186\n",
      "0.2367981628871869\n",
      "-0.11369406637087681\n",
      "0.1308512079902569\n",
      "0.15985407322910766\n",
      "0.10661717086377052\n",
      "-0.1497833457593665\n",
      "0.1453600718549334\n",
      "-0.24716399014846904\n",
      "-0.11202914160974907\n",
      "-0.10609820873163706\n",
      "0.1755013311721219\n",
      "-0.14018536810825313\n",
      "0.20888554667265177\n",
      "0.17675696641720362\n",
      "-0.10945864230024166\n",
      "-0.17502858228697565\n",
      "-0.13885440568863155\n",
      "-0.14972175021064457\n",
      "0.19844690635560627\n",
      "0.1406203397056076\n",
      "-0.1667313462559748\n",
      "0.12190336769950601\n",
      "-0.11888461616976179\n",
      "0.16064920044838643\n",
      "0.15591469219251733\n",
      "0.12664148380000434\n",
      "-0.10695138370436663\n",
      "0.1567104725560985\n",
      "0.1560586468440709\n",
      "0.17702840909307038\n",
      "-0.10559043611856087\n",
      "-0.12316647889146849\n",
      "-0.18825817883525606\n",
      "0.12443800633356311\n",
      "-0.16891527820411237\n",
      "0.19408040164349596\n",
      "-0.1496807497241787\n",
      "-0.12326757473264773\n",
      "0.1632114073980463\n",
      "0.14592495305533243\n",
      "0.16897571235223244\n",
      "0.10536884564509146\n",
      "-0.12363334019553315\n",
      "0.1842487208943692\n",
      "-0.1174798780566449\n",
      "-0.16648185915131408\n",
      "-0.12114834554700767\n",
      "-0.1366191422375547\n",
      "0.10464338491288042\n",
      "0.10742635480742407\n",
      "-0.12023051455864049\n",
      "-0.10244653823837448\n",
      "-0.13985182105967386\n",
      "0.11486557289513641\n",
      "-0.1018536906247031\n",
      "-0.1278339923745487\n",
      "-0.2636328783925219\n",
      "-0.12316501690060258\n",
      "-0.1350481250350763\n",
      "0.10168561600066256\n",
      "0.20187388189360256\n",
      "-0.21564968666576556\n",
      "-0.10250377877425695\n",
      "-0.1280950535745346\n",
      "-0.13431820650153425\n",
      "0.21459359895550123\n",
      "-0.13633449893564806\n",
      "-0.1414379670599369\n",
      "0.17619678687775317\n",
      "0.11409883434646834\n",
      "-0.15116179544010316\n",
      "0.15604402614517932\n",
      "-0.16108418887761297\n",
      "0.14268909318742812\n",
      "0.15600765600125466\n",
      "-0.12604176319469804\n",
      "-0.15441588114144447\n",
      "0.13371743901981145\n",
      "0.11384058407897865\n",
      "-0.21301445409570627\n",
      "-0.15703538189027405\n",
      "0.1462457775889621\n",
      "0.14459817917460685\n",
      "0.10718979387647956\n",
      "0.17077318983506432\n",
      "-0.10128147531298301\n",
      "0.13150996776457413\n",
      "0.13862897605510094\n",
      "0.14046072062020154\n",
      "-0.11593041672280816\n",
      "0.2570553375550447\n",
      "0.24182588245104672\n",
      "0.10864772840089611\n",
      "-0.15311618767202714\n",
      "-0.15484532006925\n",
      "-0.13875512910779234\n",
      "-0.13485654980076775\n",
      "0.10868450306601735\n",
      "0.14839944037412203\n",
      "0.12143094875261067\n",
      "0.11079481529533944\n",
      "-0.11485192912391184\n",
      "-0.11837151628573409\n",
      "-0.11492307186348433\n",
      "-0.11998185646957123\n",
      "-0.11109159694896875\n",
      "-0.14026598909173035\n",
      "-0.17823209980505866\n",
      "-0.1511610661178566\n",
      "0.12085400508772151\n",
      "-0.11340997272750104\n",
      "-0.1140850209634101\n",
      "-0.10624294986079748\n",
      "-0.10975743614071265\n",
      "-0.12133890327553543\n",
      "0.11154145120837797\n",
      "0.15764329015691286\n",
      "-0.1832048034764985\n",
      "0.1279412298818775\n",
      "-0.1494109361636361\n",
      "-0.1704254889088967\n",
      "-0.10998172973319024\n",
      "0.11926931824727335\n",
      "-0.11214630478714364\n",
      "0.11366686399107713\n",
      "-0.12049863770381972\n",
      "-0.11423701817672963\n",
      "0.21850409146476282\n",
      "0.14452361745784775\n",
      "0.19623765624251321\n",
      "0.1530527670519759\n",
      "0.1323089910111792\n",
      "0.20314049459847916\n",
      "-0.1079641913340419\n",
      "0.17790888559823578\n",
      "0.19548496704787613\n",
      "0.20171207300257435\n",
      "-0.1382559337344637\n",
      "0.2383719208360824\n",
      "-0.2576580235152579\n",
      "-0.15591998676113086\n",
      "-0.19343564265645385\n",
      "0.20694781802670054\n",
      "0.1173249429762159\n",
      "0.16695293235258113\n",
      "0.12309905282565292\n",
      "-0.12598105585225525\n",
      "-0.13608077087132273\n",
      "0.13041958138240564\n",
      "-0.1272179530928485\n",
      "-0.11685427753650894\n",
      "-0.1549935747081608\n",
      "-0.16354887693176073\n",
      "-0.14295780454250087\n",
      "-0.101568548674183\n",
      "0.16858185790476296\n",
      "0.1796035020366904\n",
      "0.16822297189299024\n",
      "-0.10378979459975746\n",
      "0.12892881133419748\n",
      "-0.1341730496571648\n",
      "-0.16293796901547\n",
      "-0.1802160745467059\n",
      "0.13025073402679532\n",
      "0.1120918786350856\n",
      "-0.12898951850872298\n",
      "-0.15498472206029285\n",
      "-0.11339046196853304\n",
      "-0.12943876918876715\n",
      "-0.17632497999999316\n",
      "-0.1728922880539043\n",
      "0.2796718341093428\n",
      "-0.14178827666120752\n",
      "0.1067595129454274\n",
      "-0.17950524600478526\n",
      "0.10970526540370278\n",
      "0.11152113306351637\n",
      "-0.10732863999405964\n",
      "-0.13575278533597263\n",
      "0.11660259735205351\n",
      "-0.12619397679799638\n",
      "0.12475307093009369\n",
      "0.16353140870444144\n",
      "-0.20541763014038966\n",
      "0.1265955834188329\n",
      "0.11277962806847923\n",
      "-0.1342907964890404\n",
      "0.17420421320449095\n",
      "-0.10544948768438434\n",
      "0.13455072460909656\n",
      "0.17641106691072858\n",
      "0.10245015978227855\n",
      "-0.17171981842416242\n",
      "-0.13303087584706721\n",
      "0.2343246905923924\n",
      "0.20246292991978732\n",
      "-0.13453346749921502\n",
      "-0.11975740260690836\n",
      "0.2500666620315076\n",
      "0.16574621789497634\n",
      "-0.11701969446428763\n",
      "-0.18500842572225754\n",
      "-0.1267505980732618\n",
      "-0.12705382051420339\n",
      "-0.12443660207982413\n",
      "0.15643864288372283\n",
      "-0.13199338697392704\n",
      "-0.11957887148695386\n",
      "0.2055832767997421\n",
      "-0.10089895317556163\n",
      "-0.12910079544491734\n",
      "0.24089009542671566\n",
      "0.10001874405267598\n",
      "-0.11517982177743769\n",
      "0.11267972787170435\n",
      "0.12151717585990218\n",
      "0.10913906693458396\n",
      "0.20609593622960462\n",
      "-0.10598518792074307\n",
      "-0.14549572844481706\n",
      "-0.1499412515984145\n",
      "0.16196905661207972\n",
      "-0.1066090660808169\n",
      "0.15342446247715125\n",
      "-0.15109752105695703\n",
      "0.17830093313671685\n",
      "0.17624085858301905\n",
      "-0.14510218840992103\n",
      "0.17257501554169838\n",
      "-0.1148848037595167\n",
      "0.20975140996142402\n",
      "0.11233243464595362\n",
      "-0.10398407890046\n",
      "-0.11487130017158921\n",
      "0.15937946724899082\n",
      "0.13580898039352868\n",
      "-0.13401540910712284\n",
      "-0.10431505941756439\n",
      "0.1573285176238676\n",
      "0.19547466041540873\n",
      "0.18312613375874914\n",
      "-0.16869676654646676\n",
      "-0.13078089190589612\n",
      "0.18658399494906108\n",
      "-0.1461614065470678\n",
      "0.1274880185954156\n",
      "0.16688713564951638\n",
      "0.10780738745114396\n",
      "0.16633437332191806\n",
      "0.15648465550952934\n",
      "-0.12734349364649653\n",
      "0.20763331788289321\n",
      "-0.10237622983295853\n",
      "-0.13272276966263544\n",
      "0.13750149909945342\n",
      "0.149988042935888\n",
      "0.20079744967693605\n",
      "0.16865569348930795\n",
      "0.21673479579063915\n",
      "0.14219724995816901\n",
      "0.15629345131356517\n",
      "0.12765106481209668\n",
      "0.15821907007627117\n",
      "-0.19092767435230162\n",
      "0.14212119488465744\n",
      "0.11850249173585817\n",
      "-0.10143778088827901\n",
      "0.1981163563737521\n",
      "0.23843874814488633\n",
      "-0.16148992972104587\n",
      "0.11701092352301858\n",
      "0.13263954939645003\n",
      "0.10381378698824607\n",
      "0.17161993422755562\n",
      "0.20657209132324744\n",
      "0.12953727404302234\n",
      "0.16387833531260132\n",
      "-0.1436204148074167\n",
      "0.11326593007496112\n",
      "0.30648109195361345\n",
      "0.1231071919661704\n",
      "0.15725813316144271\n",
      "0.1562428268204072\n",
      "-0.10065964749142517\n",
      "0.1387686591476331\n",
      "-0.1456392299691117\n",
      "-0.13610323425494691\n",
      "0.12399877290628208\n",
      "-0.15024223737350198\n",
      "-0.13579509566276318\n",
      "-0.10094009738368236\n",
      "-0.14835289187598977\n",
      "-0.12480942212729927\n",
      "0.11415759206170706\n",
      "-0.14593721029914886\n",
      "0.10730429522020758\n",
      "0.22199879894286872\n",
      "0.14908984986153667\n",
      "-0.12140261771788735\n",
      "-0.14834073766087255\n",
      "-0.10945562433284274\n",
      "0.1627450243341365\n",
      "-0.13765914789724973\n",
      "-0.13148206629937484\n",
      "-0.1489733609134975\n",
      "0.13995590113674922\n",
      "-0.14070656914434132\n",
      "-0.11182862382868083\n",
      "0.19868099118510277\n",
      "-0.12188468915479236\n",
      "-0.1199442249293025\n",
      "0.12446467301946315\n",
      "0.11067487782158418\n",
      "-0.1394557749517917\n",
      "-0.15525146552764002\n",
      "-0.1572419033527492\n",
      "0.10449348439520711\n",
      "0.221643952980003\n",
      "-0.1289976204992644\n",
      "0.1606422085764981\n",
      "-0.10647787101797668\n",
      "0.10664421094513483\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "932"
      ]
     },
     "execution_count": 397,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_features = select_by_corr(0.1, 500) #0.15, 400\n",
    "len(selected_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "id": "92843438-023c-4dac-b7a2-ae67081b90b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_mask = protein_data[\"uniprot_id\"].isin(selected_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 399,
   "id": "2f361f5d-d2c4-4e4c-8c62-d3ad1ddfcf5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_idx</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0</td>\n",
       "      <td>A0AVT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>A1L0T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>2</td>\n",
       "      <td>A3KMH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>3</td>\n",
       "      <td>A5YKK6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>4</td>\n",
       "      <td>A9QM74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8407</th>\n",
       "      <td>927</td>\n",
       "      <td>Q9Y696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8408</th>\n",
       "      <td>928</td>\n",
       "      <td>Q9Y697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8420</th>\n",
       "      <td>929</td>\n",
       "      <td>Q9Y6G9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8426</th>\n",
       "      <td>930</td>\n",
       "      <td>Q9Y6K0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8456</th>\n",
       "      <td>931</td>\n",
       "      <td>Q9Y6Y8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>932 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      protein_idx uniprot_id\n",
       "15              0     A0AVT1\n",
       "22              1     A1L0T0\n",
       "31              2     A3KMH1\n",
       "40              3     A5YKK6\n",
       "76              4     A9QM74\n",
       "...           ...        ...\n",
       "8407          927     Q9Y696\n",
       "8408          928     Q9Y697\n",
       "8420          929     Q9Y6G9\n",
       "8426          930     Q9Y6K0\n",
       "8456          931     Q9Y6Y8\n",
       "\n",
       "[932 rows x 2 columns]"
      ]
     },
     "execution_count": 399,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_protein_data = protein_data[feature_mask]\n",
    "selected_protein_data = selected_protein_data.copy()\n",
    "selected_protein_data[\"protein_idx\"] = selected_protein_data.reset_index(drop = True).index\n",
    "selected_protein_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 400,
   "id": "d48990b7-6e9e-478c-b470-a484530e6d44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>model_id</th>\n",
       "      <th>model_name</th>\n",
       "      <th>protein_intensity</th>\n",
       "      <th>zscore</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A075B6K4</td>\n",
       "      <td>SIDM00483</td>\n",
       "      <td>SK-GT-4</td>\n",
       "      <td>-0.577590</td>\n",
       "      <td>-1.284310</td>\n",
       "      <td>IGLV3-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A075B6K4</td>\n",
       "      <td>SIDM00689</td>\n",
       "      <td>JM1</td>\n",
       "      <td>3.408550</td>\n",
       "      <td>0.129406</td>\n",
       "      <td>IGLV3-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A075B6K4</td>\n",
       "      <td>SIDM01259</td>\n",
       "      <td>GR-ST</td>\n",
       "      <td>6.300060</td>\n",
       "      <td>1.154900</td>\n",
       "      <td>IGLV3-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A075B6N1</td>\n",
       "      <td>SIDM00846</td>\n",
       "      <td>HeLa</td>\n",
       "      <td>0.127721</td>\n",
       "      <td>-1.034210</td>\n",
       "      <td>TRBV19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A075B6N1</td>\n",
       "      <td>SIDM00958</td>\n",
       "      <td>CML-T1</td>\n",
       "      <td>4.283890</td>\n",
       "      <td>1.352450</td>\n",
       "      <td>TRBV19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4919677</th>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>SIDM01248</td>\n",
       "      <td>CGTH-W-1</td>\n",
       "      <td>3.812190</td>\n",
       "      <td>0.723363</td>\n",
       "      <td>SEC23IP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4919678</th>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>SIDM01251</td>\n",
       "      <td>H9</td>\n",
       "      <td>3.854180</td>\n",
       "      <td>0.787636</td>\n",
       "      <td>SEC23IP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4919679</th>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>SIDM01259</td>\n",
       "      <td>GR-ST</td>\n",
       "      <td>3.802910</td>\n",
       "      <td>0.709159</td>\n",
       "      <td>SEC23IP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4919680</th>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>SIDM01261</td>\n",
       "      <td>YMB-1-E</td>\n",
       "      <td>4.540440</td>\n",
       "      <td>1.838070</td>\n",
       "      <td>SEC23IP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4919681</th>\n",
       "      <td>Q9Y6Y8</td>\n",
       "      <td>SIDM01265</td>\n",
       "      <td>MM1S</td>\n",
       "      <td>3.873330</td>\n",
       "      <td>0.816948</td>\n",
       "      <td>SEC23IP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4919682 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         uniprot_id   model_id model_name  protein_intensity    zscore  \\\n",
       "0        A0A075B6K4  SIDM00483    SK-GT-4          -0.577590 -1.284310   \n",
       "1        A0A075B6K4  SIDM00689        JM1           3.408550  0.129406   \n",
       "2        A0A075B6K4  SIDM01259      GR-ST           6.300060  1.154900   \n",
       "3        A0A075B6N1  SIDM00846       HeLa           0.127721 -1.034210   \n",
       "4        A0A075B6N1  SIDM00958     CML-T1           4.283890  1.352450   \n",
       "...             ...        ...        ...                ...       ...   \n",
       "4919677      Q9Y6Y8  SIDM01248   CGTH-W-1           3.812190  0.723363   \n",
       "4919678      Q9Y6Y8  SIDM01251         H9           3.854180  0.787636   \n",
       "4919679      Q9Y6Y8  SIDM01259      GR-ST           3.802910  0.709159   \n",
       "4919680      Q9Y6Y8  SIDM01261    YMB-1-E           4.540440  1.838070   \n",
       "4919681      Q9Y6Y8  SIDM01265       MM1S           3.873330  0.816948   \n",
       "\n",
       "           symbol  \n",
       "0        IGLV3-10  \n",
       "1        IGLV3-10  \n",
       "2        IGLV3-10  \n",
       "3          TRBV19  \n",
       "4          TRBV19  \n",
       "...           ...  \n",
       "4919677   SEC23IP  \n",
       "4919678   SEC23IP  \n",
       "4919679   SEC23IP  \n",
       "4919680   SEC23IP  \n",
       "4919681   SEC23IP  \n",
       "\n",
       "[4919682 rows x 6 columns]"
      ]
     },
     "execution_count": 400,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteomic_to_model_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 401,
   "id": "c7168593-b7a5-43c4-936d-669f8f7ea604",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(644622, 2)"
      ]
     },
     "execution_count": 401,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_protein_merged_df = get_proteomic_and_model_merged_df(selected_protein_data, merged_drug_model_df, proteomic_to_model_df)\n",
    "model_to_protein_edge_index = model_protein_merged_df.loc[:, [\"model_idx\", \"protein_idx\"]].values.astype(np.int64)\n",
    "model_to_protein_edge_index.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 402,
   "id": "0610473b-a2cf-4a41-a19d-6d27ce51b228",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_to_protein_edge_attr = torch.from_numpy(model_protein_merged_df[\"protein_intensity\"].values).type(torch.float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 403,
   "id": "af8e41d6-eb75-43e1-9469-d0171044e3c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "torch.Size([932, 932])"
      ]
     },
     "execution_count": 403,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteomic_features = torch.eye(selected_protein_data[\"uniprot_id\"].shape[0])\n",
    "proteomic_features.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 404,
   "id": "a0e688f0-995a-44f2-af6b-bb3aec2b8189",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 1., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 1., ..., 0., 0., 0.],\n",
       "       ...,\n",
       "       [0., 0., 0., ..., 1., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 1., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 1.]])"
      ]
     },
     "execution_count": 404,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_and_target_features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8128798b-f858-40de-ac49-f5f43c6157eb",
   "metadata": {},
   "source": [
    "## Create graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 405,
   "id": "df2d0f8f-b0be-430b-8214-77c70faef03d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.data import HeteroData\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "standard_scaler = StandardScaler()\n",
    "data = HeteroData()\n",
    "data['model'].x = torch.from_numpy(standard_scaler.fit_transform(model_and_target_features)).type(torch.float)\n",
    "data['gene'].x = torch.from_numpy(gene_features).type(torch.float)\n",
    "data['protein'].x = proteomic_features\n",
    "# data['model', 'copy_number', 'gene'].edge_index = torch.transpose(torch.from_numpy(model_to_gene_copy_number_edge_index).type(torch.long), 0, -1)\n",
    "# data['model', 'mutation', 'gene'].edge_index = torch.transpose(torch.from_numpy(model_to_gene_mutations_edge_index).type(torch.long), 0, -1)\n",
    "data['model', 'expression', 'gene'].edge_index = torch.transpose(torch.from_numpy(gene_expression_model_edge_index).type(torch.long), 0, -1)\n",
    "data['model', 'expression', 'protein'].edge_index = torch.transpose(torch.from_numpy(model_to_protein_edge_index).type(torch.long), 0, -1)\n",
    "\n",
    "# # data['model', 'copy_number', 'gene'].edge_attr = torch.from_numpy(standard_scaler.fit_transform(copy_number_edge_attributes)).type(torch.float)\n",
    "# # data['model', 'mutation', 'gene'].edge_attr = torch.from_numpy(standard_scaler.fit_transform(mutations_edge_attributes)).type(torch.float)\n",
    "data['model', 'expression', 'gene'].edge_attr = torch.from_numpy(standard_scaler.fit_transform(gene_expression_model_edge_attributes)).type(torch.float)\n",
    "data['model', 'expression', 'protein'].edge_attr = model_to_protein_edge_attr.unsqueeze(-1)\n",
    "\n",
    "# data['gene', 'copy_number', 'model'].edge_index = torch.transpose(reverse_edge_index(torch.from_numpy(model_to_gene_copy_number_edge_index)).type(torch.long), 0, -1)\n",
    "# data['gene', 'mutation', 'model'].edge_index = torch.transpose(reverse_edge_index(torch.from_numpy(model_to_gene_mutations_edge_index)).type(torch.long), 0, -1)\n",
    "data['gene', 'expression', 'model'].edge_index = torch.transpose(reverse_edge_index(torch.from_numpy(gene_expression_model_edge_index)).type(torch.long), 0, -1)\n",
    "data['protein', 'expression', 'model'].edge_index = torch.transpose(reverse_edge_index(torch.from_numpy(model_to_protein_edge_index)).type(torch.long), 0, -1)\n",
    "\n",
    "# data['gene', 'copy_number', 'model'].edge_attr = torch.from_numpy(standard_scaler.fit_transform(copy_number_edge_attributes)).type(torch.float)\n",
    "# data['gene', 'mutation', 'model'].edge_attr = torch.from_numpy(standard_scaler.fit_transform(mutations_edge_attributes)).type(torch.float)\n",
    "data['gene', 'expression', 'model'].edge_attr = torch.from_numpy(standard_scaler.fit_transform(gene_expression_model_edge_attributes)).type(torch.float)\n",
    "data['protein', 'expression', 'model'].edge_attr = model_to_protein_edge_attr.unsqueeze(-1)\n",
    "\n",
    "# label_leak_cn = torch.index_select(y, 0 , data['gene', 'copy_number', 'model'].edge_index[1])\n",
    "# data['gene', 'copy_number', 'model'].edge_attr = torch.concat((data['gene', 'copy_number', 'model'].edge_attr, label_leak_cn.unsqueeze(1)), dim = 1)\n",
    "\n",
    "y = torch.from_numpy(target_labels).type(torch.float)\n",
    "data['model'].y = y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "id": "bec519fc-ede8-4416-badf-a3f897342855",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HeteroData(\n",
       "  model={\n",
       "    x=[717, 717],\n",
       "    y=[717],\n",
       "  },\n",
       "  gene={ x=[295, 295] },\n",
       "  protein={ x=[932, 932] },\n",
       "  (model, expression, gene)={\n",
       "    edge_index=[2, 210925],\n",
       "    edge_attr=[210925, 1],\n",
       "  },\n",
       "  (model, expression, protein)={\n",
       "    edge_index=[2, 644622],\n",
       "    edge_attr=[644622, 1],\n",
       "  },\n",
       "  (gene, expression, model)={\n",
       "    edge_index=[2, 210925],\n",
       "    edge_attr=[210925, 1],\n",
       "  },\n",
       "  (protein, expression, model)={\n",
       "    edge_index=[2, 644622],\n",
       "    edge_attr=[644622, 1],\n",
       "  }\n",
       ")"
      ]
     },
     "execution_count": 406,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e8ce583-a190-494e-8983-080750d0f505",
   "metadata": {},
   "source": [
    "## Filter unimportant proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "id": "f48b2acd-1b82-4397-a594-9db7128c71aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def index_to_mask(rows, index_array):\n",
    "    mask_array = np.zeros(rows, dtype=int)\n",
    "    mask_array[index_array] = 1\n",
    "    return torch.from_numpy(mask_array.astype(np.bool_))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 408,
   "id": "9384f5d7-3f9a-4570-8fe8-0d40a9b6d16e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from tqdm.notebook import tqdm\n",
    "# from torch_geometric.utils import degree\n",
    "# if bool(data[\"protein\"]):\n",
    "#     DEG_THRESHOLD = 100 \n",
    "#     high_degree_mask = degree(data[(\"protein\", \"expression\", \"model\")].edge_index[0], data[\"protein\"].x.shape[0]) >= DEG_THRESHOLD\n",
    "#     high_deg_proteins = torch.arange(data[\"protein\"].x.shape[0])[high_degree_mask]\n",
    "    \n",
    "#     var_scores = torch.zeros(high_deg_proteins.shape[0])\n",
    "#     for i, protein in tqdm(enumerate(high_deg_proteins)):\n",
    "#         protein_edge_mask = data[(\"protein\", \"expression\", \"model\")].edge_index[0] == protein\n",
    "#         protein_attr = data[(\"protein\", \"expression\", \"model\")].edge_attr[protein_edge_mask]\n",
    "#         var_scores[i] = protein_attr.var()\n",
    "    \n",
    "#     q = torch.tensor([0.95])\n",
    "#     high_deg_high_var_mask = var_scores <= torch.quantile(data[(\"protein\", \"expression\", \"model\")].edge_attr, q) #var_scores.median()\n",
    "#     unimportant_proteins = high_deg_proteins[high_deg_high_var_mask]\n",
    "    \n",
    "#     unimportant_protein_edge_mask = torch.isin(data[(\"protein\", \"expression\", \"model\")].edge_index[0], unimportant_proteins, invert=True)\n",
    "#     unimportant_protein_edge_mask_rev = torch.isin(data[(\"model\", \"expression\", \"protein\")].edge_index[1], unimportant_proteins, invert=True)\n",
    "    \n",
    "#     def mask_edge_index(edge_index, mask):\n",
    "#         source = edge_index[0][mask]\n",
    "#         target = edge_index[1][mask]\n",
    "#         return torch.stack([source, target])\n",
    "#     data[(\"protein\", \"expression\", \"model\")].edge_index = mask_edge_index(data[(\"protein\", \"expression\", \"model\")].edge_index, unimportant_protein_edge_mask)\n",
    "#     data[(\"model\", \"expression\", \"protein\")].edge_index = mask_edge_index(data[(\"model\", \"expression\", \"protein\")].edge_index, unimportant_protein_edge_mask_rev)\n",
    "    \n",
    "#     data[(\"protein\", \"expression\", \"model\")].edge_attr = data[(\"protein\", \"expression\", \"model\")].edge_attr[unimportant_protein_edge_mask]\n",
    "#     data[(\"model\", \"expression\", \"protein\")].edge_attr = data[(\"model\", \"expression\", \"protein\")].edge_attr[unimportant_protein_edge_mask_rev]\n",
    "    \n",
    "#     # relevant_protein_mask = ~index_to_mask(data[\"protein\"].x.shape[0], unimportant_proteins)\n",
    "#     # data[\"protein\"].x = torch.eye(data[\"protein\"].x[relevant_protein_mask].shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 409,
   "id": "5ed22b91-df25-4d56-b8e5-673b87d13024",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HeteroData(\n",
       "  model={\n",
       "    x=[717, 717],\n",
       "    y=[717],\n",
       "  },\n",
       "  gene={ x=[295, 295] },\n",
       "  protein={ x=[932, 932] },\n",
       "  (model, expression, gene)={\n",
       "    edge_index=[2, 210925],\n",
       "    edge_attr=[210925, 1],\n",
       "  },\n",
       "  (model, expression, protein)={\n",
       "    edge_index=[2, 644622],\n",
       "    edge_attr=[644622, 1],\n",
       "  },\n",
       "  (gene, expression, model)={\n",
       "    edge_index=[2, 210925],\n",
       "    edge_attr=[210925, 1],\n",
       "  },\n",
       "  (protein, expression, model)={\n",
       "    edge_index=[2, 644622],\n",
       "    edge_attr=[644622, 1],\n",
       "  }\n",
       ")"
      ]
     },
     "execution_count": 409,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aebfb603-80a3-4946-b4b9-d53f30f9355e",
   "metadata": {},
   "source": [
    "## train, val test split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 410,
   "id": "f18c8a70-6849-4762-b395-f065651d78e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(torch.Size([537]), torch.Size([180]))"
      ]
     },
     "execution_count": 410,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "train_indices, test_indices = train_test_split(torch.arange(y.shape[0]), random_state = 42, test_size=0.25)\n",
    "train_indices.shape, test_indices.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86b7e341-1762-4f58-9a55-6f1ecb5267cc",
   "metadata": {},
   "source": [
    "## Construct GNN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 411,
   "id": "6199da2e-d323-4040-96d0-08e6ab6f6db9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "from torch.nn import Parameter\n",
    "from torch_geometric.nn import MessagePassing\n",
    "from torch_geometric.utils import add_self_loops, degree\n",
    "from torch_geometric.nn import Linear\n",
    "from torch_geometric.utils import degree\n",
    "\n",
    "# This function is copied from https://github.com/gordicaleksa/pytorch-GAT/blob/main/The%20Annotated%20GAT%20(PPI).ipynb\n",
    "## Thanks to Aleksa Gordic for the implementation!\n",
    "def explicit_broadcast(this, other):\n",
    "    # Append singleton dimensions until this.dim() == other.dim()\n",
    "    for _ in range(this.dim(), other.dim()):\n",
    "        this = this.unsqueeze(-1)\n",
    "    return this.expand_as(other)\n",
    "\n",
    "def describe(tensor):\n",
    "    return f\"{tensor.mean()} +- {tensor.std()} [{tensor.min()}, {tensor.max()}]\"\n",
    "    \n",
    "class HeteroCustomGNN(torch.nn.Module):\n",
    "    \n",
    "    def __init__(self, hidden_dim, out_channels, x_dict, edge_attr_dict, dropout = 0.0, debug = False):\n",
    "        ## message aggregator\n",
    "        super().__init__()\n",
    "        self.dropout = torch.nn.Dropout(dropout)\n",
    "        self.hidden_dim = hidden_dim\n",
    "\n",
    "        self.in_dict = torch.nn.ModuleDict()\n",
    "        for key in x_dict:\n",
    "            self.in_dict[key] = torch.nn.Linear(x_dict[key].shape[-1], hidden_dim)\n",
    "        self.debug = debug\n",
    "\n",
    "\n",
    "    def forward(self, x_dict, edge_index_dict, edge_attr_dict):\n",
    "        for key in x_dict:\n",
    "            x_dict[key] = self.in_dict[key](x_dict[key])\n",
    "            x_dict[key] = self.dropout(x_dict[key])\n",
    "\n",
    "        \n",
    "        message_dict = dict()\n",
    "        for key in edge_index_dict:\n",
    "            source, edge, target = key\n",
    "            edge_index = edge_index_dict[key]\n",
    "            if edge_index.shape[-1] == 0:\n",
    "                continue\n",
    "            source_lift = torch.index_select(x_dict[source], 0, edge_index[0])\n",
    "            target_lift = torch.index_select(x_dict[target], 0, edge_index[1])\n",
    "\n",
    "            \n",
    "                   \n",
    "            # messages = source_lift + target_lift + edge_attr_dict[key]\n",
    "            messages = edge_attr_dict[key] * source_lift #torch.cat([source_lift, edge_attr_dict[key], target_lift], dim = 1)\n",
    "            # messages = self.dropout(messages) #*explicit_broadcast(edge_dropout, messages)\n",
    "            # messages = torch.relu(self.message_dict[str(key)](messages))\n",
    "            aggregated_messages = torch.zeros((x_dict[target].shape[0], messages.shape[1]), device = device, dtype=x_dict[target].dtype)\n",
    "            \n",
    "            \n",
    "            \n",
    "            broadcasted_target_index = explicit_broadcast(edge_index[1], messages)\n",
    "            aggregated_messages.scatter_reduce_(0, broadcasted_target_index, messages, \"sum\")\n",
    "            message_dict[key] = aggregated_messages\n",
    "            \n",
    "            if self.debug:\n",
    "                print(f\"source_lift: {source_lift.shape}\")\n",
    "                print(f\"target_lift: {target_lift.shape}\")\n",
    "                print(f\"edge_attr_dict: {edge_attr_dict[key].shape}\")\n",
    "                print(f\"mess: {messages.shape}\")\n",
    "                print(f\"agg. mess: {aggregated_messages.shape}\")\n",
    "                print(f\"{key}: {aggregated_messages.shape}\")\n",
    "            \n",
    "        ## Message aggregation\n",
    "        message_aggr_dict = dict()\n",
    "        for key in message_dict:\n",
    "            source, edge, target = key\n",
    "            messages_for_target = message_dict[key]\n",
    "            previous_message_aggr_target = message_aggr_dict.get(target, None)\n",
    "            if previous_message_aggr_target is None:\n",
    "                message_aggr_dict[target] = messages_for_target\n",
    "            else:\n",
    "                message_aggr_dict[target] = previous_message_aggr_target + messages_for_target\n",
    "\n",
    "        # Message passing\n",
    "        for key in message_aggr_dict:\n",
    "            x_dict[key] = message_aggr_dict[key] #* x_dict[key]  #torch.relu(self.passing_dict[key](messages)) # + x_dict[key]\n",
    "            if self.debug: print(f\"{key}: {x_dict[key].shape}\")\n",
    "        \n",
    "        return x_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 412,
   "id": "e02dd307-c027-4f7b-98e2-d460e6082195",
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.nn import Linear, GATv2Conv, HANConv, SAGEConv\n",
    "class GNN(torch.nn.Module):\n",
    "    def __init__(self, hidden_dim = 1, heads = 1):\n",
    "        super().__init__()\n",
    "        self.lin = Linear(-1, 1)\n",
    "        self.conv = GATv2Conv(1, 1, add_self_loops=False, edge_dim=1, heads = heads, concat=False)\n",
    "        # self.conv_2 = GATv2Conv(hidden_dim*heads, hidden_dim, add_self_loops=False, edge_dim=1, heads = heads, concat=True)\n",
    "        # self.lin_out = Linear(hidden_dim*heads, 1)\n",
    "        \n",
    "    def forward(self, x, edge_index, edge_attr):\n",
    "        x = self.lin(x)\n",
    "        x = self.conv(x, edge_index, edge_attr)\n",
    "        # x = torch.relu(x)\n",
    "        # x = self.conv_2(x, edge_index, edge_attr)\n",
    "        # x = torch.relu(x)\n",
    "        # x = self.lin_out(x)\n",
    "        return x"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f4ce4e1-d03b-48f7-82f7-1b1958cf382c",
   "metadata": {},
   "source": [
    "## K-fold cross validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 413,
   "id": "83f0d1e4-3f3f-456d-aa43-c0089d0fcb35",
   "metadata": {},
   "outputs": [],
   "source": [
    "device = torch.device(\"cuda:1\")\n",
    "loss_fn = torch.nn.MSELoss(reduction=\"mean\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 414,
   "id": "a2897733-7f47-48b3-8511-2bbb12021f68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HeteroData(\n",
       "  model={\n",
       "    x=[717, 717],\n",
       "    y=[717],\n",
       "  },\n",
       "  gene={ x=[295, 295] },\n",
       "  protein={ x=[932, 932] },\n",
       "  (model, expression, gene)={\n",
       "    edge_index=[2, 210925],\n",
       "    edge_attr=[210925, 1],\n",
       "  },\n",
       "  (model, expression, protein)={\n",
       "    edge_index=[2, 644622],\n",
       "    edge_attr=[644622, 1],\n",
       "  },\n",
       "  (gene, expression, model)={\n",
       "    edge_index=[2, 210925],\n",
       "    edge_attr=[210925, 1],\n",
       "  },\n",
       "  (protein, expression, model)={\n",
       "    edge_index=[2, 644622],\n",
       "    edge_attr=[644622, 1],\n",
       "  }\n",
       ")"
      ]
     },
     "execution_count": 414,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 415,
   "id": "f90da468-3cda-4fd7-b1ce-a7de8bb29409",
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy\n",
    "def train_k(loader, hyperparameters, data = None):\n",
    "    global device\n",
    "    model = hyperparameters[\"model\"].to(device)\n",
    "    optim = torch.optim.Adam(params=model.parameters(), lr = hyperparameters[\"lr\"], weight_decay=hyperparameters[\"wd\"])\n",
    "    loss_list = []\n",
    "    for epoch in tqdm(range(hyperparameters[\"epochs\"])):\n",
    "        acc_loss = 0\n",
    "        batch_size = 0\n",
    "        if data is None:\n",
    "            for i, batch in enumerate(loader):\n",
    "                optim.zero_grad()\n",
    "                batch = batch.to(device)\n",
    "                model.train()\n",
    "                out =  model(batch.x_dict, batch.edge_index_dict, batch.edge_attr_dict)\n",
    "                loss = loss_fn(out[\"model\"].squeeze(),batch[\"model\"].y)\n",
    "                loss.backward()\n",
    "                optim.step()\n",
    "                acc_loss += loss.item()\n",
    "                batch_size += batch[\"model\"].x.shape[0]\n",
    "                batch = batch.cpu()\n",
    "        if data is not None:\n",
    "            optim.zero_grad()\n",
    "            data = data.to(device)\n",
    "            model.train()\n",
    "            out =  model(data.x_dict,data.edge_index_dict, data.edge_attr_dict)\n",
    "            loss = loss_fn(out[\"model\"].squeeze()[data['model'].train_mask],data[\"model\"].y[data['model'].train_mask])\n",
    "            loss.backward()\n",
    "            optim.step()\n",
    "            acc_loss += loss.item()\n",
    "            batch_size += data[\"model\"].x.shape[0]\n",
    "            data = data.cpu()\n",
    "        batch_loss = acc_loss / batch_size\n",
    "        # print(f\"{batch_loss=}\")\n",
    "        loss_list.append(batch_loss)                \n",
    "        \n",
    "    return model, np.array(loss_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 416,
   "id": "0919a102-a51f-45b5-93d0-3c19f5c57863",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import mean_squared_error, mean_absolute_error\n",
    "from sklearn.metrics import r2_score\n",
    "\n",
    "def evaluate_k(model, loader, data = None, mask = None):\n",
    "    global device\n",
    "    model = model.to(device)\n",
    "    with torch.inference_mode():\n",
    "        true_target_values = []\n",
    "        pred_target_values = []\n",
    "\n",
    "        if data is None:\n",
    "            for i, batch in enumerate(loader):\n",
    "                batch = batch.to(device)\n",
    "                model.eval()\n",
    "                pred = model(batch.x_dict, batch.edge_index_dict, batch.edge_attr_dict)\n",
    "    \n",
    "                true_target_values.append(batch[\"model\"].y)\n",
    "                pred_target_values.append(pred[\"model\"].squeeze(0).flatten())\n",
    "                batch = batch.cpu()\n",
    "        if data is not None:\n",
    "            data = data.to(device)\n",
    "            model.eval()\n",
    "            pred = model(data.x_dict, data.edge_index_dict, data.edge_attr_dict)\n",
    "\n",
    "            true_target_values.append(data[\"model\"].y[mask])\n",
    "            pred_target_values.append(pred[\"model\"].squeeze(0).flatten()[mask])\n",
    "            data = data.cpu()\n",
    "        \n",
    "        true_target_values = torch.cat(true_target_values, 0).squeeze()\n",
    "        pred_target_values = torch.cat(pred_target_values, 0).squeeze()\n",
    "        \n",
    "        y_true = true_target_values.cpu().numpy()\n",
    "        y_pred = pred_target_values.cpu().numpy()\n",
    "        \n",
    "        mse = mean_squared_error(y_true,y_pred)\n",
    "        mae = mean_absolute_error(y_true,y_pred)\n",
    "        r2 = r2_score(y_true,y_pred)\n",
    "        secondary_scores = {\n",
    "            \"mse\": mse,\n",
    "            \"mae\": mae,\n",
    "            \"r2\": r2\n",
    "        }\n",
    "        return r2, secondary_scores #, (pred_target_values, true_target_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "id": "d9360fcf-61cb-4d44-9944-f66af1a80851",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[\"model\"].train_mask = index_to_mask(data[\"model\"].x.shape[0], train_indices)\n",
    "data[\"model\"].test_mask = index_to_mask(data[\"model\"].x.shape[0], test_indices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 418,
   "id": "49c9d253-ea97-4abb-85f1-548bed117255",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HeteroData(\n",
       "  model={\n",
       "    x=[717, 717],\n",
       "    y=[717],\n",
       "    train_mask=[717],\n",
       "    test_mask=[717],\n",
       "  },\n",
       "  gene={ x=[295, 295] },\n",
       "  protein={ x=[932, 932] },\n",
       "  (model, expression, gene)={\n",
       "    edge_index=[2, 210925],\n",
       "    edge_attr=[210925, 1],\n",
       "  },\n",
       "  (model, expression, protein)={\n",
       "    edge_index=[2, 644622],\n",
       "    edge_attr=[644622, 1],\n",
       "  },\n",
       "  (gene, expression, model)={\n",
       "    edge_index=[2, 210925],\n",
       "    edge_attr=[210925, 1],\n",
       "  },\n",
       "  (protein, expression, model)={\n",
       "    edge_index=[2, 644622],\n",
       "    edge_attr=[644622, 1],\n",
       "  }\n",
       ")"
      ]
     },
     "execution_count": 418,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 419,
   "id": "fa1a6142-06ca-4c03-bce0-d60b40a6c5ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.nn import to_hetero\n",
    "\n",
    "best_hyperparams = dict()\n",
    "\n",
    "m = GNN(heads = 1, hidden_dim = 1)\n",
    "m = to_hetero(m, metadata=data.metadata(), aggr = \"sum\")\n",
    "\n",
    "best_hyperparams[\"model\"] = HeteroCustomGNN( 1, 1, data.x_dict, data.edge_attr_dict, 0.1) #0.05\n",
    "best_hyperparams[\"epochs\"] = 40_000\n",
    "best_hyperparams[\"lr\"] = 1e-4 #1e-3\n",
    "best_hyperparams[\"wd\"] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 420,
   "id": "8e9c2bfc-fffb-4414-92ad-1bbd245fe630",
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.loader import NeighborLoader\n",
    "\n",
    "BATCH_SIZE = 8\n",
    "train_input_nodes = ('model', data['model'].train_mask)\n",
    "train_loader = NeighborLoader(data, num_neighbors={key: [-1] * 2 for key in data.edge_types}, shuffle=True,batch_size=int(data['model'].train_mask.sum()), input_nodes=train_input_nodes)\n",
    "test_input_nodes = ('model', data['model'].test_mask)\n",
    "test_loader = NeighborLoader(data, num_neighbors={key: [-1] * 2 for key in data.edge_types}, shuffle=True,batch_size=int(data['model'].test_mask.sum()), input_nodes=test_input_nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 562,
   "id": "4ed117e4-aa50-4d2a-808b-a4cf48d03430",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cd6d554816714a2fa0b78476180b9de2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/40000 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjIAAAHHCAYAAACle7JuAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/OQEPoAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA+pElEQVR4nO3deXhU5d3/8c/JNkkgCYFAFgk7si/KJqKAZYeqIK0b2oD9iQuoPFiruLBofRBsLbVa1C5QfUCKVsCiqBFZqoLsm2IAZVMIkSWZhJAwJPfvD8zAGJZkZsKZmbxf15XLzDlnTr7fmQn5eN9nsYwxRgAAAEEozO4CAAAAvEWQAQAAQYsgAwAAghZBBgAABC2CDAAACFoEGQAAELQIMgAAIGgRZAAAQNAiyAAAgKBFkAEQMEaOHKlGjRr5bX/Lly+XZVlavny53/YJILAQZABclGVZFfoiMAC41CLsLgBA4HvjjTc8Hr/++uvKzMwst7xVq1Y+/Zy//vWvKi0t9WkfAKoXggyAi7rjjjs8Hq9evVqZmZnllv9UYWGhYmNjK/xzIiMjvaoPQPXF1BIAv+jdu7fatm2r9evXq2fPnoqNjdXjjz8uSVq0aJGGDBmitLQ0ORwONW3aVM8884xKSko89vHTY2T27Nkjy7L0+9//Xq+99pqaNm0qh8OhLl26aO3atV7X+tZbb6lTp06KiYlRUlKS7rjjDn3//fce22RnZ2vUqFGqX7++HA6HUlNTdeONN2rPnj3ubdatW6cBAwYoKSlJMTExaty4se666y6v6wJQeYzIAPCbI0eOaNCgQbr11lt1xx13KDk5WZI0e/Zs1axZU+PHj1fNmjX1ySefaOLEiXI6nXr++ecvut+5c+cqPz9f99xzjyzL0vTp03XTTTfp22+/rfQozuzZszVq1Ch16dJFU6dO1aFDh/SnP/1Jn332mTZu3KhatWpJkoYPH64vv/xSDzzwgBo1aqScnBxlZmZq37597sf9+/dX3bp19dhjj6lWrVras2eP3nnnnUq/bgB8YACgksaMGWN++s9Hr169jCTzyiuvlNu+sLCw3LJ77rnHxMbGmqKiIveyjIwM07BhQ/fj3bt3G0mmTp065ujRo+7lixYtMpLMf/7znwvWuWzZMiPJLFu2zBhjzMmTJ029evVM27ZtzYkTJ9zbLV682EgyEydONMYYc+zYMSPJPP/88+fd94IFC4wks3bt2gvWAKBqMbUEwG8cDodGjRpVbnlMTIz7+/z8fB0+fFjXXnutCgsL9fXXX190v7fccosSExPdj6+99lpJ0rffflup+tatW6ecnBzdf//9io6Odi8fMmSIWrZsqffee89db1RUlJYvX65jx46dc19lIzeLFy+Wy+WqVB0A/IcgA8BvLrvsMkVFRZVb/uWXX2rYsGFKSEhQfHy86tat6z5QOC8v76L7bdCggcfjslBzvpBxPnv37pUktWjRoty6li1butc7HA5NmzZNS5YsUXJysnr27Knp06crOzvbvX2vXr00fPhwTZkyRUlJSbrxxhs1a9YsFRcXV6omAL4hyADwm7NHXsrk5uaqV69e2rx5s55++mn95z//UWZmpqZNmyZJFTrdOjw8/JzLjTG+FXwB48aN044dOzR16lRFR0frqaeeUqtWrbRx40ZJp6+t8/bbb2vVqlUaO3asvv/+e911113q1KmTCgoKqqwuAJ4IMgCq1PLly3XkyBHNnj1bDz30kH7+85+rb9++HlNFl0rDhg0lSVlZWeXWZWVludeXadq0qR5++GF99NFH2rZtm06ePKk//OEPHttcddVVevbZZ7Vu3TrNmTNHX375pebNm1d1TQDwQJABUKXKRlPOHj05efKk/vKXv1zyWjp37qx69erplVde8ZgCWrJkibZv364hQ4ZIOn39m6KiIo/nNm3aVHFxce7nHTt2rNyIUMeOHSWJ6SXgEuL0awBV6uqrr1ZiYqIyMjL04IMPyrIsvfHGG1U6LXQ+kZGRmjZtmkaNGqVevXrptttuc59+3ahRI/3P//yPJGnHjh3q06ePbr75ZrVu3VoRERFasGCBDh06pFtvvVWS9M9//lN/+ctfNGzYMDVt2lT5+fn661//qvj4eA0ePPiS9wZUVwQZAFWqTp06Wrx4sR5++GE9+eSTSkxM1B133KE+ffpowIABl7yekSNHKjY2Vs8995weffRR1ahRQ8OGDdO0adPcZyKlp6frtttu09KlS/XGG28oIiJCLVu21Pz58zV8+HBJpw/2XbNmjebNm6dDhw4pISFBXbt21Zw5c9S4ceNL3hdQXVnGjv8tAgAA8AOOkQEAAEGLIAMAAIIWQQYAAAQtggwAAAhaBBkAABC0CDIAACBohfx1ZEpLS3XgwAHFxcXJsiy7ywEAABVgjFF+fr7S0tIUFnb+cZeQDzIHDhxQenq63WUAAAAv7N+/X/Xr1z/v+pAPMnFxcZJOvxDx8fF+26/L5dJHH32k/v37KzIy0m/7DSSh3mOo9yeFfo/0F/xCvUf6857T6VR6err77/j5hHyQKZtOio+P93uQiY2NVXx8fEh+OKXQ7zHU+5NCv0f6C36h3iP9+e5ih4VwsC8AAAhaBBkAABC0CDIAACBoEWQAAEDQIsgAAICgRZABAABBiyADAACCFkEGAAAELYIMAAAIWgQZAAAQtAgyAAAgaBFkAABA0CLIeOlY4UkdLZacJ1x2lwIAQLVFkPHSHzJ3acqGCL2+ep/dpQAAUG0RZHxk7C4AAIBqjCDjJcv68RuSDAAAtiHIeMm6+CYAAKCKEWR8ZBiSAQDANgQZL5VNLRlyDAAAtiHIeMn6cXKJHAMAgH0IMl6yOEgGAADbEWR8xNQSAAD2Ich46czZ1yQZAADsQpDxFnNLAADYjiDjKwZkAACwja1BZuXKlbr++uuVlpYmy7K0cOFC9zqXy6VHH31U7dq1U40aNZSWlqZf/epXOnDggH0Fn4UL+wIAYD9bg8zx48fVoUMHvfzyy+XWFRYWasOGDXrqqae0YcMGvfPOO8rKytINN9xgQ6XlMbMEAID9Iuz84YMGDdKgQYPOuS4hIUGZmZkey1566SV17dpV+/btU4MGDS5FiRfFWUsAANjH1iBTWXl5ebIsS7Vq1TrvNsXFxSouLnY/djqdkk5PVblcLr/VYkpLJUklJSV+3W8gKeuL/oJXqPdIf8Ev1HukP9/3fTGWMYExpmBZlhYsWKChQ4eec31RUZF69Oihli1bas6cOefdz+TJkzVlypRyy+fOnavY2Fh/lasFe8K0/GCY+qSV6oaGpX7bLwAAOH2Iye233668vDzFx8efd7ugGJFxuVy6+eabZYzRzJkzL7jthAkTNH78ePdjp9Op9PR09e/f/4IvRGVteG+7dHC/GjVqpMGDWvptv4HE5XIpMzNT/fr1U2RkpN3l+F2o9yeFfo/0F/xCvUf6817ZjMrFBHyQKQsxe/fu1SeffHLRMOJwOORwOMotj4yM9OuLHB52+jhpKywsJD+cZ/P3axdoQr0/KfR7pL/gF+o90p93+6yIgA4yZSFm586dWrZsmerUqWN3SW5n7n4dEDNzAABUS7YGmYKCAu3atcv9ePfu3dq0aZNq166t1NRU/eIXv9CGDRu0ePFilZSUKDs7W5JUu3ZtRUVF2VU2AAAIELYGmXXr1um6665zPy47tiUjI0OTJ0/Wu+++K0nq2LGjx/OWLVum3r17X6oyAQBAgLI1yPTu3fuCUzOBPG1j/Ti3FMAlAgAQ8rjXkpe4sC8AAPYjyHjJfbCvvWUAAFCtEWR8FMjTXwAAhDqCjJesHyeXiDEAANiHIOMl7n4NAID9CDI+YmYJAAD7EGS8VDYgQ44BAMA+BBlvMbUEAIDtCDJeKjvYl7klAADsQ5DxETEGAAD7EGS8ZDEgAwCA7QgyXuIQGQAA7EeQ8ZFhcgkAANsQZLzE1BIAAPYjyHjJYnIJAADbEWR8xIAMAAD2Ich4i6klAABsR5DxEhNLAADYjyDjJcvibksAANiNIOMjppYAALAPQcZLjMcAAGA/goyXLA6SAQDAdgQZHzG1BACAfQgyXjoztUSSAQDALgQZL1nMLQEAYDuCjI+YWgIAwD4EGR+RYwAAsA9BxktnrodHlAEAwC4EGS9xiAwAAPYjyPiIARkAAOxDkPGS9eMJ2OQYAADsQ5DxElNLAADYjyDjJY71BQDAfgQZH3FlXwAA7EOQ8RJX9gUAwH4EGR8xtQQAgH0IMj4ixwAAYB+CjJfOXNnX1jIAAKjWCDJe4ggZAADsR5DxEWctAQBgH4KMl8rOWuJgXwAA7EOQ8RJTSwAA2I8g46Wyg30ZkAEAwD62BpmVK1fq+uuvV1pamizL0sKFCz3WG2M0ceJEpaamKiYmRn379tXOnTvtKfY8DHNLAADYxtYgc/z4cXXo0EEvv/zyOddPnz5dL774ol555RV98cUXqlGjhgYMGKCioqJLXGl5TC0BAGC/CDt/+KBBgzRo0KBzrjPGaMaMGXryySd14403SpJef/11JScna+HChbr11lsvZanllR3sa28VAABUa7YGmQvZvXu3srOz1bdvX/eyhIQEdevWTatWrTpvkCkuLlZxcbH7sdPplCS5XC65XC6/1VdaUnL6v6Wlft1vICnri/6CV6j3SH/BL9R7pD/f930xARtksrOzJUnJyckey5OTk93rzmXq1KmaMmVKueUfffSRYmNj/Vbf9mxLUrgOHcrR+++/77f9BqLMzEy7S6hSod6fFPo90l/wC/Ue6a/yCgsLK7RdwAYZb02YMEHjx493P3Y6nUpPT1f//v0VHx/vt59zdPUeafcO1atXT4MHX+m3/QYSl8ulzMxM9evXT5GRkXaX43eh3p8U+j3SX/AL9R7pz3tlMyoXE7BBJiUlRZJ06NAhpaamupcfOnRIHTt2PO/zHA6HHA5HueWRkZF+fZEjwk+/dJZlheSH82z+fu0CTaj3J4V+j/QX/EK9R/rzbp8VEbDXkWncuLFSUlK0dOlS9zKn06kvvvhC3bt3t7EyTxzsCwCAfWwdkSkoKNCuXbvcj3fv3q1Nmzapdu3aatCggcaNG6ff/e53at68uRo3bqynnnpKaWlpGjp0qH1F/8ji/GsAAGxna5BZt26drrvuOvfjsmNbMjIyNHv2bP32t7/V8ePHNXr0aOXm5uqaa67RBx98oOjoaLtKdivLMVwPDwAA+9gaZHr37n3BK+NalqWnn35aTz/99CWsqnK4+zUAAPYJ2GNkAp37XkvkGAAAbEOQ8RoHyQAAYDeCjJe4+zUAAPYjyPiKJAMAgG0IMl5iYgkAAPsRZLx0ZmqJIRkAAOxCkPERZy0BAGAfgoyXLCaXAACwHUHGS5y1BACA/QgyPmJqCQAA+xBkvOS+1xJjMgAA2IYg4y1ufw0AgO0IMl5yxxgGZAAAsA1BxkfkGAAA7EOQ8RIzSwAA2I8g4yX3wb6ctgQAgG0IMj4ixgAAYB+CjJcs5pYAALAdQcZLZ6aWbC0DAIBqjSDjJW5RAACA/QgyPuJgXwAA7EOQAQAAQYsg4yUO9gUAwH4EGR8xswQAgH0IMl5iPAYAAPsRZLzEWUsAANiPIOMl68cxGc5aAgDAPgQZHxFjAACwD0HGS5y0BACA/QgyXuIWBQAA2I8g4yPD5BIAALYhyHiLqSUAAGxHkPGSxfnXAADYjiDjI3IMAAD2Ich4iZklAADsR5DxkntmiSEZAABsQ5Dxkvv0ayaXAACwDUHGR4zIAABgH4KMlywu7QsAgO0IMl7iyr4AANiPIOMjcgwAAPYhyHiLmSUAAGwX0EGmpKRETz31lBo3bqyYmBg1bdpUzzzzjEwAzOdY4vxrAADsFmF3ARcybdo0zZw5U//85z/Vpk0brVu3TqNGjVJCQoIefPBBW2vjDgUAANgvoIPM559/rhtvvFFDhgyRJDVq1Ehvvvmm1qxZY3NlZzAgAwCAfQI6yFx99dV67bXXtGPHDl1++eXavHmzPv30U73wwgvnfU5xcbGKi4vdj51OpyTJ5XLJ5XL5rbaSU6cknb4gnj/3G0jK+qK/4BXqPdJf8Av1HunP931fjGUC4YCT8ygtLdXjjz+u6dOnKzw8XCUlJXr22Wc1YcKE8z5n8uTJmjJlSrnlc+fOVWxsrN9q25Fn6eWvwpUSYzShY4nf9gsAAKTCwkLdfvvtysvLU3x8/Hm3C+gRmfnz52vOnDmaO3eu2rRpo02bNmncuHFKS0tTRkbGOZ8zYcIEjR8/3v3Y6XQqPT1d/fv3v+ALUVlxO3KkrzapRs0aGjz4Gr/tN5C4XC5lZmaqX79+ioyMtLscvwv1/qTQ75H+gl+o90h/3iubUbmYgA4yjzzyiB577DHdeuutkqR27dpp7969mjp16nmDjMPhkMPhKLc8MjLSry9yZMTpl86SFZIfzrP5+7ULNKHenxT6PdJf8Av1HunPu31WRECffl1YWKiwMM8Sw8PDVVpaalNFZ3DWEgAA9gvoEZnrr79ezz77rBo0aKA2bdpo48aNeuGFF3TXXXfZXZpb4B5hBABA6AvoIPPnP/9ZTz31lO6//37l5OQoLS1N99xzjyZOnGh3aQAAIAAEdJCJi4vTjBkzNGPGDLtLKefMza8ZkgEAwC4VPkYmJyfngutPnToVUBeqq2pltyhgagkAAPtUOMikpqZ6hJl27dpp//797sdHjhxR9+7d/VtdECDHAABgnwoHmZ9eN2/Pnj3lrroXwNfW8zuLu18DAGA7v55+bVWjv+5lnVaj7AYAQMAJ6OvIBAPD5BIAALap8FlLlmUpPz9f0dHRMsbIsiwVFBS4LyFc0UsJh4rqNPoEAECgqnCQMcbo8ssv93h8xRVXeDyuTn/cmVoCAMB+FQ4yy5Ytq8o6gg+3KAAAwHYVDjK9evWqyjoAAAAqrcJB5tSpUyopKfG4s/ShQ4f0yiuv6Pjx47rhhht0zTXXVEmRgejMhX0ZkwEAwC4VDjJ33323oqKi9Oqrr0qS8vPz1aVLFxUVFSk1NVV//OMftWjRIg0ePLjKig0kZccDEWMAALBPhU+//uyzzzR8+HD349dff10lJSXauXOnNm/erPHjx+v555+vkiIDGQMyAADYp8JB5vvvv1fz5s3dj5cuXarhw4crISFBkpSRkaEvv/zS/xUGqOpzfhYAAIGrwkEmOjpaJ06ccD9evXq1unXr5rG+oKDAv9UFMIuzlgAAsF2Fg0zHjh31xhtvSJL++9//6tChQ/rZz37mXv/NN98oLS3N/xUGqDN3vybKAABglwof7Dtx4kQNGjRI8+fP18GDBzVy5Eilpqa61y9YsEA9evSokiIBAADOpVLXkVm/fr0++ugjpaSk6Je//KXH+o4dO6pr165+LzBQMbUEAID9KhxkJKlVq1Zq1arVOdeNHj3aLwUFHZIMAAC2qXCQWblyZYW269mzp9fFBCNyDAAA9qlwkOndu/eZi8Cd5wBXy7JUUlLin8oCXDW6PyYAAAGrwkEmMTFRcXFxGjlypO68804lJSVVZV0Bj7OWAACwX4VPvz548KCmTZumVatWqV27dvr1r3+tzz//XPHx8UpISHB/VTfEGAAA7FPhIBMVFaVbbrlFH374ob7++mu1b99eY8eOVXp6up544gmdOnWqKusMOEwtAQBgvwoHmbM1aNBAEydO1Mcff6zLL79czz33nJxOp79rC2hlOYaZJQAA7FPpIFNcXKy5c+eqb9++atu2rZKSkvTee++pdu3aVVFfwDpzHRmSDAAAdqnwwb5r1qzRrFmzNG/ePDVq1EijRo3S/Pnzq12AAQAAgaPCQeaqq65SgwYN9OCDD6pTp06SpE8//bTcdjfccIP/qgtgZ85asrkQAACqsUpd2Xffvn165plnzru+Ol1HRhzsCwCA7SocZEpLS6uyjqDFiAwAAPbx6qwlMCADAEAgIMh4yX27Bs5aAgDANgQZL3EdGQAA7EeQAQAAQYsg46UzF8QDAAB2qdTp12c7efKkcnJyyp3N1KBBA5+LCgbuIEOSAQDANpUOMjt37tRdd92lzz//3GO5MaZ6XUcGAADYrtJBZuTIkYqIiNDixYuVmprqPnununFf2ZfJJQAAbFPpILNp0yatX79eLVu2rIp6gof7tCVbqwAAoFqr9MG+rVu31uHDh6uilqBCjgEAwH6VDjLTpk3Tb3/7Wy1fvlxHjhyR0+n0+AIAALhUKj211LdvX0lSnz59PJZXt4N9z5y1xJgMAAB2qXSQWbZsWVXUEXTOHOwLAADsUukg06tXr6qo47y+//57Pfroo1qyZIkKCwvVrFkzzZo1S507d76kdQAAgMBToSCzZcsWtW3bVmFhYdqyZcsFt23fvr1fCpOkY8eOqUePHrruuuu0ZMkS1a1bVzt37lRiYqLffoa3uCAeAAD2q1CQ6dixo7Kzs1WvXj117NhRlmWd89gQfx8jM23aNKWnp2vWrFnuZY0bN/bb/n3BWUsAANivQkFm9+7dqlu3rvv7S+Xdd9/VgAED9Mtf/lIrVqzQZZddpvvvv1933333JavhYjjYFwAA+1QoyDRs2PCc31e1b7/9VjNnztT48eP1+OOPa+3atXrwwQcVFRWljIyMcz6nuLhYxcXF7sdlp4S7XC65XC6/1Xbq1Cn39/7cbyAp64v+gleo90h/wS/Ue6Q/3/d9MZYJ4CGFqKgode7c2eO+Tg8++KDWrl2rVatWnfM5kydP1pQpU8otnzt3rmJjY/1WW26xNGlDhMItoxeuqh6nnAMAcKkUFhbq9ttvV15enuLj48+7ndd3v74UUlNT1bp1a49lrVq10r///e/zPmfChAkaP368+7HT6VR6err69+9/wReisr47UiBt+FyWFabBgwf4bb+BxOVyKTMzU/369VNkZKTd5fhdqPcnhX6P9Bf8Qr1H+vNeRS+yG9BBpkePHsrKyvJYtmPHjgtObzkcDjkcjnLLIyMj/foiR0SeeelC8cN5Nn+/doEm1PuTQr9H+gt+od4j/Xm3z4qo9C0KLqX/+Z//0erVq/W///u/2rVrl+bOnavXXntNY8aMsbs0zloCACAAVDrI7N+/X99995378Zo1azRu3Di99tprfi1Mkrp06aIFCxbozTffVNu2bfXMM89oxowZGjFihN9/VmVZP15IJoAPMQIAIORVemrp9ttv1+jRo3XnnXcqOztb/fr1U5s2bTRnzhxlZ2dr4sSJfi3w5z//uX7+85/7dZ8AACA0VHpEZtu2berataskaf78+Wrbtq0+//xzzZkzR7Nnz/Z3fQGLqSUAAOxX6SDjcrncB9N+/PHHuuGGGyRJLVu21MGDB/1bXQDjFgUAANiv0kGmTZs2euWVV/Tf//5XmZmZGjhwoCTpwIEDqlOnjt8LDFTWxTcBAABVrNJBZtq0aXr11VfVu3dv3XbbberQoYOk07cTKJtyAgAAuBQqfbBv7969dfjwYTmdTo+7UI8ePdqvV84NeNaZMRljjPssJgAAcOlUekTmxIkTKi4udoeYvXv3asaMGcrKylK9evX8XmCgOju2cJwMAAD2qHSQufHGG/X6669LknJzc9WtWzf94Q9/0NChQzVz5ky/FwgAAHA+lQ4yGzZs0LXXXitJevvtt5WcnKy9e/fq9ddf14svvuj3AgPV2TNJDMgAAGCPSgeZwsJCxcXFSZI++ugj3XTTTQoLC9NVV12lvXv3+r3AQGXJ8xgZAABw6VU6yDRr1kwLFy7U/v379eGHH6p///6SpJycHL/eXTrQMSIDAID9Kh1kJk6cqN/85jdq1KiRunbtqu7du0s6PTpzxRVX+L1AAACA86n06de/+MUvdM011+jgwYPua8hIUp8+fTRs2DC/FhfIOGsJAAD7VTrISFJKSopSUlLcd8GuX79+tbsYnufUEkkGAAA7VHpqqbS0VE8//bQSEhLUsGFDNWzYULVq1dIzzzyj0tLSqqgRAADgnCo9IvPEE0/o73//u5577jn16NFDkvTpp59q8uTJKioq0rPPPuv3IgPT2Wct2VgGAADVWKWDzD//+U/97W9/c9/1WpLat2+vyy67TPfff3+1CTLckQAAAPtVemrp6NGjatmyZbnlLVu21NGjR/1SFAAAQEVUOsh06NBBL730UrnlL730ksdZTKGOs5YAALBfpaeWpk+friFDhujjjz92X0Nm1apV2r9/v95//32/FxioOGsJAAD7VXpEplevXtqxY4eGDRum3Nxc5ebm6qabblJWVpb7HkzVgcXBvgAA2M6r68ikpaWVO6j3u+++0+jRo/Xaa6/5pTAAAICLqfSIzPkcOXJEf//73/21u4DHvZYAALCf34JMdeN5sC9RBgAAOxBkAABA0CLIeOusuSXGYwAAsEeFD/a96aabLrg+NzfX11qCCteRAQDAfhUOMgkJCRdd/6tf/crngoIFtygAAMB+FQ4ys2bNqso6ghsjMgAA2IJjZLzkMbVEkgEAwBYEGS9ZFlf2BQDAbgQZAAAQtAgyXvKcWgIAAHYgyHjJ4xYFzC0BAGALggwAAAhaBBkvWVzZFwAA2xFk/ICZJQAA7EGQ8YH141gM15EBAMAeBBkAABC0CDL+wIAMAAC2IMj4oOxwX3IMAAD2IMgAAICgRZDxxY9DMpy1BACAPQgyPjgztUSSAQDADkEVZJ577jlZlqVx48bZXQoAAAgAQRNk1q5dq1dffVXt27e3u5RymFoCAMAeQRFkCgoKNGLECP31r39VYmKi3eW4cdYSAAD2irC7gIoYM2aMhgwZor59++p3v/vdBbctLi5WcXGx+7HT6ZQkuVwuuVwuv9XkcrncQeb0voPipayUstfLn69bIAn1/qTQ75H+gl+o90h/vu/7YixjAntiZN68eXr22We1du1aRUdHq3fv3urYsaNmzJhxzu0nT56sKVOmlFs+d+5cxcbG+rW236wOl8tYmnTlKdV2+HXXAABUa4WFhbr99tuVl5en+Pj4824X0MMI+/fv10MPPaTMzExFR0dX6DkTJkzQ+PHj3Y+dTqfS09PVv3//C74QleVyuaQvPpGM1Lv3daqfGOO3fQcKl8ulzMxM9evXT5GRkXaX43eh3p8U+j3SX/AL9R7pz3tlMyoXE9BBZv369crJydGVV17pXlZSUqKVK1fqpZdeUnFxscLDwz2e43A45HCUHx6JjIz0+4tcNrUUERERkh/QMlXx2gWSUO9PCv0e6S/4hXqP9OfdPisioINMnz59tHXrVo9lo0aNUsuWLfXoo4+WCzEAAKB6CeggExcXp7Zt23osq1GjhurUqVNuuZ0C+ygjAABCV1Ccfh2ouLIvAAD2CugRmXNZvny53SWcYV18EwAAUHUYkfEDppYAALAHQcYHXNkXAAB7EWR84A4yDMkAAGALggwAAAhaBBk/YDwGAAB7EGR8cGZqydYyAACotggyvuD0awAAbEWQ8QuGZAAAsANBxgdMLQEAYC+CjB+QYwAAsAdBBgAABC2CjA+YWgIAwF4EGV/8mGS4+zUAAPYgyPiAs68BALAXQcYPmFoCAMAeBBkfcIwMAAD2IsgAAICgRZDxAw72BQDAHgQZHzC1BACAvQgyvuC0JQAAbEWQ8QE5BgAAexFk/ICpJQAA7EGQ8QMO9gUAwB4EGR8wtQQAgL0IMn7A1BIAAPYgyPjAct80EgAA2IEg4weGIRkAAGxBkAEAAEGLIOMD95V9ba0CAIDqiyDjB8wsAQBgD4KMDzj9GgAAexFkfJDvKvuOIRkAAOxAkPFBcenpMZl1e47ZXAkAANUTQcYPPv/miN0lAABQLRFk/GDoFWl2lwAAQLVEkPFBcszpY2OS46NtrgQAgOqJIOMDzloCAMBeBBl/4KQlAABsQZDxA3IMAAD2IMj4wH2LApIMAAC2IMj44OCJ01HmyPFimysBAKB6Isj4wcZ9uXaXAABAtRTQQWbq1Knq0qWL4uLiVK9ePQ0dOlRZWVl2l+VWM+L0nNLVTevYXAkAANVTQAeZFStWaMyYMVq9erUyMzPlcrnUv39/HT9+3O7SJElJP14+hkNkAACwR4TdBVzIBx984PF49uzZqlevntavX6+ePXvaVNUZ1o9H+3KwLwAA9gjoEZmfysvLkyTVrl3b5kp+iiQDAIAdAnpE5mylpaUaN26cevToobZt2553u+LiYhUXnzmLyOl0SpJcLpdcLpff6nG5XO7Tr12nSvy670BR1lMo9iaFfn9S6PdIf8Ev1HukP9/3fTGWMcExMXLfffdpyZIl+vTTT1W/fv3zbjd58mRNmTKl3PK5c+cqNjbWrzW9uC1c3+RbGnV5iTrWCYqXEQCAoFBYWKjbb79deXl5io+PP+92QRFkxo4dq0WLFmnlypVq3LjxBbc914hMenq6Dh8+fMEXorJcLpdumPGJdjkt/enm9hrcLsVv+w4ULpdLmZmZ6tevnyIjI+0ux+9CvT8p9Hukv+AX6j3Sn/ecTqeSkpIuGmQCemrJGKMHHnhACxYs0PLlyy8aYiTJ4XDI4XCUWx4ZGVkFHyIjyVJYeHhIfkDLVM1rFzhCvT8p9Hukv+AX6j3Sn3f7rIiADjJjxozR3LlztWjRIsXFxSk7O1uSlJCQoJiYGJurO+sWBbZWAQBA9RXQZy3NnDlTeXl56t27t1JTU91f//rXv+wuzUMQzM4BABCSAnpEJtADQtl1ZAAAgD0CekQm0JXlmNIAD1wAAIQqgowP3MfIkGMAALAFQcYPCDIAANiDIOMD972W7C0DAIBqiyDjB4F+UDIAAKGKIOMDjpEBAMBeBBkfnJlaIskAAGAHgowPykZiGJEBAMAeBBkffJV7+uWbv26/zZUAAFA9EWT8YMO+XLtLAACgWiLIAACAoEWQAQAAQYsgAwAAghZBxgcNa54+Xemenk1srgQAgOqJIOODI0Wn/7uRg30BALAFQcYHBadOXxFvzZ6jNlcCAED1RJDxQULU6aml9vUTbK4EAIDqiSDjgyvqnA4yVzdNsrkSAACqJ4KMD8pePO5+DQCAPQgyvvjxppGlBBkAAGxBkPHBjzmGm0YCAGATgowPLPeIjL11AABQXRFkfPBdwen//mvtPnsLAQCgmiLI+ODrvNMv3/GTJTZXAgBA9USQAQAAQYsgAwAAghZBBgAABC2CDAAACFoEGQAAELQIMgAAIGgRZHxwU6PTp123To23uRIAAKongowPYiJO/zcpzmFvIQAAVFMEGR+UvXglpaW21gEAQHVFkPFB+I+vnquEmy0BAGAHgowPzozIEGQAALADQcYHZSMyp0qYWgIAwA4EGR+U5ZfN3+XZWwgAANUUQcYHGw5bdpcAAEC1RpDxAUfGAABgL4KMD9rXJsoAAGAngowPikrsrgAAgOqNIOOD2Igz35dyCjYAAJccQcYHLWudCS8uru4LAMAlFxRB5uWXX1ajRo0UHR2tbt26ac2aNXaXJEkKP+ukpTdW7bWvEAAAqqmADzL/+te/NH78eE2aNEkbNmxQhw4dNGDAAOXk5NhdmkeQ+d172+0rBACAasoyxgT0wR3dunVTly5d9NJLL0mSSktLlZ6ergceeECPPfbYRZ/vdDqVkJCgvLw8xcfH+60ul8ul9957X+NWR1x8Y1RIRJilU5fwWKOhHVK1bPsB5Z30/npArVPj9dVB55l9dkzTwk0Hym3XqE6s9hwp9Fh2XYu6Wpb1g8LDLDWvV1Mv3naF5n6xT7M/3yNJurJBLW3YlytJurZ5kmKjwvXhl4ckSSOvbuTeTpJ+0am+LElvrf9O9RNj1LlhomrXcCgm0tLLy7/V2N5NtPdYkQa3TZEkLdmWrXc3H9D/u6axLEv6639367JaMWqcVEOf7jqsLo0SdWuXBip0lWjP4eNKT4yRZVn6/YdZsizp9V9309rdR7Xpu1x9uvOw8k64JEmxUeEa1DZVv76msb49XKCSUqNiV6mmf/i1bumSrt4t6umlT3bpmx8KlF90SpOub62I8DBt2Z+rU6VG7209qGubJaluvEMd6tfS2+u/U2JslG7uXF9vrtmntFoxiggPU/N6NdXusgSdKjmlf3+wQpmH4xUeFqYBbVO0+psjSq0VrXDLUuHJEhkZHcwr0qC2qTrkLFKTujUUFR6mFilxmvKfr1TkKtHjg1vpX2v3q2ndGkpJiNG/1u7T2J81V3beCTmLTikqPEzr9x7TB19m6+bO9VXDEaFaMVFqnXb63xRjjKIiwpTjLFZ4mKXLk+O041C+Vuz4QZ0aJmrFjh9097VNFBMVrv9sPqAjBcVKrRWjOjWiFB0Zrg+2ZSuxRpTu6tFIy7N+UFLNKF2eHKflWYcUeXinTtZpri6N62jzd3m6qkltrdjxg/6z6YBG9miky2rFaldOgSIjLP13x2G1So1XfEyEfsgv1idf5+jh/i3UoX6CoiPD5SopVZhlaVdOgdbtPaZXVnyjyde31uUpcfohv1h7jxTKmNNXLTdGOlxQrNo1HHJEhikuOkJHC05qR06BMro31O7Dx7X3SKHq1IxSm7QElRojS9LR4yf18fYc/bxDqgqLS7R2z1HddOVlWrzloGZ9tlvhYZYe+FlzxcdEKq/wpPYfLVRa0R6dTLpchwtc+veG7/THWzoqMtxS3gmXXl35rXo2r6vOjRLVJKmm3lyzT3uOHNcTQ1rJERGuk6dKVVJqtP2gUwdyTygpzqFwy1KsI1yR4WFKTYjWHzN36KomdXRlw0TlF7k06d0v1TIlXjddcZlOlpTq811HlF47RmFhlr4/dkJx0ZFqXz9Bm/bnqmVKnLZ979Tqb4/oqiZ11LVxovYfPaEBbVJUdKpE6/ceU46zWOm1Y/TvDd/p/13TRM4ilzbtz5XzxCk1SYrRuq1ZGjfsar2z8aASa0SpYZ1YHc4/qW9+KFCPZknKyS/W3iPHdWuXBoqKsFRQXKI5q/eqab2aSoiJ1Ly1+2VJurppHXVrUkelpUYN68Tqmx+OK6lmlE6cLNHuI8eVnVek2KgIdUyvpS3f5SolIVrHi0vkiAhT0akSbT/oVGFxifJOuDSgbYr2HD6u48WnFB8TqV92SpdlSbtyCrQ8K0eHj59U83o1lRwfrQnvbNX1HdL09UGnujaurfyiU7qqSR0dKShWTUeYPt3wlV65b6AiIyO9/rf0XCr69zugg8zJkycVGxurt99+W0OHDnUvz8jIUG5urhYtWlTuOcXFxSouLnY/djqdSk9P1+HDh/0eZDIzM/XQKoIMAKB6G9H1Mk2+vo1f9+l0OpWUlHTRIBPQf4UPHz6skpISJScneyxPTk7W119/fc7nTJ06VVOmTCm3/KOPPlJsbKzfaxxUv0RLvgv3+34BXFyEZXTKcIXtiogKMzpZGlqvVZhlVFpN3/9Iy8jl5959+Yw4cvfp/ff9e6xoYWHhxTdSgAcZb0yYMEHjx493Py4bkenfv3+VjMj84a6+etHPw2mBoqzHfv36+X3IMBCEen9S6PdIf8Ev1HukP+85nc6Lb6QADzJJSUkKDw/XoUOHPJYfOnRIKSkp53yOw+GQw+EotzwyMrJKPkRVtd9AEuo9hnp/Uuj3SH/BL9R7pD/v9lkRAX3WUlRUlDp16qSlS5e6l5WWlmrp0qXq3r27jZUBAIBAENAjMpI0fvx4ZWRkqHPnzuratatmzJih48ePa9SoUXaXBgAAbBbwQeaWW27RDz/8oIkTJyo7O1sdO3bUBx98UO4AYAAAUP0EfJCRpLFjx2rs2LF2lwEAAAJMQB8jAwAAcCEEGQAAELQIMgAAIGgRZAAAQNAiyAAAgKBFkAEAAEGLIAMAAIIWQQYAAAQtggwAAAhaQXFlX18YYyRV/HbgFeVyuVRYWCin0xmydzQN9R5DvT8p9Hukv+AX6j3Sn/fK/m6X/R0/n5APMvn5+ZKk9PR0mysBAACVlZ+fr4SEhPOut8zFok6QKy0t1YEDBxQXFyfLsvy2X6fTqfT0dO3fv1/x8fF+228gCfUeQ70/KfR7pL/gF+o90p/3jDHKz89XWlqawsLOfyRMyI/IhIWFqX79+lW2//j4+JD8cJ4t1HsM9f6k0O+R/oJfqPdIf9650EhMGQ72BQAAQYsgAwAAghZBxksOh0OTJk2Sw+Gwu5QqE+o9hnp/Uuj3SH/BL9R7pL+qF/IH+wIAgNDFiAwAAAhaBBkAABC0CDIAACBoEWQAAEDQIsh46eWXX1ajRo0UHR2tbt26ac2aNXaXVM7kyZNlWZbHV8uWLd3ri4qKNGbMGNWpU0c1a9bU8OHDdejQIY997Nu3T0OGDFFsbKzq1aunRx55RKdOnfLYZvny5bryyivlcDjUrFkzzZ49u8p6Wrlypa6//nqlpaXJsiwtXLjQY70xRhMnTlRqaqpiYmLUt29f7dy502Obo0ePasSIEYqPj1etWrX061//WgUFBR7bbNmyRddee62io6OVnp6u6dOnl6vlrbfeUsuWLRUdHa127drp/fffr/L+Ro4cWe49HThwYND0N3XqVHXp0kVxcXGqV6+ehg4dqqysLI9tLuXn0t+/xxXpr3fv3uXew3vvvTco+pOkmTNnqn379u4LoHXv3l1Llixxrw/m968i/QX7+/dTzz33nCzL0rhx49zLgu49NKi0efPmmaioKPOPf/zDfPnll+buu+82tWrVMocOHbK7NA+TJk0ybdq0MQcPHnR//fDDD+719957r0lPTzdLly4169atM1dddZW5+uqr3etPnTpl2rZta/r27Ws2btxo3n//fZOUlGQmTJjg3ubbb781sbGxZvz48earr74yf/7zn014eLj54IMPqqSn999/3zzxxBPmnXfeMZLMggULPNY/99xzJiEhwSxcuNBs3rzZ3HDDDaZx48bmxIkT7m0GDhxoOnToYFavXm3++9//mmbNmpnbbrvNvT4vL88kJyebESNGmG3btpk333zTxMTEmFdffdW9zWeffWbCw8PN9OnTzVdffWWefPJJExkZabZu3Vql/WVkZJiBAwd6vKdHjx712CaQ+xswYICZNWuW2bZtm9m0aZMZPHiwadCggSkoKHBvc6k+l1Xxe1yR/nr16mXuvvtuj/cwLy8vKPozxph3333XvPfee2bHjh0mKyvLPP744yYyMtJs27bNGBPc719F+gv29+9sa9asMY0aNTLt27c3Dz30kHt5sL2HBBkvdO3a1YwZM8b9uKSkxKSlpZmpU6faWFV5kyZNMh06dDjnutzcXBMZGWneeust97Lt27cbSWbVqlXGmNN/VMPCwkx2drZ7m5kzZ5r4+HhTXFxsjDHmt7/9rWnTpo3Hvm+55RYzYMAAP3dT3k//0JeWlpqUlBTz/PPPu5fl5uYah8Nh3nzzTWOMMV999ZWRZNauXeveZsmSJcayLPP9998bY4z5y1/+YhITE909GmPMo48+alq0aOF+fPPNN5shQ4Z41NOtWzdzzz33VFl/xpwOMjfeeON5nxNM/RljTE5OjpFkVqxYYYy5tJ/LS/F7/NP+jDn9h/DsPxo/FUz9lUlMTDR/+9vfQu79+2l/xoTO+5efn2+aN29uMjMzPXoKxveQqaVKOnnypNavX6++ffu6l4WFhalv375atWqVjZWd286dO5WWlqYmTZpoxIgR2rdvnyRp/fr1crlcHn20bNlSDRo0cPexatUqtWvXTsnJye5tBgwYIKfTqS+//NK9zdn7KNvGjtdi9+7dys7O9qgnISFB3bp18+ipVq1a6ty5s3ubvn37KiwsTF988YV7m549eyoqKsq9zYABA5SVlaVjx465t7Gr7+XLl6tevXpq0aKF7rvvPh05csS9Ltj6y8vLkyTVrl1b0qX7XF6q3+Of9ldmzpw5SkpKUtu2bTVhwgQVFha61wVTfyUlJZo3b56OHz+u7t27h9z799P+yoTC+zdmzBgNGTKkXB3B+B6G/E0j/e3w4cMqKSnxeAMlKTk5WV9//bVNVZ1bt27dNHv2bLVo0UIHDx7UlClTdO2112rbtm3Kzs5WVFSUatWq5fGc5ORkZWdnS5Kys7PP2WfZugtt43Q6deLECcXExFRRd+WV1XSues6ut169eh7rIyIiVLt2bY9tGjduXG4fZesSExPP23fZPqrKwIEDddNNN6lx48b65ptv9Pjjj2vQoEFatWqVwsPDg6q/0tJSjRs3Tj169FDbtm3dP/9SfC6PHTtW5b/H5+pPkm6//XY1bNhQaWlp2rJlix599FFlZWXpnXfeCZr+tm7dqu7du6uoqEg1a9bUggUL1Lp1a23atCkk3r/z9SeFxvs3b948bdiwQWvXri23Lhh/BwkyIWzQoEHu79u3b69u3bqpYcOGmj9//iUNGPCfW2+91f19u3bt1L59ezVt2lTLly9Xnz59bKys8saMGaNt27bp008/tbuUKnG+/kaPHu3+vl27dkpNTVWfPn30zTffqGnTppe6TK+0aNFCmzZtUl5ent5++21lZGRoxYoVdpflN+frr3Xr1kH//u3fv18PPfSQMjMzFR0dbXc5fsHUUiUlJSUpPDy83BHchw4dUkpKik1VVUytWrV0+eWXa9euXUpJSdHJkyeVm5vrsc3ZfaSkpJyzz7J1F9omPj7+koelspou9N6kpKQoJyfHY/2pU6d09OhRv/R9qT8DTZo0UVJSknbt2uWuKxj6Gzt2rBYvXqxly5apfv367uWX6nNZ1b/H5+vvXLp16yZJHu9hoPcXFRWlZs2aqVOnTpo6dao6dOigP/3pTyHz/p2vv3MJtvdv/fr1ysnJ0ZVXXqmIiAhFRERoxYoVevHFFxUREaHk5OSgew8JMpUUFRWlTp06aenSpe5lpaWlWrp0qcccaiAqKCjQN998o9TUVHXq1EmRkZEefWRlZWnfvn3uPrp3766tW7d6/GHMzMxUfHy8e5i1e/fuHvso28aO16Jx48ZKSUnxqMfpdOqLL77w6Ck3N1fr1693b/PJJ5+otLTU/Q9S9+7dtXLlSrlcLvc2mZmZatGihRITE93bBELf3333nY4cOaLU1FR3XYHcnzFGY8eO1YIFC/TJJ5+Um+K6VJ/Lqvo9vlh/57Jp0yZJ8ngPA7W/8yktLVVxcXHQv38X6+9cgu3969Onj7Zu3apNmza5vzp37qwRI0a4vw+697BShwbDGHP6lDGHw2Fmz55tvvrqKzN69GhTq1YtjyO4A8HDDz9sli9fbnbv3m0+++wz07dvX5OUlGRycnKMMadPsWvQoIH55JNPzLp160z37t1N9+7d3c8vO8Wuf//+ZtOmTeaDDz4wdevWPecpdo888ojZvn27efnll6v09Ov8/HyzceNGs3HjRiPJvPDCC2bjxo1m7969xpjTp1/XqlXLLFq0yGzZssXceOON5zz9+oorrjBffPGF+fTTT03z5s09Tk/Ozc01ycnJ5s477zTbtm0z8+bNM7GxseVOT46IiDC///3vzfbt282kSZP8cnryhfrLz883v/nNb8yqVavM7t27zccff2yuvPJK07x5c1NUVBQU/d13330mISHBLF++3OP01cLCQvc2l+pzWRW/xxfrb9euXebpp58269atM7t37zaLFi0yTZo0MT179gyK/owx5rHHHjMrVqwwu3fvNlu2bDGPPfaYsSzLfPTRR8aY4H7/LtZfKLx/5/LTM7GC7T0kyHjpz3/+s2nQoIGJiooyXbt2NatXr7a7pHJuueUWk5qaaqKiosxll11mbrnlFrNr1y73+hMnTpj777/fJCYmmtjYWDNs2DBz8OBBj33s2bPHDBo0yMTExJikpCTz8MMPG5fL5bHNsmXLTMeOHU1UVJRp0qSJmTVrVpX1tGzZMiOp3FdGRoYx5vQp2E899ZRJTk42DofD9OnTx2RlZXns48iRI+a2224zNWvWNPHx8WbUqFEmPz/fY5vNmzeba665xjgcDnPZZZeZ5557rlwt8+fPN5dffrmJiooybdq0Me+9916V9ldYWGj69+9v6tatayIjI03Dhg3N3XffXe6XPpD7O1dvkjw+M5fyc+nv3+OL9bdv3z7Ts2dPU7t2beNwOEyzZs3MI4884nEdkkDuzxhj7rrrLtOwYUMTFRVl6tata/r06eMOMcYE9/t3sf5C4f07l58GmWB7Dy1jjKncGA4AAEBg4BgZAAAQtAgyAAAgaBFkAABA0CLIAACAoEWQAQAAQYsgAwAAghZBBgAABC2CDICQZ1mWFi5caHcZAKoAQQZAlRo5cqQsyyr3NXDgQLtLAxACIuwuAEDoGzhwoGbNmuWxzOFw2FQNgFDCiAyAKudwOJSSkuLxVXanbcuyNHPmTA0aNEgxMTFq0qSJ3n77bY/nb926VT/72c8UExOjOnXqaPTo0SooKPDY5h//+IfatGkjh8Oh1NRUjR071mP94cOHNWzYMMXGxqp58+Z699133euOHTumESNGqG7duoqJiVHz5s3LBS8AgYkgA8B2Tz31lIYPH67NmzdrxIgRuvXWW7V9+3ZJ0vHjxzVgwAAlJiZq7dq1euutt/Txxx97BJWZM2dqzJgxGj16tLZu3ap3331XzZo18/gZU6ZM0c0336wtW7Zo8ODBGjFihI4ePer++V999ZWWLFmi7du3a+bMmUpKSrp0LwAA71X6NpMAUAkZGRkmPDzc1KhRw+Pr2WefNcacvmP0vffe6/Gcbt26mfvuu88YY8xrr71mEhMTTUFBgXv9e++9Z8LCwtx3/k5LSzNPPPHEeWuQZJ588kn344KCAiPJLFmyxBhjzPXXX29GjRrln4YBXFIcIwOgyl133XWaOXOmx7LatWu7v+/evbvHuu7du2vTpk2SpO3bt6tDhw6qUaOGe32PHj1UWlqqrKwsWZalAwcOqE+fPhesoX379u7va9Soofj4eOXk5EiS7rvvPg0fPlwbNmxQ//79NXToUF199dVe9Qrg0iLIAKhyNWrUKDfV4y8xMTEV2i4yMtLjsWVZKi0tlSQNGjRIe/fu1fvvv6/MzEz16dNHY8aM0e9//3u/1wvAvzhGBoDtVq9eXe5xq1atJEmtWrXS5s2bdfz4cff6zz77TGFhYWrRooXi4uLUqFEjLV261Kca6tatq4yMDP3f//2fZsyYoddee82n/QG4NBiRAVDliouLlZ2d7bEsIiLCfUDtW2+9pc6dO+uaa67RnDlztGbNGv3973+XJI0YMUKTJk1SRkaGJk+erB9++EEPPPCA7rzzTiUnJ0uSJk+erHvvvVf16tXToEGDlJ+fr88++0wPPPBAheqbOHGiOnXqpDZt2qi4uFiLFy92BykAgY0gA6DKffDBB0pNTfVY1qJFC3399deSTp9RNG/ePN1///1KTU3Vm2++qdatW0uSYmNj9eGHH+qhhx5Sly5dFBsbq+HDh+uFF15w7ysjI0NFRUX64x//qN/85jdKSkrSL37xiwrXFxUVpQkTJmjPnj2KiYnRtddeq3nz5vmhcwBVzTLGGLuLAFB9WZalBQsWaOjQoXaXAiAIcYwMAAAIWgQZAAAQtDhGBoCtmN0G4AtGZAAAQNAiyAAAgKBFkAEAAEGLIAMAAIIWQQYAAAQtggwAAAhaBBkAABC0CDIAACBoEWQAAEDQ+v/DjCobpLl7yAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final evaluation score is: 0.4541\n",
      "['Final evaluation mse score is: 1.7374', 'Final evaluation mae score is: 1.0619', 'Final evaluation r2 score is: 0.4541']\n",
      "Train evaluation score after: 0.6171 training.\n",
      "['Train evaluation mse score is: 1.3499', 'Train evaluation mae score is: 0.9043', 'Train evaluation r2 score is: 0.6171']\n"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from tqdm.notebook import tqdm\n",
    "from torch_geometric.nn import to_hetero\n",
    "import copy\n",
    "\n",
    "def plot_loss(loss_list):\n",
    "    plt.plot(loss_list)\n",
    "    plt.title(\"Train loss\")\n",
    "    plt.ylabel(\"Loss in MSE\")\n",
    "    plt.xlabel(\"Epochs\")\n",
    "    plt.grid()\n",
    "    plt.show()\n",
    "\n",
    "def refit(train_loader, hyperparameters, data = None):\n",
    "    model, loss_list = train_k(train_loader, hyperparameters, data)\n",
    "    return model, loss_list\n",
    "\n",
    "def reevaluate(model, loader, data = None, mask = None):\n",
    "    score, sec_scores = evaluate_k(model, loader, data, mask)\n",
    "    return score, sec_scores#, (pred_target_values, true_target_values)\n",
    "    \n",
    "def final_train_and_evaluate(train_loader, test_loader, hyperparameters):\n",
    "    model, loss_list = refit(train_loader, hyperparameters, data)\n",
    "    plot_loss(loss_list)\n",
    "    train_score_after_training, sec_train_scores = reevaluate(model, train_loader, data, data['model'].train_mask)\n",
    "    score, sec_scores = reevaluate(model, test_loader, data, data['model'].test_mask)\n",
    "    \n",
    "    print(f\"Final evaluation score is: {format(score, '.4f')}\")\n",
    "    print([f\"Final evaluation {key} score is: {format(sec_scores[key], '.4f')}\" for key in sec_scores])\n",
    "    print(f\"Train evaluation score after: {format(train_score_after_training, '.4f')} training.\")\n",
    "    print([f\"Train evaluation {key} score is: {format(sec_train_scores[key], '.4f')}\" for key in sec_train_scores])\n",
    "    return loss_list #, (pred_target_values, true_target_values)\n",
    "\n",
    "# training_score_before_training, sec_train_scores = reevaluate(best_hyperparams[\"model\"], train_loader, data, data['model'].train_mask)\n",
    "# print(f\"Train evaluation score before: {format(training_score_before_training, '.4f')} training.\")\n",
    "# print([f\"Train evaluation {key} score is: {format(sec_train_scores[key], '.4f')}\" for key in sec_train_scores])\n",
    "\n",
    "loss_list = final_train_and_evaluate(train_loader, test_loader, best_hyperparams)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 555,
   "id": "e01fe563-2e02-447e-9358-01592313bf70",
   "metadata": {},
   "outputs": [],
   "source": [
    "## TOOD normalize based on number of neighbors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "14c086b2-058b-433b-90d9-0f536bd0d72f",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_to_go_df = pd.read_csv(\"uniprot_go_mappings.csv\")\n",
    "uniprot_to_go_df = uniprot_to_go_df.groupby(\"UniProt ID\").aggregate(list).reset_index()\n",
    "uniprot_to_go_df = uniprot_to_go_df.rename(columns={\"UniProt ID\": \"uniprot_id\", \"GO Term\": \"go\"})\n",
    "exploded_protein_go_merge = pd.merge(proteomic_to_model_df, uniprot_to_go_df, how = \"inner\", on= \"uniprot_id\").explode(\"go\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "72096058-bda4-4e93-bbff-b7c81870982f",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_to_bioprocess_df = pd.read_csv(\"uniprot_to_bioprocess.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "20bf3480-5783-4963-8577-23d6f3c6f715",
   "metadata": {},
   "outputs": [],
   "source": [
    "bioprocess_to_bioprocess_parent_df = pd.read_csv(\"bioprocess_to_bioprocess_parent.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "fc9f3530-01d0-41c4-89c7-8b9babe47642",
   "metadata": {},
   "outputs": [],
   "source": [
    "bioprocess_to_bioprocess_parent_df = bioprocess_to_bioprocess_parent_df.copy()[bioprocess_to_bioprocess_parent_df[\"bioprocess_parent\"] != 19959]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "id": "7c54cacf-7903-43f0-ac71-859bd5f5d8f2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def recurse_bioprocesses(traversed_df, MAX_HOPS, depth = 0):\n",
    "    # print(depth)\n",
    "    # print(traversed_df[\"bioprocess_id\"].unique().shape[0])\n",
    "    # print((traversed_df[\"bioprocess_id\"] == 19959).sum() / traversed_df[\"bioprocess_id\"].shape[0])\n",
    "    if depth == MAX_HOPS:\n",
    "        return traversed_df\n",
    "    merged_df = pd.merge(traversed_df, bioprocess_to_bioprocess_parent_df, on = \"bioprocess_id\", how = \"inner\")\n",
    "    merged_df.pop(\"bioprocess_id\")\n",
    "    merged_df = merged_df.rename(columns= {\"bioprocess_parent\": \"bioprocess_id\"})\n",
    "    depth += 1    \n",
    "    return recurse_bioprocesses(merged_df, MAX_HOPS, depth)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "5ad196fe-1518-4953-9725-2ce585645185",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_aggregated_proteomic_data(hops):\n",
    "    traversed_df = uniprot_to_bioprocess_df.copy()\n",
    "    uniprot_to_bioprocess = recurse_bioprocesses(traversed_df, hops)\n",
    "    proteomic_bp_to_model_merge = pd.merge(proteomic_to_model_df, uniprot_to_bioprocess, how = \"left\", on= \"uniprot_id\")\n",
    "    proteomic_bp_to_model_pivot = pd.pivot_table(proteomic_bp_to_model_merge, values='protein_intensity', index=['model_id'],\n",
    "                       columns=['bioprocess_id'], aggfunc=\"sum\", fill_value=0.0).reset_index()\n",
    "    merged_reduced_protein_data = pd.merge(merged_drug_model_df.rename(columns = {\"SANGER_MODEL_ID\": \"model_id\"}), proteomic_bp_to_model_pivot, on = \"model_id\", how = \"inner\")\n",
    "    y = merged_reduced_protein_data.pop(\"LN_IC50\")\n",
    "    merged_reduced_protein_data.pop(\"model_id\")\n",
    "    merged_reduced_protein_data.pop(\"model_idx\")\n",
    "    merged_reduced_protein_data.pop(\"MIN_CONC\")\n",
    "    return y.values, merged_reduced_protein_data.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "id": "abf549b1-0d36-45d3-98b4-dbf49d408b24",
   "metadata": {},
   "outputs": [],
   "source": [
    "y, merged_reduced_protein_data = get_aggregated_proteomic_data(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "id": "5c2f9d5a-038e-4c1b-9a37-b392b95e4edf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      0.030016\n",
       "1      0.030016\n",
       "2      0.030016\n",
       "3      0.030016\n",
       "4      0.030016\n",
       "         ...   \n",
       "691    0.030016\n",
       "692    0.030016\n",
       "693    0.030016\n",
       "694    0.030016\n",
       "695    0.030016\n",
       "Name: MIN_CONC, Length: 696, dtype: float64"
      ]
     },
     "execution_count": 336,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "traversed_df = uniprot_to_bioprocess_df.copy()\n",
    "uniprot_to_bioprocess = recurse_bioprocesses(traversed_df, 2)\n",
    "proteomic_bp_to_model_merge = pd.merge(proteomic_to_model_df, uniprot_to_bioprocess, how = \"left\", on= \"uniprot_id\")\n",
    "proteomic_bp_to_model_pivot = pd.pivot_table(proteomic_bp_to_model_merge, values='protein_intensity', index=['model_id'],\n",
    "                   columns=['bioprocess_id'], aggfunc=\"sum\", fill_value=0.0).reset_index()\n",
    "merged_reduced_protein_data = pd.merge(merged_drug_model_df.rename(columns = {\"SANGER_MODEL_ID\": \"model_id\"}), proteomic_bp_to_model_pivot, on = \"model_id\", how = \"inner\")\n",
    "y = merged_reduced_protein_data.pop(\"LN_IC50\")\n",
    "merged_reduced_protein_data.pop(\"model_id\")\n",
    "merged_reduced_protein_data.pop(\"model_idx\")\n",
    "merged_reduced_protein_data.pop(\"MIN_CONC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "id": "37706c9d-b128-408b-bbae-33ca56bbd38a",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_index, test_index = train_test_split(np.arange(merged_reduced_protein_data.shape[0]), random_state = 42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "d531f9d8-056f-4baf-998f-47f710c66c0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_data = merged_reduced_protein_data.iloc[train_index, :]\n",
    "test_data = merged_reduced_protein_data.iloc[test_index, :]\n",
    "y_train = y.values[train_index]\n",
    "y_test = y.values[test_index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "919e9f24-72c0-4f0f-9449-a669dff8cd22",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import normalize\n",
    "y_train = normalize(np.expand_dims(y_train, axis = 0)).squeeze()\n",
    "y_test = normalize(np.expand_dims(y_test, axis = 0)).squeeze()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "id": "61b6c778-cbf9-498f-b16a-e107b14de6b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(174,)"
      ]
     },
     "execution_count": 340,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_test.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "id": "1bff6c1b-8802-449c-bdc5-bda355be1892",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn import preprocessing\n",
    "\n",
    "min_max_scaler = preprocessing.StandardScaler() #MinMaxScaler()\n",
    "train_data_scaled =  min_max_scaler.fit_transform(train_data) #normalize(train_data)\n",
    "test_data_scaled = min_max_scaler.transform(test_data) #normalize(test_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "id": "b908f052-4b04-49f7-a737-9c1ce7f65f4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(522, 3887)"
      ]
     },
     "execution_count": 342,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_data_scaled.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "id": "ff878674-1cd3-40af-96b5-c83c621b4316",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3887"
      ]
     },
     "execution_count": 343,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.stats import pearsonr\n",
    "\n",
    "min_var_features = []\n",
    "for i in range(train_data_scaled.shape[-1]):\n",
    "    rows = train_data_scaled[:, i]\n",
    "    corr_value = pearsonr(rows, y_train).statistic\n",
    "    if np.abs(corr_value) >= 1e-10:\n",
    "        min_var_features.append(i)\n",
    "len(min_var_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "id": "cd1db04d-9535-48ab-b3e3-56ce1fd4051d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-32 {color: black;background-color: white;}#sk-container-id-32 pre{padding: 0;}#sk-container-id-32 div.sk-toggleable {background-color: white;}#sk-container-id-32 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-32 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-32 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-32 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-32 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-32 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-32 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-32 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-32 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-32 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-32 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-32 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-32 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-32 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-32 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-32 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-32 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-32 div.sk-item {position: relative;z-index: 1;}#sk-container-id-32 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-32 div.sk-item::before, #sk-container-id-32 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-32 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-32 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-32 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-32 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-32 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-32 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-32 div.sk-label-container {text-align: center;}#sk-container-id-32 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-32 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-32\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>XGBRegressor(alpha=100, base_score=None, booster=None, callbacks=None,\n",
       "             colsample_bylevel=None, colsample_bynode=None,\n",
       "             colsample_bytree=None, device=None, early_stopping_rounds=None,\n",
       "             enable_categorical=False, eval_metric=None, feature_types=None,\n",
       "             gamma=None, grow_policy=None, importance_type=None,\n",
       "             interaction_constraints=None, learning_rate=None, max_bin=None,\n",
       "             max_cat_threshold=None, max_cat_to_onehot=None,\n",
       "             max_delta_step=None, max_depth=None, max_leaves=None,\n",
       "             min_child_weight=None, missing=nan, monotone_constraints=None,\n",
       "             multi_strategy=None, n_estimators=None, n_jobs=None,\n",
       "             num_parallel_tree=None, ...)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-60\" type=\"checkbox\" checked><label for=\"sk-estimator-id-60\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">XGBRegressor</label><div class=\"sk-toggleable__content\"><pre>XGBRegressor(alpha=100, base_score=None, booster=None, callbacks=None,\n",
       "             colsample_bylevel=None, colsample_bynode=None,\n",
       "             colsample_bytree=None, device=None, early_stopping_rounds=None,\n",
       "             enable_categorical=False, eval_metric=None, feature_types=None,\n",
       "             gamma=None, grow_policy=None, importance_type=None,\n",
       "             interaction_constraints=None, learning_rate=None, max_bin=None,\n",
       "             max_cat_threshold=None, max_cat_to_onehot=None,\n",
       "             max_delta_step=None, max_depth=None, max_leaves=None,\n",
       "             min_child_weight=None, missing=nan, monotone_constraints=None,\n",
       "             multi_strategy=None, n_estimators=None, n_jobs=None,\n",
       "             num_parallel_tree=None, ...)</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "XGBRegressor(alpha=100, base_score=None, booster=None, callbacks=None,\n",
       "             colsample_bylevel=None, colsample_bynode=None,\n",
       "             colsample_bytree=None, device=None, early_stopping_rounds=None,\n",
       "             enable_categorical=False, eval_metric=None, feature_types=None,\n",
       "             gamma=None, grow_policy=None, importance_type=None,\n",
       "             interaction_constraints=None, learning_rate=None, max_bin=None,\n",
       "             max_cat_threshold=None, max_cat_to_onehot=None,\n",
       "             max_delta_step=None, max_depth=None, max_leaves=None,\n",
       "             min_child_weight=None, missing=nan, monotone_constraints=None,\n",
       "             multi_strategy=None, n_estimators=None, n_jobs=None,\n",
       "             num_parallel_tree=None, ...)"
      ]
     },
     "execution_count": 344,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.linear_model import LinearRegression, Ridge\n",
    "from xgboost import XGBRegressor\n",
    "\n",
    "model = XGBRegressor\n",
    "lr = model(alpha = 100 )\n",
    "lr.fit(train_data_scaled, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "id": "b491586e-08d7-4373-811e-21a4562b41dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((174, 3887), (522, 3887))"
      ]
     },
     "execution_count": 345,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data_scaled.shape, train_data_scaled.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "id": "89aa0012-5c18-4b68-b53c-3fa1cb49ae27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(174,)"
      ]
     },
     "execution_count": 346,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_test.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "id": "90128ed2-4f66-4fe1-adaa-a6aa1da1cb89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-5.908118284558839e-05, 0.17803981637019706)"
      ]
     },
     "execution_count": 347,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.metrics import r2_score\n",
    "\n",
    "r2_score(y_test, lr.predict(test_data_scaled)), r2_score(y_train, lr.predict(train_data_scaled))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "id": "7da0671f-cd74-46b7-9e4e-e1c50f189781",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-33 {color: black;background-color: white;}#sk-container-id-33 pre{padding: 0;}#sk-container-id-33 div.sk-toggleable {background-color: white;}#sk-container-id-33 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-33 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-33 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-33 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-33 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-33 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-33 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-33 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-33 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-33 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-33 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-33 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-33 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-33 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-33 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-33 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-33 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-33 div.sk-item {position: relative;z-index: 1;}#sk-container-id-33 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-33 div.sk-item::before, #sk-container-id-33 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-33 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-33 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-33 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-33 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-33 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-33 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-33 div.sk-label-container {text-align: center;}#sk-container-id-33 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-33 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-33\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>GridSearchCV(cv=3,\n",
       "             estimator=XGBRegressor(base_score=None, booster=None,\n",
       "                                    callbacks=None, colsample_bylevel=None,\n",
       "                                    colsample_bynode=None,\n",
       "                                    colsample_bytree=None, device=None,\n",
       "                                    early_stopping_rounds=None,\n",
       "                                    enable_categorical=False, eval_metric=None,\n",
       "                                    feature_types=None, gamma=None,\n",
       "                                    grow_policy=None, importance_type=None,\n",
       "                                    interaction_constraints=None,\n",
       "                                    learning_rate=None, m...\n",
       "        11.,  12.,  13.,  14.,  15.,  16.,  17.,  18.,  19.,  20.,  21.,\n",
       "        22.,  23.,  24.,  25.,  26.,  27.,  28.,  29.,  30.,  31.,  32.,\n",
       "        33.,  34.,  35.,  36.,  37.,  38.,  39.,  40.,  41.,  42.,  43.,\n",
       "        44.,  45.,  46.,  47.,  48.,  49.,  50.,  51.,  52.,  53.,  54.,\n",
       "        55.,  56.,  57.,  58.,  59.,  60.,  61.,  62.,  63.,  64.,  65.,\n",
       "        66.,  67.,  68.,  69.,  70.,  71.,  72.,  73.,  74.,  75.,  76.,\n",
       "        77.,  78.,  79.,  80.,  81.,  82.,  83.,  84.,  85.,  86.,  87.,\n",
       "        88.,  89.,  90.,  91.,  92.,  93.,  94.,  95.,  96.,  97.,  98.,\n",
       "        99., 100.])},\n",
       "             scoring=&#x27;r2&#x27;)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-61\" type=\"checkbox\" ><label for=\"sk-estimator-id-61\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">GridSearchCV</label><div class=\"sk-toggleable__content\"><pre>GridSearchCV(cv=3,\n",
       "             estimator=XGBRegressor(base_score=None, booster=None,\n",
       "                                    callbacks=None, colsample_bylevel=None,\n",
       "                                    colsample_bynode=None,\n",
       "                                    colsample_bytree=None, device=None,\n",
       "                                    early_stopping_rounds=None,\n",
       "                                    enable_categorical=False, eval_metric=None,\n",
       "                                    feature_types=None, gamma=None,\n",
       "                                    grow_policy=None, importance_type=None,\n",
       "                                    interaction_constraints=None,\n",
       "                                    learning_rate=None, m...\n",
       "        11.,  12.,  13.,  14.,  15.,  16.,  17.,  18.,  19.,  20.,  21.,\n",
       "        22.,  23.,  24.,  25.,  26.,  27.,  28.,  29.,  30.,  31.,  32.,\n",
       "        33.,  34.,  35.,  36.,  37.,  38.,  39.,  40.,  41.,  42.,  43.,\n",
       "        44.,  45.,  46.,  47.,  48.,  49.,  50.,  51.,  52.,  53.,  54.,\n",
       "        55.,  56.,  57.,  58.,  59.,  60.,  61.,  62.,  63.,  64.,  65.,\n",
       "        66.,  67.,  68.,  69.,  70.,  71.,  72.,  73.,  74.,  75.,  76.,\n",
       "        77.,  78.,  79.,  80.,  81.,  82.,  83.,  84.,  85.,  86.,  87.,\n",
       "        88.,  89.,  90.,  91.,  92.,  93.,  94.,  95.,  96.,  97.,  98.,\n",
       "        99., 100.])},\n",
       "             scoring=&#x27;r2&#x27;)</pre></div></div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-62\" type=\"checkbox\" ><label for=\"sk-estimator-id-62\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">estimator: XGBRegressor</label><div class=\"sk-toggleable__content\"><pre>XGBRegressor(base_score=None, booster=None, callbacks=None,\n",
       "             colsample_bylevel=None, colsample_bynode=None,\n",
       "             colsample_bytree=None, device=None, early_stopping_rounds=None,\n",
       "             enable_categorical=False, eval_metric=None, feature_types=None,\n",
       "             gamma=None, grow_policy=None, importance_type=None,\n",
       "             interaction_constraints=None, learning_rate=None, max_bin=None,\n",
       "             max_cat_threshold=None, max_cat_to_onehot=None,\n",
       "             max_delta_step=None, max_depth=None, max_leaves=None,\n",
       "             min_child_weight=None, missing=nan, monotone_constraints=None,\n",
       "             multi_strategy=None, n_estimators=None, n_jobs=None,\n",
       "             num_parallel_tree=None, random_state=None, ...)</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-63\" type=\"checkbox\" ><label for=\"sk-estimator-id-63\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">XGBRegressor</label><div class=\"sk-toggleable__content\"><pre>XGBRegressor(base_score=None, booster=None, callbacks=None,\n",
       "             colsample_bylevel=None, colsample_bynode=None,\n",
       "             colsample_bytree=None, device=None, early_stopping_rounds=None,\n",
       "             enable_categorical=False, eval_metric=None, feature_types=None,\n",
       "             gamma=None, grow_policy=None, importance_type=None,\n",
       "             interaction_constraints=None, learning_rate=None, max_bin=None,\n",
       "             max_cat_threshold=None, max_cat_to_onehot=None,\n",
       "             max_delta_step=None, max_depth=None, max_leaves=None,\n",
       "             min_child_weight=None, missing=nan, monotone_constraints=None,\n",
       "             multi_strategy=None, n_estimators=None, n_jobs=None,\n",
       "             num_parallel_tree=None, random_state=None, ...)</pre></div></div></div></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "GridSearchCV(cv=3,\n",
       "             estimator=XGBRegressor(base_score=None, booster=None,\n",
       "                                    callbacks=None, colsample_bylevel=None,\n",
       "                                    colsample_bynode=None,\n",
       "                                    colsample_bytree=None, device=None,\n",
       "                                    early_stopping_rounds=None,\n",
       "                                    enable_categorical=False, eval_metric=None,\n",
       "                                    feature_types=None, gamma=None,\n",
       "                                    grow_policy=None, importance_type=None,\n",
       "                                    interaction_constraints=None,\n",
       "                                    learning_rate=None, m...\n",
       "        11.,  12.,  13.,  14.,  15.,  16.,  17.,  18.,  19.,  20.,  21.,\n",
       "        22.,  23.,  24.,  25.,  26.,  27.,  28.,  29.,  30.,  31.,  32.,\n",
       "        33.,  34.,  35.,  36.,  37.,  38.,  39.,  40.,  41.,  42.,  43.,\n",
       "        44.,  45.,  46.,  47.,  48.,  49.,  50.,  51.,  52.,  53.,  54.,\n",
       "        55.,  56.,  57.,  58.,  59.,  60.,  61.,  62.,  63.,  64.,  65.,\n",
       "        66.,  67.,  68.,  69.,  70.,  71.,  72.,  73.,  74.,  75.,  76.,\n",
       "        77.,  78.,  79.,  80.,  81.,  82.,  83.,  84.,  85.,  86.,  87.,\n",
       "        88.,  89.,  90.,  91.,  92.,  93.,  94.,  95.,  96.,  97.,  98.,\n",
       "        99., 100.])},\n",
       "             scoring='r2')"
      ]
     },
     "execution_count": 348,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from  sklearn.model_selection import GridSearchCV\n",
    "\n",
    "low = 0\n",
    "high = 100\n",
    "param_grid = {\n",
    "    \"alpha\":np.linspace(low, high, high - low +1 , endpoint=True)\n",
    "}\n",
    "grid_search = GridSearchCV(\n",
    "    model(),\n",
    "    n_jobs=-1,\n",
    "    param_grid = param_grid,\n",
    "    cv = 3,\n",
    "    scoring = \"r2\"\n",
    ")\n",
    "grid_search.fit(train_data_scaled[:, min_var_features], y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "id": "4d67158e-187b-413f-a860-bc836ac41d95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "({'alpha': 43.0}, 0.05970144058082919)"
      ]
     },
     "execution_count": 349,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grid_search.best_params_, grid_search.best_score_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "id": "fb13cbab-8308-4d80-a4cd-061b0a74c8cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-0.0536518819158609, 0.6456758936623965)"
      ]
     },
     "execution_count": 350,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r2_score(y_test, grid_search.predict(test_data_scaled[:, min_var_features])), r2_score(y_train, grid_search.predict(train_data_scaled[:, min_var_features]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "f3e32cf4-aabb-4932-9ce2-c9d54645b0e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Do we have some previous knowledge where we can exclude the relevance of specific biological processes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "5caed61a-7c33-493c-ae84-e11a9643c192",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_scaled_data(X, train_index, test_index, scaler = None):\n",
    "    X_train = X[train_index]\n",
    "    X_test = X[test_index]\n",
    "    if scaler is None: return X_train, X_test\n",
    "    X_train = scaler.fit_transform(X_train)\n",
    "    X_test = scaler.transform(X_test)\n",
    "    return X_train, X_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "ddb5de56-d187-416e-9876-a646be241e43",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = XGBRegressor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "id": "240d86a7-6afd-493b-b0ff-4276d11f185b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a9136c37e06b46239b885147dd1b5c5c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9c0b4980e6e34a73b8a86cb21a3a0535",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/3 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-9.072930378453874: {'hop': 1, 'scaler': MinMaxScaler()}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Process LokyProcess-338:\n",
      "Process LokyProcess-344:\n",
      "Process LokyProcess-352:\n",
      "Process LokyProcess-335:\n",
      "Process LokyProcess-353:\n",
      "Process LokyProcess-358:\n",
      "Process LokyProcess-334:\n",
      "Process LokyProcess-355:\n",
      "Process LokyProcess-359:\n",
      "Process LokyProcess-339:\n",
      "Process LokyProcess-341:\n",
      "Process LokyProcess-346:\n",
      "Process LokyProcess-330:\n",
      "Process LokyProcess-327:\n",
      "Process LokyProcess-326:\n",
      "Process LokyProcess-332:\n",
      "Process LokyProcess-331:\n",
      "Process LokyProcess-354:\n",
      "Process LokyProcess-363:\n",
      "Process LokyProcess-325:\n",
      "Process LokyProcess-321:\n",
      "Process LokyProcess-348:\n",
      "Process LokyProcess-360:\n",
      "Process LokyProcess-357:\n",
      "Process LokyProcess-345:\n",
      "Process LokyProcess-329:\n",
      "Process LokyProcess-362:\n",
      "Process LokyProcess-361:\n",
      "Process LokyProcess-347:\n",
      "Process LokyProcess-324:\n",
      "Process LokyProcess-364:\n",
      "Process LokyProcess-350:\n",
      "Process LokyProcess-351:\n",
      "Process LokyProcess-356:\n",
      "Process LokyProcess-340:\n",
      "Process LokyProcess-322:\n",
      "Process LokyProcess-343:\n",
      "Process LokyProcess-323:\n",
      "Process LokyProcess-342:\n",
      "Process LokyProcess-349:\n",
      "Process LokyProcess-336:\n",
      "Process LokyProcess-333:\n",
      "Process LokyProcess-328:\n",
      "Process LokyProcess-337:\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 466, in _process_worker\n",
      "    with worker_exit_lock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 466, in _process_worker\n",
      "    with worker_exit_lock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 466, in _process_worker\n",
      "    with worker_exit_lock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 466, in _process_worker\n",
      "    with worker_exit_lock:\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 466, in _process_worker\n",
      "    with worker_exit_lock:\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 466, in _process_worker\n",
      "    with worker_exit_lock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 466, in _process_worker\n",
      "    with worker_exit_lock:\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 466, in _process_worker\n",
      "    with worker_exit_lock:\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 466, in _process_worker\n",
      "    with worker_exit_lock:\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "Traceback (most recent call last):\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "Traceback (most recent call last):\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "Traceback (most recent call last):\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 465, in _process_worker\n",
      "    result_queue.put(pid)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "Traceback (most recent call last):\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "KeyboardInterrupt\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "Traceback (most recent call last):\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "Traceback (most recent call last):\n",
      "KeyboardInterrupt\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "KeyboardInterrupt\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "KeyboardInterrupt\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "Traceback (most recent call last):\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "Traceback (most recent call last):\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 620, in __call__\n",
      "    return self.func(*args, **kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/parallel.py\", line 288, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "KeyboardInterrupt\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py\", line 123, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 686, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "Traceback (most recent call last):\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "KeyboardInterrupt\n",
      "Traceback (most recent call last):\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "Traceback (most recent call last):\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1086, in fit\n",
      "    self._Booster = train(\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "Traceback (most recent call last):\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "KeyboardInterrupt\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 729, in inner_f\n",
      "    return func(**kwargs)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "Traceback (most recent call last):\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "Traceback (most recent call last):\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/training.py\", line 181, in train\n",
      "    bst.update(dtrain, i, obj)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "KeyboardInterrupt\n",
      "  File \"/home/dwalke/.local/lib/python3.10/site-packages/xgboost/core.py\", line 2050, in update\n",
      "    _LIB.XGBoosterUpdateOneIter(\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 428, in _process_worker\n",
      "    r = call_item()\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 275, in __call__\n",
      "    return self.fn(*self.args, **self.kwargs)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py\", line 625, in __call__\n",
      "    raise WorkerInterrupt() from e\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "joblib.my_exceptions.WorkerInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 192, in put\n",
      "    self._writer.send_bytes(obj)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/connection.py\", line 200, in send_bytes\n",
      "    self._send_bytes(m[offset:offset + size])\n",
      "  File \"/usr/lib/python3.10/multiprocessing/connection.py\", line 411, in _send_bytes\n",
      "    self._send(header + buf)\n",
      "  File \"/usr/lib/python3.10/multiprocessing/connection.py\", line 368, in _send\n",
      "    n = write(self._handle, buf)\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/usr/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "KeyboardInterrupt\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py\", line 431, in _process_worker\n",
      "    result_queue.put(_ResultItem(call_item.work_id, exception=exc))\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/queues.py\", line 191, in put\n",
      "    with self._wlock:\n",
      "  File \"/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/backend/synchronize.py\", line 112, in __enter__\n",
      "    return self._semlock.acquire()\n",
      "KeyboardInterrupt\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/parallel.py:975\u001b[0m, in \u001b[0;36mParallel.retrieve\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    974\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backend, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124msupports_timeout\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28;01mFalse\u001b[39;00m):\n\u001b[0;32m--> 975\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_output\u001b[38;5;241m.\u001b[39mextend(\u001b[43mjob\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m    976\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py:567\u001b[0m, in \u001b[0;36mLokyBackend.wrap_future_result\u001b[0;34m(future, timeout)\u001b[0m\n\u001b[1;32m    566\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 567\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfuture\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mresult\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    568\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m CfTimeoutError \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m/usr/lib/python3.10/concurrent/futures/_base.py:453\u001b[0m, in \u001b[0;36mFuture.result\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    451\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m__get_result()\n\u001b[0;32m--> 453\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_condition\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    455\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_state \u001b[38;5;129;01min\u001b[39;00m [CANCELLED, CANCELLED_AND_NOTIFIED]:\n",
      "File \u001b[0;32m/usr/lib/python3.10/threading.py:320\u001b[0m, in \u001b[0;36mCondition.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    319\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 320\u001b[0m     \u001b[43mwaiter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43macquire\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    321\u001b[0m     gotit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/sklearn/model_selection/_search.py:874\u001b[0m, in \u001b[0;36mBaseSearchCV.fit\u001b[0;34m(self, X, y, groups, **fit_params)\u001b[0m\n\u001b[1;32m    872\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m results\n\u001b[0;32m--> 874\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run_search\u001b[49m\u001b[43m(\u001b[49m\u001b[43mevaluate_candidates\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    876\u001b[0m \u001b[38;5;66;03m# multimetric is determined here because in the case of a callable\u001b[39;00m\n\u001b[1;32m    877\u001b[0m \u001b[38;5;66;03m# self.scoring the return type is only known after calling\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/sklearn/model_selection/_search.py:1388\u001b[0m, in \u001b[0;36mGridSearchCV._run_search\u001b[0;34m(self, evaluate_candidates)\u001b[0m\n\u001b[1;32m   1387\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Search all candidates in param_grid\"\"\"\u001b[39;00m\n\u001b[0;32m-> 1388\u001b[0m \u001b[43mevaluate_candidates\u001b[49m\u001b[43m(\u001b[49m\u001b[43mParameterGrid\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparam_grid\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/sklearn/model_selection/_search.py:821\u001b[0m, in \u001b[0;36mBaseSearchCV.fit.<locals>.evaluate_candidates\u001b[0;34m(candidate_params, cv, more_results)\u001b[0m\n\u001b[1;32m    814\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\n\u001b[1;32m    815\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFitting \u001b[39m\u001b[38;5;132;01m{0}\u001b[39;00m\u001b[38;5;124m folds for each of \u001b[39m\u001b[38;5;132;01m{1}\u001b[39;00m\u001b[38;5;124m candidates,\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    816\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m totalling \u001b[39m\u001b[38;5;132;01m{2}\u001b[39;00m\u001b[38;5;124m fits\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(\n\u001b[1;32m    817\u001b[0m             n_splits, n_candidates, n_candidates \u001b[38;5;241m*\u001b[39m n_splits\n\u001b[1;32m    818\u001b[0m         )\n\u001b[1;32m    819\u001b[0m     )\n\u001b[0;32m--> 821\u001b[0m out \u001b[38;5;241m=\u001b[39m \u001b[43mparallel\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    822\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdelayed\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_fit_and_score\u001b[49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    823\u001b[0m \u001b[43m        \u001b[49m\u001b[43mclone\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbase_estimator\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    824\u001b[0m \u001b[43m        \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    825\u001b[0m \u001b[43m        \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    826\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtrain\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtrain\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    827\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtest\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtest\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    828\u001b[0m \u001b[43m        \u001b[49m\u001b[43mparameters\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparameters\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    829\u001b[0m \u001b[43m        \u001b[49m\u001b[43msplit_progress\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43msplit_idx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_splits\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    830\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcandidate_progress\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mcand_idx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_candidates\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    831\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mfit_and_score_kwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    832\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    833\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[43mcand_idx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparameters\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[43msplit_idx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[43mtrain\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtest\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mproduct\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    834\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43menumerate\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mcandidate_params\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43menumerate\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mcv\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msplit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgroups\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    835\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    836\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    838\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(out) \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m1\u001b[39m:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/sklearn/utils/parallel.py:63\u001b[0m, in \u001b[0;36mParallel.__call__\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m     59\u001b[0m iterable_with_config \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m     60\u001b[0m     (_with_config(delayed_func, config), args, kwargs)\n\u001b[1;32m     61\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m delayed_func, args, kwargs \u001b[38;5;129;01min\u001b[39;00m iterable\n\u001b[1;32m     62\u001b[0m )\n\u001b[0;32m---> 63\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__call__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43miterable_with_config\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/parallel.py:1098\u001b[0m, in \u001b[0;36mParallel.__call__\u001b[0;34m(self, iterable)\u001b[0m\n\u001b[1;32m   1097\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backend\u001b[38;5;241m.\u001b[39mretrieval_context():\n\u001b[0;32m-> 1098\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mretrieve\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1099\u001b[0m \u001b[38;5;66;03m# Make sure that we get a last message telling us we are done\u001b[39;00m\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/parallel.py:997\u001b[0m, in \u001b[0;36mParallel.retrieve\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    996\u001b[0m     ensure_ready \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_managed_backend\n\u001b[0;32m--> 997\u001b[0m     \u001b[43mbackend\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mabort_everything\u001b[49m\u001b[43m(\u001b[49m\u001b[43mensure_ready\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mensure_ready\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    998\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py:586\u001b[0m, in \u001b[0;36mLokyBackend.abort_everything\u001b[0;34m(self, ensure_ready)\u001b[0m\n\u001b[1;32m    584\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Shutdown the workers and restart a new one with the same parameters\u001b[39;00m\n\u001b[1;32m    585\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m--> 586\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_workers\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mterminate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkill_workers\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m    587\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_workers \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/executor.py:74\u001b[0m, in \u001b[0;36mMemmappingExecutor.terminate\u001b[0;34m(self, kill_workers)\u001b[0m\n\u001b[1;32m     73\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mterminate\u001b[39m(\u001b[38;5;28mself\u001b[39m, kill_workers\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m):\n\u001b[0;32m---> 74\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshutdown\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkill_workers\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkill_workers\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     75\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m kill_workers:\n\u001b[1;32m     76\u001b[0m         \u001b[38;5;66;03m# When workers are killed in such a brutal manner, they cannot\u001b[39;00m\n\u001b[1;32m     77\u001b[0m         \u001b[38;5;66;03m# execute the finalizer of their shared memmaps. The refcount of\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     83\u001b[0m \n\u001b[1;32m     84\u001b[0m         \u001b[38;5;66;03m# unregister temporary resources from all contexts\u001b[39;00m\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py:1199\u001b[0m, in \u001b[0;36mProcessPoolExecutor.shutdown\u001b[0;34m(self, wait, kill_workers)\u001b[0m\n\u001b[1;32m   1198\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m executor_manager_thread \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m wait:\n\u001b[0;32m-> 1199\u001b[0m     \u001b[43mexecutor_manager_thread\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1201\u001b[0m \u001b[38;5;66;03m# To reduce the risk of opening too many files, remove references to\u001b[39;00m\n\u001b[1;32m   1202\u001b[0m \u001b[38;5;66;03m# objects that use file descriptors.\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3.10/threading.py:1096\u001b[0m, in \u001b[0;36mThread.join\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1095\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1096\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_wait_for_tstate_lock\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1097\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1098\u001b[0m     \u001b[38;5;66;03m# the behavior of a negative timeout isn't documented, but\u001b[39;00m\n\u001b[1;32m   1099\u001b[0m     \u001b[38;5;66;03m# historically .join(timeout=x) for x<0 has acted as if timeout=0\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3.10/threading.py:1116\u001b[0m, in \u001b[0;36mThread._wait_for_tstate_lock\u001b[0;34m(self, block, timeout)\u001b[0m\n\u001b[1;32m   1115\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1116\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mlock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43macquire\u001b[49m\u001b[43m(\u001b[49m\u001b[43mblock\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m   1117\u001b[0m         lock\u001b[38;5;241m.\u001b[39mrelease()\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[279], line 31\u001b[0m\n\u001b[1;32m     24\u001b[0m param_grid \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124malpha\u001b[39m\u001b[38;5;124m\"\u001b[39m:np\u001b[38;5;241m.\u001b[39mlinspace(low, high, high \u001b[38;5;241m-\u001b[39m low \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m1\u001b[39m , endpoint\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)}\n\u001b[1;32m     25\u001b[0m grid_search \u001b[38;5;241m=\u001b[39m GridSearchCV(\n\u001b[1;32m     26\u001b[0m     model(),\n\u001b[1;32m     27\u001b[0m     n_jobs\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m,\n\u001b[1;32m     28\u001b[0m     param_grid \u001b[38;5;241m=\u001b[39m param_grid,\n\u001b[1;32m     29\u001b[0m     cv \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m3\u001b[39m,\n\u001b[1;32m     30\u001b[0m     scoring \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mr2\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 31\u001b[0m \u001b[43mgrid_search\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX_train\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my_train\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     32\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m grid_search\u001b[38;5;241m.\u001b[39mbest_score_ \u001b[38;5;241m>\u001b[39m best_score:\n\u001b[1;32m     33\u001b[0m     best_score \u001b[38;5;241m=\u001b[39m grid_search\u001b[38;5;241m.\u001b[39mbest_score_\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/sklearn/model_selection/_search.py:803\u001b[0m, in \u001b[0;36mBaseSearchCV.fit\u001b[0;34m(self, X, y, groups, **fit_params)\u001b[0m\n\u001b[1;32m    792\u001b[0m fit_and_score_kwargs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m(\n\u001b[1;32m    793\u001b[0m     scorer\u001b[38;5;241m=\u001b[39mscorers,\n\u001b[1;32m    794\u001b[0m     fit_params\u001b[38;5;241m=\u001b[39mfit_params,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    800\u001b[0m     verbose\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mverbose,\n\u001b[1;32m    801\u001b[0m )\n\u001b[1;32m    802\u001b[0m results \u001b[38;5;241m=\u001b[39m {}\n\u001b[0;32m--> 803\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m parallel:\n\u001b[1;32m    804\u001b[0m     all_candidate_params \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m    805\u001b[0m     all_out \u001b[38;5;241m=\u001b[39m []\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/parallel.py:769\u001b[0m, in \u001b[0;36mParallel.__exit__\u001b[0;34m(self, exc_type, exc_value, traceback)\u001b[0m\n\u001b[1;32m    768\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__exit__\u001b[39m(\u001b[38;5;28mself\u001b[39m, exc_type, exc_value, traceback):\n\u001b[0;32m--> 769\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_terminate_backend\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    770\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_managed_backend \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/parallel.py:799\u001b[0m, in \u001b[0;36mParallel._terminate_backend\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    797\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_terminate_backend\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    798\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backend \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 799\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_backend\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mterminate\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/_parallel_backends.py:576\u001b[0m, in \u001b[0;36mLokyBackend.terminate\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    571\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mterminate\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    572\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_workers \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    573\u001b[0m         \u001b[38;5;66;03m# Don't terminate the workers as we want to reuse them in later\u001b[39;00m\n\u001b[1;32m    574\u001b[0m         \u001b[38;5;66;03m# calls, but cleanup the temporary resources that the Parallel call\u001b[39;00m\n\u001b[1;32m    575\u001b[0m         \u001b[38;5;66;03m# created. This 'hack' requires a private, low-level operation.\u001b[39;00m\n\u001b[0;32m--> 576\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_workers\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_temp_folder_manager\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_unlink_temporary_resources\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    577\u001b[0m \u001b[43m            \u001b[49m\u001b[43mcontext_id\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparallel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_id\u001b[49m\n\u001b[1;32m    578\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    579\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_workers \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    581\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreset_batch_stats()\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/_memmapping_reducer.py:631\u001b[0m, in \u001b[0;36mTemporaryResourcesManager._unlink_temporary_resources\u001b[0;34m(self, context_id)\u001b[0m\n\u001b[1;32m    627\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m filename \u001b[38;5;129;01min\u001b[39;00m os\u001b[38;5;241m.\u001b[39mlistdir(temp_folder):\n\u001b[1;32m    628\u001b[0m     resource_tracker\u001b[38;5;241m.\u001b[39mmaybe_unlink(\n\u001b[1;32m    629\u001b[0m         os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(temp_folder, filename), \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfile\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    630\u001b[0m     )\n\u001b[0;32m--> 631\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_try_delete_folder\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    632\u001b[0m \u001b[43m    \u001b[49m\u001b[43mallow_non_empty\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcontext_id\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcontext_id\u001b[49m\n\u001b[1;32m    633\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/_memmapping_reducer.py:658\u001b[0m, in \u001b[0;36mTemporaryResourcesManager._try_delete_folder\u001b[0;34m(self, allow_non_empty, context_id)\u001b[0m\n\u001b[1;32m    656\u001b[0m temp_folder \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cached_temp_folders[context_id]\n\u001b[1;32m    657\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 658\u001b[0m     \u001b[43mdelete_folder\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    659\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtemp_folder\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mallow_non_empty\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mallow_non_empty\u001b[49m\n\u001b[1;32m    660\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    661\u001b[0m     \u001b[38;5;66;03m# Now that this folder is deleted, we can forget about it\u001b[39;00m\n\u001b[1;32m    662\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_unregister_context(context_id)\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/joblib/disk.py:136\u001b[0m, in \u001b[0;36mdelete_folder\u001b[0;34m(folder_path, onerror, allow_non_empty)\u001b[0m\n\u001b[1;32m    131\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m err_count \u001b[38;5;241m>\u001b[39m RM_SUBDIRS_N_RETRY:\n\u001b[1;32m    132\u001b[0m         \u001b[38;5;66;03m# the folder cannot be deleted right now. It maybe\u001b[39;00m\n\u001b[1;32m    133\u001b[0m         \u001b[38;5;66;03m# because some temporary files have not been deleted\u001b[39;00m\n\u001b[1;32m    134\u001b[0m         \u001b[38;5;66;03m# yet.\u001b[39;00m\n\u001b[1;32m    135\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m\n\u001b[0;32m--> 136\u001b[0m \u001b[43mtime\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msleep\u001b[49m\u001b[43m(\u001b[49m\u001b[43mRM_SUBDIRS_RETRY_TIME\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import KFold\n",
    "from tqdm.notebook import tqdm\n",
    "splits = 3\n",
    "kf = KFold(n_splits=splits,shuffle=True)\n",
    "scalers = [preprocessing.MinMaxScaler(), preprocessing.StandardScaler(), None]\n",
    "res = np.zeros(splits)\n",
    "\n",
    "for i, (train_index, test_index) in tqdm(enumerate(kf.split(np.arange(696)))):\n",
    "    best_score = float(\"-inf\")\n",
    "    best_params = None\n",
    "    best_configs = dict({\"hop\": None, \"scaler\": None})\n",
    "    \n",
    "    for hop in tqdm(range(1, 4)):\n",
    "        y, X = get_aggregated_proteomic_data(hop)\n",
    "        y_train, y_test = y[train_index], y[test_index]\n",
    "        \n",
    "        for scaler in scalers:\n",
    "            X_train, X_test = get_scaled_data(X, train_index, test_index, scaler)\n",
    "            \n",
    "            ## Scale also y???\n",
    "            \n",
    "            low = 0\n",
    "            high = 100\n",
    "            param_grid = {\"alpha\":np.linspace(low, high, high - low +1 , endpoint=True)}\n",
    "            grid_search = GridSearchCV(\n",
    "                model(),\n",
    "                n_jobs=-1,\n",
    "                param_grid = param_grid,\n",
    "                cv = 3,\n",
    "                scoring = \"r2\")\n",
    "            grid_search.fit(X_train, y_train)\n",
    "            if grid_search.best_score_ > best_score:\n",
    "                best_score = grid_search.best_score_\n",
    "                best_params = grid_search.best_params_\n",
    "                best_configs[\"hop\"] = hop\n",
    "                best_configs[\"scaler\"] = scaler\n",
    "                print(f\"{best_score}: {best_configs}\")\n",
    "    y, X = get_aggregated_proteomic_data(best_configs[\"hop\"])\n",
    "    y_train, y_test = y[train_index], y[test_index]\n",
    "    X_train, X_test = get_scaled_data(X, train_index, test_index, best_configs[\"scaler\"])\n",
    "    model.fit(**best_params)\n",
    "    r2_results = r2_score(y_test, model.predict(X_test))\n",
    "    res[i] = r2_results\n",
    "print(f\"{res.mean()} +- {res.std()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "e11c58c3-5770-4b11-8939-c41f33127c01",
   "metadata": {},
   "outputs": [],
   "source": [
    "proteome_table = pd.pivot_table(proteomic_to_model_df, values='protein_intensity', index=['model_id'],\n",
    "                       columns=['uniprot_id'], aggfunc=\"sum\", fill_value=0.0)\n",
    "proteome_table_merged = pd.merge(merged_drug_model_df.rename(columns = {\"SANGER_MODEL_ID\": \"model_id\"}), proteome_table, on = \"model_id\", how = \"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "46d3deed-3e10-4a5b-807e-a526bce7c4fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model_id</th>\n",
       "      <th>MIN_CONC</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>model_idx</th>\n",
       "      <th>A0A075B6K4</th>\n",
       "      <th>A0A075B6N1</th>\n",
       "      <th>A0A075B7B8</th>\n",
       "      <th>A0A075B7D8</th>\n",
       "      <th>A0A087X0M5</th>\n",
       "      <th>A0A0B4J1V0</th>\n",
       "      <th>...</th>\n",
       "      <th>Q9Y6V0</th>\n",
       "      <th>Q9Y6V7</th>\n",
       "      <th>Q9Y6W3</th>\n",
       "      <th>Q9Y6W5</th>\n",
       "      <th>Q9Y6X4</th>\n",
       "      <th>Q9Y6X5</th>\n",
       "      <th>Q9Y6X8</th>\n",
       "      <th>Q9Y6X9</th>\n",
       "      <th>Q9Y6Y0</th>\n",
       "      <th>Q9Y6Y8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDM01132</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>5.078148</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.38499</td>\n",
       "      <td>2.478020</td>\n",
       "      <td>4.79677</td>\n",
       "      <td>3.66727</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>3.01557</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.66191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDM00848</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>2.849687</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.21332</td>\n",
       "      <td>2.067740</td>\n",
       "      <td>4.51982</td>\n",
       "      <td>4.56456</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.26890</td>\n",
       "      <td>3.67573</td>\n",
       "      <td>2.864420</td>\n",
       "      <td>3.72880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>5.814401</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.865560</td>\n",
       "      <td>4.07119</td>\n",
       "      <td>4.05313</td>\n",
       "      <td>3.01002</td>\n",
       "      <td>0.527221</td>\n",
       "      <td>3.18820</td>\n",
       "      <td>2.59367</td>\n",
       "      <td>-0.372063</td>\n",
       "      <td>2.70822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDM00269</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>3.390164</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.24159</td>\n",
       "      <td>-0.075127</td>\n",
       "      <td>4.24225</td>\n",
       "      <td>3.82222</td>\n",
       "      <td>3.15247</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.04894</td>\n",
       "      <td>3.046890</td>\n",
       "      <td>4.16562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDM00203</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>6.030742</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-1.130470</td>\n",
       "      <td>4.16679</td>\n",
       "      <td>3.75188</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.082380</td>\n",
       "      <td>2.79088</td>\n",
       "      <td>3.58533</td>\n",
       "      <td>2.074080</td>\n",
       "      <td>2.63864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691</th>\n",
       "      <td>SIDM00214</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>4.973348</td>\n",
       "      <td>712</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.63727</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.53678</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.337140</td>\n",
       "      <td>1.50818</td>\n",
       "      <td>2.81642</td>\n",
       "      <td>3.200490</td>\n",
       "      <td>3.73990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692</th>\n",
       "      <td>SIDM00194</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>3.812519</td>\n",
       "      <td>713</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.68525</td>\n",
       "      <td>-0.059671</td>\n",
       "      <td>4.56068</td>\n",
       "      <td>3.07724</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.236390</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.30044</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.94487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>693</th>\n",
       "      <td>SIDM00193</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>2.698528</td>\n",
       "      <td>714</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.40518</td>\n",
       "      <td>1.226040</td>\n",
       "      <td>4.79602</td>\n",
       "      <td>4.80620</td>\n",
       "      <td>2.41336</td>\n",
       "      <td>0.101013</td>\n",
       "      <td>2.80196</td>\n",
       "      <td>3.06613</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.91771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>694</th>\n",
       "      <td>SIDM00498</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>7.079313</td>\n",
       "      <td>715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.18342</td>\n",
       "      <td>1.997690</td>\n",
       "      <td>3.37983</td>\n",
       "      <td>3.75180</td>\n",
       "      <td>2.36091</td>\n",
       "      <td>3.161740</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>3.53633</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.46901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>695</th>\n",
       "      <td>SIDM00049</td>\n",
       "      <td>0.030016</td>\n",
       "      <td>1.573649</td>\n",
       "      <td>716</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.39436</td>\n",
       "      <td>4.35387</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.696030</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>2.09598</td>\n",
       "      <td>2.025090</td>\n",
       "      <td>3.58070</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>696 rows × 8461 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      model_id  MIN_CONC   LN_IC50  model_idx  A0A075B6K4  A0A075B6N1  \\\n",
       "0    SIDM01132  0.030016  5.078148          0         0.0         0.0   \n",
       "1    SIDM00848  0.030016  2.849687          1         0.0         0.0   \n",
       "2    SIDM00263  0.030016  5.814401          2         0.0         0.0   \n",
       "3    SIDM00269  0.030016  3.390164          3         0.0         0.0   \n",
       "4    SIDM00203  0.030016  6.030742          4         0.0         0.0   \n",
       "..         ...       ...       ...        ...         ...         ...   \n",
       "691  SIDM00214  0.030016  4.973348        712         0.0         0.0   \n",
       "692  SIDM00194  0.030016  3.812519        713         0.0         0.0   \n",
       "693  SIDM00193  0.030016  2.698528        714         0.0         0.0   \n",
       "694  SIDM00498  0.030016  7.079313        715         0.0         0.0   \n",
       "695  SIDM00049  0.030016  1.573649        716         0.0         0.0   \n",
       "\n",
       "     A0A075B7B8  A0A075B7D8  A0A087X0M5  A0A0B4J1V0  ...   Q9Y6V0    Q9Y6V7  \\\n",
       "0           0.0         0.0         0.0         0.0  ...  3.38499  2.478020   \n",
       "1           0.0         0.0         0.0         0.0  ...  3.21332  2.067740   \n",
       "2           0.0         0.0         0.0         0.0  ...  0.00000  1.865560   \n",
       "3           0.0         0.0         0.0         0.0  ...  3.24159 -0.075127   \n",
       "4           0.0         0.0         0.0         0.0  ...  0.00000 -1.130470   \n",
       "..          ...         ...         ...         ...  ...      ...       ...   \n",
       "691         0.0         0.0         0.0         0.0  ...  4.63727  0.000000   \n",
       "692         0.0         0.0         0.0         0.0  ...  5.68525 -0.059671   \n",
       "693         0.0         0.0         0.0         0.0  ...  4.40518  1.226040   \n",
       "694         0.0         0.0         0.0         0.0  ...  4.18342  1.997690   \n",
       "695         0.0         0.0         0.0         0.0  ...  0.00000  0.000000   \n",
       "\n",
       "      Q9Y6W3   Q9Y6W5   Q9Y6X4    Q9Y6X5   Q9Y6X8   Q9Y6X9    Q9Y6Y0   Q9Y6Y8  \n",
       "0    4.79677  3.66727  0.00000  0.000000  0.00000  3.01557  0.000000  3.66191  \n",
       "1    4.51982  4.56456  0.00000  0.000000  2.26890  3.67573  2.864420  3.72880  \n",
       "2    4.07119  4.05313  3.01002  0.527221  3.18820  2.59367 -0.372063  2.70822  \n",
       "3    4.24225  3.82222  3.15247  0.000000  0.00000  1.04894  3.046890  4.16562  \n",
       "4    4.16679  3.75188  0.00000  2.082380  2.79088  3.58533  2.074080  2.63864  \n",
       "..       ...      ...      ...       ...      ...      ...       ...      ...  \n",
       "691  0.00000  4.53678  0.00000  2.337140  1.50818  2.81642  3.200490  3.73990  \n",
       "692  4.56068  3.07724  0.00000  2.236390  0.00000  1.30044  0.000000  2.94487  \n",
       "693  4.79602  4.80620  2.41336  0.101013  2.80196  3.06613  0.000000  2.91771  \n",
       "694  3.37983  3.75180  2.36091  3.161740  0.00000  3.53633  0.000000  2.46901  \n",
       "695  1.39436  4.35387  0.00000  1.696030  0.00000  2.09598  2.025090  3.58070  \n",
       "\n",
       "[696 rows x 8461 columns]"
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteome_table_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "e324db44-948b-4b90-823b-4a10f5a8dcc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3446"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import torch\n",
    "import gc\n",
    "\n",
    "# Assume tensor is a CUDA tensor you no longer need\n",
    "del train_loader\n",
    "del test_loader\n",
    "\n",
    "# Empty the CUDA cache\n",
    "torch.cuda.empty_cache()\n",
    "\n",
    "# Run garbage collection\n",
    "gc.collect()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "023ac7bd-73e0-4684-9647-bdac796b69ad",
   "metadata": {},
   "source": [
    "## Log of experiments\n",
    "1) Selected genes with custom HetGNN -> Works (Data reading, processing and training works)\n",
    "2) Selected genes with custom HetGNN with batches -> Doesnt work (batching in graphs is problematic???)\n",
    "3) Selected genes with HeteroGATConv2 -> Overfits even with one head\n",
    "4) Selected genes + all proteins with custom HetGNN without batches -> Requires too much RAM\n",
    "5) Selected genes + all proteins with custom HetGNN with batches -> Exploding/Vanishing gradients due to missing normalization in aggregation (normalize based on edge degree?)\n",
    "6) Selected genes + all proteins with custom HeteroGatConv2 with batches -> Does not work (tried many configs and doesnt even overfit), either architecture or batches are problem\n",
    "7) Selected genes + selected proteins (low degree and high degree/high var edges) with custom HeteroGatConv2 with batches -> Does not work, either architecture or batches are problem\n",
    "8) Selected genes + selected proteins with HetGNN and single batch -> Works wih r2 about 0.1/0.2 but worse than only selected genes (due to overfit might companesate with dropout, wd, andearly stopping)\n",
    "9) Selected proteins with HetGNN and single batch -> (if this would work we can try ensembles of GNNs trained in different features) overfits even with high dropouts like 0.7\n",
    "10) Pearson based protein selection with 0.2 as threshold and 300 sparsity threshold, 1e-4 lr and 10000 epochs without genes -> r2 of  0.16 (try spearman corr or other thresholds ? ), 0.1 with 500, 1e-4 and 30_000 epochs returned 0.23\n",
    "11) Similar procedures for genes and already have 0.45 for 40_000 epochs and dropout 0.1\n",
    "\n",
    "Comparison against Feature selection of Johannes Schwerdt and Emanuel Lange? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62ad78d7-5b04-4294-8ce9-fd0b1a853669",
   "metadata": {},
   "outputs": [],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9f769e4-2c47-419a-9c40-836b3ac3e2a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[\"protein\"].x.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f78258b1-a794-4512-802b-aa14544b1a81",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e01ba659-c74e-4676-b386-2e8fbedd26af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_new_features(edge_index_dict, edge_attr_dict, x_dict):\n",
    "    message_dict = dict()\n",
    "    for key in edge_index_dict:\n",
    "        source, edge, target = key\n",
    "        edge_index = edge_index_dict[key]\n",
    "               \n",
    "        messages = edge_attr_dict[key] \n",
    "        aggregated_messages = torch.zeros((x_dict[target].shape[0], messages.shape[1]), dtype=x_dict[target].dtype)\n",
    "        \n",
    "        broadcasted_target_index = explicit_broadcast(edge_index[1], messages)\n",
    "        aggregated_messages.scatter_reduce_(0, broadcasted_target_index, messages, \"sum\")\n",
    "        message_dict[key] = aggregated_messages\n",
    "        \n",
    "    ## Message aggregation\n",
    "    message_aggr_dict = dict()\n",
    "    for key in message_dict:\n",
    "        source, edge, target = key\n",
    "        messages_for_target = message_dict[key]\n",
    "        previous_message_aggr_target = message_aggr_dict.get(target, None)\n",
    "        if previous_message_aggr_target is None:\n",
    "            message_aggr_dict[target] = messages_for_target\n",
    "        else:\n",
    "            message_aggr_dict[target] = torch.cat((previous_message_aggr_target, messages_for_target), dim = 1)\n",
    "    return message_aggr_dict[\"model\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b17c3f6b-e11d-4586-98cb-620ccbf739a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_data = get_new_features(data.edge_index_dict, data.edge_attr_dict, data.x_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76a24662-cdf3-4a5b-b345-6ab81e78ed66",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import normalize, StandardScaler\n",
    "##TODO change to train,. test val indices form above\n",
    "X_train, X_test , y_train, y_test = train_test_split(new_data, target_labels, random_state=0)\n",
    "# standard_scaler = StandardScaler()\n",
    "# standard_scaler.fit(X_train)\n",
    "\n",
    "# X_train, X_test = standard_scaler.transform(X_train), standard_scaler.transform(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17e8f50e-6df2-485d-ab4e-7de7793cdbb3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from xgboost import XGBRegressor\n",
    "from sklearn.datasets import make_regression\n",
    "from sklearn.metrics import make_scorer, r2_score\n",
    "\n",
    "# Define the XGBoost regressor\n",
    "xgb_reg = XGBRegressor()\n",
    "\n",
    "# Define the grid of hyperparameters to search\n",
    "param_grid = {\n",
    "    'n_estimators': [100, 200, 300, 400],\n",
    "    'max_depth': [3, 4, 5, 6],\n",
    "    'learning_rate': [0.01, 0.05, 0.1, 0.2],\n",
    "    'gamma': [0, 0.1, 0.2, 0.3],\n",
    "    'subsample': [0.6, 0.7, 0.8, 0.9, 1.0],\n",
    "    'colsample_bytree': [0.6, 0.7, 0.8, 0.9, 1.0],\n",
    "    'min_child_weight': [1, 5, 10, 20],\n",
    "    'reg_alpha': [0, 0.01, 0.1, 0.5, 1.0],\n",
    "    'reg_lambda': [0, 0.01, 0.1, 0.5, 1.0],\n",
    "}\n",
    "r2_scorer = make_scorer(r2_score)\n",
    "# Perform grid search with 5-fold cross-validation\n",
    "grid_search = GridSearchCV(estimator=xgb_reg, param_grid=param_grid, cv=3, scoring=r2_scorer, verbose=1, n_jobs = -1)\n",
    "grid_search.fit(X_train, y_train)\n",
    "\n",
    "# Print the best parameters and best score\n",
    "print(\"Best Parameters:\", grid_search.best_params_)\n",
    "print(\"Best Score:\", -grid_search.best_score_)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "f20243bf-f1bc-4e33-a3df-ffd4a6e35d00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'train_rmse': 1.4344861423667379,\n",
       " 'train_r2': 0.41209220806541835,\n",
       " 'test_rmse': 1.8167445670845757,\n",
       " 'test_r2': -0.014697180086989503}"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from xgboost import XGBRegressor\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.neural_network import MLPRegressor\n",
    "from sklearn.metrics import root_mean_squared_error, r2_score\n",
    "\n",
    "base_model = XGBRegressor(**grid_search.best_params_)\n",
    "base_model.fit(X_train, y_train)\n",
    "\n",
    "{\n",
    "    \"train_rmse\":root_mean_squared_error(y_train, base_model.predict(X_train)),\n",
    "    \"train_r2\":r2_score(y_train, base_model.predict(X_train)),\n",
    "    \"test_rmse\": root_mean_squared_error(y_test, base_model.predict(X_test)),\n",
    "    \"test_r2\":r2_score(y_test, base_model.predict(X_test))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73bdc7d3-ac9f-42b6-b3b1-bcdc15816843",
   "metadata": {},
   "source": [
    "## Test for Cross validaion index translator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "0862edff-256a-47d1-be03-1f857c9f351c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##############################\n",
      "tensor(True)\n",
      "##############################\n",
      "tensor(True)\n",
      "##############################\n",
      "tensor(True)\n",
      "##############################\n",
      "tensor(True)\n",
      "##############################\n",
      "tensor(True)\n"
     ]
    }
   ],
   "source": [
    "k_fold = KFold(n_splits=5)\n",
    "for (idx, val_idx) in k_fold.split(train_indices):\n",
    "    print(30*\"#\")\n",
    "    # print(train_indices[idx])\n",
    "    # print(train_indices[val_idx])\n",
    "\n",
    "    ## Test\n",
    "    sorted_k_folds = torch.sort(torch.cat([train_indices[idx], train_indices[val_idx]], 0))[0]\n",
    "    sorted_all_indices = torch.sort(train_indices)[0]\n",
    "    mathing_indices = (sorted_all_indices == sorted_k_folds).sum()\n",
    "    print(mathing_indices == train_indices.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95c28dd7-e1c6-49d5-8677-272914b12ad4",
   "metadata": {},
   "source": [
    "## Empty GPU memory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "71044706-08b2-47fe-b15b-1f8b3f38bc39",
   "metadata": {},
   "outputs": [],
   "source": [
    "del model, loss_fn, optim\n",
    "torch._C._cuda_clearCublasWorkspaces()\n",
    "torch._dynamo.reset()\n",
    "import gc\n",
    "gc.collect()\n",
    "torch.cuda.empty_cache()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51da2cd6-50dd-4452-a424-420ba8855517",
   "metadata": {},
   "source": [
    "## Base model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "9c372ef3-4f54-4350-8e42-aab34a0d85c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import normalize, StandardScaler\n",
    "##TODO change to train,. test val indices form above\n",
    "X_train, X_test , y_train, y_test = train_test_split(model_and_target_features, target_labels, random_state=0)\n",
    "standard_scaler = StandardScaler()\n",
    "standard_scaler.fit(X_train)\n",
    "\n",
    "X_train, X_test = standard_scaler.transform(X_train), standard_scaler.transform(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "e8486f18-132b-4486-9f14-216f54039bf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[5.5100e+02, 3.0016e-02, 3.3800e+02, ..., 4.0000e+00, 1.2000e+01,\n",
       "        0.0000e+00],\n",
       "       [2.4000e+01, 3.0016e-02, 7.1000e+01, ..., 4.0000e+00, 1.2000e+01,\n",
       "        1.0000e+00],\n",
       "       [1.5500e+02, 3.0016e-02, 7.1300e+02, ..., 4.0000e+00, 1.2000e+01,\n",
       "        1.0000e+00],\n",
       "       ...,\n",
       "       [6.3300e+02, 3.0016e-02, 6.1000e+02, ..., 4.0000e+00, 1.3000e+01,\n",
       "        1.0000e+00],\n",
       "       [6.3600e+02, 3.0016e-02, 9.4100e+02, ..., 4.0000e+00, 1.3000e+01,\n",
       "        1.0000e+00],\n",
       "       [1.3600e+02, 3.0016e-02, 7.3300e+02, ..., 4.0000e+00, 1.2000e+01,\n",
       "        1.0000e+00]])"
      ]
     },
     "execution_count": 280,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "57197f56-1227-4af7-a1c4-ee786473989f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dwalke/.local/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-12 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: black;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-12 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-12 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-12 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-12 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-12 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-12 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: block;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-12 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-12 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-12 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-12 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-12 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-12 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-12 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-12 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-12 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-12 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-12 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-12 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 1ex;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-12 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-12 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-12 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-12\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>MLPRegressor(hidden_layer_sizes=20)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-12\" type=\"checkbox\" checked><label for=\"sk-estimator-id-12\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">&nbsp;&nbsp;MLPRegressor<a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.4/modules/generated/sklearn.neural_network.MLPRegressor.html\">?<span>Documentation for MLPRegressor</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></label><div class=\"sk-toggleable__content fitted\"><pre>MLPRegressor(hidden_layer_sizes=20)</pre></div> </div></div></div></div>"
      ],
      "text/plain": [
       "MLPRegressor(hidden_layer_sizes=20)"
      ]
     },
     "execution_count": 277,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from xgboost import XGBRegressor\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.neural_network import MLPRegressor\n",
    "\n",
    "base_model = MLPRegressor(hidden_layer_sizes=(20))\n",
    "base_model.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "9d6d301c-7e81-4eaa-adf3-df049be58a64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'train_rmse': 0.630041661641547,\n",
       " 'train_r2': 0.2358449424336172,\n",
       " 'test_rmse': 0.7127170857289105,\n",
       " 'test_r2': 0.036665393902286736}"
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.metrics import root_mean_squared_error, r2_score\n",
    "{\n",
    "    \"train_rmse\":root_mean_squared_error(y_train, base_model.predict(X_train)),\n",
    "    \"train_r2\":r2_score(y_train, base_model.predict(X_train)),\n",
    "    \"test_rmse\": root_mean_squared_error(y_test, base_model.predict(X_test)),\n",
    "    \"test_r2\":r2_score(y_test, base_model.predict(X_test))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e463cb08-96db-4a77-800c-fb61129bb8b8",
   "metadata": {},
   "source": [
    "## Old code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46b6248f-3983-4aaf-90d3-838a7979aa8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "raise Exception(\"Keep cell outputs here\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "06462886-a61d-42f9-9286-47262e1e80e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>NLME_RESULT_ID</th>\n",
       "      <th>NLME_CURVE_ID</th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>TCGA_DESC</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PUTATIVE_TARGET</th>\n",
       "      <th>...</th>\n",
       "      <th>2175</th>\n",
       "      <th>2177</th>\n",
       "      <th>2359</th>\n",
       "      <th>2360</th>\n",
       "      <th>2361</th>\n",
       "      <th>2362</th>\n",
       "      <th>2438</th>\n",
       "      <th>2439</th>\n",
       "      <th>2498</th>\n",
       "      <th>2499</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18945558</td>\n",
       "      <td>683667</td>\n",
       "      <td>PFSK-1</td>\n",
       "      <td>SIDM01132</td>\n",
       "      <td>MB</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.366388</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.846626</td>\n",
       "      <td>10.613759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18945796</td>\n",
       "      <td>684052</td>\n",
       "      <td>A673</td>\n",
       "      <td>SIDM00848</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.723942</td>\n",
       "      <td>5.240599</td>\n",
       "      <td>4.432934</td>\n",
       "      <td>4.716167</td>\n",
       "      <td>4.285911</td>\n",
       "      <td>7.793623</td>\n",
       "      <td>9.156372</td>\n",
       "      <td>7.568813</td>\n",
       "      <td>7.063871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18946078</td>\n",
       "      <td>684057</td>\n",
       "      <td>ES5</td>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>...</td>\n",
       "      <td>4.879849</td>\n",
       "      <td>1.901766</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18946335</td>\n",
       "      <td>684059</td>\n",
       "      <td>ES7</td>\n",
       "      <td>SIDM00269</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.133421</td>\n",
       "      <td>4.578782</td>\n",
       "      <td>3.893851</td>\n",
       "      <td>3.137208</td>\n",
       "      <td>2.719145</td>\n",
       "      <td>8.871504</td>\n",
       "      <td>9.302207</td>\n",
       "      <td>6.655762</td>\n",
       "      <td>9.441838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18946617</td>\n",
       "      <td>684062</td>\n",
       "      <td>EW-11</td>\n",
       "      <td>SIDM00203</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.880307</td>\n",
       "      <td>5.137711</td>\n",
       "      <td>4.863771</td>\n",
       "      <td>5.007426</td>\n",
       "      <td>5.389040</td>\n",
       "      <td>10.462131</td>\n",
       "      <td>9.801610</td>\n",
       "      <td>8.228663</td>\n",
       "      <td>10.817398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242031</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>19187490</td>\n",
       "      <td>1659928</td>\n",
       "      <td>SNU-175</td>\n",
       "      <td>SIDM00216</td>\n",
       "      <td>COREAD</td>\n",
       "      <td>2499</td>\n",
       "      <td>N-acetyl cysteine</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.310061</td>\n",
       "      <td>5.423870</td>\n",
       "      <td>5.397164</td>\n",
       "      <td>5.401524</td>\n",
       "      <td>4.973343</td>\n",
       "      <td>9.264933</td>\n",
       "      <td>9.211173</td>\n",
       "      <td>7.445385</td>\n",
       "      <td>10.134495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242032</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>19187943</td>\n",
       "      <td>1660034</td>\n",
       "      <td>SNU-407</td>\n",
       "      <td>SIDM00214</td>\n",
       "      <td>COREAD</td>\n",
       "      <td>2499</td>\n",
       "      <td>N-acetyl cysteine</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.628609</td>\n",
       "      <td>5.042005</td>\n",
       "      <td>5.055076</td>\n",
       "      <td>4.918666</td>\n",
       "      <td>4.122141</td>\n",
       "      <td>11.121460</td>\n",
       "      <td>9.528520</td>\n",
       "      <td>8.344004</td>\n",
       "      <td>8.575555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242033</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>19188201</td>\n",
       "      <td>1660035</td>\n",
       "      <td>SNU-61</td>\n",
       "      <td>SIDM00194</td>\n",
       "      <td>COREAD</td>\n",
       "      <td>2499</td>\n",
       "      <td>N-acetyl cysteine</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.866949</td>\n",
       "      <td>6.130028</td>\n",
       "      <td>5.649833</td>\n",
       "      <td>4.392357</td>\n",
       "      <td>4.382946</td>\n",
       "      <td>10.879932</td>\n",
       "      <td>10.092187</td>\n",
       "      <td>7.743591</td>\n",
       "      <td>10.520666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242034</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>19188741</td>\n",
       "      <td>1674021</td>\n",
       "      <td>SNU-C5</td>\n",
       "      <td>SIDM00498</td>\n",
       "      <td>COREAD</td>\n",
       "      <td>2499</td>\n",
       "      <td>N-acetyl cysteine</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.653533</td>\n",
       "      <td>6.151265</td>\n",
       "      <td>6.022426</td>\n",
       "      <td>5.661197</td>\n",
       "      <td>4.820414</td>\n",
       "      <td>9.652950</td>\n",
       "      <td>10.181355</td>\n",
       "      <td>7.305220</td>\n",
       "      <td>10.701430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242035</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>19189023</td>\n",
       "      <td>1789883</td>\n",
       "      <td>DiFi</td>\n",
       "      <td>SIDM00049</td>\n",
       "      <td>COREAD</td>\n",
       "      <td>2499</td>\n",
       "      <td>N-acetyl cysteine</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.537148</td>\n",
       "      <td>5.757955</td>\n",
       "      <td>5.714278</td>\n",
       "      <td>5.576271</td>\n",
       "      <td>5.645880</td>\n",
       "      <td>10.544895</td>\n",
       "      <td>9.265865</td>\n",
       "      <td>8.275000</td>\n",
       "      <td>10.038769</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>242036 rows × 314 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       DATASET  NLME_RESULT_ID  NLME_CURVE_ID  COSMIC_ID CELL_LINE_NAME  \\\n",
       "0        GDSC2             401       18945558     683667         PFSK-1   \n",
       "1        GDSC2             401       18945796     684052           A673   \n",
       "2        GDSC2             401       18946078     684057            ES5   \n",
       "3        GDSC2             401       18946335     684059            ES7   \n",
       "4        GDSC2             401       18946617     684062          EW-11   \n",
       "...        ...             ...            ...        ...            ...   \n",
       "242031   GDSC2             401       19187490    1659928        SNU-175   \n",
       "242032   GDSC2             401       19187943    1660034        SNU-407   \n",
       "242033   GDSC2             401       19188201    1660035         SNU-61   \n",
       "242034   GDSC2             401       19188741    1674021         SNU-C5   \n",
       "242035   GDSC2             401       19189023    1789883           DiFi   \n",
       "\n",
       "       SANGER_MODEL_ID     TCGA_DESC  DRUG_ID          DRUG_NAME  \\\n",
       "0            SIDM01132            MB     1003       Camptothecin   \n",
       "1            SIDM00848  UNCLASSIFIED     1003       Camptothecin   \n",
       "2            SIDM00263  UNCLASSIFIED     1003       Camptothecin   \n",
       "3            SIDM00269  UNCLASSIFIED     1003       Camptothecin   \n",
       "4            SIDM00203  UNCLASSIFIED     1003       Camptothecin   \n",
       "...                ...           ...      ...                ...   \n",
       "242031       SIDM00216        COREAD     2499  N-acetyl cysteine   \n",
       "242032       SIDM00214        COREAD     2499  N-acetyl cysteine   \n",
       "242033       SIDM00194        COREAD     2499  N-acetyl cysteine   \n",
       "242034       SIDM00498        COREAD     2499  N-acetyl cysteine   \n",
       "242035       SIDM00049        COREAD     2499  N-acetyl cysteine   \n",
       "\n",
       "       PUTATIVE_TARGET  ...      2175      2177      2359      2360      2361  \\\n",
       "0                 TOP1  ...       NaN  2.366388       NaN       NaN       NaN   \n",
       "1                 TOP1  ...       NaN  1.723942  5.240599  4.432934  4.716167   \n",
       "2                 TOP1  ...  4.879849  1.901766       NaN       NaN       NaN   \n",
       "3                 TOP1  ...       NaN  2.133421  4.578782  3.893851  3.137208   \n",
       "4                 TOP1  ...       NaN  1.880307  5.137711  4.863771  5.007426   \n",
       "...                ...  ...       ...       ...       ...       ...       ...   \n",
       "242031      Metabolism  ...       NaN  1.310061  5.423870  5.397164  5.401524   \n",
       "242032      Metabolism  ...       NaN  2.628609  5.042005  5.055076  4.918666   \n",
       "242033      Metabolism  ...       NaN  2.866949  6.130028  5.649833  4.392357   \n",
       "242034      Metabolism  ...       NaN  2.653533  6.151265  6.022426  5.661197   \n",
       "242035      Metabolism  ...       NaN  1.537148  5.757955  5.714278  5.576271   \n",
       "\n",
       "            2362       2438       2439      2498       2499  \n",
       "0            NaN        NaN        NaN  7.846626  10.613759  \n",
       "1       4.285911   7.793623   9.156372  7.568813   7.063871  \n",
       "2            NaN        NaN        NaN       NaN        NaN  \n",
       "3       2.719145   8.871504   9.302207  6.655762   9.441838  \n",
       "4       5.389040  10.462131   9.801610  8.228663  10.817398  \n",
       "...          ...        ...        ...       ...        ...  \n",
       "242031  4.973343   9.264933   9.211173  7.445385  10.134495  \n",
       "242032  4.122141  11.121460   9.528520  8.344004   8.575555  \n",
       "242033  4.382946  10.879932  10.092187  7.743591  10.520666  \n",
       "242034  4.820414   9.652950  10.181355  7.305220  10.701430  \n",
       "242035  5.645880  10.544895   9.265865  8.275000  10.038769  \n",
       "\n",
       "[242036 rows x 314 columns]"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pivot_targets = pd.pivot_table(targets, values=\"LN_IC50\", index=[\"SANGER_MODEL_ID\"],\n",
    "#                        columns=['DRUG_ID'])\n",
    "\n",
    "# pd.merge(targets.loc[:, [\"SANGER_MODEL_ID\",'PUTATIVE_TARGET', 'PATHWAY_NAME', 'MIN_CONC', 'MAX_CONC']], pivot_targets, on=[\"SANGER_MODEL_ID\",\"SANGER_MODEL_ID\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "6b7bc938-68ed-4f07-8e88-4124ed0a26f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>1003</th>\n",
       "      <th>1004</th>\n",
       "      <th>1005</th>\n",
       "      <th>1006</th>\n",
       "      <th>1007</th>\n",
       "      <th>1008</th>\n",
       "      <th>1009</th>\n",
       "      <th>1010</th>\n",
       "      <th>1011</th>\n",
       "      <th>1012</th>\n",
       "      <th>...</th>\n",
       "      <th>2175</th>\n",
       "      <th>2177</th>\n",
       "      <th>2359</th>\n",
       "      <th>2360</th>\n",
       "      <th>2361</th>\n",
       "      <th>2362</th>\n",
       "      <th>2438</th>\n",
       "      <th>2439</th>\n",
       "      <th>2498</th>\n",
       "      <th>2499</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SIDM00003</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00023</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00040</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00041</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM00042</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM01248</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM01251</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM01259</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM01261</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIDM01265</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>969 rows × 295 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "DRUG_ID          1003   1004   1005   1006  1007   1008   1009  1010  1011  \\\n",
       "SANGER_MODEL_ID                                                              \n",
       "SIDM00003        True   True   True   True  True  False  False  True  True   \n",
       "SIDM00023        True  False  False  False  True   True   True  True  True   \n",
       "SIDM00040        True  False  False  False  True   True   True  True  True   \n",
       "SIDM00041        True  False  False  False  True   True   True  True  True   \n",
       "SIDM00042        True   True   True   True  True   True  False  True  True   \n",
       "...               ...    ...    ...    ...   ...    ...    ...   ...   ...   \n",
       "SIDM01248        True   True   True   True  True   True  False  True  True   \n",
       "SIDM01251        True   True   True   True  True  False  False  True  True   \n",
       "SIDM01259        True   True   True   True  True   True  False  True  True   \n",
       "SIDM01261        True  False  False  False  True   True   True  True  True   \n",
       "SIDM01265        True   True   True   True  True   True  False  True  True   \n",
       "\n",
       "DRUG_ID          1012  ...   2175   2177   2359   2360   2361   2362   2438  \\\n",
       "SANGER_MODEL_ID        ...                                                    \n",
       "SIDM00003        True  ...  False  False  False  False  False  False  False   \n",
       "SIDM00023        True  ...   True   True  False  False  False  False  False   \n",
       "SIDM00040        True  ...   True   True  False  False  False  False  False   \n",
       "SIDM00041        True  ...   True   True  False  False  False  False  False   \n",
       "SIDM00042        True  ...  False   True   True   True   True   True   True   \n",
       "...               ...  ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "SIDM01248        True  ...  False   True   True   True   True   True   True   \n",
       "SIDM01251        True  ...  False   True   True   True   True   True   True   \n",
       "SIDM01259        True  ...  False   True   True   True   True   True   True   \n",
       "SIDM01261        True  ...   True   True  False  False  False  False  False   \n",
       "SIDM01265        True  ...  False   True   True   True   True   True   True   \n",
       "\n",
       "DRUG_ID           2439   2498   2499  \n",
       "SANGER_MODEL_ID                       \n",
       "SIDM00003        False  False  False  \n",
       "SIDM00023        False  False  False  \n",
       "SIDM00040        False  False  False  \n",
       "SIDM00041        False  False  False  \n",
       "SIDM00042         True   True   True  \n",
       "...                ...    ...    ...  \n",
       "SIDM01248         True   True   True  \n",
       "SIDM01251         True   True   True  \n",
       "SIDM01259         True   True   True  \n",
       "SIDM01261        False  False  False  \n",
       "SIDM01265         True   True   True  \n",
       "\n",
       "[969 rows x 295 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ~pivot_targets.isnull()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
