{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "11c8453c-0745-4f96-8224-1ae89f394340",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dwalke/.local/lib/python3.10/site-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.\n",
      "  _torch_pytree._register_pytree_node(\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from bs4 import BeautifulSoup\n",
    "import numpy as np\n",
    "from rdkit import Chem\n",
    "from rdkit.Chem.rdmolops import GetAdjacencyMatrix\n",
    "import torch\n",
    "from torch_geometric.data import Data\n",
    "from torch.utils.data import DataLoader\n",
    "from tqdm.notebook import tqdm\n",
    "import os\n",
    "import torch\n",
    "from torch_geometric.nn import GCNConv, GATConv\n",
    "from sklearn.model_selection import train_test_split\n",
    "from torch_geometric.loader import DataLoader\n",
    "from sklearn.metrics import r2_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e58536d8-2bad-453c-8b45-6ac30b3f3263",
   "metadata": {},
   "outputs": [],
   "source": [
    "models = pd.read_csv(\"cell_models.csv\")\n",
    "targets = pd.read_csv(\"target.csv\",sep=\";\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "542d66c9-40c7-41b4-81b1-3ffdd8233d7d",
   "metadata": {},
   "source": [
    "## 1. Get SMILES\n",
    "\n",
    "## Drug Id of GDSC2  to PubChem Id via their online table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "41e6c5f5-f99a-4a9d-b525-80f033ea43c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = []\n",
    "drug_table_html = '<table class=\"sorted-table before filter export_csv_tab_xls paginate_10_x25_50_all zebra-table headers_rotated\"><thead><tr><th class=\"header\">drug_id</th><th class=\"header\">drug_name</th><th class=\"header\">synonyms</th><th class=\"header\">pathway_name</th><th class=\"header\">targets</th><th class=\"header\">pubchem</th></tr></thead><tbody><tr class=\"even  even\"><td class=\"r\">1259</td><td class=\"r\">Talazoparib</td><td class=\"r\">BMN-673, BMN 973</td><td class=\"r\">Genome integrity</td><td class=\"r\">PARP1, PARP2</td><td class=\"r\">44819241</td></tr><tr class=\"odd  odd\"><td class=\"r\">1372</td><td class=\"r\">Trametinib</td><td class=\"r\">GSK1120212, Mekinist</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">MEK1, MEK2</td><td class=\"r\">11707110</td></tr><tr class=\"even  even\"><td class=\"r\">1559</td><td class=\"r\">Luminespib</td><td class=\"r\">AUY922, VER-52296,NVP-AUY922,  AUY</td><td class=\"r\">Protein stability and degradation</td><td class=\"r\">HSP90</td><td class=\"r\">10096043</td></tr><tr class=\"odd  odd\"><td class=\"r\">1615</td><td class=\"r\">CZC24832</td><td class=\"r\">GTPL6653</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">PI3Kgamma</td><td class=\"r\">42623951</td></tr><tr class=\"even  even\"><td class=\"r\">1620</td><td class=\"r\">PFI3</td><td class=\"r\">PFI-3, PFI 3, AOB2221</td><td class=\"r\">Chromatin other</td><td class=\"r\">Polybromo 1, SMARCA4, SMARCA2</td><td class=\"r\">78243717</td></tr><tr class=\"odd  odd\"><td class=\"r\">1622</td><td class=\"r\">Wnt-C59</td><td class=\"r\">AOB3540, AK161197</td><td class=\"r\">WNT signaling</td><td class=\"r\">PORCN</td><td class=\"r\">57519544</td></tr><tr class=\"even  even\"><td class=\"r\">1626</td><td class=\"r\">OTX015</td><td class=\"r\">Birabresib, OTX-015, MK-8628</td><td class=\"r\">Chromatin other</td><td class=\"r\">BRD2, BRD3, BRD4</td><td class=\"r\">9936746</td></tr><tr class=\"odd  odd\"><td class=\"r\">1627</td><td class=\"r\">GSK343</td><td class=\"r\">GTPL8240, AOB3680, AK175558</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">EZH2</td><td class=\"r\">71268957</td></tr><tr class=\"even  even\"><td class=\"r\">1634</td><td class=\"r\">AGI-6780</td><td class=\"r\">AK162157</td><td class=\"r\">Metabolism</td><td class=\"r\">IDH2 R140Q mutant</td><td class=\"r\">71299339</td></tr><tr class=\"odd  odd\"><td class=\"r\">1854</td><td class=\"r\">MN-64</td><td class=\"r\">-</td><td class=\"r\">WNT signaling</td><td class=\"r\">TNKS1, TNKS2</td><td class=\"r\">2802462</td></tr><tr class=\"even  even\"><td class=\"r\">1909</td><td class=\"r\">Venetoclax</td><td class=\"r\">ABT-199, Veneclexta, GDC-0199</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">BCL2</td><td class=\"r\">49846579</td></tr><tr class=\"odd  odd\"><td class=\"r\">1084</td><td class=\"r\">Rapamycin</td><td class=\"r\">AY-22989, Sirolimus, WY-090217, Torisel, Rapamune</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">MTORC1</td><td class=\"r\">5384616</td></tr><tr class=\"even  even\"><td class=\"r\">1131</td><td class=\"r\">PRIMA-1MET</td><td class=\"r\">Prima 1MET, APR-246</td><td class=\"r\">p53 pathway</td><td class=\"r\">TP53 activation</td><td class=\"r\">52918385</td></tr><tr class=\"odd  odd\"><td class=\"r\">1584</td><td class=\"r\">BDOCA000347a</td><td class=\"r\">4-bromo-N-(6-methoxy-1,3-dimethyl-2-oxo-2,3-dihydro-1H-benzo[d]imidazol-5-yl)-2-methylbenzenesulfonamide</td><td class=\"r\">Chromatin histone acetylation</td><td class=\"r\">BRPF1, BRPF2, BRPF3</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1716</td><td class=\"r\">IRAK4_4710</td><td class=\"r\">SN1050852374, IRAK4_4710</td><td class=\"r\">Other, kinases</td><td class=\"r\">IRAK4</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1812</td><td class=\"r\">Bleomycin</td><td class=\"r\">-</td><td class=\"r\">DNA replication</td><td class=\"r\">dsDNA break induction</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1817</td><td class=\"r\">Romidepsin</td><td class=\"r\">Istodax, FK228, Depsipeptide</td><td class=\"r\">Chromatin histone acetylation</td><td class=\"r\">HDAC1, HDAC2, HDAC3, HDAC8</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1834</td><td class=\"r\">743380</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1913</td><td class=\"r\">AGI-5198</td><td class=\"r\">IDH-C35</td><td class=\"r\">Metabolism</td><td class=\"r\">IDH1 (R132H)</td><td class=\"r\">56645356</td></tr><tr class=\"odd  odd\"><td class=\"r\">1915</td><td class=\"r\">AZD3759</td><td class=\"r\">-</td><td class=\"r\">EGFR signaling</td><td class=\"r\">EGFR</td><td class=\"r\">78209992</td></tr><tr class=\"even  even\"><td class=\"r\">1932</td><td class=\"r\">NVP-ADW742</td><td class=\"r\">NVP ADW742, NVPADW742</td><td class=\"r\">IGF1R signaling</td><td class=\"r\">IGF1R</td><td class=\"r\">9825149</td></tr><tr class=\"odd  odd\"><td class=\"r\">1933</td><td class=\"r\">P22077</td><td class=\"r\">P 22077, P-22077</td><td class=\"r\">Protein stability and degradation</td><td class=\"r\">USP7, USP47</td><td class=\"r\">46931953</td></tr><tr class=\"even  even\"><td class=\"r\">1997</td><td class=\"r\">WEHI-539</td><td class=\"r\">WEHI539, WEHI-539</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">BCL-XL</td><td class=\"r\">71297207</td></tr><tr class=\"odd  odd\"><td class=\"r\">2011</td><td class=\"r\">ICL-SIRT078</td><td class=\"r\">-</td><td class=\"r\">Other</td><td class=\"r\">SIR2</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2038</td><td class=\"r\">UNC0638</td><td class=\"r\">-</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">G9A, GLP</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">2040</td><td class=\"r\">Foretinib</td><td class=\"r\">GSK1363089, XL-880, EXEL-2880, GSK089</td><td class=\"r\">RTK signaling</td><td class=\"r\">MET, KDR, TIE2, VEGFR3/FLT4, RON, PDGFR, FGFR1, EGFR</td><td class=\"r\">42642645</td></tr><tr class=\"even  even\"><td class=\"r\">2106</td><td class=\"r\">Uprosertib</td><td class=\"r\">GSK2141795, GSK2141795C, GSK-2141795</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">AKT1, AKT2, AKT3</td><td class=\"r\">51042438</td></tr><tr class=\"odd  odd\"><td class=\"r\">2154</td><td class=\"r\">GSK-LSD1-2HCl </td><td class=\"r\">-</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">LSD1</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2159</td><td class=\"r\">UNC0379</td><td class=\"r\">-</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">SETD8</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">2360</td><td class=\"r\">THR-101</td><td class=\"r\">WIMM synthesis</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">Mutant RAS</td><td class=\"r\">None</td></tr><tr class=\"even  even\"><td class=\"r\">2439</td><td class=\"r\">glutathione</td><td class=\"r\">G6013, sigma</td><td class=\"r\">Other</td><td class=\"r\">anti-oxidant proteins</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1008</td><td class=\"r\">Methotrexate</td><td class=\"r\">Abitrexate, Amethopterin, Rheumatrex, Trexall, Folex</td><td class=\"r\">DNA replication</td><td class=\"r\">Antimetabolite</td><td class=\"r\">126941</td></tr><tr class=\"even  even\"><td class=\"r\">1017</td><td class=\"r\">Olaparib</td><td class=\"r\">AZD2281, KU0059436,  Lynparza</td><td class=\"r\">Genome integrity</td><td class=\"r\">PARP1, PARP2</td><td class=\"r\">23725625</td></tr><tr class=\"odd  odd\"><td class=\"r\">1019</td><td class=\"r\">Bosutinib</td><td class=\"r\">SKI-606, Bosulif</td><td class=\"r\">Other, kinases</td><td class=\"r\">SRC, ABL, TEC</td><td class=\"r\">5328940</td></tr><tr class=\"even  even\"><td class=\"r\">1021</td><td class=\"r\">Axitinib</td><td class=\"r\">AG-13736, Inlyta</td><td class=\"r\">RTK signaling</td><td class=\"r\">PDGFR, KIT, VEGFR</td><td class=\"r\">6450551</td></tr><tr class=\"odd  odd\"><td class=\"r\">1170</td><td class=\"r\">CCT-018159</td><td class=\"r\">CCT018159, CCT 018159</td><td class=\"r\">Protein stability and degradation</td><td class=\"r\">HSP90</td><td class=\"r\">5327091</td></tr><tr class=\"even  even\"><td class=\"r\">1375</td><td class=\"r\">Temozolomide</td><td class=\"r\">Temodar, Temodal, M-39831, SCH 52365</td><td class=\"r\">DNA replication</td><td class=\"r\">DNA alkylating agent</td><td class=\"r\">5394</td></tr><tr class=\"odd  odd\"><td class=\"r\">1502</td><td class=\"r\">Bicalutamide</td><td class=\"r\">ICI-176334, Casodex, Cosudex, ICI 176334</td><td class=\"r\">Hormone-related</td><td class=\"r\">AR</td><td class=\"r\">2375</td></tr><tr class=\"even  even\"><td class=\"r\">1557</td><td class=\"r\">LCL161</td><td class=\"r\">-</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">XIAP, IAP1, IAP2</td><td class=\"r\">24737642</td></tr><tr class=\"odd  odd\"><td class=\"r\">1594</td><td class=\"r\">OSI-027</td><td class=\"r\">A-1065-5</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">MTORC1, MTORC2</td><td class=\"r\">44224160</td></tr><tr class=\"even  even\"><td class=\"r\">1799</td><td class=\"r\">Ibrutinib</td><td class=\"r\">PCI-32765, CRA-032765, PCI-32765-00</td><td class=\"r\">Other, kinases</td><td class=\"r\">BTK</td><td class=\"r\">24821094</td></tr><tr class=\"odd  odd\"><td class=\"r\">1873</td><td class=\"r\">Buparlisib</td><td class=\"r\">BKM120, NVP-BKM120</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">PI3Kalpha, PI3Kdelta, PI3Kbeta, PI3Kgamma</td><td class=\"r\">66577015, 16654980</td></tr><tr class=\"even  even\"><td class=\"r\">1912</td><td class=\"r\">Afuresertib</td><td class=\"r\">GSK2110183, GSK2110183C</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">AKT1, AKT2, AKT3</td><td class=\"r\">46843057</td></tr><tr class=\"odd  odd\"><td class=\"r\">1050</td><td class=\"r\">ZM447439</td><td class=\"r\">ZM-447439, ZM 447439</td><td class=\"r\">Mitosis</td><td class=\"r\">AURKA, AURKB</td><td class=\"r\">9914412</td></tr><tr class=\"even  even\"><td class=\"r\">1059</td><td class=\"r\">AZD8055</td><td class=\"r\">AZD-8055</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">MTORC1, MTORC2</td><td class=\"r\">25262965</td></tr><tr class=\"odd  odd\"><td class=\"r\">1083</td><td class=\"r\">Crizotinib</td><td class=\"r\">Xalkori, PF2341066, PF-2341066, PF 2341066</td><td class=\"r\">RTK signaling</td><td class=\"r\">MET, ALK, ROS1</td><td class=\"r\">11626560</td></tr><tr class=\"even  even\"><td class=\"r\">1449</td><td class=\"r\">AZD1208</td><td class=\"r\">SN1028350371</td><td class=\"r\">Other, kinases</td><td class=\"r\">PIM1, PIM2, PIM3</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1580</td><td class=\"r\">GSK2801</td><td class=\"r\">1-(7-(2-methylsulfonyl-phenyl)-4-propoxy-1-aza-bicyclo[4.3.0]nona-2,4,6,8-tetraen-9-yl)-ethanone</td><td class=\"r\">Chromatin other</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1776</td><td class=\"r\">GSK2256098C</td><td class=\"r\">-</td><td class=\"r\">Cytoskeleton</td><td class=\"r\">FAK1</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1783</td><td class=\"r\">LMB_AB1</td><td class=\"r\">-</td><td class=\"r\">Other</td><td class=\"r\">ADRA1A, ADRB1</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1804</td><td class=\"r\">Acetalax</td><td class=\"r\">Oxyphenisatin acetate</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1818</td><td class=\"r\">Vincristine</td><td class=\"r\">-</td><td class=\"r\">-</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1825</td><td class=\"r\">Podophyllotoxin bromide</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1832</td><td class=\"r\">729189</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1919</td><td class=\"r\">Osimertinib</td><td class=\"r\">AZD9291, AZD 9291, AZD-9291, Tagrisso, Mereletinib</td><td class=\"r\">EGFR signaling</td><td class=\"r\">EGFR</td><td class=\"r\">71496458</td></tr><tr class=\"odd  odd\"><td class=\"r\">1936</td><td class=\"r\">Savolitinib</td><td class=\"r\">AZD6094, Volitinib, AZD-6094, AZD 6094</td><td class=\"r\">RTK signaling</td><td class=\"r\">MET</td><td class=\"r\">68289010</td></tr><tr class=\"even  even\"><td class=\"r\">2044</td><td class=\"r\">Pyridostatin</td><td class=\"r\">-</td><td class=\"r\">DNA replication</td><td class=\"r\">G-quadruplex stabiliser</td><td class=\"r\">25227847</td></tr><tr class=\"odd  odd\"><td class=\"r\">2047</td><td class=\"r\">Ulixertinib</td><td class=\"r\">BVD-523, VRT752271</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">ERK1, ERK2</td><td class=\"r\">11719003, 58641927</td></tr><tr class=\"even  even\"><td class=\"r\">2048</td><td class=\"r\">Vinorelbine</td><td class=\"r\">vinorelbine tartrate, Navelbine, Exelbine</td><td class=\"r\">Mitosis</td><td class=\"r\">Microtubule destabiliser</td><td class=\"r\">5311497</td></tr><tr class=\"odd  odd\"><td class=\"r\">2112</td><td class=\"r\">VTP-A</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2149</td><td class=\"r\">CT7033-2</td><td class=\"r\">-</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">KDM4A, KDM4C, KDM4E, KDM3A, KDM6B</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">2499</td><td class=\"r\">N-acetyl cysteine</td><td class=\"r\">NAC</td><td class=\"r\">Metabolism</td><td class=\"r\">Metabolism</td><td class=\"r\">None</td></tr><tr class=\"even  even\"><td class=\"r\">1007</td><td class=\"r\">Docetaxel</td><td class=\"r\">RP-56976, Taxotere</td><td class=\"r\">Mitosis</td><td class=\"r\">Microtubule stabiliser</td><td class=\"r\">148124</td></tr><tr class=\"odd  odd\"><td class=\"r\">1012</td><td class=\"r\">Vorinostat</td><td class=\"r\">Zolinza, SAHA, suberanilohydroxamic acid, suberoylanilide hydroxamic acid, MK-0683</td><td class=\"r\">Chromatin histone acetylation</td><td class=\"r\">HDAC inhibitor Class I, IIa, IIb, IV</td><td class=\"r\">5311</td></tr><tr class=\"even  even\"><td class=\"r\">1016</td><td class=\"r\">Temsirolimus</td><td class=\"r\">CCI-779, Torisel</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">MTOR</td><td class=\"r\">6918289</td></tr><tr class=\"odd  odd\"><td class=\"r\">1248</td><td class=\"r\">Daporinad</td><td class=\"r\">APO866,  FK866, FK866</td><td class=\"r\">Metabolism</td><td class=\"r\">NAMPT</td><td class=\"r\">6914657</td></tr><tr class=\"even  even\"><td class=\"r\">1249</td><td class=\"r\">BMS-345541</td><td class=\"r\">-</td><td class=\"r\">Other</td><td class=\"r\">IKK-1, IKK-2</td><td class=\"r\">9926054</td></tr><tr class=\"odd  odd\"><td class=\"r\">1268</td><td class=\"r\">XAV939</td><td class=\"r\">NVP-XAV939, XAV-939, XAV 939</td><td class=\"r\">WNT signaling</td><td class=\"r\">TNKS1, TNKS2</td><td class=\"r\">2726824</td></tr><tr class=\"even  even\"><td class=\"r\">1549</td><td class=\"r\">Sapitinib</td><td class=\"r\">AZD8931</td><td class=\"r\">EGFR signaling</td><td class=\"r\">EGFR, ERBB2, ERBB3</td><td class=\"r\">11488320</td></tr><tr class=\"odd  odd\"><td class=\"r\">1561</td><td class=\"r\">Taselisib</td><td class=\"r\">GDC-0032, GDC0032, RG7604</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">PI3K (beta sparing)</td><td class=\"r\">51001932</td></tr><tr class=\"even  even\"><td class=\"r\">1853</td><td class=\"r\">OF-1</td><td class=\"r\">-</td><td class=\"r\">Chromatin histone acetylation</td><td class=\"r\">BRPF1B, BRPF2</td><td class=\"r\">35397514</td></tr><tr class=\"odd  odd\"><td class=\"r\">1072</td><td class=\"r\">Avagacestat</td><td class=\"r\">BMS-708163, BMS 708163</td><td class=\"r\">Other</td><td class=\"r\">Amyloid beta20, Amyloid beta40</td><td class=\"r\">46883536</td></tr><tr class=\"even  even\"><td class=\"r\">1073</td><td class=\"r\">5-Fluorouracil</td><td class=\"r\">5-FU</td><td class=\"r\">Other</td><td class=\"r\">Antimetabolite (DNA &amp; RNA)</td><td class=\"r\">3385</td></tr><tr class=\"odd  odd\"><td class=\"r\">1177</td><td class=\"r\">Niraparib</td><td class=\"r\">MK-4827, MK4827</td><td class=\"r\">Genome integrity</td><td class=\"r\">PARP1, PARP2</td><td class=\"r\">24958200</td></tr><tr class=\"even  even\"><td class=\"r\">1463</td><td class=\"r\">AZD1332</td><td class=\"r\">SN1061387896</td><td class=\"r\">RTK signaling</td><td class=\"r\">NTRK1, NTRK2, NTRK3</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1733</td><td class=\"r\">ULK1_4989</td><td class=\"r\">SN1049257073, ULK1_4989</td><td class=\"r\">Other, kinases</td><td class=\"r\">ULK1</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1734</td><td class=\"r\">VSP34_8731</td><td class=\"r\">SN1049998399, VSP34_8731</td><td class=\"r\">Other</td><td class=\"r\">VSP34</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1736</td><td class=\"r\">Selumetinib</td><td class=\"r\">SN1103949345, AZD1480</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">MEK1, MEK2</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1784</td><td class=\"r\">LMB_AB2</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">GADD34</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1821</td><td class=\"r\">765771</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1828</td><td class=\"r\">720427</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1830</td><td class=\"r\">Gallibiscoquinazole</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1839</td><td class=\"r\">615590</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1843</td><td class=\"r\">LMP744</td><td class=\"r\">MJ-III-65</td><td class=\"r\">DNA replication</td><td class=\"r\">TOP1</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1847</td><td class=\"r\">BEN</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1922</td><td class=\"r\">Cediranib</td><td class=\"r\">AZD2171, AZD 2171, AZD-2171, Recentin</td><td class=\"r\">RTK signaling</td><td class=\"r\">VEGFR, FLT1, FLT2, FLT3, FLT4, KIT, PDGFRB</td><td class=\"r\">9933475</td></tr><tr class=\"even  even\"><td class=\"r\">1930</td><td class=\"r\">Telomerase Inhibitor IX</td><td class=\"r\">MST-312, MST 312, MST312</td><td class=\"r\">Genome integrity</td><td class=\"r\">Telomerase</td><td class=\"r\">10385095</td></tr><tr class=\"odd  odd\"><td class=\"r\">1939</td><td class=\"r\">UMI-77</td><td class=\"r\">UMI 77, UMI-77</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">MCL1</td><td class=\"r\">992586</td></tr><tr class=\"even  even\"><td class=\"r\">1941</td><td class=\"r\">Sepantronium bromide</td><td class=\"r\">YM155, YM-155, YM 155</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">BIRC5</td><td class=\"r\">11178236</td></tr><tr class=\"odd  odd\"><td class=\"r\">1996</td><td class=\"r\">MIM1</td><td class=\"r\">MIM 1, MIM-1</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">MCL1</td><td class=\"r\">16241412</td></tr><tr class=\"even  even\"><td class=\"r\">2010</td><td class=\"r\">HKMTI-1-005</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">2057</td><td class=\"r\">Remodelin</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2157</td><td class=\"r\">A-366</td><td class=\"r\">-</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">EHMT1, EHMT2</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">2171</td><td class=\"r\">BMS-754807</td><td class=\"r\">BMS-754807</td><td class=\"r\">IGF1R signaling</td><td class=\"r\">IGF1R, IR</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2177</td><td class=\"r\">SGC0946</td><td class=\"r\">-</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">DOT1L</td><td class=\"r\">56962337</td></tr><tr class=\"odd  odd\"><td class=\"r\">1003</td><td class=\"r\">Camptothecin</td><td class=\"r\">Camptothecine, (+)-Camptothecin</td><td class=\"r\">DNA replication</td><td class=\"r\">TOP1</td><td class=\"r\">24360</td></tr><tr class=\"even  even\"><td class=\"r\">1004</td><td class=\"r\">Vinblastine</td><td class=\"r\">Velban</td><td class=\"r\">Mitosis</td><td class=\"r\">Microtubule destabiliser</td><td class=\"r\">6710780</td></tr><tr class=\"odd  odd\"><td class=\"r\">1010</td><td class=\"r\">Gefitinib</td><td class=\"r\">ZD-1839, Iressa</td><td class=\"r\">EGFR signaling</td><td class=\"r\">EGFR</td><td class=\"r\">123631</td></tr><tr class=\"even  even\"><td class=\"r\">1023</td><td class=\"r\">GW441756</td><td class=\"r\">GW 441756</td><td class=\"r\">RTK signaling</td><td class=\"r\">NTRK1</td><td class=\"r\">9943465</td></tr><tr class=\"odd  odd\"><td class=\"r\">1025</td><td class=\"r\">SB216763</td><td class=\"r\">SB-216763, SB 216763</td><td class=\"r\">WNT signaling</td><td class=\"r\">GSK3A, GSK3B</td><td class=\"r\">176158</td></tr><tr class=\"even  even\"><td class=\"r\">1029</td><td class=\"r\">Motesanib</td><td class=\"r\">AMG-706, AMG 706, AMG706</td><td class=\"r\">RTK signaling</td><td class=\"r\">VEGFR, RET, KIT, PDGFR</td><td class=\"r\">11667893</td></tr><tr class=\"odd  odd\"><td class=\"r\">1034</td><td class=\"r\">Staurosporine</td><td class=\"r\">-</td><td class=\"r\">RTK signaling</td><td class=\"r\">Broad spectrum kinase inhibitor</td><td class=\"r\">several</td></tr><tr class=\"even  even\"><td class=\"r\">1038</td><td class=\"r\">NU7441</td><td class=\"r\">KU-57788, NU-7432, NU-7741</td><td class=\"r\">Genome integrity</td><td class=\"r\">DNAPK</td><td class=\"r\">11327430</td></tr><tr class=\"odd  odd\"><td class=\"r\">1031</td><td class=\"r\">Elesclomol</td><td class=\"r\">STA-4783</td><td class=\"r\">Protein stability and degradation</td><td class=\"r\">HSP90</td><td class=\"r\">300471</td></tr><tr class=\"even  even\"><td class=\"r\">1096</td><td class=\"r\">Tozasertib</td><td class=\"r\">MK 0457,MK-0457,MK-045, VX-680 VX 680 VX-68</td><td class=\"r\">Mitosis</td><td class=\"r\">AURKA, AURKB, AURKC, others</td><td class=\"r\">5494449</td></tr><tr class=\"odd  odd\"><td class=\"r\">1194</td><td class=\"r\">SB505124</td><td class=\"r\">SB 505124, SB505124</td><td class=\"r\">RTK signaling</td><td class=\"r\">TGFBR1, ACVR1B, ACVR1C</td><td class=\"r\">9858940</td></tr><tr class=\"even  even\"><td class=\"r\">1239</td><td class=\"r\">YK-4-279</td><td class=\"r\">YK 4-279</td><td class=\"r\">Other</td><td class=\"r\">RNA helicase A</td><td class=\"r\">44632017</td></tr><tr class=\"odd  odd\"><td class=\"r\">1910</td><td class=\"r\">ABT737</td><td class=\"r\">-</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">BCL2, BCL-XL,  BCL-W, BCL-B, BFL1</td><td class=\"r\">11228183</td></tr><tr class=\"even  even\"><td class=\"r\">1911</td><td class=\"r\">Dactinomycin</td><td class=\"r\">Actinomycin D, Cosmegen, Lyovac, Meractinomycin, Ac-De, ActD</td><td class=\"r\">Other</td><td class=\"r\">RNA polymerase</td><td class=\"r\">several</td></tr><tr class=\"odd  odd\"><td class=\"r\">1043</td><td class=\"r\">JNK Inhibitor VIII</td><td class=\"r\">-</td><td class=\"r\">JNK and p38 signaling</td><td class=\"r\">JNK</td><td class=\"r\">11624601</td></tr><tr class=\"even  even\"><td class=\"r\">1058</td><td class=\"r\">Pictilisib</td><td class=\"r\">GDC-0941, GDC0941, RG-7621</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">PI3K (class 1)</td><td class=\"r\">17755052</td></tr><tr class=\"odd  odd\"><td class=\"r\">1069</td><td class=\"r\">EHT-1864</td><td class=\"r\">EHT 1864</td><td class=\"r\">Cytoskeleton</td><td class=\"r\">RAC1, RAC2, RAC3</td><td class=\"r\">9938202</td></tr><tr class=\"even  even\"><td class=\"r\">1085</td><td class=\"r\">Sorafenib</td><td class=\"r\">Nexavar, 284461-73-0, BAY 43-9006</td><td class=\"r\">Other, kinases</td><td class=\"r\">PDGFR, KIT, VEGFR, RAF</td><td class=\"r\">216239</td></tr><tr class=\"odd  odd\"><td class=\"r\">1089</td><td class=\"r\">Oxaliplatin</td><td class=\"r\">Eloxatin</td><td class=\"r\">DNA replication</td><td class=\"r\">DNA alkylating agent</td><td class=\"r\">5310940</td></tr><tr class=\"even  even\"><td class=\"r\">1093</td><td class=\"r\">GSK1904529A</td><td class=\"r\">GSK-1904529A, GSK 1904529A</td><td class=\"r\">IGF1R signaling</td><td class=\"r\">IGF1R, IR</td><td class=\"r\">25124816</td></tr><tr class=\"odd  odd\"><td class=\"r\">1133</td><td class=\"r\">Serdemetan</td><td class=\"r\">JNJ-26854165</td><td class=\"r\">p53 pathway</td><td class=\"r\">MDM2</td><td class=\"r\">11609586</td></tr><tr class=\"even  even\"><td class=\"r\">1168</td><td class=\"r\">Erlotinib</td><td class=\"r\">Tarceva, RG-1415, CP-358774, OSI-774, Ro-508231, R-1415</td><td class=\"r\">EGFR signaling</td><td class=\"r\">EGFR</td><td class=\"r\">176870</td></tr><tr class=\"odd  odd\"><td class=\"r\">1179</td><td class=\"r\">MK-1775</td><td class=\"r\">AZD1775, Adavosertib</td><td class=\"r\">Cell cycle</td><td class=\"r\">WEE1, PLK1</td><td class=\"r\">24856436</td></tr><tr class=\"even  even\"><td class=\"r\">1190</td><td class=\"r\">Gemcitabine</td><td class=\"r\">Gemzar, LY-188011</td><td class=\"r\">DNA replication</td><td class=\"r\">Pyrimidine antimetabolite</td><td class=\"r\">60750</td></tr><tr class=\"odd  odd\"><td class=\"r\">1378</td><td class=\"r\">Bleomycin (50 uM)</td><td class=\"r\">-</td><td class=\"r\">DNA replication</td><td class=\"r\">dsDNA break induction</td><td class=\"r\">5460769</td></tr><tr class=\"even  even\"><td class=\"r\">1583</td><td class=\"r\">GSK-LSD1</td><td class=\"r\">N-((1R,2S)-2-phenylcyclopropyl)piperidin-4-amine</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">KDM1</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1708</td><td class=\"r\">CDK9_5576</td><td class=\"r\">SN1047226266, CDK9_5576</td><td class=\"r\">Cell cycle</td><td class=\"r\">CDK9</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1738</td><td class=\"r\">IGF1R_3801</td><td class=\"r\">SN1051640224, IGF1R_3801</td><td class=\"r\">IGF1R signaling</td><td class=\"r\">IGFR1</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1739</td><td class=\"r\">JAK_8517</td><td class=\"r\">SN1066590414, JAK_8517</td><td class=\"r\">Other, kinases</td><td class=\"r\">JAK1, JAK2</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1777</td><td class=\"r\">GSK2276186C</td><td class=\"r\">-</td><td class=\"r\">Other, kinases</td><td class=\"r\">JAK1, JAK2, JAK3</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1778</td><td class=\"r\">GSK2110183B</td><td class=\"r\">-</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">AKT1, AKT2, AKT3</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1780</td><td class=\"r\">GSK3337463A</td><td class=\"r\">-</td><td class=\"r\">Other, kinases</td><td class=\"r\">NIK</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1802</td><td class=\"r\">Zoledronate</td><td class=\"r\">Zoledronic acid, Zometa, Reclast</td><td class=\"r\">Other</td><td class=\"r\"> </td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1809</td><td class=\"r\">Teniposide</td><td class=\"r\">Vumon</td><td class=\"r\">DNA replication</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1816</td><td class=\"r\">Fulvestrant</td><td class=\"r\">-</td><td class=\"r\">Hormone-related</td><td class=\"r\">ESR</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1820</td><td class=\"r\">123829</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1824</td><td class=\"r\">123138</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1829</td><td class=\"r\">667880</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1831</td><td class=\"r\">L-Oxonoreleagnine</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1833</td><td class=\"r\">741909</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1840</td><td class=\"r\">630600</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1926</td><td class=\"r\">GNE-317</td><td class=\"r\">GNE317, GNE 317</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">PI3Kalpha</td><td class=\"r\">70676303</td></tr><tr class=\"odd  odd\"><td class=\"r\">1940</td><td class=\"r\">WIKI4</td><td class=\"r\">WIKI-4, WIKI 4</td><td class=\"r\">WNT signaling</td><td class=\"r\">TNKS1, TNKS2</td><td class=\"r\">2984337</td></tr><tr class=\"even  even\"><td class=\"r\">2045</td><td class=\"r\">AMG-319</td><td class=\"r\">AMG319</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">PI3K (beta sparing)</td><td class=\"r\">68947304</td></tr><tr class=\"odd  odd\"><td class=\"r\">2170</td><td class=\"r\">AT13148</td><td class=\"r\">-</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">AKT1</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2498</td><td class=\"r\">alpha-lipoic acid</td><td class=\"r\">aLA</td><td class=\"r\">Metabolism</td><td class=\"r\">Metabolism</td><td class=\"r\">None</td></tr><tr class=\"odd  odd\"><td class=\"r\">1022</td><td class=\"r\">AZD7762</td><td class=\"r\">AZD-7762, AZD 7762</td><td class=\"r\">Cell cycle</td><td class=\"r\">CHEK1, CHEK2</td><td class=\"r\">11152667</td></tr><tr class=\"even  even\"><td class=\"r\">2169</td><td class=\"r\">AZD6482</td><td class=\"r\">AZD 6482, AZD-6482, AK-55409</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">PI3Kbeta</td><td class=\"r\">44137675</td></tr><tr class=\"odd  odd\"><td class=\"r\">2174</td><td class=\"r\">IOX2</td><td class=\"r\">IOX-2, IOX 2, AK176060</td><td class=\"r\">Other</td><td class=\"r\">EGLN1</td><td class=\"r\">54685215</td></tr><tr class=\"even  even\"><td class=\"r\">1033</td><td class=\"r\">Vismodegib</td><td class=\"r\">GDC0449, Erivedge</td><td class=\"r\">Other</td><td class=\"r\">SMO</td><td class=\"r\">24776445</td></tr><tr class=\"odd  odd\"><td class=\"r\">1037</td><td class=\"r\">BX795</td><td class=\"r\">BX-795</td><td class=\"r\">Other, kinases</td><td class=\"r\">TBK1, PDK1 (PDPK1), IKK, AURKB, AURKC</td><td class=\"r\">10077147</td></tr><tr class=\"even  even\"><td class=\"r\">1200</td><td class=\"r\">Fulvestrant</td><td class=\"r\">Faslodex, ICI-182780</td><td class=\"r\">Hormone-related</td><td class=\"r\">ESR</td><td class=\"r\">104741</td></tr><tr class=\"odd  odd\"><td class=\"r\">1243</td><td class=\"r\">Piperlongumine</td><td class=\"r\">Piplartine</td><td class=\"r\">Other</td><td class=\"r\">Induces reactive oxygen species</td><td class=\"r\">637858</td></tr><tr class=\"even  even\"><td class=\"r\">1250</td><td class=\"r\">AZ960</td><td class=\"r\">AZ 960, AZ-960</td><td class=\"r\">Other, kinases</td><td class=\"r\">JAK2, JAK3</td><td class=\"r\">25099184</td></tr><tr class=\"odd  odd\"><td class=\"r\">1560</td><td class=\"r\">Alpelisib</td><td class=\"r\">BYL719, BYL-719, NVP-BYL719</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">PI3Kalpha</td><td class=\"r\">56649450</td></tr><tr class=\"even  even\"><td class=\"r\">1631</td><td class=\"r\">PRT062607</td><td class=\"r\">P505-15, PRT-2607, BIIB-057</td><td class=\"r\">Other, kinases</td><td class=\"r\">SYK</td><td class=\"r\">44462758</td></tr><tr class=\"odd  odd\"><td class=\"r\">1632</td><td class=\"r\">Ribociclib</td><td class=\"r\">LEE011, NVP-LEE011, LEE011-BBA</td><td class=\"r\">Cell cycle</td><td class=\"r\">CDK4, CDK6</td><td class=\"r\">44631912</td></tr><tr class=\"even  even\"><td class=\"r\">1635</td><td class=\"r\">Picolinici-acid</td><td class=\"r\">Picolinate</td><td class=\"r\">Other</td><td class=\"r\">Inflammatory related</td><td class=\"r\">1018</td></tr><tr class=\"odd  odd\"><td class=\"r\">1046</td><td class=\"r\">Wee1 Inhibitor</td><td class=\"r\">681640, Wee1 Inhibitor</td><td class=\"r\">Cell cycle</td><td class=\"r\">WEE1, CHEK1</td><td class=\"r\">10384072</td></tr><tr class=\"even  even\"><td class=\"r\">1068</td><td class=\"r\">Obatoclax Mesylate</td><td class=\"r\">GX15-070MS, Obatoclax, GX15-070</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">BCL2, BCL-XL, BCL-W, MCL1</td><td class=\"r\">11404337</td></tr><tr class=\"odd  odd\"><td class=\"r\">1180</td><td class=\"r\">Dinaciclib</td><td class=\"r\">SCH727965, SCH 727965, SCH-727-965, MK-7965</td><td class=\"r\">Cell cycle</td><td class=\"r\">CDK1, CDK2, CDK5, CDK9</td><td class=\"r\">46926350</td></tr><tr class=\"even  even\"><td class=\"r\">1730</td><td class=\"r\">PAK_5339</td><td class=\"r\">SN1046290829, PAK_5339</td><td class=\"r\">Cytoskeleton</td><td class=\"r\">PAK1, PAK2</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1803</td><td class=\"r\">Acetalax</td><td class=\"r\">Oxyphenisatin acetate</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1814</td><td class=\"r\">Nelarabine</td><td class=\"r\">Atriance, Arranon</td><td class=\"r\">DNA replication</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1827</td><td class=\"r\">Dihydrorotenone</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1836</td><td class=\"r\">150412</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">2002</td><td class=\"r\">N30652-18-1</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2110</td><td class=\"r\">GSK591</td><td class=\"r\">EPZ015866, GSK3203591</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">PMRT5</td><td class=\"r\">117072552</td></tr><tr class=\"odd  odd\"><td class=\"r\">2111</td><td class=\"r\">VE821</td><td class=\"r\">VE 821, VE-821</td><td class=\"r\">Genome integrity</td><td class=\"r\">ATR</td><td class=\"r\">51000408</td></tr><tr class=\"even  even\"><td class=\"r\">2172</td><td class=\"r\">JQ1</td><td class=\"r\">JQ-1, (+)-JQ-1</td><td class=\"r\">Chromatin other</td><td class=\"r\">BRD2, BRD3, BRD4, BRDT</td><td class=\"r\">46907787</td></tr><tr class=\"odd  odd\"><td class=\"r\">2359</td><td class=\"r\">GSK2830371</td><td class=\"r\">None</td><td class=\"r\">Other</td><td class=\"r\">WIP1</td><td class=\"r\">70983932</td></tr><tr class=\"even  even\"><td class=\"r\">2361</td><td class=\"r\">THR-102</td><td class=\"r\">WIMM synthesis</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">Mutant RAS</td><td class=\"r\">None</td></tr><tr class=\"odd  odd\"><td class=\"r\">1011</td><td class=\"r\">Navitoclax</td><td class=\"r\">ABT-263, ABT263, ABT 263</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">BCL2, BCL-XL, BCL-W</td><td class=\"r\">24978538</td></tr><tr class=\"even  even\"><td class=\"r\">1014</td><td class=\"r\">Refametinib</td><td class=\"r\">RDEA119, BAY-86-9766, BAY 869766</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">MEK1, MEK2</td><td class=\"r\">44182295</td></tr><tr class=\"odd  odd\"><td class=\"r\">1024</td><td class=\"r\">Lestaurtinib</td><td class=\"r\">CEP-701, SP-924, SPM-924, A-154475, KT-555</td><td class=\"r\">Other, kinases</td><td class=\"r\">FLT3, JAK2, NTRK1, NTRK2, NTRK3</td><td class=\"r\">126565</td></tr><tr class=\"even  even\"><td class=\"r\">2175</td><td class=\"r\">CHIR-99021</td><td class=\"r\">CT 99021, CHIR99021, CHIR 99021</td><td class=\"r\">WNT signaling</td><td class=\"r\">GSK3A, GSK3B</td><td class=\"r\">9956119</td></tr><tr class=\"odd  odd\"><td class=\"r\">1237</td><td class=\"r\">EPZ004777</td><td class=\"r\">EPZ-004777</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">DOT1L</td><td class=\"r\">56962336</td></tr><tr class=\"even  even\"><td class=\"r\">1373</td><td class=\"r\">Dabrafenib</td><td class=\"r\">GSK2118436, Tafinlar</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">BRAF</td><td class=\"r\">44462760</td></tr><tr class=\"odd  odd\"><td class=\"r\">1529</td><td class=\"r\">Pevonedistat</td><td class=\"r\">MLN4924, MLN 4924, MLN-4924</td><td class=\"r\">Other</td><td class=\"r\">NAE</td><td class=\"r\">16720766</td></tr><tr class=\"even  even\"><td class=\"r\">1598</td><td class=\"r\">LGK974</td><td class=\"r\">LGK-974</td><td class=\"r\">WNT signaling</td><td class=\"r\">PORCN</td><td class=\"r\">46926973</td></tr><tr class=\"odd  odd\"><td class=\"r\">1613</td><td class=\"r\">VE-822</td><td class=\"r\">VE 822, VE822, Berzosertib</td><td class=\"r\">Genome integrity</td><td class=\"r\">ATR</td><td class=\"r\">59472121</td></tr><tr class=\"even  even\"><td class=\"r\">1618</td><td class=\"r\">GSK2606414</td><td class=\"r\">AK175551</td><td class=\"r\">Metabolism</td><td class=\"r\">PERK</td><td class=\"r\">53469448</td></tr><tr class=\"odd  odd\"><td class=\"r\">1629</td><td class=\"r\">ML323</td><td class=\"r\">GTPL7898, AOB2313,</td><td class=\"r\">Protein stability and degradation</td><td class=\"r\">USP1, UAF1</td><td class=\"r\">60167849</td></tr><tr class=\"even  even\"><td class=\"r\">1849</td><td class=\"r\">Sabutoclax</td><td class=\"r\">BI-97C1</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">BCL2,  BCL-XL,  BFL1, MCL1</td><td class=\"r\">46236925</td></tr><tr class=\"odd  odd\"><td class=\"r\">1048</td><td class=\"r\">Mirin</td><td class=\"r\">-</td><td class=\"r\">Genome integrity</td><td class=\"r\">MRE11</td><td class=\"r\">none</td></tr><tr class=\"even  even\"><td class=\"r\">1051</td><td class=\"r\">Alisertib</td><td class=\"r\">MLN8237</td><td class=\"r\">Mitosis</td><td class=\"r\">AURKA</td><td class=\"r\">24771867</td></tr><tr class=\"odd  odd\"><td class=\"r\">1057</td><td class=\"r\">Dactolisib</td><td class=\"r\">NVP-BEZ235, BEZ235</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">PI3K (class 1), MTORC1, MTORC2</td><td class=\"r\">11977753</td></tr><tr class=\"even  even\"><td class=\"r\">1061</td><td class=\"r\">SB590885</td><td class=\"r\">SB-590885</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">BRAF</td><td class=\"r\">11316960</td></tr><tr class=\"odd  odd\"><td class=\"r\">1062</td><td class=\"r\">Selumetinib</td><td class=\"r\">AZD6244, AZD-6244, ARRY-886</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">MEK1, MEK2</td><td class=\"r\">10127622</td></tr><tr class=\"even  even\"><td class=\"r\">1091</td><td class=\"r\">BMS-536924</td><td class=\"r\">BMS 536924</td><td class=\"r\">IGF1R signaling</td><td class=\"r\">IGF1R, IR</td><td class=\"r\">10390396</td></tr><tr class=\"odd  odd\"><td class=\"r\">1175</td><td class=\"r\">Rucaparib</td><td class=\"r\">PF-01367338, AG-014699, AG-14447, AG-14699</td><td class=\"r\">Genome integrity</td><td class=\"r\">PARP1, PARP2</td><td class=\"r\">9931953</td></tr><tr class=\"even  even\"><td class=\"r\">1191</td><td class=\"r\">Bortezomib</td><td class=\"r\">PS-341, LDP-341, Velcade</td><td class=\"r\">Protein stability and degradation</td><td class=\"r\">Proteasome</td><td class=\"r\">387447</td></tr><tr class=\"odd  odd\"><td class=\"r\">1586</td><td class=\"r\">BDILV000379a</td><td class=\"r\">4-(4-(1-methyl-1H-pyrazol-4-yl)-1-(2-(1-methyl-1H-pyrazol-4-yl)ethyl)-4,5-dihydro-1H-imidazol-5-yl)benzonitrile</td><td class=\"r\">Chromatin histone acetylation</td><td class=\"r\">BAZ2A, BAZ2B</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1718</td><td class=\"r\">JAK1_8709</td><td class=\"r\">SN1049200060, JAK1_8709</td><td class=\"r\">Other, kinases</td><td class=\"r\">JAK1</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1732</td><td class=\"r\">TAF1_5496</td><td class=\"r\">SN1050393042, TAF1_5496</td><td class=\"r\">Other</td><td class=\"r\">TAF1</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1779</td><td class=\"r\">GSK626616AC</td><td class=\"r\">-</td><td class=\"r\">Other, kinases</td><td class=\"r\">DYRK1A</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1785</td><td class=\"r\">LMB_AB3</td><td class=\"r\">-</td><td class=\"r\">Other</td><td class=\"r\">PPP1R15B</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1806</td><td class=\"r\">Oxaliplatin</td><td class=\"r\">-</td><td class=\"r\">DNA replication</td><td class=\"r\">DNA alkylating agent</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1835</td><td class=\"r\">Elephantin</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1838</td><td class=\"r\">Sinularin</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1844</td><td class=\"r\">776928</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1916</td><td class=\"r\">AZD5363</td><td class=\"r\">-</td><td class=\"r\">Other, kinases</td><td class=\"r\">AKT1, AKT2, AKT3, ROCK2</td><td class=\"r\">25227436, 42602260</td></tr><tr class=\"odd  odd\"><td class=\"r\">1927</td><td class=\"r\">GSK2578215A</td><td class=\"r\">GSK 2578215A,  GSK-2578215A</td><td class=\"r\">Other, kinases</td><td class=\"r\">LRRK2</td><td class=\"r\">68107965</td></tr><tr class=\"even  even\"><td class=\"r\">2000</td><td class=\"r\">N27922-53-1</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">2046</td><td class=\"r\">MK-8776</td><td class=\"r\">SCH900776</td><td class=\"r\">Cell cycle</td><td class=\"r\">CHEK1, CHEK2, CDK2</td><td class=\"r\">16224745</td></tr><tr class=\"even  even\"><td class=\"r\">2113</td><td class=\"r\">VTP-B</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">2145</td><td class=\"r\">PBD-288</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2148</td><td class=\"r\">POMHEX</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">2173</td><td class=\"r\">PFI-1</td><td class=\"r\">-</td><td class=\"r\">Chromatin other</td><td class=\"r\">BRD4</td><td class=\"r\">71271629</td></tr><tr class=\"even  even\"><td class=\"r\">1013</td><td class=\"r\">Nilotinib</td><td class=\"r\">Tasigna, AMN 107</td><td class=\"r\">ABL signaling</td><td class=\"r\">ABL</td><td class=\"r\">644241</td></tr><tr class=\"odd  odd\"><td class=\"r\">1032</td><td class=\"r\">Afatinib</td><td class=\"r\">BIBW2992, Tovok, Gilotrif</td><td class=\"r\">EGFR signaling</td><td class=\"r\">EGFR, ERBB2</td><td class=\"r\">10184653</td></tr><tr class=\"even  even\"><td class=\"r\">1036</td><td class=\"r\">PLX-4720</td><td class=\"r\">PLX4720, PLX 4720</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">BRAF</td><td class=\"r\">24180719</td></tr><tr class=\"odd  odd\"><td class=\"r\">1149</td><td class=\"r\">TW 37</td><td class=\"r\">TW37, TW-37</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">BCL2, BCL-XL, MCL1</td><td class=\"r\">11455910</td></tr><tr class=\"even  even\"><td class=\"r\">1009</td><td class=\"r\">Tretinoin</td><td class=\"r\">ATRA, Vesanoid, Renova, Atralin, Tretin-X, Avita</td><td class=\"r\">Other</td><td class=\"r\">Retinoic acid</td><td class=\"r\">444795</td></tr><tr class=\"odd  odd\"><td class=\"r\">1039</td><td class=\"r\">SL0101</td><td class=\"r\">SL-0101, SL 0101-1</td><td class=\"r\">Other, kinases</td><td class=\"r\">RSK, AURKB, PIM1, PIM3</td><td class=\"r\">10459196</td></tr><tr class=\"even  even\"><td class=\"r\">1199</td><td class=\"r\">Tamoxifen</td><td class=\"r\">Nolvadex, Soltamox, Zynoplex, ICI-46474, Kessar</td><td class=\"r\">Hormone-related</td><td class=\"r\">ESR1</td><td class=\"r\">2733526</td></tr><tr class=\"odd  odd\"><td class=\"r\">1512</td><td class=\"r\">Cyclophosphamide</td><td class=\"r\">Cytoxan, Neosar, Cyclophosphamid, Procytox, Cyclophosphane</td><td class=\"r\">DNA replication</td><td class=\"r\">Alkylating agent</td><td class=\"r\">2907</td></tr><tr class=\"even  even\"><td class=\"r\">1558</td><td class=\"r\">Lapatinib</td><td class=\"r\">Tykerb, Tyverb</td><td class=\"r\">EGFR signaling</td><td class=\"r\">EGFR, ERBB2</td><td class=\"r\">208908</td></tr><tr class=\"odd  odd\"><td class=\"r\">1617</td><td class=\"r\">AZD5582</td><td class=\"r\">GTPL7710, AOB5560</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">XIAP, cIAP</td><td class=\"r\">49847690</td></tr><tr class=\"even  even\"><td class=\"r\">1624</td><td class=\"r\">I-BET-762</td><td class=\"r\">GSK525762A</td><td class=\"r\">Chromatin other</td><td class=\"r\">BRD2, BRD3, BRD4</td><td class=\"r\">46943432</td></tr><tr class=\"odd  odd\"><td class=\"r\">1852</td><td class=\"r\">LY2109761</td><td class=\"r\">5XE</td><td class=\"r\">Other</td><td class=\"r\">TGFB1</td><td class=\"r\">11655119</td></tr><tr class=\"even  even\"><td class=\"r\">1862</td><td class=\"r\">MG-132</td><td class=\"r\">LLL cpd, MG 132, MG132</td><td class=\"r\">Protein stability and degradation</td><td class=\"r\">Proteasome, CAPN1</td><td class=\"r\">462382</td></tr><tr class=\"odd  odd\"><td class=\"r\">1047</td><td class=\"r\">Nutlin-3a (-)</td><td class=\"r\">-</td><td class=\"r\">p53 pathway</td><td class=\"r\">MDM2</td><td class=\"r\">11433190</td></tr><tr class=\"even  even\"><td class=\"r\">1067</td><td class=\"r\">CCT007093</td><td class=\"r\">-</td><td class=\"r\">Cell cycle</td><td class=\"r\">PPM1D</td><td class=\"r\">2314623</td></tr><tr class=\"odd  odd\"><td class=\"r\">1582</td><td class=\"r\">SGC-CBP30</td><td class=\"r\">8-(3-chloro-4-methoxy-phenethyl)-4-(3,5-dimethyl-isoxazol-4-yl)-9-(2-(morpholin-4-yl)-propyl)-7,9-diaza-bicyclo[4.3.0]nona-1(6),2,4,7-tetraene</td><td class=\"r\">Chromatin histone acetylation</td><td class=\"r\">EP300, CBP</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1585</td><td class=\"r\">BDF00022089a</td><td class=\"r\">-</td><td class=\"r\">Chromatin histone acetylation</td><td class=\"r\">BRPF1, BRPF2, BRPF3</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1706</td><td class=\"r\">AZD5153</td><td class=\"r\">SN1045566955, AZD5153</td><td class=\"r\">Chromatin other</td><td class=\"r\">BRD4</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1712</td><td class=\"r\">Eg5_9814</td><td class=\"r\">SN1047613775, Eg5_9814</td><td class=\"r\">Other</td><td class=\"r\">KSP11</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1713</td><td class=\"r\">ERK_2440</td><td class=\"r\">SN1051032892, ERK_2440</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">ERK1,ERK2</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1720</td><td class=\"r\">AZD5991</td><td class=\"r\">SN1049446612, AZD5991</td><td class=\"r\">Apoptosis regulation</td><td class=\"r\">MCL1</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1808</td><td class=\"r\">Topotecan</td><td class=\"r\">-</td><td class=\"r\">DNA replication</td><td class=\"r\">TOP1</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1810</td><td class=\"r\">Mitoxantrone</td><td class=\"r\">-</td><td class=\"r\">DNA replication</td><td class=\"r\">TOP2</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1811</td><td class=\"r\">Dactinomycin</td><td class=\"r\">Actinomycin D</td><td class=\"r\">Other</td><td class=\"r\">RNA polymerase</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1815</td><td class=\"r\">Dacarbazine</td><td class=\"r\">DTIC</td><td class=\"r\">Other</td><td class=\"r\">CP11A</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1866</td><td class=\"r\">BDP-00009066</td><td class=\"r\">-</td><td class=\"r\">Cytoskeleton</td><td class=\"r\">MRCKB_HUMAN</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1918</td><td class=\"r\">AZD8186</td><td class=\"r\">AZD 8186, AZD-8186</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">PI3Kalpha, PI3Kbeta</td><td class=\"r\">52913813</td></tr><tr class=\"odd  odd\"><td class=\"r\">1924</td><td class=\"r\">Ipatasertib</td><td class=\"r\">GDC0068, GDC 0068, GDC-0068</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">AKT1, AKT, AKT3</td><td class=\"r\">24788740</td></tr><tr class=\"even  even\"><td class=\"r\">1925</td><td class=\"r\">GDC0810</td><td class=\"r\">GDC-0810, GDC 0810</td><td class=\"r\">Hormone-related</td><td class=\"r\">ESR1, ESR2</td><td class=\"r\">56941241</td></tr><tr class=\"odd  odd\"><td class=\"r\">1998</td><td class=\"r\">BPD-00008900</td><td class=\"r\">-</td><td class=\"r\">Other</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2037</td><td class=\"r\">GSK343</td><td class=\"r\">-</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">EZH2</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">2039</td><td class=\"r\">AGK2</td><td class=\"r\">-</td><td class=\"r\">Chromatin histone acetylation</td><td class=\"r\">BAZ2A, BAZ2B</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2043</td><td class=\"r\">BIBR-1532</td><td class=\"r\">-</td><td class=\"r\">Genome integrity</td><td class=\"r\">TERT</td><td class=\"r\">9927531</td></tr><tr class=\"odd  odd\"><td class=\"r\">2055</td><td class=\"r\">Mycophenolic acid</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2109</td><td class=\"r\">AZ6102</td><td class=\"r\">-</td><td class=\"r\">WNT signaling</td><td class=\"r\">TNKS1, TNKS2</td><td class=\"r\">91663328</td></tr><tr class=\"odd  odd\"><td class=\"r\">2158</td><td class=\"r\">CPI-637</td><td class=\"r\">-</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">EP300</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2438</td><td class=\"r\">ascorbate (vitamin C)</td><td class=\"r\">back-up solution from YWKim</td><td class=\"r\">Other</td><td class=\"r\">anti-oxidant proteins</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1005</td><td class=\"r\">Cisplatin</td><td class=\"r\">cis-Diammineplatinum(II) dichloride, Platinol, CIS-DDP</td><td class=\"r\">DNA replication</td><td class=\"r\">DNA crosslinker</td><td class=\"r\">84691</td></tr><tr class=\"even  even\"><td class=\"r\">1006</td><td class=\"r\">Cytarabine</td><td class=\"r\">Ara-Cytidine, Arabinosyl Cytosine, U-19920</td><td class=\"r\">Other</td><td class=\"r\">Antimetabolite</td><td class=\"r\">6253</td></tr><tr class=\"odd  odd\"><td class=\"r\">1018</td><td class=\"r\">Veliparib</td><td class=\"r\">ABT-888, ABT888, ABT 888</td><td class=\"r\">Genome integrity</td><td class=\"r\">PARP1, PARP2</td><td class=\"r\">11960529</td></tr><tr class=\"even  even\"><td class=\"r\">1020</td><td class=\"r\">Lenalidomide</td><td class=\"r\">CDC-501, CC-5013, Revlimid</td><td class=\"r\">Protein stability and degradation</td><td class=\"r\">CRBN</td><td class=\"r\">216326</td></tr><tr class=\"odd  odd\"><td class=\"r\">1026</td><td class=\"r\">Tanespimycin</td><td class=\"r\">17-AAG, BMS-722782</td><td class=\"r\">Protein stability and degradation</td><td class=\"r\">HSP90</td><td class=\"r\">6505803</td></tr><tr class=\"even  even\"><td class=\"r\">1192</td><td class=\"r\">GSK269962A</td><td class=\"r\">GSK 269962A</td><td class=\"r\">Cytoskeleton</td><td class=\"r\">ROCK1, ROCK2</td><td class=\"r\">16095342</td></tr><tr class=\"odd  odd\"><td class=\"r\">1494</td><td class=\"r\">SN-38</td><td class=\"r\">7-Ethyl-10-Hydroxy-Camptothecin, SN 38</td><td class=\"r\">DNA replication</td><td class=\"r\">TOP1</td><td class=\"r\">104842</td></tr><tr class=\"even  even\"><td class=\"r\">1507</td><td class=\"r\">Ruxolitinib</td><td class=\"r\">INCB-18424, Ruxolitinib Phosphate, Jakafi</td><td class=\"r\">Other, kinases</td><td class=\"r\">JAK1, JAK2</td><td class=\"r\">25126798</td></tr><tr class=\"odd  odd\"><td class=\"r\">1510</td><td class=\"r\">Linsitinib</td><td class=\"r\">OSI-906, ASP-7487</td><td class=\"r\">IGF1R signaling</td><td class=\"r\">IGF1R</td><td class=\"r\">11640390</td></tr><tr class=\"even  even\"><td class=\"r\">1511</td><td class=\"r\">Epirubicin</td><td class=\"r\">Ellence, Farmorubicin, IMI-28</td><td class=\"r\">DNA replication</td><td class=\"r\">Anthracycline</td><td class=\"r\">41867</td></tr><tr class=\"odd  odd\"><td class=\"r\">1553</td><td class=\"r\">Uprosertib</td><td class=\"r\">GSK2141795, GSK2141795C, GSK-2141795</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">AKT1, AKT2, AKT3</td><td class=\"r\">51042438</td></tr><tr class=\"even  even\"><td class=\"r\">1563</td><td class=\"r\">EPZ5676</td><td class=\"r\">EPZ-5676, Pinometostat</td><td class=\"r\">Chromatin histone methylation</td><td class=\"r\">DOT1L</td><td class=\"r\">57345410</td></tr><tr class=\"odd  odd\"><td class=\"r\">1564</td><td class=\"r\">SCH772984</td><td class=\"r\">CS-1421</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">ERK1, ERK2</td><td class=\"r\">24866313</td></tr><tr class=\"even  even\"><td class=\"r\">1576</td><td class=\"r\">IWP-2</td><td class=\"r\">Wnt Inhibitor IWP-2</td><td class=\"r\">WNT signaling</td><td class=\"r\">PORCN</td><td class=\"r\">2155128</td></tr><tr class=\"odd  odd\"><td class=\"r\">1578</td><td class=\"r\">Leflunomide</td><td class=\"r\">Sulol, SU-101, HWA-486, Arava</td><td class=\"r\">DNA replication</td><td class=\"r\">Pyrimidine synthesis inhibitor</td><td class=\"r\">3899</td></tr><tr class=\"even  even\"><td class=\"r\">1593</td><td class=\"r\">Entinostat</td><td class=\"r\">MS-275</td><td class=\"r\">Chromatin histone acetylation</td><td class=\"r\">HDAC1, HDAC3</td><td class=\"r\">4261</td></tr><tr class=\"odd  odd\"><td class=\"r\">1614</td><td class=\"r\">WZ4003</td><td class=\"r\">WZ 4003</td><td class=\"r\">Other, kinases</td><td class=\"r\">NUAK1, NUAK2</td><td class=\"r\">72200024</td></tr><tr class=\"even  even\"><td class=\"r\">1621</td><td class=\"r\">PCI-34051</td><td class=\"r\">AK298746</td><td class=\"r\">Chromatin histone acetylation</td><td class=\"r\">HDAC8, HDAC6, HDAC1</td><td class=\"r\">24753719</td></tr><tr class=\"odd  odd\"><td class=\"r\">1625</td><td class=\"r\">RVX-208</td><td class=\"r\">1044870-39-4, Apabetalone</td><td class=\"r\">Chromatin other</td><td class=\"r\">BRD4</td><td class=\"r\">24871506</td></tr><tr class=\"even  even\"><td class=\"r\">1630</td><td class=\"r\">Entospletinib</td><td class=\"r\">GS-9973</td><td class=\"r\">Other, kinases</td><td class=\"r\">SYK</td><td class=\"r\">59473233</td></tr><tr class=\"odd  odd\"><td class=\"r\">1786</td><td class=\"r\">AZD4547</td><td class=\"r\">1035270-39-3</td><td class=\"r\">RTK signaling</td><td class=\"r\">FGFR1, FGFR2, FGFR3</td><td class=\"r\">51039095</td></tr><tr class=\"even  even\"><td class=\"r\">1855</td><td class=\"r\">KRAS (G12C) Inhibitor-12</td><td class=\"r\">GTPL8020</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">KRAS (G12C)</td><td class=\"r\">73555129</td></tr><tr class=\"odd  odd\"><td class=\"r\">1908</td><td class=\"r\">Ulixertinib</td><td class=\"r\">BVD-523, VRT752271</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">ERK1, ERK2</td><td class=\"r\">11719003, 58641927</td></tr><tr class=\"even  even\"><td class=\"r\">1042</td><td class=\"r\">Doramapimod</td><td class=\"r\">BIRB-796, BIRB 796</td><td class=\"r\">JNK and p38 signaling</td><td class=\"r\">p38, JNK2</td><td class=\"r\">156422</td></tr><tr class=\"odd  odd\"><td class=\"r\">1049</td><td class=\"r\">PD173074</td><td class=\"r\">PD-173074, PD 173074</td><td class=\"r\">RTK signaling</td><td class=\"r\">FGFR1, FGFR2, FGFR3</td><td class=\"r\">1401</td></tr><tr class=\"even  even\"><td class=\"r\">1052</td><td class=\"r\">RO-3306</td><td class=\"r\">-</td><td class=\"r\">Cell cycle</td><td class=\"r\">CDK1</td><td class=\"r\">44450571</td></tr><tr class=\"odd  odd\"><td class=\"r\">1053</td><td class=\"r\">MK-2206</td><td class=\"r\">MK 2206, MK2206</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">AKT1, AKT2</td><td class=\"r\">46930998</td></tr><tr class=\"even  even\"><td class=\"r\">1054</td><td class=\"r\">Palbociclib</td><td class=\"r\">PD0332991, PD-0332991, PF-00080665-73</td><td class=\"r\">Cell cycle</td><td class=\"r\">CDK4, CDK6</td><td class=\"r\">5330286</td></tr><tr class=\"odd  odd\"><td class=\"r\">1060</td><td class=\"r\">PD0325901</td><td class=\"r\">PD-0325901, PD 0325901</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">MEK1, MEK2</td><td class=\"r\">9826528</td></tr><tr class=\"even  even\"><td class=\"r\">1079</td><td class=\"r\">Dasatinib</td><td class=\"r\">BMS-354825-03, BMS-354825, Sprycel</td><td class=\"r\">Other, kinases</td><td class=\"r\">ABL, SRC, Ephrins, PDGFR, KIT</td><td class=\"r\">3062316</td></tr><tr class=\"odd  odd\"><td class=\"r\">1080</td><td class=\"r\">Paclitaxel</td><td class=\"r\">BMS-181339-01, Taxol, Onxol, Paxene, Praxel, Abraxane</td><td class=\"r\">Mitosis</td><td class=\"r\">Microtubule stabiliser</td><td class=\"r\">36314</td></tr><tr class=\"even  even\"><td class=\"r\">1086</td><td class=\"r\">BI-2536</td><td class=\"r\">-</td><td class=\"r\">Cell cycle</td><td class=\"r\">PLK1, PLK2, PLK3</td><td class=\"r\">11364421</td></tr><tr class=\"odd  odd\"><td class=\"r\">1088</td><td class=\"r\">Irinotecan</td><td class=\"r\">Camptosar, (+)-Irinotecan, Irinotecanum, irinotecan hydrochloride</td><td class=\"r\">DNA replication</td><td class=\"r\">TOP1</td><td class=\"r\">60838</td></tr><tr class=\"even  even\"><td class=\"r\">1401</td><td class=\"r\">AZD5438</td><td class=\"r\">SN1057898678</td><td class=\"r\">Cell cycle</td><td class=\"r\">CDK2</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1428</td><td class=\"r\">IAP_5620</td><td class=\"r\">SN1043546339</td><td class=\"r\">Other</td><td class=\"r\">IAP</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1441</td><td class=\"r\">AZD2014</td><td class=\"r\">SN1103949359</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">mTORC1, mTORC2</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1581</td><td class=\"r\">Bromosporine</td><td class=\"r\">ethyl N-[6-(3-methanesulfonamido-4-methylphenyl)-3-methyl-[1,2,4]triazolo[4,3-b]pyridazin-8-yl]carbamate</td><td class=\"r\">Chromatin other</td><td class=\"r\">CECR2, BRD2, BRD4, BRD9</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1709</td><td class=\"r\">CDK9_5038</td><td class=\"r\">SN1047483750, CDK9_5038</td><td class=\"r\">Cell cycle</td><td class=\"r\">CDK9</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1714</td><td class=\"r\">ERK_6604</td><td class=\"r\">SN1047587618, ERK_6604</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">ERK1,ERK2</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1782</td><td class=\"r\">GSK2830371A</td><td class=\"r\">-</td><td class=\"r\">Other</td><td class=\"r\">PPM1D</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1807</td><td class=\"r\">Carmustine</td><td class=\"r\">-</td><td class=\"r\">-</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1813</td><td class=\"r\">Fludarabine</td><td class=\"r\">Fludara</td><td class=\"r\">DNA replication</td><td class=\"r\">Antimetabolite</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1819</td><td class=\"r\">Docetaxel</td><td class=\"r\">RP-56976, Taxotere</td><td class=\"r\">Mitosis</td><td class=\"r\">Microtubule stabiliser</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1826</td><td class=\"r\">50869</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">1845</td><td class=\"r\">Schweinfurthin A</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">1917</td><td class=\"r\">AZD6738</td><td class=\"r\">AZD 6738, AZD-6738</td><td class=\"r\">Genome integrity</td><td class=\"r\">ATR</td><td class=\"r\">54761306</td></tr><tr class=\"odd  odd\"><td class=\"r\">1928</td><td class=\"r\">I-BRD9</td><td class=\"r\">GSK602, GSK 602, GSK-602</td><td class=\"r\">Chromatin other</td><td class=\"r\">BRD9</td><td class=\"r\">91668541</td></tr><tr class=\"even  even\"><td class=\"r\">1931</td><td class=\"r\">MIRA-1</td><td class=\"r\">MIRA 1, MIRA1, NSC19630, NSC-19630, NSC 19630</td><td class=\"r\">p53 pathway</td><td class=\"r\">TP53</td><td class=\"r\">227681</td></tr><tr class=\"odd  odd\"><td class=\"r\">1999</td><td class=\"r\">N25720-51-A1</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2003</td><td class=\"r\">N29087-69-1</td><td class=\"r\">-</td><td class=\"r\">Unclassified</td><td class=\"r\">-</td><td class=\"r\">-</td></tr><tr class=\"odd  odd\"><td class=\"r\">2096</td><td class=\"r\">VX-11e</td><td class=\"r\">VX11e, VX11e</td><td class=\"r\">ERK MAPK signaling</td><td class=\"r\">ERK2</td><td class=\"r\">11634725</td></tr><tr class=\"even  even\"><td class=\"r\">2107</td><td class=\"r\">LJI308</td><td class=\"r\">-</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">RSK2, RSK1, RSK3</td><td class=\"r\">118704762</td></tr><tr class=\"odd  odd\"><td class=\"r\">2156</td><td class=\"r\">5-azacytidine</td><td class=\"r\">-</td><td class=\"r\">Other</td><td class=\"r\">DNA methyltransferases</td><td class=\"r\">-</td></tr><tr class=\"even  even\"><td class=\"r\">2362</td><td class=\"r\">THR-103</td><td class=\"r\">WIMM synthesis</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">Mutant RAS</td><td class=\"r\">None</td></tr><tr class=\"odd  odd\"><td class=\"r\">1030</td><td class=\"r\">KU-55933</td><td class=\"r\">KU55933</td><td class=\"r\">Genome integrity</td><td class=\"r\">ATM</td><td class=\"r\">5278396</td></tr><tr class=\"even  even\"><td class=\"r\">1129</td><td class=\"r\">PF-4708671</td><td class=\"r\">PF 4708671, PF4708671</td><td class=\"r\">PI3K/MTOR signaling</td><td class=\"r\">S6K1</td><td class=\"r\">51371303</td></tr></tbody></table>'\n",
    "soup = BeautifulSoup(drug_table_html)\n",
    "table = soup.find('table')\n",
    "table_body = table.find('tbody')\n",
    "\n",
    "rows = table_body.find_all('tr')\n",
    "for row in rows:\n",
    "    cols = row.find_all('td')\n",
    "    cols = [ele.text.strip() for ele in cols]\n",
    "    data.append([cols[0], cols[-1]])\n",
    "drug_to_pubchem_df = pd.DataFrame(data = data, columns = [\"DRUG_ID\", \"PUB_CHEM_ID\"])\n",
    "drug_to_pubchem_df[\"DRUG_ID\"] = drug_to_pubchem_df[\"DRUG_ID\"].astype(\"int\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6eb292c0-9ddd-4293-aa42-c51aef305819",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_to_pubchem_df = pd.read_csv(\"gdsc_drugs.csv\", sep = \",\").rename(columns={\"Drug Id\" : \"DRUG_ID\",\" PubCHEM\": \"PUB_CHEM_ID\" })\n",
    "drug_to_pubchem_df.loc[drug_to_pubchem_df[\"DRUG_ID\"] == 1803, \"PUB_CHEM_ID\"] = 8269\n",
    "targets_drug_pubchem_df = pd.merge(targets, drug_to_pubchem_df, on = \"DRUG_ID\", how = \"left\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "795e8c7a-5f64-459a-80fc-5b8ab65f0510",
   "metadata": {},
   "source": [
    "## Some Drugs dont have PubChem Ids (None, none, several (wtf?), -) -> Filter only those which have a valid Id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "102dde31-077d-4bba-9fbe-714f01567100",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'24360,6710780,84691,6253,148124,126941,444795,123631,24978538,5311,644241,44182295,6918289,23725625,11960529,5328940,216326,6450551,11152667,9943465,126565,176158,6505803,11667893,5278396,300471,10184653,24776445,24180719,10077147,11327430,10459196,156422,11624601,10384072,11433190,1401,9914412,24771867,44450571,46930998,5330286,11977753,17755052,25262965,9826528,11316960,10127622,2314623,11404337,9938202,46883536,3385,3062316,36314,11626560,5384616,216239,11364421,60838,5310940,10390396,25124816,5494449,51371303,52918385,11609586,11455910,176870,5327091,9931953,24958200,24856436,46926350,60750,387447,16095342,9858940,2733526,104741,56962336,44632017,637858,6914657,9926054,25099184,44819241,2726824,11707110,44462760,5394,5460769,104842,2375,25126798,11640390,41867,2907,16720766,11488320,51042438,24737642,208908,10096043,56649450,51001932,57345410,24866313,2155128,3899,4261,44224160,46926973,59472121,72200024,42623951,49847690,53469448,78243717,24753719,57519544,46943432,24871506,9936746,71268957,60167849,59473233,44462758,44631912,71299339,1018,51039095,24821094,8269,46236925,11655119,35397514,2802462,73555129,462382,49846579,11228183,46843057,56645356,78209992,54761306,52913813,71496458,9933475,24788740,56941241,70676303,68107965,91668541,10385095,227681,9825149,46931953,68289010,992586,2984337,11178236,16241412,71297207,42642645,9927531,25227847,68947304,16224745,5311497,11634725,118704762,91663328,117072552,51000408,44137675,46907787,71271629,54685215,9956119,56962337,70983932'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_only_valid_pubchem_ids(targets_drug_pubchem_df):\n",
    "    def num_conv(id):\n",
    "        try:\n",
    "            return int(id)\n",
    "        except:\n",
    "            return None\n",
    "    id_nums = list(map(num_conv, targets_drug_pubchem_df[\"PUB_CHEM_ID\"].unique()))\n",
    "    id_nums = list(filter(lambda id: id != None, id_nums))\n",
    "    return id_nums\n",
    "\",\".join(map(str, get_only_valid_pubchem_ids(targets_drug_pubchem_df))) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "427c87a8-c635-436b-b9d7-906c51c97429",
   "metadata": {},
   "source": [
    "## Transferred Pubchem Ids to SMILES via https://pubchem.ncbi.nlm.nih.gov/idexchange/idexchange.cgi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "379b4402-e41b-4eb9-9792-c8cbe475d205",
   "metadata": {},
   "outputs": [],
   "source": [
    "## TODO solution for missing drugs\n",
    "pubchem_to_smiles_df = pd.read_csv(\"pubchem_to_smiles.csv\", sep = \"\\t\", header = None)\n",
    "pubchem_to_smiles_df = pubchem_to_smiles_df.rename(columns={0:\"PUB_CHEM_ID\", 1:\"SMILES\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "863be997-7010-4508-9e15-aebc975efe3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def int_filter(entry):\n",
    "    try:\n",
    "        int(entry)\n",
    "        return True\n",
    "    except:\n",
    "        return False\n",
    "        \n",
    "targets_drug_valid_pubchem_df = targets_drug_pubchem_df[targets_drug_pubchem_df[\"PUB_CHEM_ID\"].apply(int_filter)]\n",
    "targets_drug_valid_pubchem_df = targets_drug_valid_pubchem_df.copy()\n",
    "targets_drug_valid_pubchem_df[\"PUB_CHEM_ID\"] = targets_drug_valid_pubchem_df[\"PUB_CHEM_ID\"].astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3960154c-9328-4195-8f40-4a0f202e55c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pubchem_to_smiles_df = pubchem_to_smiles_df[~pubchem_to_smiles_df[\"SMILES\"].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "834bbd36-09b9-478e-9ef3-827384e93d35",
   "metadata": {},
   "outputs": [],
   "source": [
    "targets_drug_valid_smiles_df = pd.merge(targets_drug_valid_pubchem_df, pubchem_to_smiles_df, on = \"PUB_CHEM_ID\", how = \"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0b613c27-56c5-41b0-8024-7f011a69eb3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CC(=O)OC1=CC=C(C=C1)C2(C3=CC=CC=C3NC2=O)C4=CC=C(C=C4)OC(=O)C'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets_drug_valid_smiles_df.loc[targets_drug_valid_smiles_df[\"DRUG_ID\"] == 1803, \"SMILES\"].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6c81e6f-4ac2-49cd-8e9e-16da92bd726b",
   "metadata": {},
   "source": [
    "## 2. SMILES to PyG graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "54a6c201-6e06-4fd4-8314-2efda3402106",
   "metadata": {},
   "outputs": [],
   "source": [
    "def one_hot_encoding(x, permitted_list):\n",
    "    \"\"\"\n",
    "    TODO: POINT OF IMPROVEMENT\n",
    "    Maps input elements x which are not in the permitted list to the last element\n",
    "    of the permitted list.\n",
    "    \"\"\"\n",
    "\n",
    "    if x not in permitted_list:\n",
    "        x = permitted_list[-1]\n",
    "\n",
    "    binary_encoding = [int(boolean_value) for boolean_value in list(map(lambda s: x == s, permitted_list))]\n",
    "\n",
    "    return binary_encoding\n",
    "\n",
    "def get_atom_features(atom, \n",
    "                  hydrogens_implicit = True):\n",
    "    \"\"\"\n",
    "    Takes an RDKit atom object as input and gives a 1d-numpy array of atom features as output.\n",
    "    \"\"\"    \n",
    "    # compute atom features\n",
    "\n",
    "    electronegativity_table = {\n",
    "        'H': 2.20, 'Li': 0.98, 'Be': 1.57, 'B': 2.04, 'C': 2.55, 'N': 3.04, 'O': 3.44, \n",
    "        'F': 3.98, 'Na': 0.93, 'Mg': 1.31, 'Al': 1.61, 'Si': 1.90, 'P': 2.19, 'S': 2.58, \n",
    "        'Cl': 3.16, 'K': 0.82, 'Ca': 1.00, 'Br': 2.96, 'I': 2.66\n",
    "    }\n",
    "    \n",
    "    valence_electrons_table = {\n",
    "        'H': 1, 'He': 2, 'Li': 1, 'Be': 2, 'B': 3, 'C': 4, 'N': 5, 'O': 6, 'F': 7, 'Ne': 8,\n",
    "        'Na': 1, 'Mg': 2, 'Al': 3, 'Si': 4, 'P': 5, 'S': 6, 'Cl': 7, 'Ar': 8, \n",
    "        'K': 1, 'Ca': 2, 'Br': 7, 'I': 7\n",
    "    }\n",
    "    \n",
    "    atom_number = atom.GetAtomicNum()\n",
    "    symbol = atom.GetSymbol()\n",
    "    atom_mass = atom.GetMass()\n",
    "    electronegativity = electronegativity_table[symbol]\n",
    "    valence_electrons = valence_electrons_table[symbol]\n",
    "    atom_type_enc = [atom_number, atom_mass, electronegativity, valence_electrons]\n",
    "    \n",
    "    n_heavy_neighbors_enc = [int(atom.GetDegree())] \n",
    "    formal_charge_enc = [int(atom.GetFormalCharge())]\n",
    "    hybridisation_type_enc = one_hot_encoding(str(atom.GetHybridization()), [\"S\", \"SP\", \"SP2\", \"SP3\", \"SP3D\", \"SP3D2\", \"OTHER\"])\n",
    "    \n",
    "    is_in_a_ring_enc = [int(atom.IsInRing())]\n",
    "    is_aromatic_enc = [int(atom.GetIsAromatic())]\n",
    "\n",
    "    atom_feature_vector = atom_type_enc + n_heavy_neighbors_enc + formal_charge_enc + hybridisation_type_enc + is_in_a_ring_enc + is_aromatic_enc\n",
    "    \n",
    "    if hydrogens_implicit == True:\n",
    "        n_hydrogens_enc = [int(atom.GetTotalNumHs())]\n",
    "        atom_feature_vector += n_hydrogens_enc\n",
    "\n",
    "    return np.array(atom_feature_vector)\n",
    "\n",
    "def get_bond_features(bond, \n",
    "                  use_stereochemistry = True):\n",
    "    \"\"\"\n",
    "    Takes an RDKit bond object as input and gives a 1d-numpy array of bond features as output.\n",
    "    \"\"\"\n",
    "\n",
    "    permitted_list_of_bond_types = [Chem.rdchem.BondType.SINGLE, Chem.rdchem.BondType.DOUBLE, Chem.rdchem.BondType.TRIPLE, Chem.rdchem.BondType.AROMATIC]\n",
    "\n",
    "    bond_type_enc = one_hot_encoding(bond.GetBondType(), permitted_list_of_bond_types)\n",
    "    \n",
    "    bond_is_conj_enc = [int(bond.GetIsConjugated())]\n",
    "    \n",
    "    bond_is_in_ring_enc = [int(bond.IsInRing())]\n",
    "    \n",
    "    bond_feature_vector = bond_type_enc + bond_is_conj_enc + bond_is_in_ring_enc\n",
    "    \n",
    "    if use_stereochemistry == True:\n",
    "        stereo_type_enc = one_hot_encoding(str(bond.GetStereo()), [\"STEREOZ\", \"STEREOE\", \"STEREOANY\", \"STEREONONE\"])\n",
    "        bond_feature_vector += stereo_type_enc\n",
    "\n",
    "    return np.array(bond_feature_vector)\n",
    "\n",
    "def create_pytorch_geometric_graph_data_list_from_smiles(x_smiles):\n",
    "    \"\"\"\n",
    "    Inputs:\n",
    "    \n",
    "    x_smiles = [smiles_1, smiles_2, ....] ... a list of SMILES strings\n",
    "    \n",
    "    Outputs:\n",
    "    \n",
    "    data_list = [G_1, G_2, ...] ... a list of torch_geometric.data.Data objects which represent labeled molecular graphs that can readily be used for machine learning\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    data_list = []\n",
    "    \n",
    "    for i, smiles in enumerate(x_smiles):\n",
    "        \n",
    "        # convert SMILES to RDKit mol object\n",
    "        mol = Chem.MolFromSmiles(smiles)\n",
    "\n",
    "        # get feature dimensions\n",
    "        n_nodes = mol.GetNumAtoms()\n",
    "        n_edges = 2*mol.GetNumBonds()\n",
    "        unrelated_smiles = \"O=O\"\n",
    "        unrelated_mol = Chem.MolFromSmiles(unrelated_smiles)\n",
    "        n_node_features = len(get_atom_features(unrelated_mol.GetAtomWithIdx(0)))\n",
    "        n_edge_features = len(get_bond_features(unrelated_mol.GetBondBetweenAtoms(0,1)))\n",
    "\n",
    "        # construct node feature matrix X of shape (n_nodes, n_node_features)\n",
    "        X = np.zeros((n_nodes, n_node_features))\n",
    "\n",
    "        for atom in mol.GetAtoms():\n",
    "            X[atom.GetIdx(), :] = get_atom_features(atom)\n",
    "            \n",
    "        X = torch.tensor(X, dtype = torch.float)\n",
    "        \n",
    "        # construct edge index array E of shape (2, n_edges)\n",
    "        (rows, cols) = np.nonzero(GetAdjacencyMatrix(mol))\n",
    "        torch_rows = torch.from_numpy(rows.astype(np.int64)).to(torch.long)\n",
    "        torch_cols = torch.from_numpy(cols.astype(np.int64)).to(torch.long)\n",
    "        E = torch.stack([torch_rows, torch_cols], dim = 0)\n",
    "        \n",
    "        # construct edge feature array EF of shape (n_edges, n_edge_features)\n",
    "        EF = np.zeros((n_edges, n_edge_features))\n",
    "        \n",
    "        for (k, (i,j)) in enumerate(zip(rows, cols)):\n",
    "            \n",
    "            EF[k] = get_bond_features(mol.GetBondBetweenAtoms(int(i),int(j)))\n",
    "        \n",
    "        EF = torch.tensor(EF, dtype = torch.float)\n",
    "        \n",
    "        \n",
    "        # construct Pytorch Geometric data object and append to data list\n",
    "        data_list.append(Data(x = X, edge_index = E, edge_attr = EF))\n",
    "\n",
    "    return data_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "e3118fb2-cc57-49aa-8fd9-c85a09f78b44",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_smiles = targets_drug_valid_smiles_df[\"SMILES\"].unique().tolist()\n",
    "unique_smiles_graphs = create_pytorch_geometric_graph_data_list_from_smiles(unique_smiles)\n",
    "smile_to_smiles_df = pd.DataFrame(columns = [\"SMILES\", \"SMILES_GRAPH\"], data = zip(unique_smiles, unique_smiles_graphs))\n",
    "targets_drug_valid_smiles_graphs_df = pd.merge(targets_drug_valid_smiles_df, smile_to_smiles_df, on = \"SMILES\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "ca85b17d-02fc-4e64-a454-87a8f1ee705c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>NLME_RESULT_ID</th>\n",
       "      <th>NLME_CURVE_ID</th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>TCGA_DESC</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PUTATIVE_TARGET</th>\n",
       "      <th>...</th>\n",
       "      <th>Name</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Targets</th>\n",
       "      <th>Target pathway</th>\n",
       "      <th>PUB_CHEM_ID</th>\n",
       "      <th>Datasets</th>\n",
       "      <th>number of cell lines</th>\n",
       "      <th>Screening site</th>\n",
       "      <th>SMILES</th>\n",
       "      <th>SMILES_GRAPH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18945558</td>\n",
       "      <td>683667</td>\n",
       "      <td>PFSK-1</td>\n",
       "      <td>SIDM01132</td>\n",
       "      <td>MB</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>...</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>Camptothecine, (+)-Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>24360</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>968</td>\n",
       "      <td>SANGER</td>\n",
       "      <td>CC[C@@]1(C2=C(COC1=O)C(=O)N3CC4=CC5=CC=CC=C5N=...</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18945796</td>\n",
       "      <td>684052</td>\n",
       "      <td>A673</td>\n",
       "      <td>SIDM00848</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>...</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>Camptothecine, (+)-Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>24360</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>968</td>\n",
       "      <td>SANGER</td>\n",
       "      <td>CC[C@@]1(C2=C(COC1=O)C(=O)N3CC4=CC5=CC=CC=C5N=...</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18946078</td>\n",
       "      <td>684057</td>\n",
       "      <td>ES5</td>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>...</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>Camptothecine, (+)-Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>24360</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>968</td>\n",
       "      <td>SANGER</td>\n",
       "      <td>CC[C@@]1(C2=C(COC1=O)C(=O)N3CC4=CC5=CC=CC=C5N=...</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18946335</td>\n",
       "      <td>684059</td>\n",
       "      <td>ES7</td>\n",
       "      <td>SIDM00269</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>...</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>Camptothecine, (+)-Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>24360</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>968</td>\n",
       "      <td>SANGER</td>\n",
       "      <td>CC[C@@]1(C2=C(COC1=O)C(=O)N3CC4=CC5=CC=CC=C5N=...</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>18946617</td>\n",
       "      <td>684062</td>\n",
       "      <td>EW-11</td>\n",
       "      <td>SIDM00203</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1003</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>...</td>\n",
       "      <td>Camptothecin</td>\n",
       "      <td>Camptothecine, (+)-Camptothecin</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>DNA replication</td>\n",
       "      <td>24360</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>968</td>\n",
       "      <td>SANGER</td>\n",
       "      <td>CC[C@@]1(C2=C(COC1=O)C(=O)N3CC4=CC5=CC=CC=C5N=...</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201695</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>19187483</td>\n",
       "      <td>1659928</td>\n",
       "      <td>SNU-175</td>\n",
       "      <td>SIDM00216</td>\n",
       "      <td>COREAD</td>\n",
       "      <td>2359</td>\n",
       "      <td>GSK2830371</td>\n",
       "      <td>WIP1</td>\n",
       "      <td>...</td>\n",
       "      <td>GSK2830371</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WIP1</td>\n",
       "      <td>Other</td>\n",
       "      <td>70983932</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>731</td>\n",
       "      <td>SANGER</td>\n",
       "      <td>CC1=C(C=C(C=N1)Cl)NCC2=CC=C(S2)C(=O)N[C@@H](CC...</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201696</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>19187936</td>\n",
       "      <td>1660034</td>\n",
       "      <td>SNU-407</td>\n",
       "      <td>SIDM00214</td>\n",
       "      <td>COREAD</td>\n",
       "      <td>2359</td>\n",
       "      <td>GSK2830371</td>\n",
       "      <td>WIP1</td>\n",
       "      <td>...</td>\n",
       "      <td>GSK2830371</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WIP1</td>\n",
       "      <td>Other</td>\n",
       "      <td>70983932</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>731</td>\n",
       "      <td>SANGER</td>\n",
       "      <td>CC1=C(C=C(C=N1)Cl)NCC2=CC=C(S2)C(=O)N[C@@H](CC...</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201697</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>19188194</td>\n",
       "      <td>1660035</td>\n",
       "      <td>SNU-61</td>\n",
       "      <td>SIDM00194</td>\n",
       "      <td>COREAD</td>\n",
       "      <td>2359</td>\n",
       "      <td>GSK2830371</td>\n",
       "      <td>WIP1</td>\n",
       "      <td>...</td>\n",
       "      <td>GSK2830371</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WIP1</td>\n",
       "      <td>Other</td>\n",
       "      <td>70983932</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>731</td>\n",
       "      <td>SANGER</td>\n",
       "      <td>CC1=C(C=C(C=N1)Cl)NCC2=CC=C(S2)C(=O)N[C@@H](CC...</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201698</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>19188734</td>\n",
       "      <td>1674021</td>\n",
       "      <td>SNU-C5</td>\n",
       "      <td>SIDM00498</td>\n",
       "      <td>COREAD</td>\n",
       "      <td>2359</td>\n",
       "      <td>GSK2830371</td>\n",
       "      <td>WIP1</td>\n",
       "      <td>...</td>\n",
       "      <td>GSK2830371</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WIP1</td>\n",
       "      <td>Other</td>\n",
       "      <td>70983932</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>731</td>\n",
       "      <td>SANGER</td>\n",
       "      <td>CC1=C(C=C(C=N1)Cl)NCC2=CC=C(S2)C(=O)N[C@@H](CC...</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201699</th>\n",
       "      <td>GDSC2</td>\n",
       "      <td>401</td>\n",
       "      <td>19189016</td>\n",
       "      <td>1789883</td>\n",
       "      <td>DiFi</td>\n",
       "      <td>SIDM00049</td>\n",
       "      <td>COREAD</td>\n",
       "      <td>2359</td>\n",
       "      <td>GSK2830371</td>\n",
       "      <td>WIP1</td>\n",
       "      <td>...</td>\n",
       "      <td>GSK2830371</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WIP1</td>\n",
       "      <td>Other</td>\n",
       "      <td>70983932</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>731</td>\n",
       "      <td>SANGER</td>\n",
       "      <td>CC1=C(C=C(C=N1)Cl)NCC2=CC=C(S2)C(=O)N[C@@H](CC...</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>201700 rows  29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       DATASET  NLME_RESULT_ID  NLME_CURVE_ID  COSMIC_ID CELL_LINE_NAME  \\\n",
       "0        GDSC2             401       18945558     683667         PFSK-1   \n",
       "1        GDSC2             401       18945796     684052           A673   \n",
       "2        GDSC2             401       18946078     684057            ES5   \n",
       "3        GDSC2             401       18946335     684059            ES7   \n",
       "4        GDSC2             401       18946617     684062          EW-11   \n",
       "...        ...             ...            ...        ...            ...   \n",
       "201695   GDSC2             401       19187483    1659928        SNU-175   \n",
       "201696   GDSC2             401       19187936    1660034        SNU-407   \n",
       "201697   GDSC2             401       19188194    1660035         SNU-61   \n",
       "201698   GDSC2             401       19188734    1674021         SNU-C5   \n",
       "201699   GDSC2             401       19189016    1789883           DiFi   \n",
       "\n",
       "       SANGER_MODEL_ID     TCGA_DESC  DRUG_ID     DRUG_NAME PUTATIVE_TARGET  \\\n",
       "0            SIDM01132            MB     1003  Camptothecin            TOP1   \n",
       "1            SIDM00848  UNCLASSIFIED     1003  Camptothecin            TOP1   \n",
       "2            SIDM00263  UNCLASSIFIED     1003  Camptothecin            TOP1   \n",
       "3            SIDM00269  UNCLASSIFIED     1003  Camptothecin            TOP1   \n",
       "4            SIDM00203  UNCLASSIFIED     1003  Camptothecin            TOP1   \n",
       "...                ...           ...      ...           ...             ...   \n",
       "201695       SIDM00216        COREAD     2359    GSK2830371            WIP1   \n",
       "201696       SIDM00214        COREAD     2359    GSK2830371            WIP1   \n",
       "201697       SIDM00194        COREAD     2359    GSK2830371            WIP1   \n",
       "201698       SIDM00498        COREAD     2359    GSK2830371            WIP1   \n",
       "201699       SIDM00049        COREAD     2359    GSK2830371            WIP1   \n",
       "\n",
       "        ...          Name                         Synonyms  Targets  \\\n",
       "0       ...  Camptothecin  Camptothecine, (+)-Camptothecin     TOP1   \n",
       "1       ...  Camptothecin  Camptothecine, (+)-Camptothecin     TOP1   \n",
       "2       ...  Camptothecin  Camptothecine, (+)-Camptothecin     TOP1   \n",
       "3       ...  Camptothecin  Camptothecine, (+)-Camptothecin     TOP1   \n",
       "4       ...  Camptothecin  Camptothecine, (+)-Camptothecin     TOP1   \n",
       "...     ...           ...                              ...      ...   \n",
       "201695  ...    GSK2830371                              NaN     WIP1   \n",
       "201696  ...    GSK2830371                              NaN     WIP1   \n",
       "201697  ...    GSK2830371                              NaN     WIP1   \n",
       "201698  ...    GSK2830371                              NaN     WIP1   \n",
       "201699  ...    GSK2830371                              NaN     WIP1   \n",
       "\n",
       "         Target pathway  PUB_CHEM_ID   Datasets   number of cell lines  \\\n",
       "0       DNA replication        24360      GDSC2                    968   \n",
       "1       DNA replication        24360      GDSC2                    968   \n",
       "2       DNA replication        24360      GDSC2                    968   \n",
       "3       DNA replication        24360      GDSC2                    968   \n",
       "4       DNA replication        24360      GDSC2                    968   \n",
       "...                 ...          ...        ...                    ...   \n",
       "201695            Other     70983932      GDSC2                    731   \n",
       "201696            Other     70983932      GDSC2                    731   \n",
       "201697            Other     70983932      GDSC2                    731   \n",
       "201698            Other     70983932      GDSC2                    731   \n",
       "201699            Other     70983932      GDSC2                    731   \n",
       "\n",
       "         Screening site                                             SMILES  \\\n",
       "0                SANGER  CC[C@@]1(C2=C(COC1=O)C(=O)N3CC4=CC5=CC=CC=C5N=...   \n",
       "1                SANGER  CC[C@@]1(C2=C(COC1=O)C(=O)N3CC4=CC5=CC=CC=C5N=...   \n",
       "2                SANGER  CC[C@@]1(C2=C(COC1=O)C(=O)N3CC4=CC5=CC=CC=C5N=...   \n",
       "3                SANGER  CC[C@@]1(C2=C(COC1=O)C(=O)N3CC4=CC5=CC=CC=C5N=...   \n",
       "4                SANGER  CC[C@@]1(C2=C(COC1=O)C(=O)N3CC4=CC5=CC=CC=C5N=...   \n",
       "...                 ...                                                ...   \n",
       "201695           SANGER  CC1=C(C=C(C=N1)Cl)NCC2=CC=C(S2)C(=O)N[C@@H](CC...   \n",
       "201696           SANGER  CC1=C(C=C(C=N1)Cl)NCC2=CC=C(S2)C(=O)N[C@@H](CC...   \n",
       "201697           SANGER  CC1=C(C=C(C=N1)Cl)NCC2=CC=C(S2)C(=O)N[C@@H](CC...   \n",
       "201698           SANGER  CC1=C(C=C(C=N1)Cl)NCC2=CC=C(S2)C(=O)N[C@@H](CC...   \n",
       "201699           SANGER  CC1=C(C=C(C=N1)Cl)NCC2=CC=C(S2)C(=O)N[C@@H](CC...   \n",
       "\n",
       "                                             SMILES_GRAPH  \n",
       "0       [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...  \n",
       "1       [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...  \n",
       "2       [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...  \n",
       "3       [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...  \n",
       "4       [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...  \n",
       "...                                                   ...  \n",
       "201695  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...  \n",
       "201696  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...  \n",
       "201697  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...  \n",
       "201698  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...  \n",
       "201699  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...  \n",
       "\n",
       "[201700 rows x 29 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets_drug_valid_smiles_graphs_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76a2e001-a941-418b-9ca4-0597fb4a5cca",
   "metadata": {},
   "source": [
    "## Create gene, mutation, and copy number df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "7d938606-5af4-45ae-9b9b-a359c5996858",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_transformed_merged_drug_model_df(models, targets):\n",
    "    columns_with_most_val_filled = models.columns[models.isnull().sum(axis=0)/ models.shape[0] <= 0.1]\n",
    "    filtered_columns = columns_with_most_val_filled\n",
    "    models = models[filtered_columns].fillna(\"Unkown\")\n",
    "    columns_besides_id = list(filter(lambda x: x != \"model_id\", filtered_columns))\n",
    "    models = models.rename(columns={\"model_id\":\"SANGER_MODEL_ID\"}) ##rename for join\n",
    "    limited_drug_target = targets[targets[\"DRUG_ID\"] == 1803] ##filter specific drug #1862\n",
    "    limited_drug_target_filtered = limited_drug_target.loc[:, [\"CELL_LINE_NAME\", \"SANGER_MODEL_ID\", \"MIN_CONC\", \"LN_IC50\"]]\n",
    "    model_drug_information = pd.merge(limited_drug_target_filtered, models, how=\"inner\", on=\"SANGER_MODEL_ID\") ## merge drug and model information\n",
    "\n",
    "    model_drug_information[\"model_idx\"] = model_drug_information.index ##store index for later joins to retrieve edge index\n",
    "    return model_drug_information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "7b7e7328-d112-419e-82c6-3e324f6af48b",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_drug_model_df = get_transformed_merged_drug_model_df(models,  targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "d434ad81-5856-46a4-89c0-f454927a5f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "copy_numbers = pd.read_csv(\"copy_number_variation.csv\")\n",
    "copy_numbers_df = pd.pivot_table(copy_numbers, values='total_copy_number', index=['model_id'], columns=['symbol'], aggfunc=\"sum\", fill_value=0.0)\n",
    "copy_numbers_df = copy_numbers_df.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "a8faa3c1-a5ee-44c1-a751-d346fc97113d",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations = pd.read_csv(\"mutations.csv\")\n",
    "mutations[\"one_hot\"] = np.ones(mutations.shape[0])\n",
    "## TODO possible improvement, e.g., VAF\n",
    "mutations_df = pd.pivot_table(mutations, values=\"one_hot\", index=['model_id'],\n",
    "                       columns=['gene_symbol'], aggfunc=\"sum\", fill_value=0.0).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "a4f92d27-9764-4fdc-bb54-6608a8a18bd6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "rna_seq_df = pd.read_csv(\"gene_expression.csv\",index_col=0, header = None,low_memory=False)\n",
    "\n",
    "def get_transformed_df_in_model_gene_value_format(rna_seq_df):\n",
    "    rna_seq_df_transposed = rna_seq_df.transpose().iloc[0:, :]\n",
    "    new_header = rna_seq_df_transposed.iloc[0, :] \n",
    "    rna_seq_df_transposed = rna_seq_df_transposed.iloc[1:, :] \n",
    "    rna_seq_df_transposed.columns = new_header\n",
    "    rna_seq_df_transposed = rna_seq_df_transposed.iloc[:, [0, *list(range(5, new_header.shape[0]))]]\n",
    "    rna_seq_df_transposed = rna_seq_df_transposed.rename(columns={rna_seq_df_transposed.columns[0]: \"SANGER_MODEL_ID\"})\n",
    "    rna_seq_df_transposed_stacked = rna_seq_df_transposed.set_index('SANGER_MODEL_ID').stack().reset_index(name='ExpressionValue')\n",
    "    rna_seq_df_transposed_stacked = rna_seq_df_transposed_stacked.rename(columns={1:'symbol'})\n",
    "    return rna_seq_df_transposed_stacked\n",
    "\n",
    "transformed_seq_data = get_transformed_df_in_model_gene_value_format(rna_seq_df)\n",
    "transformed_seq_data[\"ExpressionValue\"] = transformed_seq_data[\"ExpressionValue\"].astype(np.float32)\n",
    "expression_df = pd.pivot_table(transformed_seq_data, values='ExpressionValue', index=['SANGER_MODEL_ID'],\n",
    "                       columns=['symbol'], aggfunc=\"sum\", fill_value=0.0)\n",
    "expression_df = expression_df.reset_index().rename(columns = {\"SANGER_MODEL_ID\": \"model_id\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "f6ea67b4-0bbb-4d33-b2b3-63d4822b821d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rename_col(df, type:str):\n",
    "    feature_col = list(filter(lambda col: col != \"model_id\", df.columns))\n",
    "    rename_col = {col: f\"{col}_{type}\" for col in feature_col}\n",
    "    df = df.rename(columns=rename_col)\n",
    "    return df\n",
    "expression_df = rename_col(expression_df, \"exp\")\n",
    "copy_numbers_df = rename_col(copy_numbers_df, \"cnv\") \n",
    "mutations_df = rename_col(mutations_df, \"mut\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "e7cad906-3cc6-4562-bbd6-203e123a072f",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_df = pd.merge(pd.merge(expression_df, copy_numbers_df, on = \"model_id\"), mutations_df, on = \"model_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "6b5f06e3-38b9-466b-831f-442a84a8b5c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_columns = list(filter(lambda col: col != \"model_id\", features_df.columns))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8ab7f64-3fd2-4f82-8955-bbd6687f8393",
   "metadata": {},
   "source": [
    "## 3. Create Masks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "71e890fc-dc04-4983-9061-c2b8e1bd42db",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = os.listdir(\"all_features\")\n",
    "filename_to_features = dict()\n",
    "for file in files:\n",
    "    if \"csv\" not in file:\n",
    "        continue\n",
    "    file_df = pd.read_csv(os.path.join(\"all_features\", file))\n",
    "    filename_to_features[file] = file_df.values.squeeze()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "a250bd75-bbe3-44ba-8b58-e932b9970698",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename_to_mask = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "5c0fe8cb-34d9-41b5-8eca-047824370159",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "39deed73baa047bb9e9a88ac3563592e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/295 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for file in tqdm(filename_to_features):\n",
    "    if file in filename_to_mask:\n",
    "        continue\n",
    "    mask = np.zeros(len(all_columns), dtype = np.bool_)\n",
    "    for i, col in enumerate(all_columns):\n",
    "        file_cols = filename_to_features[file]\n",
    "        file_cols = list(map(lambda file_col: file_col.replace(\".1\", \"\"), file_cols)) ## For some weird reason feature names changed slightly\n",
    "        mask[i] = col in file_cols\n",
    "    assert len(file_cols) == mask.sum(), file\n",
    "    filename_to_mask[file] = mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "a470817c-ed53-45c5-ba2b-b46a0b9733aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_id_to_mask = {key[4:8]: filename_to_mask[key] for key in filename_to_mask}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c48e78a4-1ffe-4bd7-b006-71070ca0686f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1d32de8a-1d02-455c-b247-87ae4bb1e346",
   "metadata": {},
   "source": [
    "## 4. Overall dataset with SMILES PyGGRaphs and Cell model features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "cdecdf00-b0fc-4a71-ba2c-d5a96e25b493",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model_id</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>SMILES_GRAPH</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>A1BG_exp</th>\n",
       "      <th>A1BG-AS1_exp</th>\n",
       "      <th>A1CF_exp</th>\n",
       "      <th>A2M_exp</th>\n",
       "      <th>A2M-AS1_exp</th>\n",
       "      <th>A2ML1_exp</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF429_mut</th>\n",
       "      <th>ZNF521_mut</th>\n",
       "      <th>ZNF626_mut</th>\n",
       "      <th>ZNF680_mut</th>\n",
       "      <th>ZNF721_mut</th>\n",
       "      <th>ZNF814_mut</th>\n",
       "      <th>ZNF93_mut</th>\n",
       "      <th>ZNRF3_mut</th>\n",
       "      <th>ZRSR2_mut</th>\n",
       "      <th>ZXDB_mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDM01132</td>\n",
       "      <td>-1.462148</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>1003</td>\n",
       "      <td>0.98</td>\n",
       "      <td>14.04</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.980000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDM00848</td>\n",
       "      <td>-4.869447</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>1003</td>\n",
       "      <td>0.84</td>\n",
       "      <td>5.57</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.150000</td>\n",
       "      <td>0.32</td>\n",
       "      <td>11.980000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>-3.360684</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>1003</td>\n",
       "      <td>4.00</td>\n",
       "      <td>15.64</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.620001</td>\n",
       "      <td>0.61</td>\n",
       "      <td>70.669998</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDM00269</td>\n",
       "      <td>-5.045014</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>1003</td>\n",
       "      <td>2.44</td>\n",
       "      <td>17.25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.560000</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.370000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDM00203</td>\n",
       "      <td>-3.741620</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>1003</td>\n",
       "      <td>1.88</td>\n",
       "      <td>11.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>0.20</td>\n",
       "      <td>7.670000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  38979 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    model_id   LN_IC50                                       SMILES_GRAPH  \\\n",
       "0  SIDM01132 -1.462148  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...   \n",
       "1  SIDM00848 -4.869447  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...   \n",
       "2  SIDM00263 -3.360684  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...   \n",
       "3  SIDM00269 -5.045014  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...   \n",
       "4  SIDM00203 -3.741620  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...   \n",
       "\n",
       "   DRUG_ID  A1BG_exp  A1BG-AS1_exp  A1CF_exp    A2M_exp  A2M-AS1_exp  \\\n",
       "0     1003      0.98         14.04       0.0   0.980000         0.00   \n",
       "1     1003      0.84          5.57       0.0   5.150000         0.32   \n",
       "2     1003      4.00         15.64       0.0  26.620001         0.61   \n",
       "3     1003      2.44         17.25       0.0  14.560000         0.08   \n",
       "4     1003      1.88         11.06       0.0   0.080000         0.20   \n",
       "\n",
       "   A2ML1_exp  ...  ZNF429_mut  ZNF521_mut  ZNF626_mut  ZNF680_mut  ZNF721_mut  \\\n",
       "0   0.030000  ...         0.0         0.0         0.0         0.0         0.0   \n",
       "1  11.980000  ...         0.0         0.0         0.0         0.0         0.0   \n",
       "2  70.669998  ...         0.0         0.0         0.0         0.0         0.0   \n",
       "3   0.370000  ...         0.0         0.0         0.0         0.0         0.0   \n",
       "4   7.670000  ...         0.0         0.0         0.0         0.0         0.0   \n",
       "\n",
       "   ZNF814_mut  ZNF93_mut  ZNRF3_mut  ZRSR2_mut  ZXDB_mut  \n",
       "0         0.0        0.0        0.0        0.0       0.0  \n",
       "1         0.0        0.0        0.0        0.0       0.0  \n",
       "2         0.0        0.0        0.0        0.0       0.0  \n",
       "3         0.0        0.0        0.0        0.0       0.0  \n",
       "4         0.0        0.0        0.0        0.0       0.0  \n",
       "\n",
       "[5 rows x 38979 columns]"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets_drug_valid_smiles_graphs_red_df = targets_drug_valid_smiles_graphs_df.loc[:, [\"SANGER_MODEL_ID\", \"LN_IC50\", \"SMILES_GRAPH\", \"DRUG_ID\"]]\n",
    "targets_drug_valid_smiles_graphs_red_df = targets_drug_valid_smiles_graphs_red_df.rename(columns = {\"SANGER_MODEL_ID\": \"model_id\"})\n",
    "targets_drug_valid_smiles_graphs_red_features_df = pd.merge(targets_drug_valid_smiles_graphs_red_df, features_df, on = \"model_id\", suffixes = (\"\", \"_r\"))\n",
    "targets_drug_valid_smiles_graphs_red_features_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "628833d5-23d8-4e56-a0e8-cb8dec47d711",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model_id</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>SMILES_GRAPH</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>A1BG_exp</th>\n",
       "      <th>A1BG-AS1_exp</th>\n",
       "      <th>A1CF_exp</th>\n",
       "      <th>A2M_exp</th>\n",
       "      <th>A2M-AS1_exp</th>\n",
       "      <th>A2ML1_exp</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF429_mut</th>\n",
       "      <th>ZNF521_mut</th>\n",
       "      <th>ZNF626_mut</th>\n",
       "      <th>ZNF680_mut</th>\n",
       "      <th>ZNF721_mut</th>\n",
       "      <th>ZNF814_mut</th>\n",
       "      <th>ZNF93_mut</th>\n",
       "      <th>ZNRF3_mut</th>\n",
       "      <th>ZRSR2_mut</th>\n",
       "      <th>ZXDB_mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDM01132</td>\n",
       "      <td>-1.462148</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>1003</td>\n",
       "      <td>0.98</td>\n",
       "      <td>14.04</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.980000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDM00848</td>\n",
       "      <td>-4.869447</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>1003</td>\n",
       "      <td>0.84</td>\n",
       "      <td>5.57</td>\n",
       "      <td>0.00</td>\n",
       "      <td>5.150000</td>\n",
       "      <td>0.32</td>\n",
       "      <td>11.980000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>-3.360684</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>1003</td>\n",
       "      <td>4.00</td>\n",
       "      <td>15.64</td>\n",
       "      <td>0.00</td>\n",
       "      <td>26.620001</td>\n",
       "      <td>0.61</td>\n",
       "      <td>70.669998</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDM00269</td>\n",
       "      <td>-5.045014</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>1003</td>\n",
       "      <td>2.44</td>\n",
       "      <td>17.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>14.560000</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.370000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDM00203</td>\n",
       "      <td>-3.741620</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>1003</td>\n",
       "      <td>1.88</td>\n",
       "      <td>11.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>0.20</td>\n",
       "      <td>7.670000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196308</th>\n",
       "      <td>SIDM00216</td>\n",
       "      <td>5.423870</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>2359</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.01</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196309</th>\n",
       "      <td>SIDM00214</td>\n",
       "      <td>5.042005</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>2359</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.80</td>\n",
       "      <td>3.400000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196310</th>\n",
       "      <td>SIDM00194</td>\n",
       "      <td>6.130028</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>2359</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.05</td>\n",
       "      <td>3.30</td>\n",
       "      <td>0.160000</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.610000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196311</th>\n",
       "      <td>SIDM00498</td>\n",
       "      <td>6.151265</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>2359</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.070000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196312</th>\n",
       "      <td>SIDM00049</td>\n",
       "      <td>5.757955</td>\n",
       "      <td>[(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...</td>\n",
       "      <td>2359</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.16</td>\n",
       "      <td>13.85</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>196313 rows  38979 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         model_id   LN_IC50  \\\n",
       "0       SIDM01132 -1.462148   \n",
       "1       SIDM00848 -4.869447   \n",
       "2       SIDM00263 -3.360684   \n",
       "3       SIDM00269 -5.045014   \n",
       "4       SIDM00203 -3.741620   \n",
       "...           ...       ...   \n",
       "196308  SIDM00216  5.423870   \n",
       "196309  SIDM00214  5.042005   \n",
       "196310  SIDM00194  6.130028   \n",
       "196311  SIDM00498  6.151265   \n",
       "196312  SIDM00049  5.757955   \n",
       "\n",
       "                                             SMILES_GRAPH  DRUG_ID  A1BG_exp  \\\n",
       "0       [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...     1003      0.98   \n",
       "1       [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...     1003      0.84   \n",
       "2       [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...     1003      4.00   \n",
       "3       [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...     1003      2.44   \n",
       "4       [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...     1003      1.88   \n",
       "...                                                   ...      ...       ...   \n",
       "196308  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...     2359      0.26   \n",
       "196309  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...     2359      0.35   \n",
       "196310  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...     2359      0.00   \n",
       "196311  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...     2359      0.03   \n",
       "196312  [(x, [tensor([ 6.0000, 12.0110,  2.5500,  4.00...     2359      0.06   \n",
       "\n",
       "        A1BG-AS1_exp  A1CF_exp    A2M_exp  A2M-AS1_exp  A2ML1_exp  ...  \\\n",
       "0              14.04      0.00   0.980000         0.00   0.030000  ...   \n",
       "1               5.57      0.00   5.150000         0.32  11.980000  ...   \n",
       "2              15.64      0.00  26.620001         0.61  70.669998  ...   \n",
       "3              17.25      0.00  14.560000         0.08   0.370000  ...   \n",
       "4              11.06      0.00   0.080000         0.20   7.670000  ...   \n",
       "...              ...       ...        ...          ...        ...  ...   \n",
       "196308          0.46      0.23   0.000000         1.01   0.030000  ...   \n",
       "196309          0.03      0.00   0.000000         0.80   3.400000  ...   \n",
       "196310          0.05      3.30   0.160000         0.16   1.610000  ...   \n",
       "196311          0.14      0.03   0.000000         0.00   0.070000  ...   \n",
       "196312          0.16     13.85   0.000000         0.13   0.000000  ...   \n",
       "\n",
       "        ZNF429_mut  ZNF521_mut  ZNF626_mut  ZNF680_mut  ZNF721_mut  \\\n",
       "0              0.0         0.0         0.0         0.0         0.0   \n",
       "1              0.0         0.0         0.0         0.0         0.0   \n",
       "2              0.0         0.0         0.0         0.0         0.0   \n",
       "3              0.0         0.0         0.0         0.0         0.0   \n",
       "4              0.0         0.0         0.0         0.0         0.0   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "196308         0.0         0.0         0.0         0.0         0.0   \n",
       "196309         0.0         0.0         0.0         0.0         0.0   \n",
       "196310         0.0         0.0         0.0         0.0         0.0   \n",
       "196311         0.0         0.0         0.0         0.0         0.0   \n",
       "196312         0.0         0.0         0.0         0.0         0.0   \n",
       "\n",
       "        ZNF814_mut  ZNF93_mut  ZNRF3_mut  ZRSR2_mut  ZXDB_mut  \n",
       "0              0.0        0.0        0.0        0.0       0.0  \n",
       "1              0.0        0.0        0.0        0.0       0.0  \n",
       "2              0.0        0.0        0.0        0.0       0.0  \n",
       "3              0.0        0.0        0.0        0.0       0.0  \n",
       "4              0.0        0.0        0.0        0.0       0.0  \n",
       "...            ...        ...        ...        ...       ...  \n",
       "196308         0.0        0.0        0.0        0.0       0.0  \n",
       "196309         0.0        0.0        0.0        0.0       0.0  \n",
       "196310         0.0        0.0        0.0        0.0       0.0  \n",
       "196311         0.0        0.0        0.0        0.0       0.0  \n",
       "196312         0.0        0.0        0.0        0.0       0.0  \n",
       "\n",
       "[196313 rows x 38979 columns]"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets_drug_valid_smiles_graphs_red_features_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae28fbcf-5c70-4d31-8a7e-d44ec84ac003",
   "metadata": {},
   "source": [
    "## Create graphs of feature-selected drug-specifc cell models "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "64e53a43-4449-4c8b-a877-1cbb5442f783",
   "metadata": {},
   "outputs": [],
   "source": [
    "acetalax_test = False ##boolean for testing my approach"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "552d936a-630a-47c0-8bae-6ea3f774cd60",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create batch graphs and one batch represents all combinations for one drug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "ddbb7d29-51b4-4610-ab10-b13f1064538f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1003\n",
      "1004\n",
      "1005\n",
      "1006\n",
      "1007\n",
      "1008\n",
      "1009\n",
      "1010\n",
      "1011\n",
      "1012\n",
      "1013\n",
      "1014\n",
      "1016\n",
      "1017\n",
      "1018\n",
      "1019\n",
      "1020\n",
      "1021\n",
      "1022\n",
      "1023\n",
      "1024\n",
      "1025\n",
      "1026\n",
      "1029\n",
      "1030\n",
      "1031\n",
      "1032\n",
      "1033\n",
      "1036\n",
      "1037\n",
      "1038\n",
      "1039\n",
      "1042\n",
      "1043\n",
      "1046\n",
      "1047\n",
      "1049\n",
      "1050\n",
      "1051\n",
      "1053\n",
      "1054\n",
      "1057\n",
      "1058\n",
      "1059\n",
      "1060\n",
      "1062\n",
      "1067\n",
      "1068\n",
      "1069\n",
      "1072\n",
      "1073\n",
      "1079\n",
      "1080\n",
      "1083\n",
      "1085\n",
      "1086\n",
      "1088\n",
      "1089\n",
      "1093\n",
      "1096\n",
      "1129\n",
      "1131\n",
      "1133\n",
      "1149\n",
      "1168\n",
      "1175\n",
      "1177\n",
      "1179\n",
      "1180\n",
      "1190\n",
      "1191\n",
      "1192\n",
      "1194\n",
      "1199\n",
      "1200\n",
      "1237\n",
      "1239\n",
      "1243\n",
      "1248\n",
      "1249\n",
      "1250\n",
      "1372\n",
      "1373\n",
      "1375\n",
      "1378\n",
      "1494\n",
      "1502\n",
      "1507\n",
      "1510\n",
      "1511\n",
      "1512\n",
      "1529\n",
      "1549\n",
      "1553\n",
      "1557\n",
      "1558\n",
      "1560\n",
      "1561\n",
      "1563\n",
      "1564\n",
      "1576\n",
      "1578\n",
      "1593\n",
      "1598\n",
      "1613\n",
      "1614\n",
      "1615\n",
      "1617\n",
      "1618\n",
      "1620\n",
      "1621\n",
      "1622\n",
      "1624\n",
      "1626\n",
      "1627\n",
      "1629\n",
      "1630\n",
      "1631\n",
      "1632\n",
      "1634\n",
      "1635\n",
      "1786\n",
      "1799\n",
      "1803\n",
      "1849\n",
      "1852\n",
      "1853\n",
      "1854\n",
      "1855\n",
      "1862\n",
      "1909\n",
      "1910\n",
      "1912\n",
      "1913\n",
      "1915\n",
      "1917\n",
      "1918\n",
      "1919\n",
      "1922\n",
      "1924\n",
      "1925\n",
      "1926\n",
      "1927\n",
      "1928\n",
      "1930\n",
      "1931\n",
      "1932\n",
      "1933\n",
      "1936\n",
      "1939\n",
      "1940\n",
      "1941\n",
      "1997\n",
      "2040\n",
      "2043\n",
      "2044\n",
      "2045\n",
      "2046\n",
      "2048\n",
      "2096\n",
      "2106\n",
      "2107\n",
      "2110\n",
      "2111\n",
      "2169\n",
      "2172\n",
      "2173\n",
      "2174\n",
      "2175\n",
      "2177\n",
      "2359\n"
     ]
    }
   ],
   "source": [
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "batch_graphs = []\n",
    "\n",
    "class Batch:\n",
    "    def __init__(self, id, drug_graph, cell_drug_graph,  train_labels, test_labels):\n",
    "        self.id = id\n",
    "        self.drug_graph = drug_graph\n",
    "        self.cell_drug_graph = cell_drug_graph\n",
    "        self.train_labels = train_labels\n",
    "        self.test_labels = test_labels\n",
    "        \n",
    "    def __repr__(self):\n",
    "        return f\"\"\"\n",
    "        self.id: {self.id}\n",
    "        self.drug_graph: {self.drug_graph.__repr__()}\\n\n",
    "        self.cell_drug_graph: {self.cell_drug_graph.__repr__()}\\n\n",
    "        self.train_labels: {self.train_labels.shape}\\n\n",
    "        self.test_labels: {self.test_labels.shape}\\n\n",
    "        \"\"\"\n",
    "\n",
    "def add_super_node_to_graph(graph):\n",
    "    graph= graph.cpu()\n",
    "    num_nodes, num_features = graph.x.shape\n",
    "    super_node = torch.zeros((1, num_features), dtype = graph.x.dtype)\n",
    "    graph_node_sources = torch.arange(num_nodes).unsqueeze(0)\n",
    "    super_node_targets = (torch.ones((num_nodes)) * (num_nodes)).unsqueeze(0)\n",
    "    super_edge_index = torch.cat((graph_node_sources, super_node_targets), dim = 0)\n",
    "    new_edge_index = torch.cat((graph.edge_index, super_edge_index), dim  = 1)\n",
    "    new_features = torch.cat((graph.x, super_node), dim = 0)\n",
    "    new_edge_attr = torch.cat((graph.edge_attr, torch.zeros((super_edge_index.shape[-1], graph.edge_attr.shape[-1]), dtype = graph.edge_attr.dtype)), dim  = 0)\n",
    "    graph.x = new_features\n",
    "    graph.edge_index = new_edge_index.type(torch.long)\n",
    "    graph.edge_attr = new_edge_attr\n",
    "    graph.super_node_mask = torch.zeros(graph.x.shape[0],dtype =torch.bool)\n",
    "    graph.super_node_mask[-1] = 1\n",
    "\n",
    "def get_edges(group_features, id, drug_id_to_mask):\n",
    "    all_drug_specific_cell_model_features = torch.from_numpy(group_features).type(torch.float32)\n",
    "    selected_feature_index_np = np.nonzero(drug_id_to_mask[str(id)])[0] # Num_Sel_features\n",
    "    selected_feature_index = torch.from_numpy(selected_feature_index_np).type(torch.long) # Num_Sel_features\n",
    "    selected_feature_index_exp = selected_feature_index.unsqueeze(0).expand(group_features.shape[0], selected_feature_index.shape[-1]) # Num_CM x Num_Sel_features\n",
    "    source_edge_index = selected_feature_index_exp.flatten() # (Num_CM x Num_Sel_features) x 1\n",
    "    \n",
    "    target_edge_index = torch.arange(group_features.shape[0]).unsqueeze(1).expand(group_features.shape[0], selected_feature_index_exp.shape[-1])\n",
    "    target_edge_index = target_edge_index.flatten()\n",
    "    edge_index = torch.cat([source_edge_index.unsqueeze(0), target_edge_index.unsqueeze(0)], dim = 0)\n",
    "    edge_attr = all_drug_specific_cell_model_features[:, selected_feature_index_np].flatten()\n",
    "    return (edge_index, edge_attr)\n",
    "\n",
    "for id, group in targets_drug_valid_smiles_graphs_red_features_df.groupby(\"DRUG_ID\"):\n",
    "    if acetalax_test and id != 1803:\n",
    "        continue\n",
    "    print(id)\n",
    "    group_features = group.iloc[:, 4:].values\n",
    "    node_features = torch.ones(group_features.shape[-1]) # Num_CMfeatures\n",
    "\n",
    "    #TODO: We have to use the same train test splits as in feature selector -> Label leak!\n",
    "    train_df, test_df = train_test_split(group, test_size = .1) \n",
    "    train_features, test_features = train_df.iloc[:, 4:].values, test_df.iloc[:, 4:].values\n",
    "    \n",
    "    train_edge_index, train_edge_attr = get_edges(train_features, id, drug_id_to_mask)\n",
    "    test_edge_index, test_edge_attr = get_edges(test_features, id, drug_id_to_mask)\n",
    "    \n",
    "    cell_model_graph = Data(x = node_features, train_edge_attr = train_edge_attr, train_edge_index= train_edge_index, test_edge_index = test_edge_index, test_edge_attr = test_edge_attr)\n",
    "    drug_graph = group.iloc[0, :][\"SMILES_GRAPH\"]\n",
    "    train_labels = torch.from_numpy(train_df[\"LN_IC50\"].values).type(torch.float32)\n",
    "    test_labels = torch.from_numpy(test_df[\"LN_IC50\"].values).type(torch.float32)\n",
    "    add_super_node_to_graph(drug_graph)\n",
    "    batch = Batch(id, drug_graph, cell_model_graph, train_labels, test_labels)\n",
    "    batch_graphs.append(batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "56dea019-8311-4739-8d9f-8d2de25c9c24",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.linear_model import LinearRegression\n",
    "\n",
    "def evaluate_lr_results():\n",
    "    lr_results = np.zeros((len(batch_graphs), 2))\n",
    "\n",
    "    for i, batch in enumerate(batch_graphs):\n",
    "        num_cell_model_features = drug_id_to_mask[str(batch.id)].sum()\n",
    "        train_features = batch.cell_drug_graph.train_edge_attr.view(-1, num_cell_model_features)\n",
    "        test_features = batch.cell_drug_graph.test_edge_attr.view(-1, num_cell_model_features)\n",
    "        train_labels = batch.train_labels\n",
    "        test_labels = batch.test_labels\n",
    "    \n",
    "        lr = LinearRegression()\n",
    "        lr.fit(train_features, train_labels)\n",
    "        pred_test = lr.predict(test_features)\n",
    "        pred_train = lr.predict(train_features)\n",
    "        r2_train, r2_test = r2_score(train_labels, pred_train), r2_score(test_labels, pred_test)\n",
    "        lr_results[i, 0] = r2_train\t\n",
    "        lr_results[i, 1] = r2_test\n",
    "    return lr_results\n",
    "lr_res = evaluate_lr_results()\n",
    "lr_res_df = pd.DataFrame(lr_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "23b4699d-7125-4d7d-bf54-2d4e6e772fc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>171.000000</td>\n",
       "      <td>171.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.943787</td>\n",
       "      <td>0.782192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.059853</td>\n",
       "      <td>0.244394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.439207</td>\n",
       "      <td>-0.904345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.938062</td>\n",
       "      <td>0.752507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.961491</td>\n",
       "      <td>0.834210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.972919</td>\n",
       "      <td>0.913223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.988200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                0           1\n",
       "count  171.000000  171.000000\n",
       "mean     0.943787    0.782192\n",
       "std      0.059853    0.244394\n",
       "min      0.439207   -0.904345\n",
       "25%      0.938062    0.752507\n",
       "50%      0.961491    0.834210\n",
       "75%      0.972919    0.913223\n",
       "max      1.000000    0.988200"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lr_res_df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f4db919f-26fa-47a3-8b21-b4c5f8c29076",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## TODO: We have to use the same train test splits as in feature selector\n",
    "# ## Train test split rather over different cell models (i.e., each drug at least seen once?) -> if we dont know we have to check the paper\n",
    "# train_graphs, test_graphs = train_test_split(batch_graphs) if not acetalax_test else ([], [])\n",
    "# len(train_graphs), len(test_graphs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "6d2facd9-1b7d-4792-a327-6d07cfb79657",
   "metadata": {},
   "outputs": [],
   "source": [
    "if acetalax_test:\n",
    "    test_edge_index = batch_graphs[0].cell_drug_graph.edge_index[:, -(436)*200:]\n",
    "    test_edge_attr = batch_graphs[0].cell_drug_graph.edge_attr[-(436)*200:]\n",
    "    \n",
    "    train_edge_index = batch_graphs[0].cell_drug_graph.edge_index[:, :-(436)*200]\n",
    "    train_edge_attr = batch_graphs[0].cell_drug_graph.edge_attr[:-(436)*200]\n",
    "    \n",
    "    train_graph = Batch(batch_graphs[0].drug_graph.clone(),\n",
    "                        batch_graphs[0].cell_drug_graph.clone(),\n",
    "                        torch.clone(batch_graphs[0].labels))\n",
    "    test_graph = Batch(batch_graphs[0].drug_graph.clone(),\n",
    "                        batch_graphs[0].cell_drug_graph.clone(),\n",
    "                        torch.clone(batch_graphs[0].labels))\n",
    "    \n",
    "    train_graph.cell_drug_graph.edge_index = train_edge_index\n",
    "    train_graph.cell_drug_graph.edge_attr = train_edge_attr\n",
    "    train_graph.labels = train_graph.labels[:-200]\n",
    "    \n",
    "    test_graph.cell_drug_graph.edge_index = test_edge_index\n",
    "    test_graph.cell_drug_graph.edge_attr = test_edge_attr\n",
    "    test_graph.cell_drug_graph.edge_index[1] = test_graph.cell_drug_graph.edge_index[1] -501\n",
    "    test_graph.labels = test_graph.labels[-200:]\n",
    "    \n",
    "    train_graphs = [train_graph]\n",
    "    test_graphs = [test_graph]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "21af8767-ccfb-47ea-9eb5-6c0e1f89bd5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "# This function is copied from https://github.com/gordicaleksa/pytorch-GAT/blob/main/The%20Annotated%20GAT%20(PPI).ipynb\n",
    "## Thanks to Aleksa Gordic for the implementation!\n",
    "def explicit_broadcast(this, other):\n",
    "    # Append singleton dimensions until this.dim() == other.dim()\n",
    "    for _ in range(this.dim(), other.dim()):\n",
    "        this = this.unsqueeze(-1)\n",
    "    return this.expand_as(other)\n",
    "    \n",
    "ACTIVATION_FUNS = {\n",
    "    \"RELU\": torch.nn.ReLU,\n",
    "    \"LEAKY_RELU\": torch.nn.LeakyReLU,\n",
    "    \"ELU\":  torch.nn.ELU\n",
    "}\n",
    "class GNN(torch.nn.Module):\n",
    "    def __init__(self, hidden_dim, dropout, heads, act):\n",
    "        super(GNN, self).__init__()\n",
    "        hidden_dim = int(hidden_dim) ## TODO add further params\n",
    "        self.conv_0 = GATConv(-1, hidden_dim, heads=heads, concat=True, add_self_loops=True, dropout=dropout)\n",
    "        self.conv_1 = GATConv(hidden_dim*heads, hidden_dim, heads = heads, concat=False, add_self_loops= True, dropout=dropout)\n",
    "        \n",
    "        self.dropout = torch.nn.Dropout(p = dropout)\n",
    "        self.activate = ACTIVATION_FUNS[act]()\n",
    "    def forward(self, x, edge_index, edge_attr):\n",
    "        ##TODO check if dropout in input might make sense\n",
    "        x = self.conv_0(x, edge_index, edge_attr)\n",
    "        x = self.activate(x)\n",
    "        x = self.dropout(x)\n",
    "        x = self.conv_1(x, edge_index, edge_attr)\n",
    "        return x\n",
    "        \n",
    "class CustomGNN(torch.nn.Module):\n",
    "    def __init__(self, num_cell_model_features, hidden_dim_cell_models, hidden_dim_drug, dropout, heads, act, device):\n",
    "        super(CustomGNN, self).__init__()\n",
    "        self.lin = torch.nn.Linear(num_cell_model_features, hidden_dim_cell_models)\n",
    "        \n",
    "        self.gnn_drug = GNN(hidden_dim_drug, dropout, heads, act) # (Atoms + 1) x H_d\n",
    "        self.lin_end = torch.nn.Linear(hidden_dim_cell_models + hidden_dim_drug, 1)\n",
    "        self.device = device\n",
    "\n",
    "    def forward(self, drug_graph, cell_drug_graph):\n",
    "        ## Drug-sepcific CellModel Information\n",
    "        cell_node_features =  self.lin.weight.transpose(1,0).clone() * cell_drug_graph.x.unsqueeze(1).clone()\n",
    "        cell_node_features_lift = torch.index_select(cell_node_features.clone(), 0, cell_drug_graph.edge_index[0].clone()) # E_cd x H_c\n",
    "        \n",
    "        drug_specific_cell_features = cell_node_features_lift.clone() * cell_drug_graph.edge_attr.unsqueeze(1).clone() # E_cd x H_c\n",
    "        drug_specific_cell_features_aggregated = torch.zeros((cell_drug_graph.edge_index[1].max() + 1, drug_specific_cell_features.shape[1]), device = self.device, dtype=drug_specific_cell_features.dtype) #CxH_c\n",
    "        broadcasted_target_index = explicit_broadcast(cell_drug_graph.edge_index[1], drug_specific_cell_features)\n",
    "        drug_specific_cell_features_aggregated.scatter_reduce_(0, broadcasted_target_index.clone(), drug_specific_cell_features.clone(), \"sum\") # C x H_c\n",
    "        drug_specific_cell_features_aggregated = drug_specific_cell_features_aggregated + self.lin.bias\n",
    "        \n",
    "        ## Drug-sepcfic information\n",
    "        drug_node_features = self.gnn_drug(drug_graph.x.clone(), drug_graph.edge_index.clone(), drug_graph.edge_attr.clone()) # (Atoms + 1) x H_d\n",
    "        drug_super_node_features = drug_node_features[-1, :] # 1 x H_d\n",
    "        drug_super_node_features_exp = drug_super_node_features.unsqueeze(0).expand(drug_specific_cell_features_aggregated.shape[0], drug_super_node_features.shape[-1]) # CxH_d\n",
    "\n",
    "        ## Combine drug and cell information\n",
    "        drug_cell_features = torch.cat([drug_super_node_features_exp, drug_specific_cell_features_aggregated], 1) # C x (H_d + H_c)\n",
    "        drug_cell_features = torch.nn.functional.relu(drug_cell_features)\n",
    "        out = self.lin_end(drug_cell_features) # C x 1\n",
    "        \n",
    "        return out # drug_specific_cell_features_aggregated\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "8ec64eba-bc1c-46a3-8676-d970cd9199ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "device = torch.device(\"cuda:0\")\n",
    "heads = 2\n",
    "dropout = 0.1\n",
    "model = CustomGNN(38_975, 512, 64, dropout, heads, \"ELU\", device).to(device) #312, 512, .4, 4\n",
    "loss_fun = torch.nn.MSELoss()\n",
    "lr = 1e-4 #1e-4\n",
    "optim = torch.optim.Adam(params= model.parameters(), lr = lr) #, weight_decay=1e-5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "437e86e9-3cbb-4af1-a62e-f1890ff60e84",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d7478c71d3b247df98e014bb2a2c2ca6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/200 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.67\n",
      "0.69\n",
      "0.67\n",
      "0.66\n",
      "0.71\n",
      "0.73\n",
      "0.67\n",
      "0.65\n",
      "0.67\n",
      "0.65\n",
      "0.62\n",
      "0.60\n",
      "0.60\n",
      "0.70\n",
      "0.68\n",
      "0.82\n",
      "0.70\n",
      "0.66\n",
      "0.65\n",
      "0.64\n",
      "0.70\n",
      "0.65\n",
      "0.66\n",
      "0.65\n",
      "0.65\n",
      "0.65\n",
      "0.68\n",
      "0.65\n",
      "0.62\n",
      "0.61\n",
      "0.63\n",
      "0.63\n",
      "0.62\n",
      "0.66\n",
      "0.64\n",
      "0.61\n",
      "0.62\n",
      "0.61\n",
      "0.57\n",
      "0.64\n",
      "0.69\n",
      "0.70\n",
      "0.71\n",
      "0.67\n",
      "0.66\n",
      "0.61\n",
      "0.59\n",
      "0.60\n",
      "0.64\n",
      "0.66\n",
      "0.57\n",
      "0.58\n",
      "0.65\n",
      "0.62\n",
      "0.67\n",
      "0.62\n",
      "0.56\n",
      "0.58\n",
      "0.59\n",
      "0.67\n",
      "0.66\n",
      "0.70\n",
      "0.63\n",
      "0.62\n",
      "0.73\n",
      "0.72\n",
      "0.63\n",
      "0.56\n",
      "0.56\n",
      "0.65\n",
      "0.65\n",
      "0.65\n",
      "0.61\n",
      "0.61\n",
      "0.61\n",
      "0.57\n",
      "0.57\n",
      "0.57\n",
      "0.55\n",
      "0.59\n",
      "0.60\n",
      "0.62\n",
      "0.60\n",
      "0.61\n",
      "0.59\n",
      "0.60\n",
      "0.57\n",
      "0.60\n",
      "0.64\n",
      "0.59\n",
      "0.57\n",
      "0.57\n",
      "0.55\n",
      "0.60\n",
      "0.61\n",
      "0.57\n",
      "0.51\n",
      "0.55\n",
      "0.65\n",
      "0.61\n",
      "0.60\n",
      "0.62\n",
      "0.60\n",
      "0.57\n",
      "0.54\n",
      "0.59\n",
      "0.62\n",
      "0.63\n",
      "0.62\n",
      "0.59\n",
      "0.58\n",
      "0.58\n",
      "0.53\n",
      "0.54\n",
      "0.54\n",
      "0.55\n",
      "0.61\n",
      "0.56\n",
      "0.55\n",
      "0.56\n",
      "0.57\n",
      "0.62\n",
      "0.65\n",
      "0.62\n",
      "0.59\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[184], line 8\u001b[0m\n\u001b[1;32m      6\u001b[0m batch\u001b[38;5;241m.\u001b[39mcell_drug_graph\u001b[38;5;241m.\u001b[39medge_index \u001b[38;5;241m=\u001b[39m batch\u001b[38;5;241m.\u001b[39mcell_drug_graph\u001b[38;5;241m.\u001b[39mtrain_edge_index\n\u001b[1;32m      7\u001b[0m batch\u001b[38;5;241m.\u001b[39mcell_drug_graph\u001b[38;5;241m.\u001b[39medge_attr \u001b[38;5;241m=\u001b[39m batch\u001b[38;5;241m.\u001b[39mcell_drug_graph\u001b[38;5;241m.\u001b[39mtrain_edge_attr\n\u001b[0;32m----> 8\u001b[0m out \u001b[38;5;241m=\u001b[39m \u001b[43mmodel\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbatch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdrug_graph\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbatch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcell_drug_graph\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      9\u001b[0m loss \u001b[38;5;241m=\u001b[39m loss_fun(out, batch\u001b[38;5;241m.\u001b[39mtrain_labels\u001b[38;5;241m.\u001b[39munsqueeze(\u001b[38;5;241m1\u001b[39m)\u001b[38;5;241m.\u001b[39mto(device))\n\u001b[1;32m     10\u001b[0m optim\u001b[38;5;241m.\u001b[39mzero_grad()\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/torch/nn/modules/module.py:1532\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1530\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1531\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1532\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/torch/nn/modules/module.py:1541\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1536\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1537\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1538\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1539\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1540\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1541\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1543\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1544\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "Cell \u001b[0;32mIn[152], line 47\u001b[0m, in \u001b[0;36mCustomGNN.forward\u001b[0;34m(self, drug_graph, cell_drug_graph)\u001b[0m\n\u001b[1;32m     44\u001b[0m cell_node_features_lift \u001b[38;5;241m=\u001b[39m torch\u001b[38;5;241m.\u001b[39mindex_select(cell_node_features\u001b[38;5;241m.\u001b[39mclone(), \u001b[38;5;241m0\u001b[39m, cell_drug_graph\u001b[38;5;241m.\u001b[39medge_index[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39mclone()) \u001b[38;5;66;03m# E_cd x H_c\u001b[39;00m\n\u001b[1;32m     46\u001b[0m drug_specific_cell_features \u001b[38;5;241m=\u001b[39m cell_node_features_lift\u001b[38;5;241m.\u001b[39mclone() \u001b[38;5;241m*\u001b[39m cell_drug_graph\u001b[38;5;241m.\u001b[39medge_attr\u001b[38;5;241m.\u001b[39munsqueeze(\u001b[38;5;241m1\u001b[39m)\u001b[38;5;241m.\u001b[39mclone() \u001b[38;5;66;03m# E_cd x H_c\u001b[39;00m\n\u001b[0;32m---> 47\u001b[0m drug_specific_cell_features_aggregated \u001b[38;5;241m=\u001b[39m \u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mzeros\u001b[49m\u001b[43m(\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcell_drug_graph\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43medge_index\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdrug_specific_cell_features\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshape\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdevice\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdrug_specific_cell_features\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;66;03m#CxH_c\u001b[39;00m\n\u001b[1;32m     48\u001b[0m broadcasted_target_index \u001b[38;5;241m=\u001b[39m explicit_broadcast(cell_drug_graph\u001b[38;5;241m.\u001b[39medge_index[\u001b[38;5;241m1\u001b[39m], drug_specific_cell_features)\n\u001b[1;32m     49\u001b[0m drug_specific_cell_features_aggregated\u001b[38;5;241m.\u001b[39mscatter_reduce_(\u001b[38;5;241m0\u001b[39m, broadcasted_target_index\u001b[38;5;241m.\u001b[39mclone(), drug_specific_cell_features\u001b[38;5;241m.\u001b[39mclone(), \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msum\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;66;03m# C x H_c\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "epochs = 200  #800\n",
    "for epoch in tqdm(range(epochs)):\n",
    "    acc_loss = 0\n",
    "    model.train()\n",
    "    for batch in batch_graphs:\n",
    "        batch.cell_drug_graph.edge_index = batch.cell_drug_graph.train_edge_index\n",
    "        batch.cell_drug_graph.edge_attr = batch.cell_drug_graph.train_edge_attr\n",
    "        out = model(batch.drug_graph.to(device), batch.cell_drug_graph.to(device))\n",
    "        loss = loss_fun(out, batch.train_labels.unsqueeze(1).to(device))\n",
    "        optim.zero_grad()\n",
    "        loss.backward()\n",
    "        acc_loss += loss.item()\n",
    "        optim.step()\n",
    "        # break\n",
    "    print(f\"{acc_loss/len(batch_graphs):.2f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "b476a4d4-a2b1-4315-9084-bc5a3defb703",
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate(batch, use_train_edges = False):\n",
    "    with torch.inference_mode():\n",
    "        model.eval()\n",
    "        \n",
    "        if use_train_edges:\n",
    "            print(\"Use Train edges\")\n",
    "            batch.cell_drug_graph.edge_index = batch.cell_drug_graph.train_edge_index\n",
    "            batch.cell_drug_graph.edge_attr = batch.cell_drug_graph.train_edge_attr\n",
    "            batch.labels = batch.train_labels\n",
    "        if not use_train_edges:\n",
    "            print(\"Use Test edges\")\n",
    "            batch.cell_drug_graph.edge_index = batch.cell_drug_graph.test_edge_index\n",
    "            batch.cell_drug_graph.edge_attr = batch.cell_drug_graph.test_edge_attr\n",
    "            batch.labels = batch.test_labels\n",
    "            \n",
    "        out = model(batch.drug_graph.to(device), batch.cell_drug_graph.to(device))\n",
    "        r2 = r2_score(batch.labels, out.squeeze().cpu())\n",
    "    return r2\n",
    "    \n",
    "if acetalax_test:\n",
    "    print(evaluate(train_graph))\n",
    "    print(evaluate(test_graph))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "0ca57b32-959b-4844-bd27-a655eb091be5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor(176431, device='cuda:0')"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_train_instances = 0\n",
    "for batch in batch_graphs:\n",
    "    max_train_idx = batch.cell_drug_graph.train_edge_index[1, :].max()\n",
    "    num_train_instances += max_train_idx\n",
    "num_train_instances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "7ba7219d-adf9-4c74-b933-92ab636932e1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Use Test edges\n",
      "0.6016890253069137\n",
      "Use Test edges\n",
      "0.8285669371037201\n",
      "Use Test edges\n",
      "0.6158692468284103\n",
      "Use Test edges\n",
      "0.9412408300635664\n",
      "Use Test edges\n",
      "0.6348940625366772\n",
      "Use Test edges\n",
      "0.7215185195173673\n",
      "Use Test edges\n",
      "0.49533084613130185\n",
      "Use Test edges\n",
      "0.6522992556353004\n",
      "Use Test edges\n",
      "0.6143441665110734\n",
      "Use Test edges\n",
      "0.7325962249936453\n",
      "Use Test edges\n",
      "0.8966740837608608\n",
      "Use Test edges\n",
      "0.5994266451804824\n",
      "Use Test edges\n",
      "0.7364467114126593\n",
      "Use Test edges\n",
      "0.7503043207955296\n",
      "Use Test edges\n",
      "0.664351723216921\n",
      "Use Test edges\n",
      "0.7389551589170131\n",
      "Use Test edges\n",
      "0.4503127399718677\n",
      "Use Test edges\n",
      "0.8354514250936601\n",
      "Use Test edges\n",
      "0.5747474441511878\n",
      "Use Test edges\n",
      "0.6341509563997137\n",
      "Use Test edges\n",
      "0.7305905229250997\n",
      "Use Test edges\n",
      "0.37547306609086395\n",
      "Use Test edges\n",
      "0.788058487042244\n",
      "Use Test edges\n",
      "0.5724143342200176\n",
      "Use Test edges\n",
      "0.6825477911141518\n",
      "Use Test edges\n",
      "0.8724219647353203\n",
      "Use Test edges\n",
      "0.674392519205703\n",
      "Use Test edges\n",
      "0.624289061911222\n",
      "Use Test edges\n",
      "0.7550597404491607\n",
      "Use Test edges\n",
      "0.7041146187751135\n",
      "Use Test edges\n",
      "0.5537338491955239\n",
      "Use Test edges\n",
      "0.9240125449654102\n",
      "Use Test edges\n",
      "0.7740700819628259\n",
      "Use Test edges\n",
      "0.5108622648164325\n",
      "Use Test edges\n",
      "0.8945068861862573\n",
      "Use Test edges\n",
      "0.8496829053444677\n",
      "Use Test edges\n",
      "0.757869923633232\n",
      "Use Test edges\n",
      "0.7813670724961177\n",
      "Use Test edges\n",
      "0.6674452051749309\n",
      "Use Test edges\n",
      "0.8199229951659168\n",
      "Use Test edges\n",
      "0.8277299775184084\n",
      "Use Test edges\n",
      "0.49691775158004814\n",
      "Use Test edges\n",
      "-0.03629187988035554\n",
      "Use Test edges\n",
      "-0.09510155087913796\n",
      "Use Test edges\n",
      "0.7614532077401718\n",
      "Use Test edges\n",
      "0.01621963283265604\n",
      "Use Test edges\n",
      "0.7589564028278173\n",
      "Use Test edges\n",
      "0.4459461576902243\n",
      "Use Test edges\n",
      "0.3815609438868003\n",
      "Use Test edges\n",
      "0.5867436185116881\n",
      "Use Test edges\n",
      "0.5971111317879134\n",
      "Use Test edges\n",
      "0.5203572514240558\n",
      "Use Test edges\n",
      "0.28542042796712197\n",
      "Use Test edges\n",
      "0.6094335450004069\n",
      "Use Test edges\n",
      "0.662612963277412\n",
      "Use Test edges\n",
      "0.7561840580995494\n",
      "Use Test edges\n",
      "0.48119626247816993\n",
      "Use Test edges\n",
      "0.5959091579974738\n",
      "Use Test edges\n",
      "0.5045190167926701\n",
      "Use Test edges\n",
      "0.6958074935852274\n",
      "Use Test edges\n",
      "0.7121638199818803\n",
      "Use Test edges\n",
      "0.5196673460363279\n",
      "Use Test edges\n",
      "0.6652765115786884\n",
      "Use Test edges\n",
      "0.763190025760975\n",
      "Use Test edges\n",
      "0.5605127564302206\n",
      "Use Test edges\n",
      "0.5198466576592593\n",
      "Use Test edges\n",
      "0.35264673641103395\n",
      "Use Test edges\n",
      "0.6314990365418856\n",
      "Use Test edges\n",
      "0.390381221729475\n",
      "Use Test edges\n",
      "0.6057870504623006\n",
      "Use Test edges\n",
      "-0.4099516226980193\n",
      "Use Test edges\n",
      "0.4016212551164189\n",
      "Use Test edges\n",
      "0.854954446823825\n",
      "Use Test edges\n",
      "0.6912976416003227\n",
      "Use Test edges\n",
      "0.3245847385784678\n",
      "Use Test edges\n",
      "0.5563237368149146\n",
      "Use Test edges\n",
      "0.06863238914220549\n",
      "Use Test edges\n",
      "0.8097786109698861\n",
      "Use Test edges\n",
      "0.6940630702832886\n",
      "Use Test edges\n",
      "-0.4956873682478655\n",
      "Use Test edges\n",
      "0.6410328015909731\n",
      "Use Test edges\n",
      "0.8269965103791919\n",
      "Use Test edges\n",
      "0.6473622740829817\n",
      "Use Test edges\n",
      "0.6733620542888354\n",
      "Use Test edges\n",
      "0.7410165481041666\n",
      "Use Test edges\n",
      "0.8033243887047685\n",
      "Use Test edges\n",
      "0.6285542069830805\n",
      "Use Test edges\n",
      "0.21713510371110267\n",
      "Use Test edges\n",
      "0.2568699140945909\n",
      "Use Test edges\n",
      "0.5461931679272847\n",
      "Use Test edges\n",
      "0.29951906744239776\n",
      "Use Test edges\n",
      "-0.15422584323554012\n",
      "Use Test edges\n",
      "0.5664677774563125\n",
      "Use Test edges\n",
      "0.6585568130910937\n",
      "Use Test edges\n",
      "0.2887086267994792\n",
      "Use Test edges\n",
      "0.5311729642515114\n",
      "Use Test edges\n",
      "0.3970598820296497\n",
      "Use Test edges\n",
      "0.5076660357207947\n",
      "Use Test edges\n",
      "0.16556294536353333\n",
      "Use Test edges\n",
      "0.6223480812814789\n",
      "Use Test edges\n",
      "0.6529277870265549\n",
      "Use Test edges\n",
      "0.39342122979025085\n",
      "Use Test edges\n",
      "0.28965821442209505\n",
      "Use Test edges\n",
      "0.26452788543070105\n",
      "Use Test edges\n",
      "0.6491790722766151\n",
      "Use Test edges\n",
      "0.5246529982232948\n",
      "Use Test edges\n",
      "0.5827924903995096\n",
      "Use Test edges\n",
      "0.33576769439202336\n",
      "Use Test edges\n",
      "0.4167067197058191\n",
      "Use Test edges\n",
      "0.40255998768085666\n",
      "Use Test edges\n",
      "0.5894584690199678\n",
      "Use Test edges\n",
      "0.3169034672953138\n",
      "Use Test edges\n",
      "0.5746688141878653\n",
      "Use Test edges\n",
      "0.592864306647332\n",
      "Use Test edges\n",
      "0.5598511615990367\n",
      "Use Test edges\n",
      "0.18963413633427728\n",
      "Use Test edges\n",
      "0.45153475511497443\n",
      "Use Test edges\n",
      "0.5062851893977216\n",
      "Use Test edges\n",
      "0.5677018064070154\n",
      "Use Test edges\n",
      "0.3522726731745105\n",
      "Use Test edges\n",
      "-0.6515264872750526\n",
      "Use Test edges\n",
      "0.3366554515823854\n",
      "Use Test edges\n",
      "0.6415374444702462\n",
      "Use Test edges\n",
      "0.7287347262446987\n",
      "Use Test edges\n",
      "0.5487164156086851\n",
      "Use Test edges\n",
      "0.38508642198753706\n",
      "Use Test edges\n",
      "0.4269801221146047\n",
      "Use Test edges\n",
      "0.4474934708671159\n",
      "Use Test edges\n",
      "0.6551694356666823\n",
      "Use Test edges\n",
      "-1.8703805091511567\n",
      "Use Test edges\n",
      "0.6593278397180312\n",
      "Use Test edges\n",
      "0.5861959066084106\n",
      "Use Test edges\n",
      "0.5880847983204016\n",
      "Use Test edges\n",
      "0.31591268201307865\n",
      "Use Test edges\n",
      "0.18526519414576414\n",
      "Use Test edges\n",
      "0.6796014360601992\n",
      "Use Test edges\n",
      "0.41121588047592217\n",
      "Use Test edges\n",
      "0.12900378860528006\n",
      "Use Test edges\n",
      "0.511489711502724\n",
      "Use Test edges\n",
      "0.6529012092067277\n",
      "Use Test edges\n",
      "0.29974382246559383\n",
      "Use Test edges\n",
      "0.48127072752365163\n",
      "Use Test edges\n",
      "0.22986767753345005\n",
      "Use Test edges\n",
      "0.5356257040856419\n",
      "Use Test edges\n",
      "0.569765065556114\n",
      "Use Test edges\n",
      "-0.2484460063941396\n",
      "Use Test edges\n",
      "0.55206511070369\n",
      "Use Test edges\n",
      "0.4297323796104008\n",
      "Use Test edges\n",
      "0.5793713918516513\n",
      "Use Test edges\n",
      "0.14493735889985337\n",
      "Use Test edges\n",
      "-0.33782099230795204\n",
      "Use Test edges\n",
      "0.2796786336756829\n",
      "Use Test edges\n",
      "0.4989712005340944\n",
      "Use Test edges\n",
      "0.5743153952261918\n",
      "Use Test edges\n",
      "-0.07957260117812148\n",
      "Use Test edges\n",
      "0.3625588334175115\n",
      "Use Test edges\n",
      "0.48114274207529284\n",
      "Use Test edges\n",
      "0.6037542870938459\n",
      "Use Test edges\n",
      "0.4462867670918301\n",
      "Use Test edges\n",
      "0.5379929011159912\n",
      "Use Test edges\n",
      "0.4251701586894264\n",
      "Use Test edges\n",
      "0.2914727414432746\n",
      "Use Test edges\n",
      "0.38739503299892186\n",
      "Use Test edges\n",
      "0.6827052653147094\n",
      "Use Test edges\n",
      "0.5725622308070925\n",
      "Use Test edges\n",
      "0.6320496470968467\n",
      "Use Test edges\n",
      "0.5269565658587797\n",
      "Use Test edges\n",
      "-0.6521206272458189\n",
      "Use Test edges\n",
      "0.0058579936860105875\n",
      "Use Test edges\n",
      "0.3955882762973979\n",
      "Use Test edges\n",
      "0.23585032173127818\n"
     ]
    }
   ],
   "source": [
    "ids = []\n",
    "r2s = []\n",
    "for batch in batch_graphs:\n",
    "    r2 = evaluate(batch, False)\n",
    "    ids.append(batch.id)\n",
    "    r2s.append(r2)\n",
    "    print(r2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "b02e4bef-8f75-4d2e-ad67-8823c1632f8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## TODO multi label evaluation in sepaarte notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "89be34a7-f32f-4957-8502-3bdc9225f57a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug_ID</th>\n",
       "      <th>R2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>171.000000</td>\n",
       "      <td>171.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1456.625731</td>\n",
       "      <td>0.485331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>398.755137</td>\n",
       "      <td>0.333397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1003.000000</td>\n",
       "      <td>-1.870381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1058.500000</td>\n",
       "      <td>0.388888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1494.000000</td>\n",
       "      <td>0.566468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1854.500000</td>\n",
       "      <td>0.664814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2359.000000</td>\n",
       "      <td>0.941241</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Drug_ID          R2\n",
       "count   171.000000  171.000000\n",
       "mean   1456.625731    0.485331\n",
       "std     398.755137    0.333397\n",
       "min    1003.000000   -1.870381\n",
       "25%    1058.500000    0.388888\n",
       "50%    1494.000000    0.566468\n",
       "75%    1854.500000    0.664814\n",
       "max    2359.000000    0.941241"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {\n",
    "  \"Drug_ID\": ids,\n",
    "  \"R2\": r2s\n",
    "}\n",
    "\n",
    "results_df_new = pd.DataFrame(data=data)\n",
    "results_df_new.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "714c3016-661b-4b44-a47a-e82d95a69ba3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug_ID</th>\n",
       "      <th>R2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>171.000000</td>\n",
       "      <td>171.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1456.625731</td>\n",
       "      <td>0.472283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>398.755137</td>\n",
       "      <td>0.272487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1003.000000</td>\n",
       "      <td>-1.312067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1058.500000</td>\n",
       "      <td>0.362395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1494.000000</td>\n",
       "      <td>0.520782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1854.500000</td>\n",
       "      <td>0.643588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2359.000000</td>\n",
       "      <td>0.922170</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Drug_ID          R2\n",
       "count   171.000000  171.000000\n",
       "mean   1456.625731    0.472283\n",
       "std     398.755137    0.272487\n",
       "min    1003.000000   -1.312067\n",
       "25%    1058.500000    0.362395\n",
       "50%    1494.000000    0.520782\n",
       "75%    1854.500000    0.643588\n",
       "max    2359.000000    0.922170"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df_new.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b16663b7-84ad-4804-ab90-47fbd471b61d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# results_df_new.to_csv(\"r2_results_drugs_gat.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1058e63f-ae32-49d5-90fc-7ca4f1e638b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "for batch in batch_graphs:\n",
    "    print(evaluate(batch, True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de5a74f5-5b4c-4aad-8303-85d57ac9a2bd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for graph in test_graphs:\n",
    "    print(evaluate(graph))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53cd1f11-bcd7-4af3-9014-04582048c427",
   "metadata": {},
   "outputs": [],
   "source": [
    "raise Exception(\"Only old code below this cell.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
